// openclaw/dist/extensions/slack/test-api.js
import { n as vi } from "../../test.p_J6dB8a-sNRqGTRJ.js";
import { A as primeChannelOutboundSendMock } from "../../testing-CGqjYcO9.js";
import { o as createSlackActions, r as slackOutbound, t as slackPlugin } from "../../channel-CbApvJYP.js";
import { r as setSlackRuntime } from "../../runtime-BMyLMT2T.js";
import { t as sendMessageSlack } from "../../send-pJvSRc2i.js";
import { s as createSlackMonitorContext, t as prepareSlackMessage } from "../../prepare-DuSOQlR5.js";
//#region extensions/slack/src/outbound-payload.test-harness.ts
function createSlackOutboundPayloadHarness(params) {
	const sendSlack = vi.fn();
	primeChannelOutboundSendMock(sendSlack, {
		messageId: "sl-1",
		channelId: "C12345",
		ts: "1234.5678"
	}, params.sendResults);
	const ctx = {
		cfg: {},
		to: "C12345",
		text: "",
		payload: params.payload,
		deps: { sendSlack }
	};
	return {
		run: async () => await slackOutbound.sendPayload(ctx),
		sendMock: sendSlack,
		to: ctx.to
	};
}
//#endregion
//#region extensions/slack/src/monitor/message-handler/prepare.test-helpers.ts
function createInboundSlackTestContext(params) {
	return createSlackMonitorContext({
		cfg: params.cfg,
		accountId: "default",
		botToken: "token",
		app: { client: params.appClient ?? {} },
		runtime: {},
		botUserId: "B1",
		teamId: "T1",
		apiAppId: "A1",
		historyLimit: 0,
		sessionScope: "per-sender",
		mainKey: "main",
		dmEnabled: true,
		dmPolicy: "open",
		allowFrom: [],
		allowNameMatching: false,
		groupDmEnabled: true,
		groupDmChannels: [],
		defaultRequireMention: params.defaultRequireMention ?? true,
		channelsConfig: params.channelsConfig,
		groupPolicy: "open",
		useAccessGroups: false,
		reactionMode: "off",
		reactionAllowlist: [],
		replyToMode: params.replyToMode ?? "off",
		threadHistoryScope: "thread",
		threadInheritParent: false,
		threadRequireExplicitMention: params.threadRequireExplicitMention ?? false,
		slashCommand: {
			enabled: false,
			name: "openclaw",
			sessionPrefix: "slack:slash",
			ephemeral: true
		},
		textLimit: 4e3,
		ackReactionScope: "group-mentions",
		typingReaction: "",
		mediaMaxBytes: 1024,
		removeAckAfterReply: false
	});
}
//#endregion
export { createInboundSlackTestContext, createSlackActions, createSlackOutboundPayloadHarness, prepareSlackMessage, sendMessageSlack, setSlackRuntime, slackOutbound, slackPlugin };
