// openclaw/dist/extensions/matrix/index.js
import { r as formatErrorMessage } from "../../errors-CGMew-1w.js";
import "../../error-runtime-BNzEFRqy.js";
import { t as defineBundledChannelEntry } from "../../channel-entry-contract-DWlZ12A9.js";
import { n as registerMatrixCliMetadata } from "../../cli-metadata-BYEhNube.js";
//#region extensions/matrix/index.ts
var matrix_default = defineBundledChannelEntry({
	id: "matrix",
	name: "Matrix",
	description: "Matrix channel plugin (matrix-js-sdk)",
	importMetaUrl: import.meta.url,
	plugin: {
		specifier: "./channel-plugin-api.js",
		exportName: "matrixPlugin"
	},
	secrets: {
		specifier: "./secret-contract-api.js",
		exportName: "channelSecrets"
	},
	runtime: {
		specifier: "./runtime-api.js",
		exportName: "setMatrixRuntime"
	},
	registerCliMetadata: registerMatrixCliMetadata,
	registerFull(api) {
		import("./plugin-entry.handlers.runtime.js").then(({ ensureMatrixCryptoRuntime }) => ensureMatrixCryptoRuntime({ log: api.logger.info }).catch((err) => {
			const message = formatErrorMessage(err);
			api.logger.warn?.(`matrix: crypto runtime bootstrap failed: ${message}`);
		})).catch((err) => {
			const message = formatErrorMessage(err);
			api.logger.warn?.(`matrix: failed loading crypto bootstrap runtime: ${message}`);
		});
		api.registerGatewayMethod("matrix.verify.recoveryKey", async (ctx) => {
			const { handleVerifyRecoveryKey } = await import("./plugin-entry.handlers.runtime.js");
			await handleVerifyRecoveryKey(ctx);
		});
		api.registerGatewayMethod("matrix.verify.bootstrap", async (ctx) => {
			const { handleVerificationBootstrap } = await import("./plugin-entry.handlers.runtime.js");
			await handleVerificationBootstrap(ctx);
		});
		api.registerGatewayMethod("matrix.verify.status", async (ctx) => {
			const { handleVerificationStatus } = await import("./plugin-entry.handlers.runtime.js");
			await handleVerificationStatus(ctx);
		});
	}
});
//#endregion
export { matrix_default as default };
