// openclaw/dist/extensions/matrix/plugin-entry.handlers.runtime.js
import { s as normalizeOptionalString } from "../../string-coerce-CIT1-v_a.js";
import "../../text-runtime-BngW26kR.js";
import { t as formatMatrixErrorMessage } from "../../errors-Be2FRQ_f.js";
//#region extensions/matrix/src/plugin-entry.runtime.ts
function sendError(respond, err) {
	respond(false, { error: formatMatrixErrorMessage(err) });
}
async function ensureMatrixCryptoRuntime(...args) {
	const { ensureMatrixCryptoRuntime: ensureRuntime } = await import("../../deps-BMFItPlt.js");
	await ensureRuntime(...args);
}
async function handleVerifyRecoveryKey({ params, respond }) {
	try {
		const { verifyMatrixRecoveryKey } = await import("../../verification-O8-NdsXC.js");
		const key = normalizeOptionalString(params?.key);
		if (!key) {
			respond(false, { error: "key required" });
			return;
		}
		const result = await verifyMatrixRecoveryKey(key, { accountId: normalizeOptionalString(params?.accountId) });
		respond(result.success, result);
	} catch (err) {
		sendError(respond, err);
	}
}
async function handleVerificationBootstrap({ params, respond }) {
	try {
		const { bootstrapMatrixVerification } = await import("../../verification-O8-NdsXC.js");
		const result = await bootstrapMatrixVerification({
			accountId: normalizeOptionalString(params?.accountId),
			recoveryKey: typeof params?.recoveryKey === "string" ? params.recoveryKey : void 0,
			forceResetCrossSigning: params?.forceResetCrossSigning === true
		});
		respond(result.success, result);
	} catch (err) {
		sendError(respond, err);
	}
}
async function handleVerificationStatus({ params, respond }) {
	try {
		const { getMatrixVerificationStatus } = await import("../../verification-O8-NdsXC.js");
		respond(true, await getMatrixVerificationStatus({
			accountId: normalizeOptionalString(params?.accountId),
			includeRecoveryKey: params?.includeRecoveryKey === true
		}));
	} catch (err) {
		sendError(respond, err);
	}
}
//#endregion
export { ensureMatrixCryptoRuntime, handleVerificationBootstrap, handleVerificationStatus, handleVerifyRecoveryKey };
