// @line/bot-sdk/dist/cjs/line-bot-client.factory.generated.js
"use strict";
/**
 * This file is autogenerated.
 *
 * Generated by scripts/line-bot-client-generator/main.mjs
 * Do not edit this file directly.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.createLineBotClientDelegates = createLineBotClientDelegates;
const insight = require("./insight/api.js");
const liff = require("./liff/api.js");
const manageAudience = require("./manage-audience/api.js");
const messagingApi = require("./messaging-api/api.js");
const moduleOperation = require("./module/api.js");
const moduleAttach = require("./module-attach/api.js");
const shop = require("./shop/api.js");
function createLineBotClientDelegates(config) {
    return {
        insight: new insight.InsightClient({
            baseURL: config.apiBaseURL,
            channelAccessToken: config.channelAccessToken,
            defaultHeaders: config.defaultHeaders,
        }),
        liff: new liff.LiffClient({
            baseURL: config.apiBaseURL,
            channelAccessToken: config.channelAccessToken,
            defaultHeaders: config.defaultHeaders,
        }),
        manageAudience: new manageAudience.ManageAudienceClient({
            baseURL: config.apiBaseURL,
            channelAccessToken: config.channelAccessToken,
            defaultHeaders: config.defaultHeaders,
        }),
        manageAudienceBlob: new manageAudience.ManageAudienceBlobClient({
            baseURL: config.dataApiBaseURL,
            channelAccessToken: config.channelAccessToken,
            defaultHeaders: config.defaultHeaders,
        }),
        messagingApi: new messagingApi.MessagingApiClient({
            baseURL: config.apiBaseURL,
            channelAccessToken: config.channelAccessToken,
            defaultHeaders: config.defaultHeaders,
        }),
        messagingApiBlob: new messagingApi.MessagingApiBlobClient({
            baseURL: config.dataApiBaseURL,
            channelAccessToken: config.channelAccessToken,
            defaultHeaders: config.defaultHeaders,
        }),
        lineModule: new moduleOperation.LineModuleClient({
            baseURL: config.apiBaseURL,
            channelAccessToken: config.channelAccessToken,
            defaultHeaders: config.defaultHeaders,
        }),
        lineModuleAttach: new moduleAttach.LineModuleAttachClient({
            baseURL: config.managerBaseURL,
            channelAccessToken: config.channelAccessToken,
            defaultHeaders: config.defaultHeaders,
        }),
        shop: new shop.ShopClient({
            baseURL: config.apiBaseURL,
            channelAccessToken: config.channelAccessToken,
            defaultHeaders: config.defaultHeaders,
        }),
    };
}
