// @line/bot-sdk/dist/cjs/line-bot-client.generated.js
"use strict";
/**
 * This file is autogenerated.
 *
 * Generated by scripts/line-bot-client-generator/main.mjs
 * Do not edit this file directly.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.LineBotClientBase = void 0;
class LineBotClientBase {
    /**
     * Retrieves the demographic attributes for a LINE Official Account\'s friends.You can only retrieve information about friends for LINE Official Accounts created by users in Japan (JP), Thailand (TH), Taiwan (TW) and Indonesia (ID).
     *
     * @see <a href="https://developers.line.biz/en/reference/messaging-api/#get-demographic"> Documentation</a>
     */
    async getFriendsDemographics() {
        return this.clients.insight.getFriendsDemographics();
    }
    /**
     * Retrieves the demographic attributes for a LINE Official Account\'s friends.You can only retrieve information about friends for LINE Official Accounts created by users in Japan (JP), Thailand (TH), Taiwan (TW) and Indonesia (ID). .
     * This method includes HttpInfo object to return additional information.
     *
     * @see <a href="https://developers.line.biz/en/reference/messaging-api/#get-demographic"> Documentation</a>
     */
    async getFriendsDemographicsWithHttpInfo() {
        return this.clients.insight.getFriendsDemographicsWithHttpInfo();
    }
    /**
     * Returns statistics about how users interact with narrowcast messages or broadcast messages sent from your LINE Official Account.
     * @summary Get user interaction statistics
     * @param requestId Request ID of a narrowcast message or broadcast message. Each Messaging API request has a request ID.
     *
     * @see <a href="https://developers.line.biz/en/reference/messaging-api/#get-message-event">Get user interaction statistics Documentation</a>
     */
    async getMessageEvent(requestId) {
        return this.clients.insight.getMessageEvent(requestId);
    }
    /**
     * Returns statistics about how users interact with narrowcast messages or broadcast messages sent from your LINE Official Account. .
     * This method includes HttpInfo object to return additional information.
     * @summary Get user interaction statistics
     * @param requestId Request ID of a narrowcast message or broadcast message. Each Messaging API request has a request ID.
     *
     * @see <a href="https://developers.line.biz/en/reference/messaging-api/#get-message-event">Get user interaction statistics Documentation</a>
     */
    async getMessageEventWithHttpInfo(requestId) {
        return this.clients.insight.getMessageEventWithHttpInfo(requestId);
    }
    /**
     * Returns the number of users who have added the LINE Official Account on or before a specified date.
     * @summary Get number of followers
     * @param date Date for which to retrieve the number of followers.  Format: yyyyMMdd (e.g. 20191231) Timezone: UTC+9
     *
     * @see <a href="https://developers.line.biz/en/reference/messaging-api/#get-number-of-followers">Get number of followers Documentation</a>
     */
    async getNumberOfFollowers(date) {
        return this.clients.insight.getNumberOfFollowers(date);
    }
    /**
     * Returns the number of users who have added the LINE Official Account on or before a specified date. .
     * This method includes HttpInfo object to return additional information.
     * @summary Get number of followers
     * @param date Date for which to retrieve the number of followers.  Format: yyyyMMdd (e.g. 20191231) Timezone: UTC+9
     *
     * @see <a href="https://developers.line.biz/en/reference/messaging-api/#get-number-of-followers">Get number of followers Documentation</a>
     */
    async getNumberOfFollowersWithHttpInfo(date) {
        return this.clients.insight.getNumberOfFollowersWithHttpInfo(date);
    }
    /**
     * Returns the number of messages sent from LINE Official Account on a specified day.
     * @summary Get number of message deliveries
     * @param date Date for which to retrieve number of sent messages. - Format: yyyyMMdd (e.g. 20191231) - Timezone: UTC+9
     *
     * @see <a href="https://developers.line.biz/en/reference/messaging-api/#get-number-of-delivery-messages">Get number of message deliveries Documentation</a>
     */
    async getNumberOfMessageDeliveries(date) {
        return this.clients.insight.getNumberOfMessageDeliveries(date);
    }
    /**
     * Returns the number of messages sent from LINE Official Account on a specified day. .
     * This method includes HttpInfo object to return additional information.
     * @summary Get number of message deliveries
     * @param date Date for which to retrieve number of sent messages. - Format: yyyyMMdd (e.g. 20191231) - Timezone: UTC+9
     *
     * @see <a href="https://developers.line.biz/en/reference/messaging-api/#get-number-of-delivery-messages">Get number of message deliveries Documentation</a>
     */
    async getNumberOfMessageDeliveriesWithHttpInfo(date) {
        return this.clients.insight.getNumberOfMessageDeliveriesWithHttpInfo(date);
    }
    /**
     * You can check the per-unit statistics of how users interact with push messages and multicast messages sent from your LINE Official Account.
     * @param customAggregationUnit Name of aggregation unit specified when sending the message. Case-sensitive. For example, `Promotion_a` and `Promotion_A` are regarded as different unit names.
     * @param from Start date of aggregation period.  Format: yyyyMMdd (e.g. 20210301) Time zone: UTC+9
     * @param to End date of aggregation period. The end date can be specified for up to 30 days later. For example, if the start date is 20210301, the latest end date is 20210331.  Format: yyyyMMdd (e.g. 20210301) Time zone: UTC+9
     *
     * @see <a href="https://developers.line.biz/en/reference/messaging-api/#get-statistics-per-unit"> Documentation</a>
     */
    async getStatisticsPerUnit(customAggregationUnit, from, to) {
        return this.clients.insight.getStatisticsPerUnit(customAggregationUnit, from, to);
    }
    /**
     * You can check the per-unit statistics of how users interact with push messages and multicast messages sent from your LINE Official Account. .
     * This method includes HttpInfo object to return additional information.
     * @param customAggregationUnit Name of aggregation unit specified when sending the message. Case-sensitive. For example, `Promotion_a` and `Promotion_A` are regarded as different unit names.
     * @param from Start date of aggregation period.  Format: yyyyMMdd (e.g. 20210301) Time zone: UTC+9
     * @param to End date of aggregation period. The end date can be specified for up to 30 days later. For example, if the start date is 20210301, the latest end date is 20210331.  Format: yyyyMMdd (e.g. 20210301) Time zone: UTC+9
     *
     * @see <a href="https://developers.line.biz/en/reference/messaging-api/#get-statistics-per-unit"> Documentation</a>
     */
    async getStatisticsPerUnitWithHttpInfo(customAggregationUnit, from, to) {
        return this.clients.insight.getStatisticsPerUnitWithHttpInfo(customAggregationUnit, from, to);
    }
    /**
     * Adding the LIFF app to a channel
     * @summary Create LIFF app
     * @param addLiffAppRequest
     *
     * @see <a href="https://developers.line.biz/en/reference/liff-server/#add-liff-app">Create LIFF app Documentation</a>
     */
    async addLIFFApp(addLiffAppRequest) {
        return this.clients.liff.addLIFFApp(addLiffAppRequest);
    }
    /**
     * Adding the LIFF app to a channel.
     * This method includes HttpInfo object to return additional information.
     * @summary Create LIFF app
     * @param addLiffAppRequest
     *
     * @see <a href="https://developers.line.biz/en/reference/liff-server/#add-liff-app">Create LIFF app Documentation</a>
     */
    async addLIFFAppWithHttpInfo(addLiffAppRequest) {
        return this.clients.liff.addLIFFAppWithHttpInfo(addLiffAppRequest);
    }
    /**
     * Deletes a LIFF app from a channel.
     * @summary Delete LIFF app from a channel
     * @param liffId ID of the LIFF app to be updated
     *
     * @see <a href="https://developers.line.biz/en/reference/liff-server/#delete-liff-app">Delete LIFF app from a channel Documentation</a>
     */
    async deleteLIFFApp(liffId) {
        return this.clients.liff.deleteLIFFApp(liffId);
    }
    /**
     * Deletes a LIFF app from a channel. .
     * This method includes HttpInfo object to return additional information.
     * @summary Delete LIFF app from a channel
     * @param liffId ID of the LIFF app to be updated
     *
     * @see <a href="https://developers.line.biz/en/reference/liff-server/#delete-liff-app">Delete LIFF app from a channel Documentation</a>
     */
    async deleteLIFFAppWithHttpInfo(liffId) {
        return this.clients.liff.deleteLIFFAppWithHttpInfo(liffId);
    }
    /**
     * Gets information on all the LIFF apps added to the channel.
     * @summary Get all LIFF apps
     *
     * @see <a href="https://developers.line.biz/en/reference/liff-server/#get-all-liff-apps">Get all LIFF apps Documentation</a>
     */
    async getAllLIFFApps() {
        return this.clients.liff.getAllLIFFApps();
    }
    /**
     * Gets information on all the LIFF apps added to the channel..
     * This method includes HttpInfo object to return additional information.
     * @summary Get all LIFF apps
     *
     * @see <a href="https://developers.line.biz/en/reference/liff-server/#get-all-liff-apps">Get all LIFF apps Documentation</a>
     */
    async getAllLIFFAppsWithHttpInfo() {
        return this.clients.liff.getAllLIFFAppsWithHttpInfo();
    }
    /**
     * Update LIFF app settings
     * @summary Update LIFF app from a channel
     * @param liffId ID of the LIFF app to be updated
     * @param updateLiffAppRequest
     *
     * @see <a href="https://developers.line.biz/en/reference/liff-server/#update-liff-app">Update LIFF app from a channel Documentation</a>
     */
    async updateLIFFApp(liffId, updateLiffAppRequest) {
        return this.clients.liff.updateLIFFApp(liffId, updateLiffAppRequest);
    }
    /**
     * Update LIFF app settings.
     * This method includes HttpInfo object to return additional information.
     * @summary Update LIFF app from a channel
     * @param liffId ID of the LIFF app to be updated
     * @param updateLiffAppRequest
     *
     * @see <a href="https://developers.line.biz/en/reference/liff-server/#update-liff-app">Update LIFF app from a channel Documentation</a>
     */
    async updateLIFFAppWithHttpInfo(liffId, updateLiffAppRequest) {
        return this.clients.liff.updateLIFFAppWithHttpInfo(liffId, updateLiffAppRequest);
    }
    /**
     * Add user IDs or Identifiers for Advertisers (IFAs) to an audience for uploading user IDs (by JSON)
     * @param addAudienceToAudienceGroupRequest
     *
     * @see <a href="https://developers.line.biz/en/reference/messaging-api/#update-upload-audience-group"> Documentation</a>
     */
    async addAudienceToAudienceGroup(addAudienceToAudienceGroupRequest) {
        return this.clients.manageAudience.addAudienceToAudienceGroup(addAudienceToAudienceGroupRequest);
    }
    /**
     * Add user IDs or Identifiers for Advertisers (IFAs) to an audience for uploading user IDs (by JSON).
     * This method includes HttpInfo object to return additional information.
     * @param addAudienceToAudienceGroupRequest
     *
     * @see <a href="https://developers.line.biz/en/reference/messaging-api/#update-upload-audience-group"> Documentation</a>
     */
    async addAudienceToAudienceGroupWithHttpInfo(addAudienceToAudienceGroupRequest) {
        return this.clients.manageAudience.addAudienceToAudienceGroupWithHttpInfo(addAudienceToAudienceGroupRequest);
    }
    /**
     * Create audience for uploading user IDs (by JSON)
     * @param createAudienceGroupRequest
     *
     * @see <a href="https://developers.line.biz/en/reference/messaging-api/#create-upload-audience-group"> Documentation</a>
     */
    async createAudienceGroup(createAudienceGroupRequest) {
        return this.clients.manageAudience.createAudienceGroup(createAudienceGroupRequest);
    }
    /**
     * Create audience for uploading user IDs (by JSON).
     * This method includes HttpInfo object to return additional information.
     * @param createAudienceGroupRequest
     *
     * @see <a href="https://developers.line.biz/en/reference/messaging-api/#create-upload-audience-group"> Documentation</a>
     */
    async createAudienceGroupWithHttpInfo(createAudienceGroupRequest) {
        return this.clients.manageAudience.createAudienceGroupWithHttpInfo(createAudienceGroupRequest);
    }
    /**
     * Create audience for click-based retargeting
     * @param createClickBasedAudienceGroupRequest
     *
     * @see <a href="https://developers.line.biz/en/reference/messaging-api/#create-click-audience-group"> Documentation</a>
     */
    async createClickBasedAudienceGroup(createClickBasedAudienceGroupRequest) {
        return this.clients.manageAudience.createClickBasedAudienceGroup(createClickBasedAudienceGroupRequest);
    }
    /**
     * Create audience for click-based retargeting.
     * This method includes HttpInfo object to return additional information.
     * @param createClickBasedAudienceGroupRequest
     *
     * @see <a href="https://developers.line.biz/en/reference/messaging-api/#create-click-audience-group"> Documentation</a>
     */
    async createClickBasedAudienceGroupWithHttpInfo(createClickBasedAudienceGroupRequest) {
        return this.clients.manageAudience.createClickBasedAudienceGroupWithHttpInfo(createClickBasedAudienceGroupRequest);
    }
    /**
     * Create audience for impression-based retargeting
     * @param createImpBasedAudienceGroupRequest
     *
     * @see <a href="https://developers.line.biz/en/reference/messaging-api/#create-imp-audience-group"> Documentation</a>
     */
    async createImpBasedAudienceGroup(createImpBasedAudienceGroupRequest) {
        return this.clients.manageAudience.createImpBasedAudienceGroup(createImpBasedAudienceGroupRequest);
    }
    /**
     * Create audience for impression-based retargeting.
     * This method includes HttpInfo object to return additional information.
     * @param createImpBasedAudienceGroupRequest
     *
     * @see <a href="https://developers.line.biz/en/reference/messaging-api/#create-imp-audience-group"> Documentation</a>
     */
    async createImpBasedAudienceGroupWithHttpInfo(createImpBasedAudienceGroupRequest) {
        return this.clients.manageAudience.createImpBasedAudienceGroupWithHttpInfo(createImpBasedAudienceGroupRequest);
    }
    /**
     * Delete audience
     * @param audienceGroupId The audience ID.
     *
     * @see <a href="https://developers.line.biz/en/reference/messaging-api/#delete-audience-group"> Documentation</a>
     */
    async deleteAudienceGroup(audienceGroupId) {
        return this.clients.manageAudience.deleteAudienceGroup(audienceGroupId);
    }
    /**
     * Delete audience.
     * This method includes HttpInfo object to return additional information.
     * @param audienceGroupId The audience ID.
     *
     * @see <a href="https://developers.line.biz/en/reference/messaging-api/#delete-audience-group"> Documentation</a>
     */
    async deleteAudienceGroupWithHttpInfo(audienceGroupId) {
        return this.clients.manageAudience.deleteAudienceGroupWithHttpInfo(audienceGroupId);
    }
    /**
     * Gets audience data.
     * @param audienceGroupId The audience ID.
     *
     * @see <a href="https://developers.line.biz/en/reference/messaging-api/#get-audience-group"> Documentation</a>
     */
    async getAudienceData(audienceGroupId) {
        return this.clients.manageAudience.getAudienceData(audienceGroupId);
    }
    /**
     * Gets audience data..
     * This method includes HttpInfo object to return additional information.
     * @param audienceGroupId The audience ID.
     *
     * @see <a href="https://developers.line.biz/en/reference/messaging-api/#get-audience-group"> Documentation</a>
     */
    async getAudienceDataWithHttpInfo(audienceGroupId) {
        return this.clients.manageAudience.getAudienceDataWithHttpInfo(audienceGroupId);
    }
    /**
     * Gets data for more than one audience.
     * @param page The page to return when getting (paginated) results. Must be 1 or higher.
     * @param description The name of the audience(s) to return. You can search for partial matches. This is case-insensitive, meaning AUDIENCE and audience are considered identical. If omitted, the name of the audience(s) will not be used as a search criterion.
     * @param status The status of the audience(s) to return. If omitted, the status of the audience(s) will not be used as a search criterion.
     * @param size The number of audiences per page. Default: 20 Max: 40
     * @param includesExternalPublicGroups true (default): Get public audiences created in all channels linked to the same bot. false: Get audiences created in the same channel.
     * @param createRoute How the audience was created. If omitted, all audiences are included.  `OA_MANAGER`: Return only audiences created with LINE Official Account Manager (opens new window). `MESSAGING_API`: Return only audiences created with Messaging API.
     *
     * @see <a href="https://developers.line.biz/en/reference/messaging-api/#get-audience-groups"> Documentation</a>
     */
    async getAudienceGroups(page, description, status, size, includesExternalPublicGroups, createRoute) {
        return this.clients.manageAudience.getAudienceGroups(page, description, status, size, includesExternalPublicGroups, createRoute);
    }
    /**
     * Gets data for more than one audience..
     * This method includes HttpInfo object to return additional information.
     * @param page The page to return when getting (paginated) results. Must be 1 or higher.
     * @param description The name of the audience(s) to return. You can search for partial matches. This is case-insensitive, meaning AUDIENCE and audience are considered identical. If omitted, the name of the audience(s) will not be used as a search criterion.
     * @param status The status of the audience(s) to return. If omitted, the status of the audience(s) will not be used as a search criterion.
     * @param size The number of audiences per page. Default: 20 Max: 40
     * @param includesExternalPublicGroups true (default): Get public audiences created in all channels linked to the same bot. false: Get audiences created in the same channel.
     * @param createRoute How the audience was created. If omitted, all audiences are included.  `OA_MANAGER`: Return only audiences created with LINE Official Account Manager (opens new window). `MESSAGING_API`: Return only audiences created with Messaging API.
     *
     * @see <a href="https://developers.line.biz/en/reference/messaging-api/#get-audience-groups"> Documentation</a>
     */
    async getAudienceGroupsWithHttpInfo(page, description, status, size, includesExternalPublicGroups, createRoute) {
        return this.clients.manageAudience.getAudienceGroupsWithHttpInfo(page, description, status, size, includesExternalPublicGroups, createRoute);
    }
    /**
     * Gets audience data.
     * @param audienceGroupId The audience ID.
     *
     * @see <a href="https://developers.line.biz/en/reference/messaging-api/#get-shared-audience"> Documentation</a>
     */
    async getSharedAudienceData(audienceGroupId) {
        return this.clients.manageAudience.getSharedAudienceData(audienceGroupId);
    }
    /**
     * Gets audience data..
     * This method includes HttpInfo object to return additional information.
     * @param audienceGroupId The audience ID.
     *
     * @see <a href="https://developers.line.biz/en/reference/messaging-api/#get-shared-audience"> Documentation</a>
     */
    async getSharedAudienceDataWithHttpInfo(audienceGroupId) {
        return this.clients.manageAudience.getSharedAudienceDataWithHttpInfo(audienceGroupId);
    }
    /**
     * Gets data for more than one audience, including those shared by the Business Manager.
     * @param page The page to return when getting (paginated) results. Must be 1 or higher.
     * @param description The name of the audience(s) to return. You can search for partial matches. This is case-insensitive, meaning AUDIENCE and audience are considered identical. If omitted, the name of the audience(s) will not be used as a search criterion.
     * @param status The status of the audience(s) to return. If omitted, the status of the audience(s) will not be used as a search criterion.
     * @param size The number of audiences per page. Default: 20 Max: 40
     * @param createRoute How the audience was created. If omitted, all audiences are included.  `OA_MANAGER`: Return only audiences created with LINE Official Account Manager (opens new window). `MESSAGING_API`: Return only audiences created with Messaging API.
     * @param includesOwnedAudienceGroups true: Include audienceGroups owned by LINE Official Account Manager false: Respond only audienceGroups shared by Business Manager
     *
     * @see <a href="https://developers.line.biz/en/reference/messaging-api/#get-shared-audience-list"> Documentation</a>
     */
    async getSharedAudienceGroups(page, description, status, size, createRoute, includesOwnedAudienceGroups) {
        return this.clients.manageAudience.getSharedAudienceGroups(page, description, status, size, createRoute, includesOwnedAudienceGroups);
    }
    /**
     * Gets data for more than one audience, including those shared by the Business Manager..
     * This method includes HttpInfo object to return additional information.
     * @param page The page to return when getting (paginated) results. Must be 1 or higher.
     * @param description The name of the audience(s) to return. You can search for partial matches. This is case-insensitive, meaning AUDIENCE and audience are considered identical. If omitted, the name of the audience(s) will not be used as a search criterion.
     * @param status The status of the audience(s) to return. If omitted, the status of the audience(s) will not be used as a search criterion.
     * @param size The number of audiences per page. Default: 20 Max: 40
     * @param createRoute How the audience was created. If omitted, all audiences are included.  `OA_MANAGER`: Return only audiences created with LINE Official Account Manager (opens new window). `MESSAGING_API`: Return only audiences created with Messaging API.
     * @param includesOwnedAudienceGroups true: Include audienceGroups owned by LINE Official Account Manager false: Respond only audienceGroups shared by Business Manager
     *
     * @see <a href="https://developers.line.biz/en/reference/messaging-api/#get-shared-audience-list"> Documentation</a>
     */
    async getSharedAudienceGroupsWithHttpInfo(page, description, status, size, createRoute, includesOwnedAudienceGroups) {
        return this.clients.manageAudience.getSharedAudienceGroupsWithHttpInfo(page, description, status, size, createRoute, includesOwnedAudienceGroups);
    }
    /**
     * Renames an existing audience.
     * @param audienceGroupId The audience ID.
     * @param updateAudienceGroupDescriptionRequest
     *
     * @see <a href="https://developers.line.biz/en/reference/messaging-api/#set-description-audience-group"> Documentation</a>
     */
    async updateAudienceGroupDescription(audienceGroupId, updateAudienceGroupDescriptionRequest) {
        return this.clients.manageAudience.updateAudienceGroupDescription(audienceGroupId, updateAudienceGroupDescriptionRequest);
    }
    /**
     * Renames an existing audience..
     * This method includes HttpInfo object to return additional information.
     * @param audienceGroupId The audience ID.
     * @param updateAudienceGroupDescriptionRequest
     *
     * @see <a href="https://developers.line.biz/en/reference/messaging-api/#set-description-audience-group"> Documentation</a>
     */
    async updateAudienceGroupDescriptionWithHttpInfo(audienceGroupId, updateAudienceGroupDescriptionRequest) {
        return this.clients.manageAudience.updateAudienceGroupDescriptionWithHttpInfo(audienceGroupId, updateAudienceGroupDescriptionRequest);
    }
    /**
     * Add user IDs or Identifiers for Advertisers (IFAs) to an audience for uploading user IDs (by file).
     * @param file A text file with one user ID or IFA entered per line. Specify text/plain as Content-Type. Max file number: 1 Max number: 1,500,000
     * @param audienceGroupId The audience ID.
     * @param uploadDescription The description to register with the job
     *
     * @see <a href="https://developers.line.biz/en/reference/messaging-api/#update-upload-audience-group-by-file"> Documentation</a>
     */
    async addUserIdsToAudience(file, audienceGroupId, uploadDescription) {
        return this.clients.manageAudienceBlob.addUserIdsToAudience(file, audienceGroupId, uploadDescription);
    }
    /**
     * Add user IDs or Identifiers for Advertisers (IFAs) to an audience for uploading user IDs (by file)..
     * This method includes HttpInfo object to return additional information.
     * @param file A text file with one user ID or IFA entered per line. Specify text/plain as Content-Type. Max file number: 1 Max number: 1,500,000
     * @param audienceGroupId The audience ID.
     * @param uploadDescription The description to register with the job
     *
     * @see <a href="https://developers.line.biz/en/reference/messaging-api/#update-upload-audience-group-by-file"> Documentation</a>
     */
    async addUserIdsToAudienceWithHttpInfo(file, audienceGroupId, uploadDescription) {
        return this.clients.manageAudienceBlob.addUserIdsToAudienceWithHttpInfo(file, audienceGroupId, uploadDescription);
    }
    /**
     * Create audience for uploading user IDs (by file).
     * @param file A text file with one user ID or IFA entered per line. Specify text/plain as Content-Type. Max file number: 1 Max number: 1,500,000
     * @param description The audience\\\'s name. This is case-insensitive, meaning AUDIENCE and audience are considered identical. Max character limit: 120
     * @param isIfaAudience To specify recipients by IFAs: set `true`. To specify recipients by user IDs: set `false` or omit isIfaAudience property.
     * @param uploadDescription The description to register for the job (in `jobs[].description`).
     *
     * @see <a href="https://developers.line.biz/en/reference/messaging-api/#create-upload-audience-group-by-file"> Documentation</a>
     */
    async createAudienceForUploadingUserIds(file, description, isIfaAudience, uploadDescription) {
        return this.clients.manageAudienceBlob.createAudienceForUploadingUserIds(file, description, isIfaAudience, uploadDescription);
    }
    /**
     * Create audience for uploading user IDs (by file)..
     * This method includes HttpInfo object to return additional information.
     * @param file A text file with one user ID or IFA entered per line. Specify text/plain as Content-Type. Max file number: 1 Max number: 1,500,000
     * @param description The audience\\\'s name. This is case-insensitive, meaning AUDIENCE and audience are considered identical. Max character limit: 120
     * @param isIfaAudience To specify recipients by IFAs: set `true`. To specify recipients by user IDs: set `false` or omit isIfaAudience property.
     * @param uploadDescription The description to register for the job (in `jobs[].description`).
     *
     * @see <a href="https://developers.line.biz/en/reference/messaging-api/#create-upload-audience-group-by-file"> Documentation</a>
     */
    async createAudienceForUploadingUserIdsWithHttpInfo(file, description, isIfaAudience, uploadDescription) {
        return this.clients.manageAudienceBlob.createAudienceForUploadingUserIdsWithHttpInfo(file, description, isIfaAudience, uploadDescription);
    }
    /**
     * Sends a message to multiple users at any time.
     * @param broadcastRequest
     * @param xLineRetryKey Retry key. Specifies the UUID in hexadecimal format (e.g., `123e4567-e89b-12d3-a456-426614174000`) generated by any method. The retry key isn\'t generated by LINE. Each developer must generate their own retry key.
     *
     * @see <a href="https://developers.line.biz/en/reference/messaging-api/#send-broadcast-message"> Documentation</a>
     */
    async broadcast(broadcastRequest, xLineRetryKey) {
        return this.clients.messagingApi.broadcast(broadcastRequest, xLineRetryKey);
    }
    /**
     * Sends a message to multiple users at any time..
     * This method includes HttpInfo object to return additional information.
     * @param broadcastRequest
     * @param xLineRetryKey Retry key. Specifies the UUID in hexadecimal format (e.g., `123e4567-e89b-12d3-a456-426614174000`) generated by any method. The retry key isn\'t generated by LINE. Each developer must generate their own retry key.
     *
     * @see <a href="https://developers.line.biz/en/reference/messaging-api/#send-broadcast-message"> Documentation</a>
     */
    async broadcastWithHttpInfo(broadcastRequest, xLineRetryKey) {
        return this.clients.messagingApi.broadcastWithHttpInfo(broadcastRequest, xLineRetryKey);
    }
    /**
     * Cancel default rich menu
     *
     * @see <a href="https://developers.line.biz/en/reference/messaging-api/#cancel-default-rich-menu"> Documentation</a>
     */
    async cancelDefaultRichMenu() {
        return this.clients.messagingApi.cancelDefaultRichMenu();
    }
    /**
     * Cancel default rich menu.
     * This method includes HttpInfo object to return additional information.
     *
     * @see <a href="https://developers.line.biz/en/reference/messaging-api/#cancel-default-rich-menu"> Documentation</a>
     */
    async cancelDefaultRichMenuWithHttpInfo() {
        return this.clients.messagingApi.cancelDefaultRichMenuWithHttpInfo();
    }
    /**
     * Close coupon
     * @param couponId
     *
     * @see <a href="https://developers.line.biz/en/reference/messaging-api/#discontinue-coupon"> Documentation</a>
     */
    async closeCoupon(couponId) {
        return this.clients.messagingApi.closeCoupon(couponId);
    }
    /**
     * Close coupon.
     * This method includes HttpInfo object to return additional information.
     * @param couponId
     *
     * @see <a href="https://developers.line.biz/en/reference/messaging-api/#discontinue-coupon"> Documentation</a>
     */
    async closeCouponWithHttpInfo(couponId) {
        return this.clients.messagingApi.closeCouponWithHttpInfo(couponId);
    }
    /**
     * Create a new coupon. Define coupon details such as type, title, and validity period.
     * @param couponCreateRequest
     *
     * @see <a href="https://developers.line.biz/en/reference/messaging-api/#create-coupon"> Documentation</a>
     */
    async createCoupon(couponCreateRequest) {
        return this.clients.messagingApi.createCoupon(couponCreateRequest);
    }
    /**
     * Create a new coupon. Define coupon details such as type, title, and validity period..
     * This method includes HttpInfo object to return additional information.
     * @param couponCreateRequest
     *
     * @see <a href="https://developers.line.biz/en/reference/messaging-api/#create-coupon"> Documentation</a>
     */
    async createCouponWithHttpInfo(couponCreateRequest) {
        return this.clients.messagingApi.createCouponWithHttpInfo(couponCreateRequest);
    }
    /**
     * Create rich menu
     * @param richMenuRequest
     *
     * @see <a href="https://developers.line.biz/en/reference/messaging-api/#create-rich-menu"> Documentation</a>
     */
    async createRichMenu(richMenuRequest) {
        return this.clients.messagingApi.createRichMenu(richMenuRequest);
    }
    /**
     * Create rich menu.
     * This method includes HttpInfo object to return additional information.
     * @param richMenuRequest
     *
     * @see <a href="https://developers.line.biz/en/reference/messaging-api/#create-rich-menu"> Documentation</a>
     */
    async createRichMenuWithHttpInfo(richMenuRequest) {
        return this.clients.messagingApi.createRichMenuWithHttpInfo(richMenuRequest);
    }
    /**
     * Create rich menu alias
     * @param createRichMenuAliasRequest
     *
     * @see <a href="https://developers.line.biz/en/reference/messaging-api/#create-rich-menu-alias"> Documentation</a>
     */
    async createRichMenuAlias(createRichMenuAliasRequest) {
        return this.clients.messagingApi.createRichMenuAlias(createRichMenuAliasRequest);
    }
    /**
     * Create rich menu alias.
     * This method includes HttpInfo object to return additional information.
     * @param createRichMenuAliasRequest
     *
     * @see <a href="https://developers.line.biz/en/reference/messaging-api/#create-rich-menu-alias"> Documentation</a>
     */
    async createRichMenuAliasWithHttpInfo(createRichMenuAliasRequest) {
        return this.clients.messagingApi.createRichMenuAliasWithHttpInfo(createRichMenuAliasRequest);
    }
    /**
     * Deletes a rich menu.
     * @param richMenuId ID of a rich menu
     *
     * @see <a href="https://developers.line.biz/en/reference/messaging-api/#delete-rich-menu"> Documentation</a>
     */
    async deleteRichMenu(richMenuId) {
        return this.clients.messagingApi.deleteRichMenu(richMenuId);
    }
    /**
     * Deletes a rich menu..
     * This method includes HttpInfo object to return additional information.
     * @param richMenuId ID of a rich menu
     *
     * @see <a href="https://developers.line.biz/en/reference/messaging-api/#delete-rich-menu"> Documentation</a>
     */
    async deleteRichMenuWithHttpInfo(richMenuId) {
        return this.clients.messagingApi.deleteRichMenuWithHttpInfo(richMenuId);
    }
    /**
     * Delete rich menu alias
     * @param richMenuAliasId Rich menu alias ID that you want to delete.
     *
     * @see <a href="https://developers.line.biz/en/reference/messaging-api/#delete-rich-menu-alias"> Documentation</a>
     */
    async deleteRichMenuAlias(richMenuAliasId) {
        return this.clients.messagingApi.deleteRichMenuAlias(richMenuAliasId);
    }
    /**
     * Delete rich menu alias.
     * This method includes HttpInfo object to return additional information.
     * @param richMenuAliasId Rich menu alias ID that you want to delete.
     *
     * @see <a href="https://developers.line.biz/en/reference/messaging-api/#delete-rich-menu-alias"> Documentation</a>
     */
    async deleteRichMenuAliasWithHttpInfo(richMenuAliasId) {
        return this.clients.messagingApi.deleteRichMenuAliasWithHttpInfo(richMenuAliasId);
    }
    /**
     * Get name list of units used this month
     * @param limit The maximum number of aggregation units you can get per request.
     * @param start Value of the continuation token found in the next property of the JSON object returned in the response. If you can\'t get all the aggregation units in one request, include this parameter to get the remaining array.
     *
     * @see <a href="https://developers.line.biz/en/reference/messaging-api/#get-name-list-of-units-used-this-month"> Documentation</a>
     */
    async getAggregationUnitNameList(limit, start) {
        return this.clients.messagingApi.getAggregationUnitNameList(limit, start);
    }
    /**
     * Get name list of units used this month.
     * This method includes HttpInfo object to return additional information.
     * @param limit The maximum number of aggregation units you can get per request.
     * @param start Value of the continuation token found in the next property of the JSON object returned in the response. If you can\'t get all the aggregation units in one request, include this parameter to get the remaining array.
     *
     * @see <a href="https://developers.line.biz/en/reference/messaging-api/#get-name-list-of-units-used-this-month"> Documentation</a>
     */
    async getAggregationUnitNameListWithHttpInfo(limit, start) {
        return this.clients.messagingApi.getAggregationUnitNameListWithHttpInfo(limit, start);
    }
    /**
     * Get number of units used this month
     *
     * @see <a href="https://developers.line.biz/en/reference/messaging-api/#get-number-of-units-used-this-month"> Documentation</a>
     */
    async getAggregationUnitUsage() {
        return this.clients.messagingApi.getAggregationUnitUsage();
    }
    /**
     * Get number of units used this month.
     * This method includes HttpInfo object to return additional information.
     *
     * @see <a href="https://developers.line.biz/en/reference/messaging-api/#get-number-of-units-used-this-month"> Documentation</a>
     */
    async getAggregationUnitUsageWithHttpInfo() {
        return this.clients.messagingApi.getAggregationUnitUsageWithHttpInfo();
    }
    /**
     * Get bot info
     *
     * @see <a href="https://developers.line.biz/en/reference/messaging-api/#get-bot-info"> Documentation</a>
     */
    async getBotInfo() {
        return this.clients.messagingApi.getBotInfo();
    }
    /**
     * Get bot info.
     * This method includes HttpInfo object to return additional information.
     *
     * @see <a href="https://developers.line.biz/en/reference/messaging-api/#get-bot-info"> Documentation</a>
     */
    async getBotInfoWithHttpInfo() {
        return this.clients.messagingApi.getBotInfoWithHttpInfo();
    }
    /**
     * Get coupon detail
     * @param couponId
     *
     * @see <a href="https://developers.line.biz/en/reference/messaging-api/#get-coupon"> Documentation</a>
     */
    async getCouponDetail(couponId) {
        return this.clients.messagingApi.getCouponDetail(couponId);
    }
    /**
     * Get coupon detail.
     * This method includes HttpInfo object to return additional information.
     * @param couponId
     *
     * @see <a href="https://developers.line.biz/en/reference/messaging-api/#get-coupon"> Documentation</a>
     */
    async getCouponDetailWithHttpInfo(couponId) {
        return this.clients.messagingApi.getCouponDetailWithHttpInfo(couponId);
    }
    /**
     * Gets the ID of the default rich menu set with the Messaging API.
     *
     * @see <a href="https://developers.line.biz/en/reference/messaging-api/#get-default-rich-menu-id"> Documentation</a>
     */
    async getDefaultRichMenuId() {
        return this.clients.messagingApi.getDefaultRichMenuId();
    }
    /**
     * Gets the ID of the default rich menu set with the Messaging API..
     * This method includes HttpInfo object to return additional information.
     *
     * @see <a href="https://developers.line.biz/en/reference/messaging-api/#get-default-rich-menu-id"> Documentation</a>
     */
    async getDefaultRichMenuIdWithHttpInfo() {
        return this.clients.messagingApi.getDefaultRichMenuIdWithHttpInfo();
    }
    /**
     * Get a list of users who added your LINE Official Account as a friend
     * @param start Value of the continuation token found in the next property of the JSON object returned in the response. Include this parameter to get the next array of user IDs.
     * @param limit The maximum number of user IDs to retrieve in a single request.
     *
     * @see <a href="https://developers.line.biz/en/reference/messaging-api/#get-follower-ids"> Documentation</a>
     */
    async getFollowers(start, limit) {
        return this.clients.messagingApi.getFollowers(start, limit);
    }
    /**
     * Get a list of users who added your LINE Official Account as a friend.
     * This method includes HttpInfo object to return additional information.
     * @param start Value of the continuation token found in the next property of the JSON object returned in the response. Include this parameter to get the next array of user IDs.
     * @param limit The maximum number of user IDs to retrieve in a single request.
     *
     * @see <a href="https://developers.line.biz/en/reference/messaging-api/#get-follower-ids"> Documentation</a>
     */
    async getFollowersWithHttpInfo(start, limit) {
        return this.clients.messagingApi.getFollowersWithHttpInfo(start, limit);
    }
    /**
     * Get number of users in a group chat
     * @param groupId Group ID
     *
     * @see <a href="https://developers.line.biz/en/reference/messaging-api/#get-members-group-count"> Documentation</a>
     */
    async getGroupMemberCount(groupId) {
        return this.clients.messagingApi.getGroupMemberCount(groupId);
    }
    /**
     * Get number of users in a group chat.
     * This method includes HttpInfo object to return additional information.
     * @param groupId Group ID
     *
     * @see <a href="https://developers.line.biz/en/reference/messaging-api/#get-members-group-count"> Documentation</a>
     */
    async getGroupMemberCountWithHttpInfo(groupId) {
        return this.clients.messagingApi.getGroupMemberCountWithHttpInfo(groupId);
    }
    /**
     * Get group chat member profile
     * @param groupId Group ID
     * @param userId User ID
     *
     * @see <a href="https://developers.line.biz/en/reference/messaging-api/#get-group-member-profile"> Documentation</a>
     */
    async getGroupMemberProfile(groupId, userId) {
        return this.clients.messagingApi.getGroupMemberProfile(groupId, userId);
    }
    /**
     * Get group chat member profile.
     * This method includes HttpInfo object to return additional information.
     * @param groupId Group ID
     * @param userId User ID
     *
     * @see <a href="https://developers.line.biz/en/reference/messaging-api/#get-group-member-profile"> Documentation</a>
     */
    async getGroupMemberProfileWithHttpInfo(groupId, userId) {
        return this.clients.messagingApi.getGroupMemberProfileWithHttpInfo(groupId, userId);
    }
    /**
     * Get group chat member user IDs
     * @param groupId Group ID
     * @param start Value of the continuation token found in the `next` property of the JSON object returned in the response. Include this parameter to get the next array of user IDs for the members of the group.
     *
     * @see <a href="https://developers.line.biz/en/reference/messaging-api/#get-group-member-user-ids"> Documentation</a>
     */
    async getGroupMembersIds(groupId, start) {
        return this.clients.messagingApi.getGroupMembersIds(groupId, start);
    }
    /**
     * Get group chat member user IDs.
     * This method includes HttpInfo object to return additional information.
     * @param groupId Group ID
     * @param start Value of the continuation token found in the `next` property of the JSON object returned in the response. Include this parameter to get the next array of user IDs for the members of the group.
     *
     * @see <a href="https://developers.line.biz/en/reference/messaging-api/#get-group-member-user-ids"> Documentation</a>
     */
    async getGroupMembersIdsWithHttpInfo(groupId, start) {
        return this.clients.messagingApi.getGroupMembersIdsWithHttpInfo(groupId, start);
    }
    /**
     * Get group chat summary
     * @param groupId Group ID
     *
     * @see <a href="https://developers.line.biz/en/reference/messaging-api/#get-group-summary"> Documentation</a>
     */
    async getGroupSummary(groupId) {
        return this.clients.messagingApi.getGroupSummary(groupId);
    }
    /**
     * Get group chat summary.
     * This method includes HttpInfo object to return additional information.
     * @param groupId Group ID
     *
     * @see <a href="https://developers.line.biz/en/reference/messaging-api/#get-group-summary"> Documentation</a>
     */
    async getGroupSummaryWithHttpInfo(groupId) {
        return this.clients.messagingApi.getGroupSummaryWithHttpInfo(groupId);
    }
    /**
     * Get a list of user IDs who joined the membership.
     * @param membershipId Membership plan ID.
     * @param start A continuation token to get next remaining membership user IDs. Returned only when there are remaining user IDs that weren\'t returned in the userIds property in the previous request. The continuation token expires in 24 hours (86,400 seconds).
     * @param limit The max number of items to return for this API call. The value is set to 300 by default, but the max acceptable value is 1000.
     *
     * @see <a href="https://developers.line.biz/en/reference/messaging-api/#get-membership-user-ids"> Documentation</a>
     */
    async getJoinedMembershipUsers(membershipId, start, limit) {
        return this.clients.messagingApi.getJoinedMembershipUsers(membershipId, start, limit);
    }
    /**
     * Get a list of user IDs who joined the membership..
     * This method includes HttpInfo object to return additional information.
     * @param membershipId Membership plan ID.
     * @param start A continuation token to get next remaining membership user IDs. Returned only when there are remaining user IDs that weren\'t returned in the userIds property in the previous request. The continuation token expires in 24 hours (86,400 seconds).
     * @param limit The max number of items to return for this API call. The value is set to 300 by default, but the max acceptable value is 1000.
     *
     * @see <a href="https://developers.line.biz/en/reference/messaging-api/#get-membership-user-ids"> Documentation</a>
     */
    async getJoinedMembershipUsersWithHttpInfo(membershipId, start, limit) {
        return this.clients.messagingApi.getJoinedMembershipUsersWithHttpInfo(membershipId, start, limit);
    }
    /**
     * Get a list of memberships.
     *
     * @see <a href="https://developers.line.biz/en/reference/messaging-api/#get-membership-plans"> Documentation</a>
     */
    async getMembershipList() {
        return this.clients.messagingApi.getMembershipList();
    }
    /**
     * Get a list of memberships..
     * This method includes HttpInfo object to return additional information.
     *
     * @see <a href="https://developers.line.biz/en/reference/messaging-api/#get-membership-plans"> Documentation</a>
     */
    async getMembershipListWithHttpInfo() {
        return this.clients.messagingApi.getMembershipListWithHttpInfo();
    }
    /**
     * Get a user\'s membership subscription.
     * @param userId User ID
     *
     * @see <a href="https://developers.line.biz/en/reference/messaging-api/#get-a-users-membership-subscription-status"> Documentation</a>
     */
    async getMembershipSubscription(userId) {
        return this.clients.messagingApi.getMembershipSubscription(userId);
    }
    /**
     * Get a user\'s membership subscription..
     * This method includes HttpInfo object to return additional information.
     * @param userId User ID
     *
     * @see <a href="https://developers.line.biz/en/reference/messaging-api/#get-a-users-membership-subscription-status"> Documentation</a>
     */
    async getMembershipSubscriptionWithHttpInfo(userId) {
        return this.clients.messagingApi.getMembershipSubscriptionWithHttpInfo(userId);
    }
    /**
     * Gets the target limit for sending messages in the current month. The total number of the free messages and the additional messages is returned.
     *
     * @see <a href="https://developers.line.biz/en/reference/messaging-api/#get-quota"> Documentation</a>
     */
    async getMessageQuota() {
        return this.clients.messagingApi.getMessageQuota();
    }
    /**
     * Gets the target limit for sending messages in the current month. The total number of the free messages and the additional messages is returned..
     * This method includes HttpInfo object to return additional information.
     *
     * @see <a href="https://developers.line.biz/en/reference/messaging-api/#get-quota"> Documentation</a>
     */
    async getMessageQuotaWithHttpInfo() {
        return this.clients.messagingApi.getMessageQuotaWithHttpInfo();
    }
    /**
     * Gets the number of messages sent in the current month.
     *
     * @see <a href="https://developers.line.biz/en/reference/messaging-api/#get-consumption"> Documentation</a>
     */
    async getMessageQuotaConsumption() {
        return this.clients.messagingApi.getMessageQuotaConsumption();
    }
    /**
     * Gets the number of messages sent in the current month..
     * This method includes HttpInfo object to return additional information.
     *
     * @see <a href="https://developers.line.biz/en/reference/messaging-api/#get-consumption"> Documentation</a>
     */
    async getMessageQuotaConsumptionWithHttpInfo() {
        return this.clients.messagingApi.getMessageQuotaConsumptionWithHttpInfo();
    }
    /**
     * Gets the status of a narrowcast message.
     * @param requestId The narrowcast message\'s request ID. Each Messaging API request has a request ID.
     *
     * @see <a href="https://developers.line.biz/en/reference/messaging-api/#get-narrowcast-progress-status"> Documentation</a>
     */
    async getNarrowcastProgress(requestId) {
        return this.clients.messagingApi.getNarrowcastProgress(requestId);
    }
    /**
     * Gets the status of a narrowcast message..
     * This method includes HttpInfo object to return additional information.
     * @param requestId The narrowcast message\'s request ID. Each Messaging API request has a request ID.
     *
     * @see <a href="https://developers.line.biz/en/reference/messaging-api/#get-narrowcast-progress-status"> Documentation</a>
     */
    async getNarrowcastProgressWithHttpInfo(requestId) {
        return this.clients.messagingApi.getNarrowcastProgressWithHttpInfo(requestId);
    }
    /**
     * Get number of sent broadcast messages
     * @param date Date the messages were sent  Format: yyyyMMdd (e.g. 20191231) Timezone: UTC+9
     *
     * @see <a href="https://developers.line.biz/en/reference/messaging-api/#get-number-of-broadcast-messages"> Documentation</a>
     */
    async getNumberOfSentBroadcastMessages(date) {
        return this.clients.messagingApi.getNumberOfSentBroadcastMessages(date);
    }
    /**
     * Get number of sent broadcast messages.
     * This method includes HttpInfo object to return additional information.
     * @param date Date the messages were sent  Format: yyyyMMdd (e.g. 20191231) Timezone: UTC+9
     *
     * @see <a href="https://developers.line.biz/en/reference/messaging-api/#get-number-of-broadcast-messages"> Documentation</a>
     */
    async getNumberOfSentBroadcastMessagesWithHttpInfo(date) {
        return this.clients.messagingApi.getNumberOfSentBroadcastMessagesWithHttpInfo(date);
    }
    /**
     * Get number of sent multicast messages
     * @param date Date the messages were sent  Format: `yyyyMMdd` (e.g. `20191231`) Timezone: UTC+9
     *
     * @see <a href="https://developers.line.biz/en/reference/messaging-api/#get-number-of-multicast-messages"> Documentation</a>
     */
    async getNumberOfSentMulticastMessages(date) {
        return this.clients.messagingApi.getNumberOfSentMulticastMessages(date);
    }
    /**
     * Get number of sent multicast messages.
     * This method includes HttpInfo object to return additional information.
     * @param date Date the messages were sent  Format: `yyyyMMdd` (e.g. `20191231`) Timezone: UTC+9
     *
     * @see <a href="https://developers.line.biz/en/reference/messaging-api/#get-number-of-multicast-messages"> Documentation</a>
     */
    async getNumberOfSentMulticastMessagesWithHttpInfo(date) {
        return this.clients.messagingApi.getNumberOfSentMulticastMessagesWithHttpInfo(date);
    }
    /**
     * Get number of sent push messages
     * @param date Date the messages were sent  Format: `yyyyMMdd` (e.g. `20191231`) Timezone: UTC+9
     *
     * @see <a href="https://developers.line.biz/en/reference/messaging-api/#get-number-of-push-messages"> Documentation</a>
     */
    async getNumberOfSentPushMessages(date) {
        return this.clients.messagingApi.getNumberOfSentPushMessages(date);
    }
    /**
     * Get number of sent push messages.
     * This method includes HttpInfo object to return additional information.
     * @param date Date the messages were sent  Format: `yyyyMMdd` (e.g. `20191231`) Timezone: UTC+9
     *
     * @see <a href="https://developers.line.biz/en/reference/messaging-api/#get-number-of-push-messages"> Documentation</a>
     */
    async getNumberOfSentPushMessagesWithHttpInfo(date) {
        return this.clients.messagingApi.getNumberOfSentPushMessagesWithHttpInfo(date);
    }
    /**
     * Get number of sent reply messages
     * @param date Date the messages were sent  Format: `yyyyMMdd` (e.g. `20191231`) Timezone: UTC+9
     *
     * @see <a href="https://developers.line.biz/en/reference/messaging-api/#get-number-of-reply-messages"> Documentation</a>
     */
    async getNumberOfSentReplyMessages(date) {
        return this.clients.messagingApi.getNumberOfSentReplyMessages(date);
    }
    /**
     * Get number of sent reply messages.
     * This method includes HttpInfo object to return additional information.
     * @param date Date the messages were sent  Format: `yyyyMMdd` (e.g. `20191231`) Timezone: UTC+9
     *
     * @see <a href="https://developers.line.biz/en/reference/messaging-api/#get-number-of-reply-messages"> Documentation</a>
     */
    async getNumberOfSentReplyMessagesWithHttpInfo(date) {
        return this.clients.messagingApi.getNumberOfSentReplyMessagesWithHttpInfo(date);
    }
    /**
     * Get number of sent LINE notification messages
     * @param date Date the message was sent  Format: `yyyyMMdd` (Example:`20211231`) Time zone: UTC+9
     *
     * @see <a href="https://developers.line.biz/en/reference/partner-docs/#get-number-of-sent-line-notification-messages"> Documentation</a>
     */
    async getPNPMessageStatistics(date) {
        return this.clients.messagingApi.getPNPMessageStatistics(date);
    }
    /**
     * Get number of sent LINE notification messages　.
     * This method includes HttpInfo object to return additional information.
     * @param date Date the message was sent  Format: `yyyyMMdd` (Example:`20211231`) Time zone: UTC+9
     *
     * @see <a href="https://developers.line.biz/en/reference/partner-docs/#get-number-of-sent-line-notification-messages"> Documentation</a>
     */
    async getPNPMessageStatisticsWithHttpInfo(date) {
        return this.clients.messagingApi.getPNPMessageStatisticsWithHttpInfo(date);
    }
    /**
     * Get profile
     * @param userId User ID
     *
     * @see <a href="https://developers.line.biz/en/reference/messaging-api/#get-profile"> Documentation</a>
     */
    async getProfile(userId) {
        return this.clients.messagingApi.getProfile(userId);
    }
    /**
     * Get profile.
     * This method includes HttpInfo object to return additional information.
     * @param userId User ID
     *
     * @see <a href="https://developers.line.biz/en/reference/messaging-api/#get-profile"> Documentation</a>
     */
    async getProfileWithHttpInfo(userId) {
        return this.clients.messagingApi.getProfileWithHttpInfo(userId);
    }
    /**
     * Gets a rich menu via a rich menu ID.
     * @param richMenuId ID of a rich menu
     *
     * @see <a href="https://developers.line.biz/en/reference/messaging-api/#get-rich-menu"> Documentation</a>
     */
    async getRichMenu(richMenuId) {
        return this.clients.messagingApi.getRichMenu(richMenuId);
    }
    /**
     * Gets a rich menu via a rich menu ID..
     * This method includes HttpInfo object to return additional information.
     * @param richMenuId ID of a rich menu
     *
     * @see <a href="https://developers.line.biz/en/reference/messaging-api/#get-rich-menu"> Documentation</a>
     */
    async getRichMenuWithHttpInfo(richMenuId) {
        return this.clients.messagingApi.getRichMenuWithHttpInfo(richMenuId);
    }
    /**
     * Get rich menu alias information
     * @param richMenuAliasId The rich menu alias ID whose information you want to obtain.
     *
     * @see <a href="https://developers.line.biz/en/reference/messaging-api/#get-rich-menu-alias-by-id"> Documentation</a>
     */
    async getRichMenuAlias(richMenuAliasId) {
        return this.clients.messagingApi.getRichMenuAlias(richMenuAliasId);
    }
    /**
     * Get rich menu alias information.
     * This method includes HttpInfo object to return additional information.
     * @param richMenuAliasId The rich menu alias ID whose information you want to obtain.
     *
     * @see <a href="https://developers.line.biz/en/reference/messaging-api/#get-rich-menu-alias-by-id"> Documentation</a>
     */
    async getRichMenuAliasWithHttpInfo(richMenuAliasId) {
        return this.clients.messagingApi.getRichMenuAliasWithHttpInfo(richMenuAliasId);
    }
    /**
     * Get list of rich menu alias
     *
     * @see <a href="https://developers.line.biz/en/reference/messaging-api/#get-rich-menu-alias-list"> Documentation</a>
     */
    async getRichMenuAliasList() {
        return this.clients.messagingApi.getRichMenuAliasList();
    }
    /**
     * Get list of rich menu alias.
     * This method includes HttpInfo object to return additional information.
     *
     * @see <a href="https://developers.line.biz/en/reference/messaging-api/#get-rich-menu-alias-list"> Documentation</a>
     */
    async getRichMenuAliasListWithHttpInfo() {
        return this.clients.messagingApi.getRichMenuAliasListWithHttpInfo();
    }
    /**
     * Get the status of Replace or unlink a linked rich menus in batches.
     * @param requestId A request ID used to batch control the rich menu linked to the user. Each Messaging API request has a request ID.
     *
     * @see <a href="https://developers.line.biz/en/reference/messaging-api/#get-batch-control-rich-menus-progress-status"> Documentation</a>
     */
    async getRichMenuBatchProgress(requestId) {
        return this.clients.messagingApi.getRichMenuBatchProgress(requestId);
    }
    /**
     * Get the status of Replace or unlink a linked rich menus in batches..
     * This method includes HttpInfo object to return additional information.
     * @param requestId A request ID used to batch control the rich menu linked to the user. Each Messaging API request has a request ID.
     *
     * @see <a href="https://developers.line.biz/en/reference/messaging-api/#get-batch-control-rich-menus-progress-status"> Documentation</a>
     */
    async getRichMenuBatchProgressWithHttpInfo(requestId) {
        return this.clients.messagingApi.getRichMenuBatchProgressWithHttpInfo(requestId);
    }
    /**
     * Get rich menu ID of user
     * @param userId User ID. Found in the `source` object of webhook event objects. Do not use the LINE ID used in LINE.
     *
     * @see <a href="https://developers.line.biz/en/reference/messaging-api/#get-rich-menu-id-of-user"> Documentation</a>
     */
    async getRichMenuIdOfUser(userId) {
        return this.clients.messagingApi.getRichMenuIdOfUser(userId);
    }
    /**
     * Get rich menu ID of user.
     * This method includes HttpInfo object to return additional information.
     * @param userId User ID. Found in the `source` object of webhook event objects. Do not use the LINE ID used in LINE.
     *
     * @see <a href="https://developers.line.biz/en/reference/messaging-api/#get-rich-menu-id-of-user"> Documentation</a>
     */
    async getRichMenuIdOfUserWithHttpInfo(userId) {
        return this.clients.messagingApi.getRichMenuIdOfUserWithHttpInfo(userId);
    }
    /**
     * Get rich menu list
     *
     * @see <a href="https://developers.line.biz/en/reference/messaging-api/#get-rich-menu-list"> Documentation</a>
     */
    async getRichMenuList() {
        return this.clients.messagingApi.getRichMenuList();
    }
    /**
     * Get rich menu list.
     * This method includes HttpInfo object to return additional information.
     *
     * @see <a href="https://developers.line.biz/en/reference/messaging-api/#get-rich-menu-list"> Documentation</a>
     */
    async getRichMenuListWithHttpInfo() {
        return this.clients.messagingApi.getRichMenuListWithHttpInfo();
    }
    /**
     * Get number of users in a multi-person chat
     * @param roomId Room ID
     *
     * @see <a href="https://developers.line.biz/en/reference/messaging-api/#get-members-room-count"> Documentation</a>
     */
    async getRoomMemberCount(roomId) {
        return this.clients.messagingApi.getRoomMemberCount(roomId);
    }
    /**
     * Get number of users in a multi-person chat.
     * This method includes HttpInfo object to return additional information.
     * @param roomId Room ID
     *
     * @see <a href="https://developers.line.biz/en/reference/messaging-api/#get-members-room-count"> Documentation</a>
     */
    async getRoomMemberCountWithHttpInfo(roomId) {
        return this.clients.messagingApi.getRoomMemberCountWithHttpInfo(roomId);
    }
    /**
     * Get multi-person chat member profile
     * @param roomId Room ID
     * @param userId User ID
     *
     * @see <a href="https://developers.line.biz/en/reference/messaging-api/#get-room-member-profile"> Documentation</a>
     */
    async getRoomMemberProfile(roomId, userId) {
        return this.clients.messagingApi.getRoomMemberProfile(roomId, userId);
    }
    /**
     * Get multi-person chat member profile.
     * This method includes HttpInfo object to return additional information.
     * @param roomId Room ID
     * @param userId User ID
     *
     * @see <a href="https://developers.line.biz/en/reference/messaging-api/#get-room-member-profile"> Documentation</a>
     */
    async getRoomMemberProfileWithHttpInfo(roomId, userId) {
        return this.clients.messagingApi.getRoomMemberProfileWithHttpInfo(roomId, userId);
    }
    /**
     * Get multi-person chat member user IDs
     * @param roomId Room ID
     * @param start Value of the continuation token found in the `next` property of the JSON object returned in the response. Include this parameter to get the next array of user IDs for the members of the group.
     *
     * @see <a href="https://developers.line.biz/en/reference/messaging-api/#get-room-member-user-ids"> Documentation</a>
     */
    async getRoomMembersIds(roomId, start) {
        return this.clients.messagingApi.getRoomMembersIds(roomId, start);
    }
    /**
     * Get multi-person chat member user IDs.
     * This method includes HttpInfo object to return additional information.
     * @param roomId Room ID
     * @param start Value of the continuation token found in the `next` property of the JSON object returned in the response. Include this parameter to get the next array of user IDs for the members of the group.
     *
     * @see <a href="https://developers.line.biz/en/reference/messaging-api/#get-room-member-user-ids"> Documentation</a>
     */
    async getRoomMembersIdsWithHttpInfo(roomId, start) {
        return this.clients.messagingApi.getRoomMembersIdsWithHttpInfo(roomId, start);
    }
    /**
     * Get webhook endpoint information
     *
     * @see <a href="https://developers.line.biz/en/reference/messaging-api/#get-webhook-endpoint-information"> Documentation</a>
     */
    async getWebhookEndpoint() {
        return this.clients.messagingApi.getWebhookEndpoint();
    }
    /**
     * Get webhook endpoint information.
     * This method includes HttpInfo object to return additional information.
     *
     * @see <a href="https://developers.line.biz/en/reference/messaging-api/#get-webhook-endpoint-information"> Documentation</a>
     */
    async getWebhookEndpointWithHttpInfo() {
        return this.clients.messagingApi.getWebhookEndpointWithHttpInfo();
    }
    /**
     * Issue link token
     * @param userId User ID for the LINE account to be linked. Found in the `source` object of account link event objects. Do not use the LINE ID used in LINE.
     *
     * @see <a href="https://developers.line.biz/en/reference/messaging-api/#issue-link-token"> Documentation</a>
     */
    async issueLinkToken(userId) {
        return this.clients.messagingApi.issueLinkToken(userId);
    }
    /**
     * Issue link token.
     * This method includes HttpInfo object to return additional information.
     * @param userId User ID for the LINE account to be linked. Found in the `source` object of account link event objects. Do not use the LINE ID used in LINE.
     *
     * @see <a href="https://developers.line.biz/en/reference/messaging-api/#issue-link-token"> Documentation</a>
     */
    async issueLinkTokenWithHttpInfo(userId) {
        return this.clients.messagingApi.issueLinkTokenWithHttpInfo(userId);
    }
    /**
     * Leave group chat
     * @param groupId Group ID
     *
     * @see <a href="https://developers.line.biz/en/reference/messaging-api/#leave-group"> Documentation</a>
     */
    async leaveGroup(groupId) {
        return this.clients.messagingApi.leaveGroup(groupId);
    }
    /**
     * Leave group chat.
     * This method includes HttpInfo object to return additional information.
     * @param groupId Group ID
     *
     * @see <a href="https://developers.line.biz/en/reference/messaging-api/#leave-group"> Documentation</a>
     */
    async leaveGroupWithHttpInfo(groupId) {
        return this.clients.messagingApi.leaveGroupWithHttpInfo(groupId);
    }
    /**
     * Leave multi-person chat
     * @param roomId Room ID
     *
     * @see <a href="https://developers.line.biz/en/reference/messaging-api/#leave-room"> Documentation</a>
     */
    async leaveRoom(roomId) {
        return this.clients.messagingApi.leaveRoom(roomId);
    }
    /**
     * Leave multi-person chat.
     * This method includes HttpInfo object to return additional information.
     * @param roomId Room ID
     *
     * @see <a href="https://developers.line.biz/en/reference/messaging-api/#leave-room"> Documentation</a>
     */
    async leaveRoomWithHttpInfo(roomId) {
        return this.clients.messagingApi.leaveRoomWithHttpInfo(roomId);
    }
    /**
     * Link rich menu to user.
     * @param userId User ID. Found in the `source` object of webhook event objects. Do not use the LINE ID used in LINE.
     * @param richMenuId ID of a rich menu
     *
     * @see <a href="https://developers.line.biz/en/reference/messaging-api/#link-rich-menu-to-user"> Documentation</a>
     */
    async linkRichMenuIdToUser(userId, richMenuId) {
        return this.clients.messagingApi.linkRichMenuIdToUser(userId, richMenuId);
    }
    /**
     * Link rich menu to user..
     * This method includes HttpInfo object to return additional information.
     * @param userId User ID. Found in the `source` object of webhook event objects. Do not use the LINE ID used in LINE.
     * @param richMenuId ID of a rich menu
     *
     * @see <a href="https://developers.line.biz/en/reference/messaging-api/#link-rich-menu-to-user"> Documentation</a>
     */
    async linkRichMenuIdToUserWithHttpInfo(userId, richMenuId) {
        return this.clients.messagingApi.linkRichMenuIdToUserWithHttpInfo(userId, richMenuId);
    }
    /**
     * Link rich menu to multiple users
     * @param richMenuBulkLinkRequest
     *
     * @see <a href="https://developers.line.biz/en/reference/messaging-api/#link-rich-menu-to-users"> Documentation</a>
     */
    async linkRichMenuIdToUsers(richMenuBulkLinkRequest) {
        return this.clients.messagingApi.linkRichMenuIdToUsers(richMenuBulkLinkRequest);
    }
    /**
     * Link rich menu to multiple users.
     * This method includes HttpInfo object to return additional information.
     * @param richMenuBulkLinkRequest
     *
     * @see <a href="https://developers.line.biz/en/reference/messaging-api/#link-rich-menu-to-users"> Documentation</a>
     */
    async linkRichMenuIdToUsersWithHttpInfo(richMenuBulkLinkRequest) {
        return this.clients.messagingApi.linkRichMenuIdToUsersWithHttpInfo(richMenuBulkLinkRequest);
    }
    /**
     * Get a paginated list of coupons.
     * @param status Filter coupons by their status.
     * @param start Pagination token to retrieve the next page of results.
     * @param limit Maximum number of coupons to return per request.
     *
     * @see <a href="https://developers.line.biz/en/reference/messaging-api/#get-coupons-list"> Documentation</a>
     */
    async listCoupon(status, start, limit) {
        return this.clients.messagingApi.listCoupon(status, start, limit);
    }
    /**
     * Get a paginated list of coupons..
     * This method includes HttpInfo object to return additional information.
     * @param status Filter coupons by their status.
     * @param start Pagination token to retrieve the next page of results.
     * @param limit Maximum number of coupons to return per request.
     *
     * @see <a href="https://developers.line.biz/en/reference/messaging-api/#get-coupons-list"> Documentation</a>
     */
    async listCouponWithHttpInfo(status, start, limit) {
        return this.clients.messagingApi.listCouponWithHttpInfo(status, start, limit);
    }
    /**
     * Mark messages from users as read
     * @param markMessagesAsReadRequest
     *
     * @see <a href="https://developers.line.biz/en/reference/partner-docs/#mark-messages-from-users-as-read"> Documentation</a>
     */
    async markMessagesAsRead(markMessagesAsReadRequest) {
        return this.clients.messagingApi.markMessagesAsRead(markMessagesAsReadRequest);
    }
    /**
     * Mark messages from users as read.
     * This method includes HttpInfo object to return additional information.
     * @param markMessagesAsReadRequest
     *
     * @see <a href="https://developers.line.biz/en/reference/partner-docs/#mark-messages-from-users-as-read"> Documentation</a>
     */
    async markMessagesAsReadWithHttpInfo(markMessagesAsReadRequest) {
        return this.clients.messagingApi.markMessagesAsReadWithHttpInfo(markMessagesAsReadRequest);
    }
    /**
     * Mark messages from users as read by token
     * @param markMessagesAsReadByTokenRequest
     *
     * @see <a href="https://developers.line.biz/en/reference/messaging-api/#mark-as-read"> Documentation</a>
     */
    async markMessagesAsReadByToken(markMessagesAsReadByTokenRequest) {
        return this.clients.messagingApi.markMessagesAsReadByToken(markMessagesAsReadByTokenRequest);
    }
    /**
     * Mark messages from users as read by token.
     * This method includes HttpInfo object to return additional information.
     * @param markMessagesAsReadByTokenRequest
     *
     * @see <a href="https://developers.line.biz/en/reference/messaging-api/#mark-as-read"> Documentation</a>
     */
    async markMessagesAsReadByTokenWithHttpInfo(markMessagesAsReadByTokenRequest) {
        return this.clients.messagingApi.markMessagesAsReadByTokenWithHttpInfo(markMessagesAsReadByTokenRequest);
    }
    /**
     * An API that efficiently sends the same message to multiple user IDs. You can\'t send messages to group chats or multi-person chats.
     * @param multicastRequest
     * @param xLineRetryKey Retry key. Specifies the UUID in hexadecimal format (e.g., `123e4567-e89b-12d3-a456-426614174000`) generated by any method. The retry key isn\'t generated by LINE. Each developer must generate their own retry key.
     *
     * @see <a href="https://developers.line.biz/en/reference/messaging-api/#send-multicast-message"> Documentation</a>
     */
    async multicast(multicastRequest, xLineRetryKey) {
        return this.clients.messagingApi.multicast(multicastRequest, xLineRetryKey);
    }
    /**
     * An API that efficiently sends the same message to multiple user IDs. You can\'t send messages to group chats or multi-person chats..
     * This method includes HttpInfo object to return additional information.
     * @param multicastRequest
     * @param xLineRetryKey Retry key. Specifies the UUID in hexadecimal format (e.g., `123e4567-e89b-12d3-a456-426614174000`) generated by any method. The retry key isn\'t generated by LINE. Each developer must generate their own retry key.
     *
     * @see <a href="https://developers.line.biz/en/reference/messaging-api/#send-multicast-message"> Documentation</a>
     */
    async multicastWithHttpInfo(multicastRequest, xLineRetryKey) {
        return this.clients.messagingApi.multicastWithHttpInfo(multicastRequest, xLineRetryKey);
    }
    /**
     * Send narrowcast message
     * @param narrowcastRequest
     * @param xLineRetryKey Retry key. Specifies the UUID in hexadecimal format (e.g., `123e4567-e89b-12d3-a456-426614174000`) generated by any method. The retry key isn\'t generated by LINE. Each developer must generate their own retry key.
     *
     * @see <a href="https://developers.line.biz/en/reference/messaging-api/#send-narrowcast-message"> Documentation</a>
     */
    async narrowcast(narrowcastRequest, xLineRetryKey) {
        return this.clients.messagingApi.narrowcast(narrowcastRequest, xLineRetryKey);
    }
    /**
     * Send narrowcast message.
     * This method includes HttpInfo object to return additional information.
     * @param narrowcastRequest
     * @param xLineRetryKey Retry key. Specifies the UUID in hexadecimal format (e.g., `123e4567-e89b-12d3-a456-426614174000`) generated by any method. The retry key isn\'t generated by LINE. Each developer must generate their own retry key.
     *
     * @see <a href="https://developers.line.biz/en/reference/messaging-api/#send-narrowcast-message"> Documentation</a>
     */
    async narrowcastWithHttpInfo(narrowcastRequest, xLineRetryKey) {
        return this.clients.messagingApi.narrowcastWithHttpInfo(narrowcastRequest, xLineRetryKey);
    }
    /**
     * Sends a message to a user, group chat, or multi-person chat at any time.
     * @param pushMessageRequest
     * @param xLineRetryKey Retry key. Specifies the UUID in hexadecimal format (e.g., `123e4567-e89b-12d3-a456-426614174000`) generated by any method. The retry key isn\'t generated by LINE. Each developer must generate their own retry key.
     *
     * @see <a href="https://developers.line.biz/en/reference/messaging-api/#send-push-message"> Documentation</a>
     */
    async pushMessage(pushMessageRequest, xLineRetryKey) {
        return this.clients.messagingApi.pushMessage(pushMessageRequest, xLineRetryKey);
    }
    /**
     * Sends a message to a user, group chat, or multi-person chat at any time..
     * This method includes HttpInfo object to return additional information.
     * @param pushMessageRequest
     * @param xLineRetryKey Retry key. Specifies the UUID in hexadecimal format (e.g., `123e4567-e89b-12d3-a456-426614174000`) generated by any method. The retry key isn\'t generated by LINE. Each developer must generate their own retry key.
     *
     * @see <a href="https://developers.line.biz/en/reference/messaging-api/#send-push-message"> Documentation</a>
     */
    async pushMessageWithHttpInfo(pushMessageRequest, xLineRetryKey) {
        return this.clients.messagingApi.pushMessageWithHttpInfo(pushMessageRequest, xLineRetryKey);
    }
    /**
     * Send LINE notification message
     * @param pnpMessagesRequest
     * @param xLineDeliveryTag String returned in the delivery.data property of the delivery completion event via Webhook.
     *
     * @see <a href="https://developers.line.biz/en/reference/partner-docs/#send-line-notification-message"> Documentation</a>
     */
    async pushMessagesByPhone(pnpMessagesRequest, xLineDeliveryTag) {
        return this.clients.messagingApi.pushMessagesByPhone(pnpMessagesRequest, xLineDeliveryTag);
    }
    /**
     * Send LINE notification message.
     * This method includes HttpInfo object to return additional information.
     * @param pnpMessagesRequest
     * @param xLineDeliveryTag String returned in the delivery.data property of the delivery completion event via Webhook.
     *
     * @see <a href="https://developers.line.biz/en/reference/partner-docs/#send-line-notification-message"> Documentation</a>
     */
    async pushMessagesByPhoneWithHttpInfo(pnpMessagesRequest, xLineDeliveryTag) {
        return this.clients.messagingApi.pushMessagesByPhoneWithHttpInfo(pnpMessagesRequest, xLineDeliveryTag);
    }
    /**
     * Send reply message
     * @param replyMessageRequest
     *
     * @see <a href="https://developers.line.biz/en/reference/messaging-api/#send-reply-message"> Documentation</a>
     */
    async replyMessage(replyMessageRequest) {
        return this.clients.messagingApi.replyMessage(replyMessageRequest);
    }
    /**
     * Send reply message.
     * This method includes HttpInfo object to return additional information.
     * @param replyMessageRequest
     *
     * @see <a href="https://developers.line.biz/en/reference/messaging-api/#send-reply-message"> Documentation</a>
     */
    async replyMessageWithHttpInfo(replyMessageRequest) {
        return this.clients.messagingApi.replyMessageWithHttpInfo(replyMessageRequest);
    }
    /**
     * You can use this endpoint to batch control the rich menu linked to the users using the endpoint such as Link rich menu to user. The following operations are available:  1. Replace a rich menu with another rich menu for all users linked to a specific rich menu 2. Unlink a rich menu for all users linked to a specific rich menu 3. Unlink a rich menu for all users linked the rich menu
     * @param richMenuBatchRequest
     *
     * @see <a href="https://developers.line.biz/en/reference/messaging-api/#batch-control-rich-menus-of-users"> Documentation</a>
     */
    async richMenuBatch(richMenuBatchRequest) {
        return this.clients.messagingApi.richMenuBatch(richMenuBatchRequest);
    }
    /**
     * You can use this endpoint to batch control the rich menu linked to the users using the endpoint such as Link rich menu to user. The following operations are available:  1. Replace a rich menu with another rich menu for all users linked to a specific rich menu 2. Unlink a rich menu for all users linked to a specific rich menu 3. Unlink a rich menu for all users linked the rich menu .
     * This method includes HttpInfo object to return additional information.
     * @param richMenuBatchRequest
     *
     * @see <a href="https://developers.line.biz/en/reference/messaging-api/#batch-control-rich-menus-of-users"> Documentation</a>
     */
    async richMenuBatchWithHttpInfo(richMenuBatchRequest) {
        return this.clients.messagingApi.richMenuBatchWithHttpInfo(richMenuBatchRequest);
    }
    /**
     * Set default rich menu
     * @param richMenuId ID of a rich menu
     *
     * @see <a href="https://developers.line.biz/en/reference/messaging-api/#set-default-rich-menu"> Documentation</a>
     */
    async setDefaultRichMenu(richMenuId) {
        return this.clients.messagingApi.setDefaultRichMenu(richMenuId);
    }
    /**
     * Set default rich menu.
     * This method includes HttpInfo object to return additional information.
     * @param richMenuId ID of a rich menu
     *
     * @see <a href="https://developers.line.biz/en/reference/messaging-api/#set-default-rich-menu"> Documentation</a>
     */
    async setDefaultRichMenuWithHttpInfo(richMenuId) {
        return this.clients.messagingApi.setDefaultRichMenuWithHttpInfo(richMenuId);
    }
    /**
     * Set webhook endpoint URL
     * @param setWebhookEndpointRequest
     *
     * @see <a href="https://developers.line.biz/en/reference/messaging-api/#set-webhook-endpoint-url"> Documentation</a>
     */
    async setWebhookEndpoint(setWebhookEndpointRequest) {
        return this.clients.messagingApi.setWebhookEndpoint(setWebhookEndpointRequest);
    }
    /**
     * Set webhook endpoint URL.
     * This method includes HttpInfo object to return additional information.
     * @param setWebhookEndpointRequest
     *
     * @see <a href="https://developers.line.biz/en/reference/messaging-api/#set-webhook-endpoint-url"> Documentation</a>
     */
    async setWebhookEndpointWithHttpInfo(setWebhookEndpointRequest) {
        return this.clients.messagingApi.setWebhookEndpointWithHttpInfo(setWebhookEndpointRequest);
    }
    /**
     * Display a loading animation in one-on-one chats between users and LINE Official Accounts.
     * @param showLoadingAnimationRequest
     *
     * @see <a href="https://developers.line.biz/en/reference/messaging-api/#display-a-loading-indicator"> Documentation</a>
     */
    async showLoadingAnimation(showLoadingAnimationRequest) {
        return this.clients.messagingApi.showLoadingAnimation(showLoadingAnimationRequest);
    }
    /**
     * Display a loading animation in one-on-one chats between users and LINE Official Accounts..
     * This method includes HttpInfo object to return additional information.
     * @param showLoadingAnimationRequest
     *
     * @see <a href="https://developers.line.biz/en/reference/messaging-api/#display-a-loading-indicator"> Documentation</a>
     */
    async showLoadingAnimationWithHttpInfo(showLoadingAnimationRequest) {
        return this.clients.messagingApi.showLoadingAnimationWithHttpInfo(showLoadingAnimationRequest);
    }
    /**
     * Test webhook endpoint
     * @param testWebhookEndpointRequest
     *
     * @see <a href="https://developers.line.biz/en/reference/messaging-api/#test-webhook-endpoint"> Documentation</a>
     */
    async testWebhookEndpoint(testWebhookEndpointRequest) {
        return this.clients.messagingApi.testWebhookEndpoint(testWebhookEndpointRequest);
    }
    /**
     * Test webhook endpoint.
     * This method includes HttpInfo object to return additional information.
     * @param testWebhookEndpointRequest
     *
     * @see <a href="https://developers.line.biz/en/reference/messaging-api/#test-webhook-endpoint"> Documentation</a>
     */
    async testWebhookEndpointWithHttpInfo(testWebhookEndpointRequest) {
        return this.clients.messagingApi.testWebhookEndpointWithHttpInfo(testWebhookEndpointRequest);
    }
    /**
     * Unlink rich menu from user
     * @param userId User ID. Found in the `source` object of webhook event objects. Do not use the LINE ID used in LINE.
     *
     * @see <a href="https://developers.line.biz/en/reference/messaging-api/#unlink-rich-menu-from-user"> Documentation</a>
     */
    async unlinkRichMenuIdFromUser(userId) {
        return this.clients.messagingApi.unlinkRichMenuIdFromUser(userId);
    }
    /**
     * Unlink rich menu from user.
     * This method includes HttpInfo object to return additional information.
     * @param userId User ID. Found in the `source` object of webhook event objects. Do not use the LINE ID used in LINE.
     *
     * @see <a href="https://developers.line.biz/en/reference/messaging-api/#unlink-rich-menu-from-user"> Documentation</a>
     */
    async unlinkRichMenuIdFromUserWithHttpInfo(userId) {
        return this.clients.messagingApi.unlinkRichMenuIdFromUserWithHttpInfo(userId);
    }
    /**
     * Unlink rich menus from multiple users
     * @param richMenuBulkUnlinkRequest
     *
     * @see <a href="https://developers.line.biz/en/reference/messaging-api/#unlink-rich-menu-from-users"> Documentation</a>
     */
    async unlinkRichMenuIdFromUsers(richMenuBulkUnlinkRequest) {
        return this.clients.messagingApi.unlinkRichMenuIdFromUsers(richMenuBulkUnlinkRequest);
    }
    /**
     * Unlink rich menus from multiple users.
     * This method includes HttpInfo object to return additional information.
     * @param richMenuBulkUnlinkRequest
     *
     * @see <a href="https://developers.line.biz/en/reference/messaging-api/#unlink-rich-menu-from-users"> Documentation</a>
     */
    async unlinkRichMenuIdFromUsersWithHttpInfo(richMenuBulkUnlinkRequest) {
        return this.clients.messagingApi.unlinkRichMenuIdFromUsersWithHttpInfo(richMenuBulkUnlinkRequest);
    }
    /**
     * Update rich menu alias
     * @param richMenuAliasId The rich menu alias ID you want to update.
     * @param updateRichMenuAliasRequest
     *
     * @see <a href="https://developers.line.biz/en/reference/messaging-api/#update-rich-menu-alias"> Documentation</a>
     */
    async updateRichMenuAlias(richMenuAliasId, updateRichMenuAliasRequest) {
        return this.clients.messagingApi.updateRichMenuAlias(richMenuAliasId, updateRichMenuAliasRequest);
    }
    /**
     * Update rich menu alias.
     * This method includes HttpInfo object to return additional information.
     * @param richMenuAliasId The rich menu alias ID you want to update.
     * @param updateRichMenuAliasRequest
     *
     * @see <a href="https://developers.line.biz/en/reference/messaging-api/#update-rich-menu-alias"> Documentation</a>
     */
    async updateRichMenuAliasWithHttpInfo(richMenuAliasId, updateRichMenuAliasRequest) {
        return this.clients.messagingApi.updateRichMenuAliasWithHttpInfo(richMenuAliasId, updateRichMenuAliasRequest);
    }
    /**
     * Validate message objects of a broadcast message
     * @param validateMessageRequest
     *
     * @see <a href="https://developers.line.biz/en/reference/messaging-api/#validate-message-objects-of-broadcast-message"> Documentation</a>
     */
    async validateBroadcast(validateMessageRequest) {
        return this.clients.messagingApi.validateBroadcast(validateMessageRequest);
    }
    /**
     * Validate message objects of a broadcast message.
     * This method includes HttpInfo object to return additional information.
     * @param validateMessageRequest
     *
     * @see <a href="https://developers.line.biz/en/reference/messaging-api/#validate-message-objects-of-broadcast-message"> Documentation</a>
     */
    async validateBroadcastWithHttpInfo(validateMessageRequest) {
        return this.clients.messagingApi.validateBroadcastWithHttpInfo(validateMessageRequest);
    }
    /**
     * Validate message objects of a multicast message
     * @param validateMessageRequest
     *
     * @see <a href="https://developers.line.biz/en/reference/messaging-api/#validate-message-objects-of-multicast-message"> Documentation</a>
     */
    async validateMulticast(validateMessageRequest) {
        return this.clients.messagingApi.validateMulticast(validateMessageRequest);
    }
    /**
     * Validate message objects of a multicast message.
     * This method includes HttpInfo object to return additional information.
     * @param validateMessageRequest
     *
     * @see <a href="https://developers.line.biz/en/reference/messaging-api/#validate-message-objects-of-multicast-message"> Documentation</a>
     */
    async validateMulticastWithHttpInfo(validateMessageRequest) {
        return this.clients.messagingApi.validateMulticastWithHttpInfo(validateMessageRequest);
    }
    /**
     * Validate message objects of a narrowcast message
     * @param validateMessageRequest
     *
     * @see <a href="https://developers.line.biz/en/reference/messaging-api/#validate-message-objects-of-narrowcast-message"> Documentation</a>
     */
    async validateNarrowcast(validateMessageRequest) {
        return this.clients.messagingApi.validateNarrowcast(validateMessageRequest);
    }
    /**
     * Validate message objects of a narrowcast message.
     * This method includes HttpInfo object to return additional information.
     * @param validateMessageRequest
     *
     * @see <a href="https://developers.line.biz/en/reference/messaging-api/#validate-message-objects-of-narrowcast-message"> Documentation</a>
     */
    async validateNarrowcastWithHttpInfo(validateMessageRequest) {
        return this.clients.messagingApi.validateNarrowcastWithHttpInfo(validateMessageRequest);
    }
    /**
     * Validate message objects of a push message
     * @param validateMessageRequest
     *
     * @see <a href="https://developers.line.biz/en/reference/messaging-api/#validate-message-objects-of-push-message"> Documentation</a>
     */
    async validatePush(validateMessageRequest) {
        return this.clients.messagingApi.validatePush(validateMessageRequest);
    }
    /**
     * Validate message objects of a push message.
     * This method includes HttpInfo object to return additional information.
     * @param validateMessageRequest
     *
     * @see <a href="https://developers.line.biz/en/reference/messaging-api/#validate-message-objects-of-push-message"> Documentation</a>
     */
    async validatePushWithHttpInfo(validateMessageRequest) {
        return this.clients.messagingApi.validatePushWithHttpInfo(validateMessageRequest);
    }
    /**
     * Validate message objects of a reply message
     * @param validateMessageRequest
     *
     * @see <a href="https://developers.line.biz/en/reference/messaging-api/#validate-message-objects-of-reply-message"> Documentation</a>
     */
    async validateReply(validateMessageRequest) {
        return this.clients.messagingApi.validateReply(validateMessageRequest);
    }
    /**
     * Validate message objects of a reply message.
     * This method includes HttpInfo object to return additional information.
     * @param validateMessageRequest
     *
     * @see <a href="https://developers.line.biz/en/reference/messaging-api/#validate-message-objects-of-reply-message"> Documentation</a>
     */
    async validateReplyWithHttpInfo(validateMessageRequest) {
        return this.clients.messagingApi.validateReplyWithHttpInfo(validateMessageRequest);
    }
    /**
     * Validate a request body of the Replace or unlink the linked rich menus in batches endpoint.
     * @param richMenuBatchRequest
     *
     * @see <a href="https://developers.line.biz/en/reference/messaging-api/#validate-batch-control-rich-menus-request"> Documentation</a>
     */
    async validateRichMenuBatchRequest(richMenuBatchRequest) {
        return this.clients.messagingApi.validateRichMenuBatchRequest(richMenuBatchRequest);
    }
    /**
     * Validate a request body of the Replace or unlink the linked rich menus in batches endpoint..
     * This method includes HttpInfo object to return additional information.
     * @param richMenuBatchRequest
     *
     * @see <a href="https://developers.line.biz/en/reference/messaging-api/#validate-batch-control-rich-menus-request"> Documentation</a>
     */
    async validateRichMenuBatchRequestWithHttpInfo(richMenuBatchRequest) {
        return this.clients.messagingApi.validateRichMenuBatchRequestWithHttpInfo(richMenuBatchRequest);
    }
    /**
     * Validate rich menu object
     * @param richMenuRequest
     *
     * @see <a href="https://developers.line.biz/en/reference/messaging-api/#validate-rich-menu-object"> Documentation</a>
     */
    async validateRichMenuObject(richMenuRequest) {
        return this.clients.messagingApi.validateRichMenuObject(richMenuRequest);
    }
    /**
     * Validate rich menu object.
     * This method includes HttpInfo object to return additional information.
     * @param richMenuRequest
     *
     * @see <a href="https://developers.line.biz/en/reference/messaging-api/#validate-rich-menu-object"> Documentation</a>
     */
    async validateRichMenuObjectWithHttpInfo(richMenuRequest) {
        return this.clients.messagingApi.validateRichMenuObjectWithHttpInfo(richMenuRequest);
    }
    /**
     * Download image, video, and audio data sent from users.
     * @param messageId Message ID of video or audio
     *
     * @see <a href="https://developers.line.biz/en/reference/messaging-api/#get-content"> Documentation</a>
     */
    async getMessageContent(messageId) {
        return this.clients.messagingApiBlob.getMessageContent(messageId);
    }
    /**
     * Download image, video, and audio data sent from users..
     * This method includes HttpInfo object to return additional information.
     * @param messageId Message ID of video or audio
     *
     * @see <a href="https://developers.line.biz/en/reference/messaging-api/#get-content"> Documentation</a>
     */
    async getMessageContentWithHttpInfo(messageId) {
        return this.clients.messagingApiBlob.getMessageContentWithHttpInfo(messageId);
    }
    /**
     * Get a preview image of the image or video
     * @param messageId Message ID of image or video
     *
     * @see <a href="https://developers.line.biz/en/reference/messaging-api/#get-image-or-video-preview"> Documentation</a>
     */
    async getMessageContentPreview(messageId) {
        return this.clients.messagingApiBlob.getMessageContentPreview(messageId);
    }
    /**
     * Get a preview image of the image or video.
     * This method includes HttpInfo object to return additional information.
     * @param messageId Message ID of image or video
     *
     * @see <a href="https://developers.line.biz/en/reference/messaging-api/#get-image-or-video-preview"> Documentation</a>
     */
    async getMessageContentPreviewWithHttpInfo(messageId) {
        return this.clients.messagingApiBlob.getMessageContentPreviewWithHttpInfo(messageId);
    }
    /**
     * Verify the preparation status of a video or audio for getting
     * @param messageId Message ID of video or audio
     *
     * @see <a href="https://developers.line.biz/en/reference/messaging-api/#verify-video-or-audio-preparation-status"> Documentation</a>
     */
    async getMessageContentTranscodingByMessageId(messageId) {
        return this.clients.messagingApiBlob.getMessageContentTranscodingByMessageId(messageId);
    }
    /**
     * Verify the preparation status of a video or audio for getting.
     * This method includes HttpInfo object to return additional information.
     * @param messageId Message ID of video or audio
     *
     * @see <a href="https://developers.line.biz/en/reference/messaging-api/#verify-video-or-audio-preparation-status"> Documentation</a>
     */
    async getMessageContentTranscodingByMessageIdWithHttpInfo(messageId) {
        return this.clients.messagingApiBlob.getMessageContentTranscodingByMessageIdWithHttpInfo(messageId);
    }
    /**
     * Download rich menu image.
     * @param richMenuId ID of the rich menu with the image to be downloaded
     *
     * @see <a href="https://developers.line.biz/en/reference/messaging-api/#download-rich-menu-image"> Documentation</a>
     */
    async getRichMenuImage(richMenuId) {
        return this.clients.messagingApiBlob.getRichMenuImage(richMenuId);
    }
    /**
     * Download rich menu image..
     * This method includes HttpInfo object to return additional information.
     * @param richMenuId ID of the rich menu with the image to be downloaded
     *
     * @see <a href="https://developers.line.biz/en/reference/messaging-api/#download-rich-menu-image"> Documentation</a>
     */
    async getRichMenuImageWithHttpInfo(richMenuId) {
        return this.clients.messagingApiBlob.getRichMenuImageWithHttpInfo(richMenuId);
    }
    /**
     * Upload rich menu image
     * @param richMenuId The ID of the rich menu to attach the image to
     * @param body
     *
     * @see <a href="https://developers.line.biz/en/reference/messaging-api/#upload-rich-menu-image"> Documentation</a>
     */
    async setRichMenuImage(richMenuId, body) {
        return this.clients.messagingApiBlob.setRichMenuImage(richMenuId, body);
    }
    /**
     * Upload rich menu image.
     * This method includes HttpInfo object to return additional information.
     * @param richMenuId The ID of the rich menu to attach the image to
     * @param body
     *
     * @see <a href="https://developers.line.biz/en/reference/messaging-api/#upload-rich-menu-image"> Documentation</a>
     */
    async setRichMenuImageWithHttpInfo(richMenuId, body) {
        return this.clients.messagingApiBlob.setRichMenuImageWithHttpInfo(richMenuId, body);
    }
    /**
     * If the Standby Channel wants to take the initiative (Chat Control), it calls the Acquire Control API. The channel that was previously an Active Channel will automatically switch to a Standby Channel.
     * @param chatId The `userId`, `roomId`, or `groupId`
     * @param acquireChatControlRequest
     *
     * @see <a href="https://developers.line.biz/en/reference/partner-docs/#acquire-control-api"> Documentation</a>
     */
    async acquireChatControl(chatId, acquireChatControlRequest) {
        return this.clients.lineModule.acquireChatControl(chatId, acquireChatControlRequest);
    }
    /**
     * If the Standby Channel wants to take the initiative (Chat Control), it calls the Acquire Control API. The channel that was previously an Active Channel will automatically switch to a Standby Channel. .
     * This method includes HttpInfo object to return additional information.
     * @param chatId The `userId`, `roomId`, or `groupId`
     * @param acquireChatControlRequest
     *
     * @see <a href="https://developers.line.biz/en/reference/partner-docs/#acquire-control-api"> Documentation</a>
     */
    async acquireChatControlWithHttpInfo(chatId, acquireChatControlRequest) {
        return this.clients.lineModule.acquireChatControlWithHttpInfo(chatId, acquireChatControlRequest);
    }
    /**
     * The module channel admin calls the Detach API to detach the module channel from a LINE Official Account.
     * @param detachModuleRequest
     *
     * @see <a href="https://developers.line.biz/en/reference/partner-docs/#unlink-detach-module-channel-by-operation-mc-admin"> Documentation</a>
     */
    async detachModule(detachModuleRequest) {
        return this.clients.lineModule.detachModule(detachModuleRequest);
    }
    /**
     * The module channel admin calls the Detach API to detach the module channel from a LINE Official Account..
     * This method includes HttpInfo object to return additional information.
     * @param detachModuleRequest
     *
     * @see <a href="https://developers.line.biz/en/reference/partner-docs/#unlink-detach-module-channel-by-operation-mc-admin"> Documentation</a>
     */
    async detachModuleWithHttpInfo(detachModuleRequest) {
        return this.clients.lineModule.detachModuleWithHttpInfo(detachModuleRequest);
    }
    /**
     * Gets a list of basic information about the bots of multiple LINE Official Accounts that have attached module channels.
     * @param start Value of the continuation token found in the next property of the JSON object returned in the response. If you can\'t get all basic information about the bots in one request, include this parameter to get the remaining array.
     * @param limit Specify the maximum number of bots that you get basic information from. The default value is 100. Max value: 100
     *
     * @see <a href="https://developers.line.biz/en/reference/partner-docs/#get-multiple-bot-info-api"> Documentation</a>
     */
    async getModules(start, limit) {
        return this.clients.lineModule.getModules(start, limit);
    }
    /**
     * Gets a list of basic information about the bots of multiple LINE Official Accounts that have attached module channels..
     * This method includes HttpInfo object to return additional information.
     * @param start Value of the continuation token found in the next property of the JSON object returned in the response. If you can\'t get all basic information about the bots in one request, include this parameter to get the remaining array.
     * @param limit Specify the maximum number of bots that you get basic information from. The default value is 100. Max value: 100
     *
     * @see <a href="https://developers.line.biz/en/reference/partner-docs/#get-multiple-bot-info-api"> Documentation</a>
     */
    async getModulesWithHttpInfo(start, limit) {
        return this.clients.lineModule.getModulesWithHttpInfo(start, limit);
    }
    /**
     * To return the initiative (Chat Control) of Active Channel to Primary Channel, call the Release Control API.
     * @param chatId The `userId`, `roomId`, or `groupId`
     *
     * @see <a href="https://developers.line.biz/en/reference/partner-docs/#release-control-api"> Documentation</a>
     */
    async releaseChatControl(chatId) {
        return this.clients.lineModule.releaseChatControl(chatId);
    }
    /**
     * To return the initiative (Chat Control) of Active Channel to Primary Channel, call the Release Control API. .
     * This method includes HttpInfo object to return additional information.
     * @param chatId The `userId`, `roomId`, or `groupId`
     *
     * @see <a href="https://developers.line.biz/en/reference/partner-docs/#release-control-api"> Documentation</a>
     */
    async releaseChatControlWithHttpInfo(chatId) {
        return this.clients.lineModule.releaseChatControlWithHttpInfo(chatId);
    }
    /**
     * Attach by operation of the module channel provider
     * @param grantType authorization_code
     * @param code Authorization code received from the LINE Platform.
     * @param redirectUri Specify the redirect_uri specified in the URL for authentication and authorization.
     * @param codeVerifier Specify when using PKCE (Proof Key for Code Exchange) defined in the OAuth 2.0 extension specification as a countermeasure against authorization code interception attacks.
     * @param clientId Instead of using Authorization header, you can use this parameter to specify the channel ID of the module channel. You can find the channel ID of the module channel in the LINE Developers Console.
     * @param clientSecret Instead of using Authorization header, you can use this parameter to specify the channel secret of the module channel. You can find the channel secret of the module channel in the LINE Developers Console.
     * @param region If you specified a value for region in the URL for authentication and authorization, specify the same value.
     * @param basicSearchId If you specified a value for basic_search_id in the URL for authentication and authorization, specify the same value.
     * @param scope If you specified a value for scope in the URL for authentication and authorization, specify the same value.
     * @param brandType If you specified a value for brand_type in the URL for authentication and authorization, specify the same value.
     *
     * @see <a href="https://developers.line.biz/en/reference/partner-docs/#link-attach-by-operation-module-channel-provider"> Documentation</a>
     */
    async attachModule(grantType, code, redirectUri, codeVerifier, clientId, clientSecret, region, basicSearchId, scope, brandType) {
        return this.clients.lineModuleAttach.attachModule(grantType, code, redirectUri, codeVerifier, clientId, clientSecret, region, basicSearchId, scope, brandType);
    }
    /**
     * Attach by operation of the module channel provider.
     * This method includes HttpInfo object to return additional information.
     * @param grantType authorization_code
     * @param code Authorization code received from the LINE Platform.
     * @param redirectUri Specify the redirect_uri specified in the URL for authentication and authorization.
     * @param codeVerifier Specify when using PKCE (Proof Key for Code Exchange) defined in the OAuth 2.0 extension specification as a countermeasure against authorization code interception attacks.
     * @param clientId Instead of using Authorization header, you can use this parameter to specify the channel ID of the module channel. You can find the channel ID of the module channel in the LINE Developers Console.
     * @param clientSecret Instead of using Authorization header, you can use this parameter to specify the channel secret of the module channel. You can find the channel secret of the module channel in the LINE Developers Console.
     * @param region If you specified a value for region in the URL for authentication and authorization, specify the same value.
     * @param basicSearchId If you specified a value for basic_search_id in the URL for authentication and authorization, specify the same value.
     * @param scope If you specified a value for scope in the URL for authentication and authorization, specify the same value.
     * @param brandType If you specified a value for brand_type in the URL for authentication and authorization, specify the same value.
     *
     * @see <a href="https://developers.line.biz/en/reference/partner-docs/#link-attach-by-operation-module-channel-provider"> Documentation</a>
     */
    async attachModuleWithHttpInfo(grantType, code, redirectUri, codeVerifier, clientId, clientSecret, region, basicSearchId, scope, brandType) {
        return this.clients.lineModuleAttach.attachModuleWithHttpInfo(grantType, code, redirectUri, codeVerifier, clientId, clientSecret, region, basicSearchId, scope, brandType);
    }
    /**
     * Sends a mission sticker.
     * @param missionStickerRequest
     *
     * @see <a href="https://developers.line.biz/en/reference/partner-docs/#send-mission-stickers-v3"> Documentation</a>
     */
    async missionStickerV3(missionStickerRequest) {
        return this.clients.shop.missionStickerV3(missionStickerRequest);
    }
    /**
     * Sends a mission sticker..
     * This method includes HttpInfo object to return additional information.
     * @param missionStickerRequest
     *
     * @see <a href="https://developers.line.biz/en/reference/partner-docs/#send-mission-stickers-v3"> Documentation</a>
     */
    async missionStickerV3WithHttpInfo(missionStickerRequest) {
        return this.clients.shop.missionStickerV3WithHttpInfo(missionStickerRequest);
    }
}
exports.LineBotClientBase = LineBotClientBase;
