// @agentclientprotocol/sdk/dist/schema/index.js
// This file is auto-generated by @hey-api/openapi-ts
export const AGENT_METHODS = {
    authenticate: "authenticate",
    document_did_change: "document/didChange",
    document_did_close: "document/didClose",
    document_did_focus: "document/didFocus",
    document_did_open: "document/didOpen",
    document_did_save: "document/didSave",
    initialize: "initialize",
    logout: "logout",
    nes_accept: "nes/accept",
    nes_close: "nes/close",
    nes_reject: "nes/reject",
    nes_start: "nes/start",
    nes_suggest: "nes/suggest",
    session_cancel: "session/cancel",
    session_close: "session/close",
    session_fork: "session/fork",
    session_list: "session/list",
    session_load: "session/load",
    session_new: "session/new",
    session_prompt: "session/prompt",
    session_resume: "session/resume",
    session_set_config_option: "session/set_config_option",
    session_set_mode: "session/set_mode",
    session_set_model: "session/set_model",
};
export const CLIENT_METHODS = {
    fs_read_text_file: "fs/read_text_file",
    fs_write_text_file: "fs/write_text_file",
    session_elicitation: "session/elicitation",
    session_elicitation_complete: "session/elicitation/complete",
    session_request_permission: "session/request_permission",
    session_update: "session/update",
    terminal_create: "terminal/create",
    terminal_kill: "terminal/kill",
    terminal_output: "terminal/output",
    terminal_release: "terminal/release",
    terminal_wait_for_exit: "terminal/wait_for_exit",
};
export const PROTOCOL_VERSION = 1;
