// @agentclientprotocol/sdk/dist/schema/zod.gen.js
// This file is auto-generated by @hey-api/openapi-ts
import { z } from "zod/v4";
/**
 * **UNSTABLE**
 *
 * This capability is not part of the spec yet, and may be removed or changed at any point.
 *
 * Authentication capabilities supported by the client.
 *
 * Advertised during initialization to inform the agent which authentication
 * method types the client can handle. This governs opt-in types that require
 * additional client-side support.
 *
 * @experimental
 */
export const zAuthCapabilities = z.looseObject({
    _meta: z.record(z.string(), z.unknown()).nullish(),
    terminal: z.boolean().optional().default(false),
});
/**
 * **UNSTABLE**
 *
 * This capability is not part of the spec yet, and may be removed or changed at any point.
 *
 * Describes a single environment variable for an [`AuthMethodEnvVar`] authentication method.
 *
 * @experimental
 */
export const zAuthEnvVar = z.looseObject({
    _meta: z.record(z.string(), z.unknown()).nullish(),
    label: z.string().nullish(),
    name: z.string(),
    optional: z.boolean().optional().default(false),
    secret: z.boolean().optional().default(true),
});
/**
 * Agent handles authentication itself.
 *
 * This is the default authentication method type.
 */
export const zAuthMethodAgent = z.looseObject({
    _meta: z.record(z.string(), z.unknown()).nullish(),
    description: z.string().nullish(),
    id: z.string(),
    name: z.string(),
});
/**
 * **UNSTABLE**
 *
 * This capability is not part of the spec yet, and may be removed or changed at any point.
 *
 * Environment variable authentication method.
 *
 * The user provides credentials that the client passes to the agent as environment variables.
 *
 * @experimental
 */
export const zAuthMethodEnvVar = z.looseObject({
    _meta: z.record(z.string(), z.unknown()).nullish(),
    description: z.string().nullish(),
    id: z.string(),
    link: z.string().nullish(),
    name: z.string(),
    vars: z.array(zAuthEnvVar),
});
/**
 * **UNSTABLE**
 *
 * This capability is not part of the spec yet, and may be removed or changed at any point.
 *
 * Terminal-based authentication method.
 *
 * The client runs an interactive terminal for the user to authenticate via a TUI.
 *
 * @experimental
 */
export const zAuthMethodTerminal = z.looseObject({
    _meta: z.record(z.string(), z.unknown()).nullish(),
    args: z.array(z.string()).optional(),
    description: z.string().nullish(),
    env: z.record(z.string(), z.string()).optional(),
    id: z.string(),
    name: z.string(),
});
/**
 * Describes an available authentication method.
 *
 * The `type` field acts as the discriminator in the serialized JSON form.
 * When no `type` is present, the method is treated as `agent`.
 */
export const zAuthMethod = z.union([
    zAuthMethodEnvVar.and(z.looseObject({
        type: z.literal("env_var"),
    })),
    zAuthMethodTerminal.and(z.looseObject({
        type: z.literal("terminal"),
    })),
    zAuthMethodAgent,
]);
/**
 * Request parameters for the authenticate method.
 *
 * Specifies which authentication method to use.
 */
export const zAuthenticateRequest = z.looseObject({
    _meta: z.record(z.string(), z.unknown()).nullish(),
    methodId: z.string(),
});
/**
 * Response to the `authenticate` method.
 */
export const zAuthenticateResponse = z.looseObject({
    _meta: z.record(z.string(), z.unknown()).nullish(),
});
/**
 * Binary resource contents.
 */
export const zBlobResourceContents = z.looseObject({
    _meta: z.record(z.string(), z.unknown()).nullish(),
    blob: z.string(),
    mimeType: z.string().nullish(),
    uri: z.string(),
});
/**
 * Schema for boolean properties in an elicitation form.
 */
export const zBooleanPropertySchema = z.looseObject({
    default: z.boolean().nullish(),
    description: z.string().nullish(),
    title: z.string().nullish(),
});
/**
 * Response from closing an NES session.
 */
export const zCloseNesResponse = z.looseObject({
    _meta: z.record(z.string(), z.unknown()).nullish(),
});
/**
 * **UNSTABLE**
 *
 * This capability is not part of the spec yet, and may be removed or changed at any point.
 *
 * Response from closing a session.
 *
 * @experimental
 */
export const zCloseSessionResponse = z.looseObject({
    _meta: z.record(z.string(), z.unknown()).nullish(),
});
/**
 * **UNSTABLE**
 *
 * This capability is not part of the spec yet, and may be removed or changed at any point.
 *
 * Cost information for a session.
 *
 * @experimental
 */
export const zCost = z.looseObject({
    amount: z.number(),
    currency: z.string(),
});
/**
 * Response containing the ID of the created terminal.
 */
export const zCreateTerminalResponse = z.looseObject({
    _meta: z.record(z.string(), z.unknown()).nullish(),
    terminalId: z.string(),
});
/**
 * A diff representing file modifications.
 *
 * Shows changes to files in a format suitable for display in the client UI.
 *
 * See protocol docs: [Content](https://agentclientprotocol.com/protocol/tool-calls#content)
 */
export const zDiff = z.looseObject({
    _meta: z.record(z.string(), z.unknown()).nullish(),
    newText: z.string(),
    oldText: z.string().nullish(),
    path: z.string(),
});
export const zElicitationContentValue = z.union([
    z.string(),
    z.number(),
    z.number(),
    z.boolean(),
    z.array(z.string()),
]);
/**
 * **UNSTABLE**
 *
 * This capability is not part of the spec yet, and may be removed or changed at any point.
 *
 * The user accepted the elicitation and provided content.
 *
 * @experimental
 */
export const zElicitationAcceptAction = z.looseObject({
    content: z.record(z.string(), zElicitationContentValue).nullish(),
});
/**
 * **UNSTABLE**
 *
 * This capability is not part of the spec yet, and may be removed or changed at any point.
 *
 * The user's action in response to an elicitation.
 *
 * @experimental
 */
export const zElicitationAction = z.union([
    zElicitationAcceptAction.and(z.looseObject({
        action: z.literal("accept"),
    })),
    z.looseObject({
        action: z.literal("decline"),
    }),
    z.looseObject({
        action: z.literal("cancel"),
    }),
]);
/**
 * **UNSTABLE**
 *
 * This capability is not part of the spec yet, and may be removed or changed at any point.
 *
 * Form-based elicitation capabilities.
 *
 * @experimental
 */
export const zElicitationFormCapabilities = z.looseObject({
    _meta: z.record(z.string(), z.unknown()).nullish(),
});
/**
 * **UNSTABLE**
 *
 * This capability is not part of the spec yet, and may be removed or changed at any point.
 *
 * Unique identifier for an elicitation.
 *
 * @experimental
 */
export const zElicitationId = z.string();
/**
 * **UNSTABLE**
 *
 * This capability is not part of the spec yet, and may be removed or changed at any point.
 *
 * Notification sent by the agent when a URL-based elicitation is complete.
 *
 * @experimental
 */
export const zElicitationCompleteNotification = z.looseObject({
    _meta: z.record(z.string(), z.unknown()).nullish(),
    elicitationId: zElicitationId,
});
/**
 * **UNSTABLE**
 *
 * This capability is not part of the spec yet, and may be removed or changed at any point.
 *
 * Response from the client to an elicitation request.
 *
 * @experimental
 */
export const zElicitationResponse = z.looseObject({
    _meta: z.record(z.string(), z.unknown()).nullish(),
    action: zElicitationAction,
});
/**
 * Object schema type.
 */
export const zElicitationSchemaType = z.literal("object");
/**
 * String schema type.
 */
export const zElicitationStringType = z.literal("string");
/**
 * **UNSTABLE**
 *
 * This capability is not part of the spec yet, and may be removed or changed at any point.
 *
 * URL-based elicitation capabilities.
 *
 * @experimental
 */
export const zElicitationUrlCapabilities = z.looseObject({
    _meta: z.record(z.string(), z.unknown()).nullish(),
});
/**
 * **UNSTABLE**
 *
 * This capability is not part of the spec yet, and may be removed or changed at any point.
 *
 * Elicitation capabilities supported by the client.
 *
 * @experimental
 */
export const zElicitationCapabilities = z.looseObject({
    _meta: z.record(z.string(), z.unknown()).nullish(),
    form: zElicitationFormCapabilities.nullish(),
    url: zElicitationUrlCapabilities.nullish(),
});
/**
 * **UNSTABLE**
 *
 * This capability is not part of the spec yet, and may be removed or changed at any point.
 *
 * URL-based elicitation mode where the client directs the user to a URL.
 *
 * @experimental
 */
export const zElicitationUrlMode = z.looseObject({
    elicitationId: zElicitationId,
    url: z.string().url(),
});
/**
 * A titled enum option with a const value and human-readable title.
 */
export const zEnumOption = z.looseObject({
    const: z.string(),
    title: z.string(),
});
/**
 * An environment variable to set when launching an MCP server.
 */
export const zEnvVariable = z.looseObject({
    _meta: z.record(z.string(), z.unknown()).nullish(),
    name: z.string(),
    value: z.string(),
});
/**
 * Predefined error codes for common JSON-RPC and ACP-specific errors.
 *
 * These codes follow the JSON-RPC 2.0 specification for standard errors
 * and use the reserved range (-32000 to -32099) for protocol-specific errors.
 */
export const zErrorCode = z.union([
    z.literal(-32700),
    z.literal(-32600),
    z.literal(-32601),
    z.literal(-32602),
    z.literal(-32603),
    z.literal(-32800),
    z.literal(-32000),
    z.literal(-32002),
    z.literal(-32042),
    z
        .number()
        .int()
        .min(-2147483648, {
        message: "Invalid value: Expected int32 to be >= -2147483648",
    })
        .max(2147483647, {
        message: "Invalid value: Expected int32 to be <= 2147483647",
    }),
]);
/**
 * JSON-RPC error object.
 *
 * Represents an error that occurred during method execution, following the
 * JSON-RPC 2.0 error object specification with optional additional data.
 *
 * See protocol docs: [JSON-RPC Error Object](https://www.jsonrpc.org/specification#error_object)
 */
export const zError = z.looseObject({
    code: zErrorCode,
    data: z.unknown().optional(),
    message: z.string(),
});
/**
 * Allows the Agent to send an arbitrary notification that is not part of the ACP spec.
 * Extension notifications provide a way to send one-way messages for custom functionality
 * while maintaining protocol compatibility.
 *
 * See protocol docs: [Extensibility](https://agentclientprotocol.com/protocol/extensibility)
 */
export const zExtNotification = z.unknown();
/**
 * Allows for sending an arbitrary request that is not part of the ACP spec.
 * Extension methods provide a way to add custom functionality while maintaining
 * protocol compatibility.
 *
 * See protocol docs: [Extensibility](https://agentclientprotocol.com/protocol/extensibility)
 */
export const zExtRequest = z.unknown();
/**
 * Allows for sending an arbitrary response to an [`ExtRequest`] that is not part of the ACP spec.
 * Extension methods provide a way to add custom functionality while maintaining
 * protocol compatibility.
 *
 * See protocol docs: [Extensibility](https://agentclientprotocol.com/protocol/extensibility)
 */
export const zExtResponse = z.unknown();
/**
 * File system capabilities that a client may support.
 *
 * See protocol docs: [FileSystem](https://agentclientprotocol.com/protocol/initialization#filesystem)
 */
export const zFileSystemCapabilities = z.looseObject({
    _meta: z.record(z.string(), z.unknown()).nullish(),
    readTextFile: z.boolean().optional().default(false),
    writeTextFile: z.boolean().optional().default(false),
});
/**
 * An HTTP header to set when making requests to the MCP server.
 */
export const zHttpHeader = z.looseObject({
    _meta: z.record(z.string(), z.unknown()).nullish(),
    name: z.string(),
    value: z.string(),
});
/**
 * Metadata about the implementation of the client or agent.
 * Describes the name and version of an MCP implementation, with an optional
 * title for UI representation.
 */
export const zImplementation = z.looseObject({
    _meta: z.record(z.string(), z.unknown()).nullish(),
    name: z.string(),
    title: z.string().nullish(),
    version: z.string(),
});
/**
 * Schema for integer properties in an elicitation form.
 */
export const zIntegerPropertySchema = z.looseObject({
    default: z.number().nullish(),
    description: z.string().nullish(),
    maximum: z.number().nullish(),
    minimum: z.number().nullish(),
    title: z.string().nullish(),
});
/**
 * Response to `terminal/kill` method
 */
export const zKillTerminalResponse = z.looseObject({
    _meta: z.record(z.string(), z.unknown()).nullish(),
});
/**
 * Request parameters for listing existing sessions.
 *
 * Only available if the Agent supports the `sessionCapabilities.list` capability.
 */
export const zListSessionsRequest = z.looseObject({
    _meta: z.record(z.string(), z.unknown()).nullish(),
    additionalDirectories: z.array(z.string()).optional(),
    cursor: z.string().nullish(),
    cwd: z.string().nullish(),
});
/**
 * **UNSTABLE**
 *
 * This capability is not part of the spec yet, and may be removed or changed at any point.
 *
 * Logout capabilities supported by the agent.
 *
 * By supplying `{}` it means that the agent supports the logout method.
 *
 * @experimental
 */
export const zLogoutCapabilities = z.looseObject({
    _meta: z.record(z.string(), z.unknown()).nullish(),
});
/**
 * **UNSTABLE**
 *
 * This capability is not part of the spec yet, and may be removed or changed at any point.
 *
 * Authentication-related capabilities supported by the agent.
 *
 * @experimental
 */
export const zAgentAuthCapabilities = z.looseObject({
    _meta: z.record(z.string(), z.unknown()).nullish(),
    logout: zLogoutCapabilities.nullish(),
});
/**
 * **UNSTABLE**
 *
 * This capability is not part of the spec yet, and may be removed or changed at any point.
 *
 * Request parameters for the logout method.
 *
 * Terminates the current authenticated session.
 *
 * @experimental
 */
export const zLogoutRequest = z.looseObject({
    _meta: z.record(z.string(), z.unknown()).nullish(),
});
/**
 * **UNSTABLE**
 *
 * This capability is not part of the spec yet, and may be removed or changed at any point.
 *
 * Response to the `logout` method.
 *
 * @experimental
 */
export const zLogoutResponse = z.looseObject({
    _meta: z.record(z.string(), z.unknown()).nullish(),
});
/**
 * MCP capabilities supported by the agent
 */
export const zMcpCapabilities = z.looseObject({
    _meta: z.record(z.string(), z.unknown()).nullish(),
    http: z.boolean().optional().default(false),
    sse: z.boolean().optional().default(false),
});
/**
 * HTTP transport configuration for MCP.
 */
export const zMcpServerHttp = z.looseObject({
    _meta: z.record(z.string(), z.unknown()).nullish(),
    headers: z.array(zHttpHeader),
    name: z.string(),
    url: z.string(),
});
/**
 * SSE transport configuration for MCP.
 */
export const zMcpServerSse = z.looseObject({
    _meta: z.record(z.string(), z.unknown()).nullish(),
    headers: z.array(zHttpHeader),
    name: z.string(),
    url: z.string(),
});
/**
 * Stdio transport configuration for MCP.
 */
export const zMcpServerStdio = z.looseObject({
    _meta: z.record(z.string(), z.unknown()).nullish(),
    args: z.array(z.string()),
    command: z.string(),
    env: z.array(zEnvVariable),
    name: z.string(),
});
/**
 * Configuration for connecting to an MCP (Model Context Protocol) server.
 *
 * MCP servers provide tools and context that the agent can use when
 * processing prompts.
 *
 * See protocol docs: [MCP Servers](https://agentclientprotocol.com/protocol/session-setup#mcp-servers)
 */
export const zMcpServer = z.union([
    zMcpServerHttp.and(z.looseObject({
        type: z.literal("http"),
    })),
    zMcpServerSse.and(z.looseObject({
        type: z.literal("sse"),
    })),
    zMcpServerStdio,
]);
/**
 * **UNSTABLE**
 *
 * This capability is not part of the spec yet, and may be removed or changed at any point.
 *
 * A unique identifier for a model.
 *
 * @experimental
 */
export const zModelId = z.string();
/**
 * **UNSTABLE**
 *
 * This capability is not part of the spec yet, and may be removed or changed at any point.
 *
 * Information about a selectable model.
 *
 * @experimental
 */
export const zModelInfo = z.looseObject({
    _meta: z.record(z.string(), z.unknown()).nullish(),
    description: z.string().nullish(),
    modelId: zModelId,
    name: z.string(),
});
/**
 * Severity of a diagnostic.
 */
export const zNesDiagnosticSeverity = z.union([
    z.literal("error"),
    z.literal("warning"),
    z.literal("information"),
    z.literal("hint"),
]);
/**
 * Capabilities for diagnostics context.
 */
export const zNesDiagnosticsCapabilities = z.looseObject({
    _meta: z.record(z.string(), z.unknown()).nullish(),
});
/**
 * Marker for `document/didClose` capability support.
 */
export const zNesDocumentDidCloseCapabilities = z.looseObject({
    _meta: z.record(z.string(), z.unknown()).nullish(),
});
/**
 * Marker for `document/didFocus` capability support.
 */
export const zNesDocumentDidFocusCapabilities = z.looseObject({
    _meta: z.record(z.string(), z.unknown()).nullish(),
});
/**
 * Marker for `document/didOpen` capability support.
 */
export const zNesDocumentDidOpenCapabilities = z.looseObject({
    _meta: z.record(z.string(), z.unknown()).nullish(),
});
/**
 * Marker for `document/didSave` capability support.
 */
export const zNesDocumentDidSaveCapabilities = z.looseObject({
    _meta: z.record(z.string(), z.unknown()).nullish(),
});
/**
 * Capabilities for edit history context.
 */
export const zNesEditHistoryCapabilities = z.looseObject({
    _meta: z.record(z.string(), z.unknown()).nullish(),
    maxCount: z
        .number()
        .int()
        .gte(0)
        .max(4294967295, {
        message: "Invalid value: Expected uint32 to be <= 4294967295",
    })
        .nullish(),
});
/**
 * An entry in the edit history.
 */
export const zNesEditHistoryEntry = z.looseObject({
    diff: z.string(),
    uri: z.string(),
});
/**
 * A code excerpt from a file.
 */
export const zNesExcerpt = z.looseObject({
    endLine: z.number().int().gte(0).max(4294967295, {
        message: "Invalid value: Expected uint32 to be <= 4294967295",
    }),
    startLine: z.number().int().gte(0).max(4294967295, {
        message: "Invalid value: Expected uint32 to be <= 4294967295",
    }),
    text: z.string(),
});
/**
 * Marker for jump suggestion support.
 */
export const zNesJumpCapabilities = z.looseObject({
    _meta: z.record(z.string(), z.unknown()).nullish(),
});
/**
 * Capabilities for open files context.
 */
export const zNesOpenFilesCapabilities = z.looseObject({
    _meta: z.record(z.string(), z.unknown()).nullish(),
});
/**
 * A recently accessed file.
 */
export const zNesRecentFile = z.looseObject({
    languageId: z.string(),
    text: z.string(),
    uri: z.string(),
});
/**
 * Capabilities for recent files context.
 */
export const zNesRecentFilesCapabilities = z.looseObject({
    _meta: z.record(z.string(), z.unknown()).nullish(),
    maxCount: z
        .number()
        .int()
        .gte(0)
        .max(4294967295, {
        message: "Invalid value: Expected uint32 to be <= 4294967295",
    })
        .nullish(),
});
/**
 * The reason a suggestion was rejected.
 */
export const zNesRejectReason = z.union([
    z.literal("rejected"),
    z.literal("ignored"),
    z.literal("replaced"),
    z.literal("cancelled"),
]);
/**
 * A related code snippet from a file.
 */
export const zNesRelatedSnippet = z.looseObject({
    excerpts: z.array(zNesExcerpt),
    uri: z.string(),
});
/**
 * Capabilities for related snippets context.
 */
export const zNesRelatedSnippetsCapabilities = z.looseObject({
    _meta: z.record(z.string(), z.unknown()).nullish(),
});
/**
 * Marker for rename suggestion support.
 */
export const zNesRenameCapabilities = z.looseObject({
    _meta: z.record(z.string(), z.unknown()).nullish(),
});
/**
 * Repository metadata for an NES session.
 */
export const zNesRepository = z.looseObject({
    name: z.string(),
    owner: z.string(),
    remoteUrl: z.string(),
});
/**
 * Marker for search and replace suggestion support.
 */
export const zNesSearchAndReplaceCapabilities = z.looseObject({
    _meta: z.record(z.string(), z.unknown()).nullish(),
});
/**
 * NES capabilities advertised by the client during initialization.
 */
export const zClientNesCapabilities = z.looseObject({
    _meta: z.record(z.string(), z.unknown()).nullish(),
    jump: zNesJumpCapabilities.nullish(),
    rename: zNesRenameCapabilities.nullish(),
    searchAndReplace: zNesSearchAndReplaceCapabilities.nullish(),
});
/**
 * A search-and-replace suggestion.
 */
export const zNesSearchAndReplaceSuggestion = z.looseObject({
    id: z.string(),
    isRegex: z.boolean().nullish(),
    replace: z.string(),
    search: z.string(),
    uri: z.string(),
});
/**
 * What triggered the suggestion request.
 */
export const zNesTriggerKind = z.union([
    z.literal("automatic"),
    z.literal("diagnostic"),
    z.literal("manual"),
]);
/**
 * Capabilities for user actions context.
 */
export const zNesUserActionsCapabilities = z.looseObject({
    _meta: z.record(z.string(), z.unknown()).nullish(),
    maxCount: z
        .number()
        .int()
        .gte(0)
        .max(4294967295, {
        message: "Invalid value: Expected uint32 to be <= 4294967295",
    })
        .nullish(),
});
/**
 * Context capabilities the agent wants attached to each suggestion request.
 */
export const zNesContextCapabilities = z.looseObject({
    _meta: z.record(z.string(), z.unknown()).nullish(),
    diagnostics: zNesDiagnosticsCapabilities.nullish(),
    editHistory: zNesEditHistoryCapabilities.nullish(),
    openFiles: zNesOpenFilesCapabilities.nullish(),
    recentFiles: zNesRecentFilesCapabilities.nullish(),
    relatedSnippets: zNesRelatedSnippetsCapabilities.nullish(),
    userActions: zNesUserActionsCapabilities.nullish(),
});
/**
 * Request parameters for creating a new session.
 *
 * See protocol docs: [Creating a Session](https://agentclientprotocol.com/protocol/session-setup#creating-a-session)
 */
export const zNewSessionRequest = z.looseObject({
    _meta: z.record(z.string(), z.unknown()).nullish(),
    additionalDirectories: z.array(z.string()).optional(),
    cwd: z.string(),
    mcpServers: z.array(zMcpServer),
});
/**
 * Schema for number (floating-point) properties in an elicitation form.
 */
export const zNumberPropertySchema = z.looseObject({
    default: z.number().nullish(),
    description: z.string().nullish(),
    maximum: z.number().nullish(),
    minimum: z.number().nullish(),
    title: z.string().nullish(),
});
/**
 * Unique identifier for a permission option.
 */
export const zPermissionOptionId = z.string();
/**
 * The type of permission option being presented to the user.
 *
 * Helps clients choose appropriate icons and UI treatment.
 */
export const zPermissionOptionKind = z.union([
    z.literal("allow_once"),
    z.literal("allow_always"),
    z.literal("reject_once"),
    z.literal("reject_always"),
]);
/**
 * An option presented to the user when requesting permission.
 */
export const zPermissionOption = z.looseObject({
    _meta: z.record(z.string(), z.unknown()).nullish(),
    kind: zPermissionOptionKind,
    name: z.string(),
    optionId: zPermissionOptionId,
});
/**
 * Priority levels for plan entries.
 *
 * Used to indicate the relative importance or urgency of different
 * tasks in the execution plan.
 * See protocol docs: [Plan Entries](https://agentclientprotocol.com/protocol/agent-plan#plan-entries)
 */
export const zPlanEntryPriority = z.union([
    z.literal("high"),
    z.literal("medium"),
    z.literal("low"),
]);
/**
 * Status of a plan entry in the execution flow.
 *
 * Tracks the lifecycle of each task from planning through completion.
 * See protocol docs: [Plan Entries](https://agentclientprotocol.com/protocol/agent-plan#plan-entries)
 */
export const zPlanEntryStatus = z.union([
    z.literal("pending"),
    z.literal("in_progress"),
    z.literal("completed"),
]);
/**
 * A single entry in the execution plan.
 *
 * Represents a task or goal that the assistant intends to accomplish
 * as part of fulfilling the user's request.
 * See protocol docs: [Plan Entries](https://agentclientprotocol.com/protocol/agent-plan#plan-entries)
 */
export const zPlanEntry = z.looseObject({
    _meta: z.record(z.string(), z.unknown()).nullish(),
    content: z.string(),
    priority: zPlanEntryPriority,
    status: zPlanEntryStatus,
});
/**
 * An execution plan for accomplishing complex tasks.
 *
 * Plans consist of multiple entries representing individual tasks or goals.
 * Agents report plans to clients to provide visibility into their execution strategy.
 * Plans can evolve during execution as the agent discovers new requirements or completes tasks.
 *
 * See protocol docs: [Agent Plan](https://agentclientprotocol.com/protocol/agent-plan)
 */
export const zPlan = z.looseObject({
    _meta: z.record(z.string(), z.unknown()).nullish(),
    entries: z.array(zPlanEntry),
});
/**
 * A zero-based position in a text document.
 *
 * The meaning of `character` depends on the negotiated position encoding.
 */
export const zPosition = z.looseObject({
    character: z.number().int().gte(0).max(4294967295, {
        message: "Invalid value: Expected uint32 to be <= 4294967295",
    }),
    line: z.number().int().gte(0).max(4294967295, {
        message: "Invalid value: Expected uint32 to be <= 4294967295",
    }),
});
/**
 * A jump-to-location suggestion.
 */
export const zNesJumpSuggestion = z.looseObject({
    id: z.string(),
    position: zPosition,
    uri: z.string(),
});
/**
 * A rename symbol suggestion.
 */
export const zNesRenameSuggestion = z.looseObject({
    id: z.string(),
    newName: z.string(),
    position: zPosition,
    uri: z.string(),
});
/**
 * A user action (typing, cursor movement, etc.).
 */
export const zNesUserAction = z.looseObject({
    action: z.string(),
    position: zPosition,
    timestampMs: z.number(),
    uri: z.string(),
});
/**
 * The encoding used for character offsets in positions.
 *
 * Follows the same conventions as LSP 3.17. The default is UTF-16.
 */
export const zPositionEncodingKind = z.union([
    z.literal("utf-16"),
    z.literal("utf-32"),
    z.literal("utf-8"),
]);
/**
 * Capabilities supported by the client.
 *
 * Advertised during initialization to inform the agent about
 * available features and methods.
 *
 * See protocol docs: [Client Capabilities](https://agentclientprotocol.com/protocol/initialization#client-capabilities)
 */
export const zClientCapabilities = z.looseObject({
    _meta: z.record(z.string(), z.unknown()).nullish(),
    auth: zAuthCapabilities.optional().default({ terminal: false }),
    elicitation: zElicitationCapabilities.nullish(),
    fs: zFileSystemCapabilities
        .optional()
        .default({ readTextFile: false, writeTextFile: false }),
    nes: zClientNesCapabilities.nullish(),
    positionEncodings: z.array(zPositionEncodingKind).optional(),
    terminal: z.boolean().optional().default(false),
});
/**
 * Prompt capabilities supported by the agent in `session/prompt` requests.
 *
 * Baseline agent functionality requires support for [`ContentBlock::Text`]
 * and [`ContentBlock::ResourceLink`] in prompt requests.
 *
 * Other variants must be explicitly opted in to.
 * Capabilities for different types of content in prompt requests.
 *
 * Indicates which content types beyond the baseline (text and resource links)
 * the agent can process.
 *
 * See protocol docs: [Prompt Capabilities](https://agentclientprotocol.com/protocol/initialization#prompt-capabilities)
 */
export const zPromptCapabilities = z.looseObject({
    _meta: z.record(z.string(), z.unknown()).nullish(),
    audio: z.boolean().optional().default(false),
    embeddedContext: z.boolean().optional().default(false),
    image: z.boolean().optional().default(false),
});
/**
 * Protocol version identifier.
 *
 * This version is only bumped for breaking changes.
 * Non-breaking changes should be introduced via capabilities.
 */
export const zProtocolVersion = z.number().int().gte(0).lte(65535);
/**
 * Request parameters for the initialize method.
 *
 * Sent by the client to establish connection and negotiate capabilities.
 *
 * See protocol docs: [Initialization](https://agentclientprotocol.com/protocol/initialization)
 */
export const zInitializeRequest = z.looseObject({
    _meta: z.record(z.string(), z.unknown()).nullish(),
    clientCapabilities: zClientCapabilities.optional().default({
        auth: { terminal: false },
        fs: { readTextFile: false, writeTextFile: false },
        terminal: false,
    }),
    clientInfo: zImplementation.nullish(),
    protocolVersion: zProtocolVersion,
});
/**
 * A range in a text document, expressed as start and end positions.
 */
export const zRange = z.looseObject({
    end: zPosition,
    start: zPosition,
});
/**
 * A diagnostic (error, warning, etc.).
 */
export const zNesDiagnostic = z.looseObject({
    message: z.string(),
    range: zRange,
    severity: zNesDiagnosticSeverity,
    uri: z.string(),
});
/**
 * An open file in the editor.
 */
export const zNesOpenFile = z.looseObject({
    languageId: z.string(),
    lastFocusedMs: z.number().nullish(),
    uri: z.string(),
    visibleRange: zRange.nullish(),
});
/**
 * Context attached to a suggestion request.
 */
export const zNesSuggestContext = z.looseObject({
    _meta: z.record(z.string(), z.unknown()).nullish(),
    diagnostics: z.array(zNesDiagnostic).nullish(),
    editHistory: z.array(zNesEditHistoryEntry).nullish(),
    openFiles: z.array(zNesOpenFile).nullish(),
    recentFiles: z.array(zNesRecentFile).nullish(),
    relatedSnippets: z.array(zNesRelatedSnippet).nullish(),
    userActions: z.array(zNesUserAction).nullish(),
});
/**
 * A text edit within a suggestion.
 */
export const zNesTextEdit = z.looseObject({
    newText: z.string(),
    range: zRange,
});
/**
 * A text edit suggestion.
 */
export const zNesEditSuggestion = z.looseObject({
    cursorPosition: zPosition.nullish(),
    edits: z.array(zNesTextEdit),
    id: z.string(),
    uri: z.string(),
});
/**
 * A suggestion returned by the agent.
 */
export const zNesSuggestion = z.union([
    zNesEditSuggestion.and(z.looseObject({
        kind: z.literal("edit"),
    })),
    zNesJumpSuggestion.and(z.looseObject({
        kind: z.literal("jump"),
    })),
    zNesRenameSuggestion.and(z.looseObject({
        kind: z.literal("rename"),
    })),
    zNesSearchAndReplaceSuggestion.and(z.looseObject({
        kind: z.literal("searchAndReplace"),
    })),
]);
/**
 * Response containing the contents of a text file.
 */
export const zReadTextFileResponse = z.looseObject({
    _meta: z.record(z.string(), z.unknown()).nullish(),
    content: z.string(),
});
/**
 * Response to terminal/release method
 */
export const zReleaseTerminalResponse = z.looseObject({
    _meta: z.record(z.string(), z.unknown()).nullish(),
});
/**
 * JSON RPC Request Id
 *
 * An identifier established by the Client that MUST contain a String, Number, or NULL value if included. If it is not included it is assumed to be a notification. The value SHOULD normally not be Null [1] and Numbers SHOULD NOT contain fractional parts [2]
 *
 * The Server MUST reply with the same value in the Response object if included. This member is used to correlate the context between the two objects.
 *
 * [1] The use of Null as a value for the id member in a Request object is discouraged, because this specification uses a value of Null for Responses with an unknown id. Also, because JSON-RPC 1.0 uses an id value of Null for Notifications this could cause confusion in handling.
 *
 * [2] Fractional parts may be problematic, since many decimal fractions cannot be represented exactly as binary fractions.
 */
export const zRequestId = z.union([z.number(), z.string()]).nullable();
/**
 * **UNSTABLE**
 *
 * This capability is not part of the spec yet, and may be removed or changed at any point.
 *
 * Notification to cancel an ongoing request.
 *
 * See protocol docs: [Cancellation](https://agentclientprotocol.com/protocol/cancellation)
 *
 * @experimental
 */
export const zCancelRequestNotification = z.looseObject({
    _meta: z.record(z.string(), z.unknown()).nullish(),
    requestId: zRequestId,
});
/**
 * The sender or recipient of messages and data in a conversation.
 */
export const zRole = z.enum(["assistant", "user"]);
/**
 * Optional annotations for the client. The client can use annotations to inform how objects are used or displayed
 */
export const zAnnotations = z.looseObject({
    _meta: z.record(z.string(), z.unknown()).nullish(),
    audience: z.array(zRole).nullish(),
    lastModified: z.string().nullish(),
    priority: z.number().nullish(),
});
/**
 * Audio provided to or from an LLM.
 */
export const zAudioContent = z.looseObject({
    _meta: z.record(z.string(), z.unknown()).nullish(),
    annotations: zAnnotations.nullish(),
    data: z.string(),
    mimeType: z.string(),
});
/**
 * An image provided to or from an LLM.
 */
export const zImageContent = z.looseObject({
    _meta: z.record(z.string(), z.unknown()).nullish(),
    annotations: zAnnotations.nullish(),
    data: z.string(),
    mimeType: z.string(),
    uri: z.string().nullish(),
});
/**
 * A resource that the server is capable of reading, included in a prompt or tool call result.
 */
export const zResourceLink = z.looseObject({
    _meta: z.record(z.string(), z.unknown()).nullish(),
    annotations: zAnnotations.nullish(),
    description: z.string().nullish(),
    mimeType: z.string().nullish(),
    name: z.string(),
    size: z.number().nullish(),
    title: z.string().nullish(),
    uri: z.string(),
});
/**
 * The user selected one of the provided options.
 */
export const zSelectedPermissionOutcome = z.looseObject({
    _meta: z.record(z.string(), z.unknown()).nullish(),
    optionId: zPermissionOptionId,
});
/**
 * The outcome of a permission request.
 */
export const zRequestPermissionOutcome = z.union([
    z.looseObject({
        outcome: z.literal("cancelled"),
    }),
    zSelectedPermissionOutcome.and(z.looseObject({
        outcome: z.literal("selected"),
    })),
]);
/**
 * Response to a permission request.
 */
export const zRequestPermissionResponse = z.looseObject({
    _meta: z.record(z.string(), z.unknown()).nullish(),
    outcome: zRequestPermissionOutcome,
});
/**
 * **UNSTABLE**
 *
 * This capability is not part of the spec yet, and may be removed or changed at any point.
 *
 * Capabilities for additional session directories support.
 *
 * By supplying `{}` it means that the agent supports the `additionalDirectories` field on
 * supported session lifecycle requests and `session/list`.
 *
 * @experimental
 */
export const zSessionAdditionalDirectoriesCapabilities = z.looseObject({
    _meta: z.record(z.string(), z.unknown()).nullish(),
});
/**
 * **UNSTABLE**
 *
 * This capability is not part of the spec yet, and may be removed or changed at any point.
 *
 * Capabilities for the `session/close` method.
 *
 * By supplying `{}` it means that the agent supports closing of sessions.
 *
 * @experimental
 */
export const zSessionCloseCapabilities = z.looseObject({
    _meta: z.record(z.string(), z.unknown()).nullish(),
});
/**
 * **UNSTABLE**
 *
 * This capability is not part of the spec yet, and may be removed or changed at any point.
 *
 * A boolean on/off toggle session configuration option payload.
 *
 * @experimental
 */
export const zSessionConfigBoolean = z.looseObject({
    currentValue: z.boolean(),
});
/**
 * Unique identifier for a session configuration option value group.
 */
export const zSessionConfigGroupId = z.string();
/**
 * Unique identifier for a session configuration option.
 */
export const zSessionConfigId = z.string();
/**
 * Semantic category for a session configuration option.
 *
 * This is intended to help Clients distinguish broadly common selectors (e.g. model selector vs
 * session mode selector vs thought/reasoning level) for UX purposes (keyboard shortcuts, icons,
 * placement). It MUST NOT be required for correctness. Clients MUST handle missing or unknown
 * categories gracefully.
 *
 * Category names beginning with `_` are free for custom use, like other ACP extension methods.
 * Category names that do not begin with `_` are reserved for the ACP spec.
 */
export const zSessionConfigOptionCategory = z.union([
    z.literal("mode"),
    z.literal("model"),
    z.literal("thought_level"),
    z.string(),
]);
/**
 * Unique identifier for a session configuration option value.
 */
export const zSessionConfigValueId = z.string();
/**
 * A possible value for a session configuration option.
 */
export const zSessionConfigSelectOption = z.looseObject({
    _meta: z.record(z.string(), z.unknown()).nullish(),
    description: z.string().nullish(),
    name: z.string(),
    value: zSessionConfigValueId,
});
/**
 * A group of possible values for a session configuration option.
 */
export const zSessionConfigSelectGroup = z.looseObject({
    _meta: z.record(z.string(), z.unknown()).nullish(),
    group: zSessionConfigGroupId,
    name: z.string(),
    options: z.array(zSessionConfigSelectOption),
});
/**
 * Possible values for a session configuration option.
 */
export const zSessionConfigSelectOptions = z.union([
    z.array(zSessionConfigSelectOption),
    z.array(zSessionConfigSelectGroup),
]);
/**
 * A single-value selector (dropdown) session configuration option payload.
 */
export const zSessionConfigSelect = z.looseObject({
    currentValue: zSessionConfigValueId,
    options: zSessionConfigSelectOptions,
});
export const zSessionConfigOption = z.intersection(z.union([
    zSessionConfigSelect.and(z.looseObject({
        type: z.literal("select"),
    })),
    zSessionConfigBoolean.and(z.looseObject({
        type: z.literal("boolean"),
    })),
]), z.looseObject({
    _meta: z.record(z.string(), z.unknown()).nullish(),
    category: zSessionConfigOptionCategory.nullish(),
    description: z.string().nullish(),
    id: zSessionConfigId,
    name: z.string(),
}));
/**
 * Session configuration options have been updated.
 */
export const zConfigOptionUpdate = z.looseObject({
    _meta: z.record(z.string(), z.unknown()).nullish(),
    configOptions: z.array(zSessionConfigOption),
});
/**
 * **UNSTABLE**
 *
 * This capability is not part of the spec yet, and may be removed or changed at any point.
 *
 * Capabilities for the `session/fork` method.
 *
 * By supplying `{}` it means that the agent supports forking of sessions.
 *
 * @experimental
 */
export const zSessionForkCapabilities = z.looseObject({
    _meta: z.record(z.string(), z.unknown()).nullish(),
});
/**
 * A unique identifier for a conversation session between a client and agent.
 *
 * Sessions maintain their own context, conversation history, and state,
 * allowing multiple independent interactions with the same agent.
 *
 * See protocol docs: [Session ID](https://agentclientprotocol.com/protocol/session-setup#session-id)
 */
export const zSessionId = z.string();
/**
 * Notification sent when a suggestion is accepted.
 */
export const zAcceptNesNotification = z.looseObject({
    _meta: z.record(z.string(), z.unknown()).nullish(),
    id: z.string(),
    sessionId: zSessionId,
});
/**
 * Notification to cancel ongoing operations for a session.
 *
 * See protocol docs: [Cancellation](https://agentclientprotocol.com/protocol/prompt-turn#cancellation)
 */
export const zCancelNotification = z.looseObject({
    _meta: z.record(z.string(), z.unknown()).nullish(),
    sessionId: zSessionId,
});
/**
 * Request to close an NES session.
 *
 * The agent **must** cancel any ongoing work related to the NES session
 * and then free up any resources associated with the session.
 */
export const zCloseNesRequest = z.looseObject({
    _meta: z.record(z.string(), z.unknown()).nullish(),
    sessionId: zSessionId,
});
/**
 * **UNSTABLE**
 *
 * This capability is not part of the spec yet, and may be removed or changed at any point.
 *
 * Request parameters for closing an active session.
 *
 * If supported, the agent **must** cancel any ongoing work related to the session
 * (treat it as if `session/cancel` was called) and then free up any resources
 * associated with the session.
 *
 * Only available if the Agent supports the `session.close` capability.
 *
 * @experimental
 */
export const zCloseSessionRequest = z.looseObject({
    _meta: z.record(z.string(), z.unknown()).nullish(),
    sessionId: zSessionId,
});
/**
 * Request to create a new terminal and execute a command.
 */
export const zCreateTerminalRequest = z.looseObject({
    _meta: z.record(z.string(), z.unknown()).nullish(),
    args: z.array(z.string()).optional(),
    command: z.string(),
    cwd: z.string().nullish(),
    env: z.array(zEnvVariable).optional(),
    outputByteLimit: z.number().nullish(),
    sessionId: zSessionId,
});
/**
 * Notification sent when a file is closed.
 */
export const zDidCloseDocumentNotification = z.looseObject({
    _meta: z.record(z.string(), z.unknown()).nullish(),
    sessionId: zSessionId,
    uri: z.string(),
});
/**
 * Notification sent when a file becomes the active editor tab.
 */
export const zDidFocusDocumentNotification = z.looseObject({
    _meta: z.record(z.string(), z.unknown()).nullish(),
    position: zPosition,
    sessionId: zSessionId,
    uri: z.string(),
    version: z.number(),
    visibleRange: zRange,
});
/**
 * Notification sent when a file is opened in the editor.
 */
export const zDidOpenDocumentNotification = z.looseObject({
    _meta: z.record(z.string(), z.unknown()).nullish(),
    languageId: z.string(),
    sessionId: zSessionId,
    text: z.string(),
    uri: z.string(),
    version: z.number(),
});
/**
 * Notification sent when a file is saved.
 */
export const zDidSaveDocumentNotification = z.looseObject({
    _meta: z.record(z.string(), z.unknown()).nullish(),
    sessionId: zSessionId,
    uri: z.string(),
});
/**
 * **UNSTABLE**
 *
 * This capability is not part of the spec yet, and may be removed or changed at any point.
 *
 * Request parameters for forking an existing session.
 *
 * Creates a new session based on the context of an existing one, allowing
 * operations like generating summaries without affecting the original session's history.
 *
 * Only available if the Agent supports the `session.fork` capability.
 *
 * @experimental
 */
export const zForkSessionRequest = z.looseObject({
    _meta: z.record(z.string(), z.unknown()).nullish(),
    additionalDirectories: z.array(z.string()).optional(),
    cwd: z.string(),
    mcpServers: z.array(zMcpServer).optional(),
    sessionId: zSessionId,
});
/**
 * Request to kill a terminal without releasing it.
 */
export const zKillTerminalRequest = z.looseObject({
    _meta: z.record(z.string(), z.unknown()).nullish(),
    sessionId: zSessionId,
    terminalId: z.string(),
});
/**
 * Request parameters for loading an existing session.
 *
 * Only available if the Agent supports the `loadSession` capability.
 *
 * See protocol docs: [Loading Sessions](https://agentclientprotocol.com/protocol/session-setup#loading-sessions)
 */
export const zLoadSessionRequest = z.looseObject({
    _meta: z.record(z.string(), z.unknown()).nullish(),
    additionalDirectories: z.array(z.string()).optional(),
    cwd: z.string(),
    mcpServers: z.array(zMcpServer),
    sessionId: zSessionId,
});
/**
 * Request to read content from a text file.
 *
 * Only available if the client supports the `fs.readTextFile` capability.
 */
export const zReadTextFileRequest = z.looseObject({
    _meta: z.record(z.string(), z.unknown()).nullish(),
    limit: z
        .number()
        .int()
        .gte(0)
        .max(4294967295, {
        message: "Invalid value: Expected uint32 to be <= 4294967295",
    })
        .nullish(),
    line: z
        .number()
        .int()
        .gte(0)
        .max(4294967295, {
        message: "Invalid value: Expected uint32 to be <= 4294967295",
    })
        .nullish(),
    path: z.string(),
    sessionId: zSessionId,
});
/**
 * Notification sent when a suggestion is rejected.
 */
export const zRejectNesNotification = z.looseObject({
    _meta: z.record(z.string(), z.unknown()).nullish(),
    id: z.string(),
    reason: zNesRejectReason.nullish(),
    sessionId: zSessionId,
});
/**
 * Request to release a terminal and free its resources.
 */
export const zReleaseTerminalRequest = z.looseObject({
    _meta: z.record(z.string(), z.unknown()).nullish(),
    sessionId: zSessionId,
    terminalId: z.string(),
});
/**
 * **UNSTABLE**
 *
 * This capability is not part of the spec yet, and may be removed or changed at any point.
 *
 * Request parameters for resuming an existing session.
 *
 * Resumes an existing session without returning previous messages (unlike `session/load`).
 * This is useful for agents that can resume sessions but don't implement full session loading.
 *
 * Only available if the Agent supports the `session.resume` capability.
 *
 * @experimental
 */
export const zResumeSessionRequest = z.looseObject({
    _meta: z.record(z.string(), z.unknown()).nullish(),
    additionalDirectories: z.array(z.string()).optional(),
    cwd: z.string(),
    mcpServers: z.array(zMcpServer).optional(),
    sessionId: zSessionId,
});
/**
 * Information about a session returned by session/list
 */
export const zSessionInfo = z.looseObject({
    _meta: z.record(z.string(), z.unknown()).nullish(),
    additionalDirectories: z.array(z.string()).optional(),
    cwd: z.string(),
    sessionId: zSessionId,
    title: z.string().nullish(),
    updatedAt: z.string().nullish(),
});
/**
 * Response from listing sessions.
 */
export const zListSessionsResponse = z.looseObject({
    _meta: z.record(z.string(), z.unknown()).nullish(),
    nextCursor: z.string().nullish(),
    sessions: z.array(zSessionInfo),
});
/**
 * Update to session metadata. All fields are optional to support partial updates.
 *
 * Agents send this notification to update session information like title or custom metadata.
 * This allows clients to display dynamic session names and track session state changes.
 */
export const zSessionInfoUpdate = z.looseObject({
    _meta: z.record(z.string(), z.unknown()).nullish(),
    title: z.string().nullish(),
    updatedAt: z.string().nullish(),
});
/**
 * Capabilities for the `session/list` method.
 *
 * By supplying `{}` it means that the agent supports listing of sessions.
 */
export const zSessionListCapabilities = z.looseObject({
    _meta: z.record(z.string(), z.unknown()).nullish(),
});
/**
 * Unique identifier for a Session Mode.
 */
export const zSessionModeId = z.string();
/**
 * The current mode of the session has changed
 *
 * See protocol docs: [Session Modes](https://agentclientprotocol.com/protocol/session-modes)
 */
export const zCurrentModeUpdate = z.looseObject({
    _meta: z.record(z.string(), z.unknown()).nullish(),
    currentModeId: zSessionModeId,
});
/**
 * A mode the agent can operate in.
 *
 * See protocol docs: [Session Modes](https://agentclientprotocol.com/protocol/session-modes)
 */
export const zSessionMode = z.looseObject({
    _meta: z.record(z.string(), z.unknown()).nullish(),
    description: z.string().nullish(),
    id: zSessionModeId,
    name: z.string(),
});
/**
 * The set of modes and the one currently active.
 */
export const zSessionModeState = z.looseObject({
    _meta: z.record(z.string(), z.unknown()).nullish(),
    availableModes: z.array(zSessionMode),
    currentModeId: zSessionModeId,
});
/**
 * **UNSTABLE**
 *
 * This capability is not part of the spec yet, and may be removed or changed at any point.
 *
 * The set of models and the one currently active.
 *
 * @experimental
 */
export const zSessionModelState = z.looseObject({
    _meta: z.record(z.string(), z.unknown()).nullish(),
    availableModels: z.array(zModelInfo),
    currentModelId: zModelId,
});
/**
 * **UNSTABLE**
 *
 * This capability is not part of the spec yet, and may be removed or changed at any point.
 *
 * Response from forking an existing session.
 *
 * @experimental
 */
export const zForkSessionResponse = z.looseObject({
    _meta: z.record(z.string(), z.unknown()).nullish(),
    configOptions: z.array(zSessionConfigOption).nullish(),
    models: zSessionModelState.nullish(),
    modes: zSessionModeState.nullish(),
    sessionId: zSessionId,
});
/**
 * Response from loading an existing session.
 */
export const zLoadSessionResponse = z.looseObject({
    _meta: z.record(z.string(), z.unknown()).nullish(),
    configOptions: z.array(zSessionConfigOption).nullish(),
    models: zSessionModelState.nullish(),
    modes: zSessionModeState.nullish(),
});
/**
 * Response from creating a new session.
 *
 * See protocol docs: [Creating a Session](https://agentclientprotocol.com/protocol/session-setup#creating-a-session)
 */
export const zNewSessionResponse = z.looseObject({
    _meta: z.record(z.string(), z.unknown()).nullish(),
    configOptions: z.array(zSessionConfigOption).nullish(),
    models: zSessionModelState.nullish(),
    modes: zSessionModeState.nullish(),
    sessionId: zSessionId,
});
/**
 * **UNSTABLE**
 *
 * This capability is not part of the spec yet, and may be removed or changed at any point.
 *
 * Response from resuming an existing session.
 *
 * @experimental
 */
export const zResumeSessionResponse = z.looseObject({
    _meta: z.record(z.string(), z.unknown()).nullish(),
    configOptions: z.array(zSessionConfigOption).nullish(),
    models: zSessionModelState.nullish(),
    modes: zSessionModeState.nullish(),
});
/**
 * **UNSTABLE**
 *
 * This capability is not part of the spec yet, and may be removed or changed at any point.
 *
 * Capabilities for the `session/resume` method.
 *
 * By supplying `{}` it means that the agent supports resuming of sessions.
 *
 * @experimental
 */
export const zSessionResumeCapabilities = z.looseObject({
    _meta: z.record(z.string(), z.unknown()).nullish(),
});
/**
 * Session capabilities supported by the agent.
 *
 * As a baseline, all Agents **MUST** support `session/new`, `session/prompt`, `session/cancel`, and `session/update`.
 *
 * Optionally, they **MAY** support other session methods and notifications by specifying additional capabilities.
 *
 * Note: `session/load` is still handled by the top-level `load_session` capability. This will be unified in future versions of the protocol.
 *
 * See protocol docs: [Session Capabilities](https://agentclientprotocol.com/protocol/initialization#session-capabilities)
 */
export const zSessionCapabilities = z.looseObject({
    _meta: z.record(z.string(), z.unknown()).nullish(),
    additionalDirectories: zSessionAdditionalDirectoriesCapabilities.nullish(),
    close: zSessionCloseCapabilities.nullish(),
    fork: zSessionForkCapabilities.nullish(),
    list: zSessionListCapabilities.nullish(),
    resume: zSessionResumeCapabilities.nullish(),
});
export const zSetSessionConfigOptionRequest = z.intersection(z.union([
    z.looseObject({
        type: z.literal("boolean"),
        value: z.boolean(),
    }),
    z.looseObject({
        value: zSessionConfigValueId,
    }),
]), z.looseObject({
    _meta: z.record(z.string(), z.unknown()).nullish(),
    configId: zSessionConfigId,
    sessionId: zSessionId,
}));
/**
 * Response to `session/set_config_option` method.
 */
export const zSetSessionConfigOptionResponse = z.looseObject({
    _meta: z.record(z.string(), z.unknown()).nullish(),
    configOptions: z.array(zSessionConfigOption),
});
/**
 * Request parameters for setting a session mode.
 */
export const zSetSessionModeRequest = z.looseObject({
    _meta: z.record(z.string(), z.unknown()).nullish(),
    modeId: zSessionModeId,
    sessionId: zSessionId,
});
/**
 * Response to `session/set_mode` method.
 */
export const zSetSessionModeResponse = z.looseObject({
    _meta: z.record(z.string(), z.unknown()).nullish(),
});
/**
 * **UNSTABLE**
 *
 * This capability is not part of the spec yet, and may be removed or changed at any point.
 *
 * Request parameters for setting a session model.
 *
 * @experimental
 */
export const zSetSessionModelRequest = z.looseObject({
    _meta: z.record(z.string(), z.unknown()).nullish(),
    modelId: zModelId,
    sessionId: zSessionId,
});
/**
 * **UNSTABLE**
 *
 * This capability is not part of the spec yet, and may be removed or changed at any point.
 *
 * Response to `session/set_model` method.
 *
 * @experimental
 */
export const zSetSessionModelResponse = z.looseObject({
    _meta: z.record(z.string(), z.unknown()).nullish(),
});
/**
 * Response to `nes/start`.
 */
export const zStartNesResponse = z.looseObject({
    _meta: z.record(z.string(), z.unknown()).nullish(),
    sessionId: zSessionId,
});
/**
 * Reasons why an agent stops processing a prompt turn.
 *
 * See protocol docs: [Stop Reasons](https://agentclientprotocol.com/protocol/prompt-turn#stop-reasons)
 */
export const zStopReason = z.union([
    z.literal("end_turn"),
    z.literal("max_tokens"),
    z.literal("max_turn_requests"),
    z.literal("refusal"),
    z.literal("cancelled"),
]);
/**
 * String format types for string properties in elicitation schemas.
 */
export const zStringFormat = z.union([
    z.literal("email"),
    z.literal("uri"),
    z.literal("date"),
    z.literal("date-time"),
]);
/**
 * Schema for string properties in an elicitation form.
 *
 * When `enum` or `oneOf` is set, this represents a single-select enum
 * with `"type": "string"`.
 */
export const zStringPropertySchema = z.looseObject({
    default: z.string().nullish(),
    description: z.string().nullish(),
    enum: z.array(z.string()).nullish(),
    format: zStringFormat.nullish(),
    maxLength: z
        .number()
        .int()
        .gte(0)
        .max(4294967295, {
        message: "Invalid value: Expected uint32 to be <= 4294967295",
    })
        .nullish(),
    minLength: z
        .number()
        .int()
        .gte(0)
        .max(4294967295, {
        message: "Invalid value: Expected uint32 to be <= 4294967295",
    })
        .nullish(),
    oneOf: z.array(zEnumOption).nullish(),
    pattern: z.string().nullish(),
    title: z.string().nullish(),
});
/**
 * Request for a code suggestion.
 */
export const zSuggestNesRequest = z.looseObject({
    _meta: z.record(z.string(), z.unknown()).nullish(),
    context: zNesSuggestContext.nullish(),
    position: zPosition,
    selection: zRange.nullish(),
    sessionId: zSessionId,
    triggerKind: zNesTriggerKind,
    uri: z.string(),
    version: z.number(),
});
/**
 * Response to `nes/suggest`.
 */
export const zSuggestNesResponse = z.looseObject({
    _meta: z.record(z.string(), z.unknown()).nullish(),
    suggestions: z.array(zNesSuggestion),
});
/**
 * Embed a terminal created with `terminal/create` by its id.
 *
 * The terminal must be added before calling `terminal/release`.
 *
 * See protocol docs: [Terminal](https://agentclientprotocol.com/protocol/terminals)
 */
export const zTerminal = z.looseObject({
    _meta: z.record(z.string(), z.unknown()).nullish(),
    terminalId: z.string(),
});
/**
 * Exit status of a terminal command.
 */
export const zTerminalExitStatus = z.looseObject({
    _meta: z.record(z.string(), z.unknown()).nullish(),
    exitCode: z
        .number()
        .int()
        .gte(0)
        .max(4294967295, {
        message: "Invalid value: Expected uint32 to be <= 4294967295",
    })
        .nullish(),
    signal: z.string().nullish(),
});
/**
 * Request to get the current output and status of a terminal.
 */
export const zTerminalOutputRequest = z.looseObject({
    _meta: z.record(z.string(), z.unknown()).nullish(),
    sessionId: zSessionId,
    terminalId: z.string(),
});
/**
 * Response containing the terminal output and exit status.
 */
export const zTerminalOutputResponse = z.looseObject({
    _meta: z.record(z.string(), z.unknown()).nullish(),
    exitStatus: zTerminalExitStatus.nullish(),
    output: z.string(),
    truncated: z.boolean(),
});
/**
 * Text provided to or from an LLM.
 */
export const zTextContent = z.looseObject({
    _meta: z.record(z.string(), z.unknown()).nullish(),
    annotations: zAnnotations.nullish(),
    text: z.string(),
});
/**
 * A content change event for a document.
 *
 * When `range` is `None`, `text` is the full content of the document.
 * When `range` is `Some`, `text` replaces the given range.
 */
export const zTextDocumentContentChangeEvent = z.looseObject({
    range: zRange.nullish(),
    text: z.string(),
});
/**
 * Notification sent when a file is edited.
 */
export const zDidChangeDocumentNotification = z.looseObject({
    _meta: z.record(z.string(), z.unknown()).nullish(),
    contentChanges: z.array(zTextDocumentContentChangeEvent),
    sessionId: zSessionId,
    uri: z.string(),
    version: z.number(),
});
export const zClientNotification = z.looseObject({
    method: z.string(),
    params: z
        .union([
        zCancelNotification,
        zDidOpenDocumentNotification,
        zDidChangeDocumentNotification,
        zDidCloseDocumentNotification,
        zDidSaveDocumentNotification,
        zDidFocusDocumentNotification,
        zAcceptNesNotification,
        zRejectNesNotification,
        zExtNotification,
    ])
        .nullish(),
});
/**
 * How the agent wants document changes delivered.
 */
export const zTextDocumentSyncKind = z.union([
    z.literal("full"),
    z.literal("incremental"),
]);
/**
 * Capabilities for `document/didChange` events.
 */
export const zNesDocumentDidChangeCapabilities = z.looseObject({
    _meta: z.record(z.string(), z.unknown()).nullish(),
    syncKind: zTextDocumentSyncKind,
});
/**
 * Document event capabilities the agent wants to receive.
 */
export const zNesDocumentEventCapabilities = z.looseObject({
    _meta: z.record(z.string(), z.unknown()).nullish(),
    didChange: zNesDocumentDidChangeCapabilities.nullish(),
    didClose: zNesDocumentDidCloseCapabilities.nullish(),
    didFocus: zNesDocumentDidFocusCapabilities.nullish(),
    didOpen: zNesDocumentDidOpenCapabilities.nullish(),
    didSave: zNesDocumentDidSaveCapabilities.nullish(),
});
/**
 * Event capabilities the agent can consume.
 */
export const zNesEventCapabilities = z.looseObject({
    _meta: z.record(z.string(), z.unknown()).nullish(),
    document: zNesDocumentEventCapabilities.nullish(),
});
/**
 * NES capabilities advertised by the agent during initialization.
 */
export const zNesCapabilities = z.looseObject({
    _meta: z.record(z.string(), z.unknown()).nullish(),
    context: zNesContextCapabilities.nullish(),
    events: zNesEventCapabilities.nullish(),
});
/**
 * Capabilities supported by the agent.
 *
 * Advertised during initialization to inform the client about
 * available features and content types.
 *
 * See protocol docs: [Agent Capabilities](https://agentclientprotocol.com/protocol/initialization#agent-capabilities)
 */
export const zAgentCapabilities = z.looseObject({
    _meta: z.record(z.string(), z.unknown()).nullish(),
    auth: zAgentAuthCapabilities.optional().default({}),
    loadSession: z.boolean().optional().default(false),
    mcpCapabilities: zMcpCapabilities
        .optional()
        .default({ http: false, sse: false }),
    nes: zNesCapabilities.nullish(),
    positionEncoding: zPositionEncodingKind.nullish(),
    promptCapabilities: zPromptCapabilities.optional().default({
        audio: false,
        embeddedContext: false,
        image: false,
    }),
    sessionCapabilities: zSessionCapabilities.optional().default({}),
});
/**
 * Response to the `initialize` method.
 *
 * Contains the negotiated protocol version and agent capabilities.
 *
 * See protocol docs: [Initialization](https://agentclientprotocol.com/protocol/initialization)
 */
export const zInitializeResponse = z.looseObject({
    _meta: z.record(z.string(), z.unknown()).nullish(),
    agentCapabilities: zAgentCapabilities.optional().default({
        auth: {},
        loadSession: false,
        mcpCapabilities: { http: false, sse: false },
        promptCapabilities: {
            audio: false,
            embeddedContext: false,
            image: false,
        },
        sessionCapabilities: {},
    }),
    agentInfo: zImplementation.nullish(),
    authMethods: z.array(zAuthMethod).optional().default([]),
    protocolVersion: zProtocolVersion,
});
/**
 * Text-based resource contents.
 */
export const zTextResourceContents = z.looseObject({
    _meta: z.record(z.string(), z.unknown()).nullish(),
    mimeType: z.string().nullish(),
    text: z.string(),
    uri: z.string(),
});
/**
 * Resource content that can be embedded in a message.
 */
export const zEmbeddedResourceResource = z.union([
    zTextResourceContents,
    zBlobResourceContents,
]);
/**
 * The contents of a resource, embedded into a prompt or tool call result.
 */
export const zEmbeddedResource = z.looseObject({
    _meta: z.record(z.string(), z.unknown()).nullish(),
    annotations: zAnnotations.nullish(),
    resource: zEmbeddedResourceResource,
});
/**
 * Content blocks represent displayable information in the Agent Client Protocol.
 *
 * They provide a structured way to handle various types of user-facing content—whether
 * it's text from language models, images for analysis, or embedded resources for context.
 *
 * Content blocks appear in:
 * - User prompts sent via `session/prompt`
 * - Language model output streamed through `session/update` notifications
 * - Progress updates and results from tool calls
 *
 * This structure is compatible with the Model Context Protocol (MCP), enabling
 * agents to seamlessly forward content from MCP tool outputs without transformation.
 *
 * See protocol docs: [Content](https://agentclientprotocol.com/protocol/content)
 */
export const zContentBlock = z.union([
    zTextContent.and(z.looseObject({
        type: z.literal("text"),
    })),
    zImageContent.and(z.looseObject({
        type: z.literal("image"),
    })),
    zAudioContent.and(z.looseObject({
        type: z.literal("audio"),
    })),
    zResourceLink.and(z.looseObject({
        type: z.literal("resource_link"),
    })),
    zEmbeddedResource.and(z.looseObject({
        type: z.literal("resource"),
    })),
]);
/**
 * Standard content block (text, images, resources).
 */
export const zContent = z.looseObject({
    _meta: z.record(z.string(), z.unknown()).nullish(),
    content: zContentBlock,
});
/**
 * A streamed item of content
 */
export const zContentChunk = z.looseObject({
    _meta: z.record(z.string(), z.unknown()).nullish(),
    content: zContentBlock,
    messageId: z.string().nullish(),
});
/**
 * Request parameters for sending a user prompt to the agent.
 *
 * Contains the user's message and any additional context.
 *
 * See protocol docs: [User Message](https://agentclientprotocol.com/protocol/prompt-turn#1-user-message)
 */
export const zPromptRequest = z.looseObject({
    _meta: z.record(z.string(), z.unknown()).nullish(),
    messageId: z.string().nullish(),
    prompt: z.array(zContentBlock),
    sessionId: zSessionId,
});
/**
 * Items definition for titled multi-select enum properties.
 */
export const zTitledMultiSelectItems = z.looseObject({
    anyOf: z.array(zEnumOption),
});
/**
 * Content produced by a tool call.
 *
 * Tool calls can produce different types of content including
 * standard content blocks (text, images) or file diffs.
 *
 * See protocol docs: [Content](https://agentclientprotocol.com/protocol/tool-calls#content)
 */
export const zToolCallContent = z.union([
    zContent.and(z.looseObject({
        type: z.literal("content"),
    })),
    zDiff.and(z.looseObject({
        type: z.literal("diff"),
    })),
    zTerminal.and(z.looseObject({
        type: z.literal("terminal"),
    })),
]);
/**
 * Unique identifier for a tool call within a session.
 */
export const zToolCallId = z.string();
/**
 * A file location being accessed or modified by a tool.
 *
 * Enables clients to implement "follow-along" features that track
 * which files the agent is working with in real-time.
 *
 * See protocol docs: [Following the Agent](https://agentclientprotocol.com/protocol/tool-calls#following-the-agent)
 */
export const zToolCallLocation = z.looseObject({
    _meta: z.record(z.string(), z.unknown()).nullish(),
    line: z
        .number()
        .int()
        .gte(0)
        .max(4294967295, {
        message: "Invalid value: Expected uint32 to be <= 4294967295",
    })
        .nullish(),
    path: z.string(),
});
/**
 * Execution status of a tool call.
 *
 * Tool calls progress through different statuses during their lifecycle.
 *
 * See protocol docs: [Status](https://agentclientprotocol.com/protocol/tool-calls#status)
 */
export const zToolCallStatus = z.union([
    z.literal("pending"),
    z.literal("in_progress"),
    z.literal("completed"),
    z.literal("failed"),
]);
/**
 * Categories of tools that can be invoked.
 *
 * Tool kinds help clients choose appropriate icons and optimize how they
 * display tool execution progress.
 *
 * See protocol docs: [Creating](https://agentclientprotocol.com/protocol/tool-calls#creating)
 */
export const zToolKind = z.union([
    z.literal("read"),
    z.literal("edit"),
    z.literal("delete"),
    z.literal("move"),
    z.literal("search"),
    z.literal("execute"),
    z.literal("think"),
    z.literal("fetch"),
    z.literal("switch_mode"),
    z.literal("other"),
]);
/**
 * Represents a tool call that the language model has requested.
 *
 * Tool calls are actions that the agent executes on behalf of the language model,
 * such as reading files, executing code, or fetching data from external sources.
 *
 * See protocol docs: [Tool Calls](https://agentclientprotocol.com/protocol/tool-calls)
 */
export const zToolCall = z.looseObject({
    _meta: z.record(z.string(), z.unknown()).nullish(),
    content: z.array(zToolCallContent).optional(),
    kind: zToolKind.optional(),
    locations: z.array(zToolCallLocation).optional(),
    rawInput: z.unknown().optional(),
    rawOutput: z.unknown().optional(),
    status: zToolCallStatus.optional(),
    title: z.string(),
    toolCallId: zToolCallId,
});
/**
 * An update to an existing tool call.
 *
 * Used to report progress and results as tools execute. All fields except
 * the tool call ID are optional - only changed fields need to be included.
 *
 * See protocol docs: [Updating](https://agentclientprotocol.com/protocol/tool-calls#updating)
 */
export const zToolCallUpdate = z.looseObject({
    _meta: z.record(z.string(), z.unknown()).nullish(),
    content: z.array(zToolCallContent).nullish(),
    kind: zToolKind.nullish(),
    locations: z.array(zToolCallLocation).nullish(),
    rawInput: z.unknown().optional(),
    rawOutput: z.unknown().optional(),
    status: zToolCallStatus.nullish(),
    title: z.string().nullish(),
    toolCallId: zToolCallId,
});
/**
 * Request for user permission to execute a tool call.
 *
 * Sent when the agent needs authorization before performing a sensitive operation.
 *
 * See protocol docs: [Requesting Permission](https://agentclientprotocol.com/protocol/tool-calls#requesting-permission)
 */
export const zRequestPermissionRequest = z.looseObject({
    _meta: z.record(z.string(), z.unknown()).nullish(),
    options: z.array(zPermissionOption),
    sessionId: zSessionId,
    toolCall: zToolCallUpdate,
});
/**
 * All text that was typed after the command name is provided as input.
 */
export const zUnstructuredCommandInput = z.looseObject({
    _meta: z.record(z.string(), z.unknown()).nullish(),
    hint: z.string(),
});
/**
 * unstructured
 *
 * All text that was typed after the command name is provided as input.
 */
export const zAvailableCommandInput = zUnstructuredCommandInput;
/**
 * Information about a command.
 */
export const zAvailableCommand = z.looseObject({
    _meta: z.record(z.string(), z.unknown()).nullish(),
    description: z.string(),
    input: zAvailableCommandInput.nullish(),
    name: z.string(),
});
/**
 * Available commands are ready or have changed
 */
export const zAvailableCommandsUpdate = z.looseObject({
    _meta: z.record(z.string(), z.unknown()).nullish(),
    availableCommands: z.array(zAvailableCommand),
});
/**
 * Items definition for untitled multi-select enum properties.
 */
export const zUntitledMultiSelectItems = z.looseObject({
    enum: z.array(z.string()),
    type: zElicitationStringType,
});
/**
 * Items for a multi-select (array) property schema.
 */
export const zMultiSelectItems = z.union([
    zUntitledMultiSelectItems,
    zTitledMultiSelectItems,
]);
/**
 * Schema for multi-select (array) properties in an elicitation form.
 */
export const zMultiSelectPropertySchema = z.looseObject({
    default: z.array(z.string()).nullish(),
    description: z.string().nullish(),
    items: zMultiSelectItems,
    maxItems: z.number().nullish(),
    minItems: z.number().nullish(),
    title: z.string().nullish(),
});
/**
 * Property schema for elicitation form fields.
 *
 * Each variant corresponds to a JSON Schema `"type"` value.
 * Single-select enums use the `String` variant with `enum` or `oneOf` set.
 * Multi-select enums use the `Array` variant.
 */
export const zElicitationPropertySchema = z.union([
    zStringPropertySchema.and(z.looseObject({
        type: z.literal("string"),
    })),
    zNumberPropertySchema.and(z.looseObject({
        type: z.literal("number"),
    })),
    zIntegerPropertySchema.and(z.looseObject({
        type: z.literal("integer"),
    })),
    zBooleanPropertySchema.and(z.looseObject({
        type: z.literal("boolean"),
    })),
    zMultiSelectPropertySchema.and(z.looseObject({
        type: z.literal("array"),
    })),
]);
/**
 * Type-safe elicitation schema for requesting structured user input.
 *
 * This represents a JSON Schema object with primitive-typed properties,
 * as required by the elicitation specification.
 */
export const zElicitationSchema = z.looseObject({
    description: z.string().nullish(),
    properties: z
        .record(z.string(), zElicitationPropertySchema)
        .optional()
        .default({}),
    required: z.array(z.string()).nullish(),
    title: z.string().nullish(),
    type: zElicitationSchemaType.optional().default("object"),
});
/**
 * **UNSTABLE**
 *
 * This capability is not part of the spec yet, and may be removed or changed at any point.
 *
 * Form-based elicitation mode where the client renders a form from the provided schema.
 *
 * @experimental
 */
export const zElicitationFormMode = z.looseObject({
    requestedSchema: zElicitationSchema,
});
export const zElicitationRequest = z.intersection(z.union([
    zElicitationFormMode.and(z.looseObject({
        mode: z.literal("form"),
    })),
    zElicitationUrlMode.and(z.looseObject({
        mode: z.literal("url"),
    })),
]), z.looseObject({
    _meta: z.record(z.string(), z.unknown()).nullish(),
    message: z.string(),
    sessionId: zSessionId,
}));
/**
 * **UNSTABLE**
 *
 * This capability is not part of the spec yet, and may be removed or changed at any point.
 *
 * Token usage information for a prompt turn.
 *
 * @experimental
 */
export const zUsage = z.looseObject({
    cachedReadTokens: z.number().nullish(),
    cachedWriteTokens: z.number().nullish(),
    inputTokens: z.number(),
    outputTokens: z.number(),
    thoughtTokens: z.number().nullish(),
    totalTokens: z.number(),
});
/**
 * Response from processing a user prompt.
 *
 * See protocol docs: [Check for Completion](https://agentclientprotocol.com/protocol/prompt-turn#4-check-for-completion)
 */
export const zPromptResponse = z.looseObject({
    _meta: z.record(z.string(), z.unknown()).nullish(),
    stopReason: zStopReason,
    usage: zUsage.nullish(),
    userMessageId: z.string().nullish(),
});
export const zAgentResponse = z.union([
    z.looseObject({
        id: zRequestId,
        result: z.union([
            zInitializeResponse,
            zAuthenticateResponse,
            zLogoutResponse,
            zNewSessionResponse,
            zLoadSessionResponse,
            zListSessionsResponse,
            zForkSessionResponse,
            zResumeSessionResponse,
            zCloseSessionResponse,
            zSetSessionModeResponse,
            zSetSessionConfigOptionResponse,
            zPromptResponse,
            zSetSessionModelResponse,
            zStartNesResponse,
            zSuggestNesResponse,
            zCloseNesResponse,
            zExtResponse,
        ]),
    }),
    z.looseObject({
        error: zError,
        id: zRequestId,
    }),
]);
/**
 * **UNSTABLE**
 *
 * This capability is not part of the spec yet, and may be removed or changed at any point.
 *
 * Context window and cost update for a session.
 *
 * @experimental
 */
export const zUsageUpdate = z.looseObject({
    _meta: z.record(z.string(), z.unknown()).nullish(),
    cost: zCost.nullish(),
    size: z.number(),
    used: z.number(),
});
/**
 * Different types of updates that can be sent during session processing.
 *
 * These updates provide real-time feedback about the agent's progress.
 *
 * See protocol docs: [Agent Reports Output](https://agentclientprotocol.com/protocol/prompt-turn#3-agent-reports-output)
 */
export const zSessionUpdate = z.union([
    zContentChunk.and(z.looseObject({
        sessionUpdate: z.literal("user_message_chunk"),
    })),
    zContentChunk.and(z.looseObject({
        sessionUpdate: z.literal("agent_message_chunk"),
    })),
    zContentChunk.and(z.looseObject({
        sessionUpdate: z.literal("agent_thought_chunk"),
    })),
    zToolCall.and(z.looseObject({
        sessionUpdate: z.literal("tool_call"),
    })),
    zToolCallUpdate.and(z.looseObject({
        sessionUpdate: z.literal("tool_call_update"),
    })),
    zPlan.and(z.looseObject({
        sessionUpdate: z.literal("plan"),
    })),
    zAvailableCommandsUpdate.and(z.looseObject({
        sessionUpdate: z.literal("available_commands_update"),
    })),
    zCurrentModeUpdate.and(z.looseObject({
        sessionUpdate: z.literal("current_mode_update"),
    })),
    zConfigOptionUpdate.and(z.looseObject({
        sessionUpdate: z.literal("config_option_update"),
    })),
    zSessionInfoUpdate.and(z.looseObject({
        sessionUpdate: z.literal("session_info_update"),
    })),
    zUsageUpdate.and(z.looseObject({
        sessionUpdate: z.literal("usage_update"),
    })),
]);
/**
 * Notification containing a session update from the agent.
 *
 * Used to stream real-time progress and results during prompt processing.
 *
 * See protocol docs: [Agent Reports Output](https://agentclientprotocol.com/protocol/prompt-turn#3-agent-reports-output)
 */
export const zSessionNotification = z.looseObject({
    _meta: z.record(z.string(), z.unknown()).nullish(),
    sessionId: zSessionId,
    update: zSessionUpdate,
});
export const zAgentNotification = z.looseObject({
    method: z.string(),
    params: z
        .union([
        zSessionNotification,
        zElicitationCompleteNotification,
        zExtNotification,
    ])
        .nullish(),
});
/**
 * Request to wait for a terminal command to exit.
 */
export const zWaitForTerminalExitRequest = z.looseObject({
    _meta: z.record(z.string(), z.unknown()).nullish(),
    sessionId: zSessionId,
    terminalId: z.string(),
});
/**
 * Response containing the exit status of a terminal command.
 */
export const zWaitForTerminalExitResponse = z.looseObject({
    _meta: z.record(z.string(), z.unknown()).nullish(),
    exitCode: z
        .number()
        .int()
        .gte(0)
        .max(4294967295, {
        message: "Invalid value: Expected uint32 to be <= 4294967295",
    })
        .nullish(),
    signal: z.string().nullish(),
});
/**
 * A workspace folder.
 */
export const zWorkspaceFolder = z.looseObject({
    name: z.string(),
    uri: z.string(),
});
/**
 * Request to start an NES session.
 */
export const zStartNesRequest = z.looseObject({
    _meta: z.record(z.string(), z.unknown()).nullish(),
    repository: zNesRepository.nullish(),
    workspaceFolders: z.array(zWorkspaceFolder).nullish(),
    workspaceUri: z.string().nullish(),
});
export const zClientRequest = z.looseObject({
    id: zRequestId,
    method: z.string(),
    params: z
        .union([
        zInitializeRequest,
        zAuthenticateRequest,
        zLogoutRequest,
        zNewSessionRequest,
        zLoadSessionRequest,
        zListSessionsRequest,
        zForkSessionRequest,
        zResumeSessionRequest,
        zCloseSessionRequest,
        zSetSessionModeRequest,
        zSetSessionConfigOptionRequest,
        zPromptRequest,
        zSetSessionModelRequest,
        zStartNesRequest,
        zSuggestNesRequest,
        zCloseNesRequest,
        zExtRequest,
    ])
        .nullish(),
});
/**
 * Request to write content to a text file.
 *
 * Only available if the client supports the `fs.writeTextFile` capability.
 */
export const zWriteTextFileRequest = z.looseObject({
    _meta: z.record(z.string(), z.unknown()).nullish(),
    content: z.string(),
    path: z.string(),
    sessionId: zSessionId,
});
export const zAgentRequest = z.looseObject({
    id: zRequestId,
    method: z.string(),
    params: z
        .union([
        zWriteTextFileRequest,
        zReadTextFileRequest,
        zRequestPermissionRequest,
        zCreateTerminalRequest,
        zTerminalOutputRequest,
        zReleaseTerminalRequest,
        zWaitForTerminalExitRequest,
        zKillTerminalRequest,
        zElicitationRequest,
        zExtRequest,
    ])
        .nullish(),
});
/**
 * Response to `fs/write_text_file`
 */
export const zWriteTextFileResponse = z.looseObject({
    _meta: z.record(z.string(), z.unknown()).nullish(),
});
export const zClientResponse = z.union([
    z.looseObject({
        id: zRequestId,
        result: z.union([
            zWriteTextFileResponse,
            zReadTextFileResponse,
            zRequestPermissionResponse,
            zCreateTerminalResponse,
            zTerminalOutputResponse,
            zReleaseTerminalResponse,
            zWaitForTerminalExitResponse,
            zKillTerminalResponse,
            zElicitationResponse,
            zExtResponse,
        ]),
    }),
    z.looseObject({
        error: zError,
        id: zRequestId,
    }),
]);
