// @mariozechner/pi-coding-agent/dist/core/tools/index.js
export { bashTool, bashToolDefinition, createBashTool, createBashToolDefinition, createLocalBashOperations, } from "./bash.js";
export { createEditTool, createEditToolDefinition, editTool, editToolDefinition, } from "./edit.js";
export { withFileMutationQueue } from "./file-mutation-queue.js";
export { createFindTool, createFindToolDefinition, findTool, findToolDefinition, } from "./find.js";
export { createGrepTool, createGrepToolDefinition, grepTool, grepToolDefinition, } from "./grep.js";
export { createLsTool, createLsToolDefinition, lsTool, lsToolDefinition, } from "./ls.js";
export { createReadTool, createReadToolDefinition, readTool, readToolDefinition, } from "./read.js";
export { DEFAULT_MAX_BYTES, DEFAULT_MAX_LINES, formatSize, truncateHead, truncateLine, truncateTail, } from "./truncate.js";
export { createWriteTool, createWriteToolDefinition, writeTool, writeToolDefinition, } from "./write.js";
import { bashTool, bashToolDefinition, createBashTool, createBashToolDefinition, } from "./bash.js";
import { createEditTool, createEditToolDefinition, editTool, editToolDefinition } from "./edit.js";
import { createFindTool, createFindToolDefinition, findTool, findToolDefinition } from "./find.js";
import { createGrepTool, createGrepToolDefinition, grepTool, grepToolDefinition } from "./grep.js";
import { createLsTool, createLsToolDefinition, lsTool, lsToolDefinition } from "./ls.js";
import { createReadTool, createReadToolDefinition, readTool, readToolDefinition, } from "./read.js";
import { createWriteTool, createWriteToolDefinition, writeTool, writeToolDefinition } from "./write.js";
export const codingTools = [readTool, bashTool, editTool, writeTool];
export const readOnlyTools = [readTool, grepTool, findTool, lsTool];
export const allTools = {
    read: readTool,
    bash: bashTool,
    edit: editTool,
    write: writeTool,
    grep: grepTool,
    find: findTool,
    ls: lsTool,
};
export const allToolDefinitions = {
    read: readToolDefinition,
    bash: bashToolDefinition,
    edit: editToolDefinition,
    write: writeToolDefinition,
    grep: grepToolDefinition,
    find: findToolDefinition,
    ls: lsToolDefinition,
};
export function createCodingToolDefinitions(cwd, options) {
    return [
        createReadToolDefinition(cwd, options?.read),
        createBashToolDefinition(cwd, options?.bash),
        createEditToolDefinition(cwd),
        createWriteToolDefinition(cwd),
    ];
}
export function createReadOnlyToolDefinitions(cwd, options) {
    return [
        createReadToolDefinition(cwd, options?.read),
        createGrepToolDefinition(cwd),
        createFindToolDefinition(cwd),
        createLsToolDefinition(cwd),
    ];
}
export function createAllToolDefinitions(cwd, options) {
    return {
        read: createReadToolDefinition(cwd, options?.read),
        bash: createBashToolDefinition(cwd, options?.bash),
        edit: createEditToolDefinition(cwd),
        write: createWriteToolDefinition(cwd),
        grep: createGrepToolDefinition(cwd),
        find: createFindToolDefinition(cwd),
        ls: createLsToolDefinition(cwd),
    };
}
export function createCodingTools(cwd, options) {
    return [
        createReadTool(cwd, options?.read),
        createBashTool(cwd, options?.bash),
        createEditTool(cwd),
        createWriteTool(cwd),
    ];
}
export function createReadOnlyTools(cwd, options) {
    return [createReadTool(cwd, options?.read), createGrepTool(cwd), createFindTool(cwd), createLsTool(cwd)];
}
export function createAllTools(cwd, options) {
    return {
        read: createReadTool(cwd, options?.read),
        bash: createBashTool(cwd, options?.bash),
        edit: createEditTool(cwd),
        write: createWriteTool(cwd),
        grep: createGrepTool(cwd),
        find: createFindTool(cwd),
        ls: createLsTool(cwd),
    };
}
