// @pierre/diffs/dist/components/AdvancedVirtualizer.js
import { DEFAULT_THEMES, DEFAULT_VIRTUAL_FILE_METRICS, DIFFS_TAG_NAME } from "../constants.js";
import { queueRender } from "../managers/UniversalRenderingManager.js";
import { createWindowFromScrollPosition } from "../utils/createWindowFromScrollPosition.js";
import { AdvancedVirtualizedFileDiff } from "./AdvancedVirtualizedFileDiff.js";

//#region src/components/AdvancedVirtualizer.ts
const OVERSCROLL_SIZE = 500;
var AdvancedVirtualizer = class AdvancedVirtualizer {
	static __STOP = false;
	static __lastScrollPosition = 0;
	type = "advanced";
	files = [];
	totalHeightUnified = 0;
	totalHeightSplit = 0;
	rendered = /* @__PURE__ */ new Map();
	containerOffset = 0;
	scrollTop = 0;
	lastRenderedScrollY = -1;
	height = 0;
	scrollHeight = 0;
	initialized = false;
	stickyContainer;
	stickyOffset;
	constructor(container, fileOptions = {
		theme: DEFAULT_THEMES,
		enableLineSelection: true,
		disableVirtualizationBuffers: true,
		diffStyle: "split"
	}, metrics = DEFAULT_VIRTUAL_FILE_METRICS, workerManager) {
		this.container = container;
		this.fileOptions = fileOptions;
		this.metrics = metrics;
		this.workerManager = workerManager;
		this.stickyOffset = document.createElement("div");
		this.stickyOffset.style.contain = "layout size";
		this.stickyContainer = document.createElement("div");
		this.stickyContainer.style.contain = "strict";
		this.stickyContainer.style.position = "sticky";
		this.stickyContainer.style.width = "100%";
		this.stickyContainer.style.contain = "strict";
		this.stickyContainer.style.isolation = "isolate";
		this.container.appendChild(this.stickyOffset);
		this.container.appendChild(this.stickyContainer);
		this.handleScroll();
		this.handleResize();
		this.containerOffset = this.container.getBoundingClientRect().top + this.scrollTop;
		window.__INSTANCE = this;
		window.__TOGGLE = () => {
			if (AdvancedVirtualizer.__STOP) {
				AdvancedVirtualizer.__STOP = false;
				window.scrollTo({ top: AdvancedVirtualizer.__lastScrollPosition });
				queueRender(this._render);
			} else {
				AdvancedVirtualizer.__lastScrollPosition = window.scrollY;
				AdvancedVirtualizer.__STOP = true;
			}
		};
	}
	reset() {
		this.files.length = 0;
		this.totalHeightSplit = 0;
		this.totalHeightUnified = 0;
		for (const [, item] of Array.from(this.rendered)) cleanupRenderedItem(item);
		this.rendered.clear();
		this.stickyContainer.innerHTML = "";
		this.stickyOffset.style.height = "";
		this.initialized = false;
		this.container.style.height = "";
		this.scrollHeight = 0;
		window.removeEventListener("scroll", this.handleScroll);
		window.removeEventListener("resize", this.handleResize);
	}
	addFiles(parsedPatches) {
		for (const patch of parsedPatches) for (const fileDiff of patch.files) {
			const vFileDiff = new AdvancedVirtualizedFileDiff({
				unifiedTop: this.totalHeightUnified,
				splitTop: this.totalHeightSplit,
				fileDiff
			}, this.fileOptions, this.metrics, this.workerManager);
			vFileDiff.cleanUp(true);
			this.files.push(vFileDiff);
			this.totalHeightUnified += vFileDiff.unifiedHeight + this.metrics.fileGap;
			this.totalHeightSplit += vFileDiff.splitHeight + this.metrics.fileGap;
		}
	}
	render() {
		this.setupContainer();
		queueRender(this._render);
	}
	_render = () => {
		if (this.files.length === 0 || AdvancedVirtualizer.__STOP) return;
		const { diffStyle = "split" } = this.fileOptions;
		const { scrollTop, height, scrollHeight, containerOffset } = this;
		const fitPerfectly = this.lastRenderedScrollY === -1 || Math.abs(scrollTop - this.lastRenderedScrollY) > height + OVERSCROLL_SIZE * 2;
		const { top, bottom } = createWindowFromScrollPosition({
			scrollTop,
			height,
			scrollHeight,
			containerOffset,
			fitPerfectly,
			overscrollSize: OVERSCROLL_SIZE
		});
		this.lastRenderedScrollY = scrollTop;
		for (const [renderedInstance, item] of Array.from(this.rendered)) if (!(getInstanceSpecs(renderedInstance, diffStyle).top > top - getInstanceSpecs(renderedInstance, diffStyle).height && getInstanceSpecs(renderedInstance, diffStyle).top <= bottom)) {
			cleanupRenderedItem(item);
			this.rendered.delete(renderedInstance);
		}
		let prevElement;
		let firstInstance;
		let lastInstance;
		for (const instance of this.files) {
			if (getInstanceSpecs(instance, diffStyle).top > bottom) break;
			if (getInstanceSpecs(instance, diffStyle).top < top - getInstanceSpecs(instance, diffStyle).height) continue;
			const rendered = this.rendered.get(instance);
			if (rendered == null) {
				const fileContainer = document.createElement(DIFFS_TAG_NAME);
				if (prevElement == null) this.stickyContainer.prepend(fileContainer);
				else if (prevElement.nextSibling !== fileContainer) prevElement.after(fileContainer);
				instance.virtualizedSetup();
				this.rendered.set(instance, {
					element: fileContainer,
					instance
				});
				instance.virtualizedRender({
					fileContainer,
					renderWindow: {
						top,
						bottom
					}
				});
				prevElement = fileContainer;
			} else {
				prevElement = rendered.element;
				rendered.instance.virtualizedRender({ renderWindow: {
					top,
					bottom
				} });
			}
			firstInstance ??= instance;
			lastInstance = instance;
		}
		if (firstInstance?.renderedRange != null && lastInstance?.renderedRange != null) {
			const firstSpecs = getInstanceSpecs(firstInstance, diffStyle);
			const lastSpecs = getInstanceSpecs(lastInstance, diffStyle);
			const stickyTop = Math.max(Math.min(firstSpecs.top + firstInstance.renderedRange.bufferBefore), 0);
			const lastBuffer = lastInstance.renderedRange.totalLines === 0 ? lastInstance.renderedRange.bufferBefore : lastInstance.renderedRange.bufferAfter;
			const totalHeight = Math.max(0, lastSpecs.top + lastSpecs.height - lastBuffer) - stickyTop;
			this.stickyOffset.style.height = `${stickyTop}px`;
			this.stickyContainer.style.top = `${-totalHeight + height + this.metrics.fileGap}px`;
			this.stickyContainer.style.bottom = `${-totalHeight + height}px`;
			this.stickyContainer.style.height = `${totalHeight}px`;
		}
		if (fitPerfectly) queueRender(this._render);
	};
	setupContainer() {
		const { diffStyle = "split" } = this.fileOptions;
		this.container.style.height = `${diffStyle === "split" ? this.totalHeightSplit : this.totalHeightUnified}px`;
		this.scrollHeight = document.documentElement.scrollHeight;
		if (!this.initialized) {
			window.addEventListener("scroll", this.handleScroll, { passive: true });
			window.addEventListener("resize", this.handleResize);
			this.initialized = true;
		}
	}
	handleScroll = () => {
		let { scrollY: scrollTop } = window;
		scrollTop = Math.max(scrollTop, 0);
		if (this.scrollTop === scrollTop) return;
		this.scrollTop = scrollTop;
		if (this.files.length === 0) return;
		queueRender(this._render);
	};
	handleResize = () => {
		const { innerHeight: height } = window;
		const { scrollHeight } = document.documentElement;
		if (this.height === height && this.scrollHeight === scrollHeight) return;
		this.height = height;
		this.scrollHeight = scrollHeight;
		if (this.files.length > 0) queueRender(this._render);
	};
};
function cleanupRenderedItem(item) {
	item.instance.cleanUp(true);
	item.element.remove();
	item.element.innerHTML = "";
	if (item.element.shadowRoot != null) item.element.shadowRoot.innerHTML = "";
}
function getInstanceSpecs(instance, diffStyle = "split") {
	if (diffStyle === "split") return {
		top: instance.splitTop,
		height: instance.splitHeight
	};
	return {
		top: instance.unifiedTop,
		height: instance.unifiedHeight
	};
}

//#endregion
export { AdvancedVirtualizer };
