// openclaw/dist/extensions/anthropic-vertex/index.js
import { l as buildNativeAnthropicReplayPolicyForModel } from "../../provider-model-shared-BTeN02cH.js";
import { t as definePluginEntry } from "../../plugin-entry-BHB2yf-a.js";
import { i as resolveAnthropicVertexConfigApiKey } from "../../region-DXZdM64_.js";
import { n as resolveImplicitAnthropicVertexProvider, t as mergeImplicitAnthropicVertexProvider } from "../../api-CUbCCO7i.js";
//#region extensions/anthropic-vertex/index.ts
const PROVIDER_ID = "anthropic-vertex";
var anthropic_vertex_default = definePluginEntry({
	id: PROVIDER_ID,
	name: "Anthropic Vertex Provider",
	description: "Bundled Anthropic Vertex provider plugin",
	register(api) {
		api.registerProvider({
			id: PROVIDER_ID,
			label: "Anthropic Vertex",
			docsPath: "/providers/models",
			auth: [],
			catalog: {
				order: "simple",
				run: async (ctx) => {
					const implicit = resolveImplicitAnthropicVertexProvider({ env: ctx.env });
					if (!implicit) return null;
					return { provider: mergeImplicitAnthropicVertexProvider({
						existing: ctx.config.models?.providers?.[PROVIDER_ID],
						implicit
					}) };
				}
			},
			resolveConfigApiKey: ({ env }) => resolveAnthropicVertexConfigApiKey(env),
			buildReplayPolicy: ({ modelId }) => buildNativeAnthropicReplayPolicyForModel(modelId)
		});
	}
});
//#endregion
export { anthropic_vertex_default as default };
