// openclaw/dist/extensions/telegram/test-support.js
import { w as getChatChannelMeta } from "../../message-channel-BzmQarxU.js";
import { n as buildDmGroupAccountAllowlistAdapter } from "../../allowlist-config-edit-C2yMYg69.js";
import "../../channel-plugin-common-Bn4hPVye.js";
import { c as resolveTelegramAccount } from "../../preview-streaming-CyybH9Ui.js";
import { i as telegramConfigAdapter } from "../../shared-f6x8ABfJ.js";
import { t as telegramApprovalCapability } from "../../approval-native-Cc28xvR9.js";
//#region extensions/telegram/test-support.ts
const telegramCommandTestPlugin = {
	id: "telegram",
	meta: getChatChannelMeta("telegram"),
	capabilities: {
		chatTypes: [
			"direct",
			"group",
			"channel",
			"thread"
		],
		reactions: true,
		threads: true,
		media: true,
		polls: true,
		nativeCommands: true,
		blockStreaming: true
	},
	config: telegramConfigAdapter,
	approvalCapability: telegramApprovalCapability,
	pairing: { idLabel: "telegramUserId" },
	allowlist: buildDmGroupAccountAllowlistAdapter({
		channelId: "telegram",
		resolveAccount: resolveTelegramAccount,
		normalize: ({ cfg, accountId, values }) => telegramConfigAdapter.formatAllowFrom({
			cfg,
			accountId,
			allowFrom: values
		}),
		resolveDmAllowFrom: (account) => account.config.allowFrom,
		resolveGroupAllowFrom: (account) => account.config.groupAllowFrom,
		resolveDmPolicy: (account) => account.config.dmPolicy,
		resolveGroupPolicy: (account) => account.config.groupPolicy
	})
};
//#endregion
export { telegramCommandTestPlugin };
