// openshell/dist/bots/stream-edit.js
function sleep(ms) {
    return new Promise((resolve) => {
        setTimeout(resolve, ms);
    });
}
export function createStreamEditController(edit, intervalMs) {
    let pending = null;
    let lastApplied = "";
    let lastEditAt = 0;
    let processing = false;
    async function run() {
        if (processing) {
            return;
        }
        processing = true;
        try {
            while (pending !== null) {
                const text = pending;
                pending = null;
                if (!text || text === lastApplied) {
                    continue;
                }
                const wait = intervalMs - (Date.now() - lastEditAt);
                if (wait > 0) {
                    await sleep(wait);
                }
                await edit(text);
                lastApplied = text;
                lastEditAt = Date.now();
            }
        }
        finally {
            processing = false;
        }
    }
    return {
        queue(text) {
            pending = text;
            void run();
        },
        async flush(text) {
            pending = text;
            await run();
        },
    };
}
export function truncateForPlatform(text, limit, suffix = "\n\n...[truncated]") {
    if (text.length <= limit) {
        return text;
    }
    const head = Math.max(0, limit - suffix.length);
    return `${text.slice(0, head)}${suffix}`;
}
