// openclaw/dist/accounts-D6ejN1bS.js
import { s as normalizeOptionalString } from "./string-coerce-CIT1-v_a.js";
import { n as normalizeAccountId } from "./account-id-BfovDWrj.js";
import { t as resolveAccountEntry } from "./account-lookup-CLbzaC7e.js";
import "./text-runtime-BngW26kR.js";
import { t as createAccountActionGate } from "./account-action-gate-BfpiVhSh.js";
import { s as resolveMergedAccountConfig, t as createAccountListHelpers } from "./account-helpers-C8gd-jmO.js";
import "./routing-CC84LXmn.js";
import { n as resolveDiscordToken } from "./token-_VkCzRE-.js";
//#region extensions/discord/src/accounts.ts
const { listAccountIds, resolveDefaultAccountId } = createAccountListHelpers("discord");
const listDiscordAccountIds = listAccountIds;
const resolveDefaultDiscordAccountId = resolveDefaultAccountId;
function resolveDiscordAccountConfig(cfg, accountId) {
	return resolveAccountEntry(cfg.channels?.discord?.accounts, accountId);
}
function mergeDiscordAccountConfig(cfg, accountId) {
	return resolveMergedAccountConfig({
		channelConfig: cfg.channels?.discord,
		accounts: cfg.channels?.discord?.accounts,
		accountId
	});
}
function createDiscordActionGate(params) {
	const accountId = normalizeAccountId(params.accountId ?? resolveDefaultDiscordAccountId(params.cfg));
	return createAccountActionGate({
		baseActions: params.cfg.channels?.discord?.actions,
		accountActions: resolveDiscordAccountConfig(params.cfg, accountId)?.actions
	});
}
function resolveDiscordAccount(params) {
	const accountId = normalizeAccountId(params.accountId ?? resolveDefaultDiscordAccountId(params.cfg));
	const baseEnabled = params.cfg.channels?.discord?.enabled !== false;
	const merged = mergeDiscordAccountConfig(params.cfg, accountId);
	const accountEnabled = merged.enabled !== false;
	const enabled = baseEnabled && accountEnabled;
	const tokenResolution = resolveDiscordToken(params.cfg, { accountId });
	return {
		accountId,
		enabled,
		name: normalizeOptionalString(merged.name),
		token: tokenResolution.token,
		tokenSource: tokenResolution.source,
		config: merged
	};
}
function resolveDiscordMaxLinesPerMessage(params) {
	if (typeof params.discordConfig?.maxLinesPerMessage === "number") return params.discordConfig.maxLinesPerMessage;
	return resolveDiscordAccount({
		cfg: params.cfg,
		accountId: params.accountId
	}).config.maxLinesPerMessage;
}
function listEnabledDiscordAccounts(cfg) {
	return listDiscordAccountIds(cfg).map((accountId) => resolveDiscordAccount({
		cfg,
		accountId
	})).filter((account) => account.enabled);
}
//#endregion
export { resolveDefaultDiscordAccountId as a, resolveDiscordMaxLinesPerMessage as c, mergeDiscordAccountConfig as i, listDiscordAccountIds as n, resolveDiscordAccount as o, listEnabledDiscordAccounts as r, resolveDiscordAccountConfig as s, createDiscordActionGate as t };
