// openclaw/dist/action-spawn-DNsZAtA6.js
import { s as normalizeOptionalString } from "./string-coerce-CIT1-v_a.js";
import { P as spawnSubagentDirect } from "./pi-embedded-Vw-lS5ti.js";
import { p as stopWithText } from "./shared-DVPQDU12.js";
//#region src/auto-reply/reply/commands-subagents/action-spawn.ts
async function handleSubagentsSpawnAction(ctx) {
	const { params, requesterKey, restTokens } = ctx;
	const agentId = restTokens[0];
	const taskParts = [];
	let model;
	let thinking;
	for (let i = 1; i < restTokens.length; i++) if (restTokens[i] === "--model" && i + 1 < restTokens.length) {
		i += 1;
		model = restTokens[i];
	} else if (restTokens[i] === "--thinking" && i + 1 < restTokens.length) {
		i += 1;
		thinking = restTokens[i];
	} else taskParts.push(restTokens[i]);
	const task = taskParts.join(" ").trim();
	if (!agentId || !task) return stopWithText("Usage: /subagents spawn <agentId> <task> [--model <model>] [--thinking <level>]");
	const commandTo = normalizeOptionalString(params.command.to) ?? "";
	const originatingTo = normalizeOptionalString(params.ctx.OriginatingTo) ?? "";
	const fallbackTo = normalizeOptionalString(params.ctx.To) ?? "";
	const normalizedTo = originatingTo || commandTo || fallbackTo || void 0;
	const result = await spawnSubagentDirect({
		task,
		agentId,
		model,
		thinking,
		mode: "run",
		cleanup: "keep",
		expectsCompletionMessage: true
	}, {
		agentSessionKey: requesterKey,
		agentChannel: params.ctx.OriginatingChannel ?? params.command.channel,
		agentAccountId: params.ctx.AccountId,
		agentTo: normalizedTo,
		agentThreadId: params.ctx.MessageThreadId,
		agentGroupId: params.sessionEntry?.groupId ?? null,
		agentGroupChannel: params.sessionEntry?.groupChannel ?? null,
		agentGroupSpace: params.sessionEntry?.space ?? null
	});
	if (result.status === "accepted") return stopWithText(`Spawned subagent ${agentId} (session ${result.childSessionKey}, run ${result.runId?.slice(0, 8)}).`);
	return stopWithText(`Spawn failed: ${result.error ?? result.status}`);
}
//#endregion
export { handleSubagentsSpawnAction };
