// openclaw/dist/auth-choice-Dic171kf.js
import { n as ensureAuthProfileStore } from "./store-DlwTtWHa.js";
import "./auth-profiles-BPRQRiAA.js";
import { f as resolveDefaultModelForAgent } from "./model-selection-FjGH2z4R.js";
import { n as listProfilesForProvider } from "./profiles-B8tnoQUl.js";
import { t as resolveEnvApiKey } from "./model-auth-env-DQOHJ5-D.js";
import { r as loadModelCatalog } from "./model-catalog-D2kYsUen.js";
import { o as hasUsableCustomProviderApiKey } from "./model-auth-RU1Lwgn8.js";
import { t as buildProviderAuthRecoveryHint } from "./provider-auth-guidance-CL_N-SXG.js";
import { r as normalizeLegacyOnboardAuthChoice } from "./auth-choice-legacy-JypF_SN4.js";
import { t as applyAuthChoiceLoadedPluginProvider } from "./provider-auth-choice-B5qhDMJ4.js";
import { n as applyAuthChoiceApiProviders, r as normalizeApiKeyTokenProviderAuthChoice } from "./provider-auth-choice-preference-CukioL9v.js";
//#region src/commands/auth-choice.apply.oauth.ts
async function applyAuthChoiceOAuth(_params) {
	return null;
}
//#endregion
//#region src/commands/auth-choice.apply.ts
async function applyAuthChoice(params) {
	const normalizedProviderAuthChoice = normalizeApiKeyTokenProviderAuthChoice({
		authChoice: normalizeLegacyOnboardAuthChoice(params.authChoice, {
			config: params.config,
			env: params.env
		}) ?? params.authChoice,
		tokenProvider: params.opts?.tokenProvider,
		config: params.config,
		env: params.env
	});
	const normalizedParams = normalizedProviderAuthChoice === params.authChoice ? params : {
		...params,
		authChoice: normalizedProviderAuthChoice
	};
	const handlers = [
		applyAuthChoiceLoadedPluginProvider,
		applyAuthChoiceOAuth,
		applyAuthChoiceApiProviders
	];
	for (const handler of handlers) {
		const result = await handler(normalizedParams);
		if (result) return result;
	}
	if (normalizedParams.authChoice === "token" || normalizedParams.authChoice === "setup-token") throw new Error([`Auth choice "${normalizedParams.authChoice}" was not matched to a provider setup flow.`, "For Anthropic legacy token auth, use \"setup-token\" with tokenProvider=\"anthropic\" or choose the Anthropic setup-token entry explicitly."].join("\n"));
	if (normalizedParams.authChoice === "oauth") throw new Error("Auth choice \"oauth\" is no longer supported directly. Use \"setup-token\" for Anthropic legacy token auth or a provider-specific OAuth entry.");
	return { config: normalizedParams.config };
}
//#endregion
//#region src/commands/auth-choice.model-check.ts
async function warnIfModelConfigLooksOff(config, prompter, options) {
	const ref = resolveDefaultModelForAgent({
		cfg: config,
		agentId: options?.agentId
	});
	const warnings = [];
	const catalog = await loadModelCatalog({
		config,
		useCache: false
	});
	if (catalog.length > 0) {
		if (!catalog.some((entry) => entry.provider === ref.provider && entry.id === ref.model)) warnings.push(`Model not found: ${ref.provider}/${ref.model}. Update agents.defaults.model or run /models list.`);
	}
	const hasProfile = listProfilesForProvider(ensureAuthProfileStore(options?.agentDir), ref.provider).length > 0;
	const envKey = resolveEnvApiKey(ref.provider);
	const hasCustomKey = hasUsableCustomProviderApiKey(config, ref.provider);
	if (!hasProfile && !envKey && !hasCustomKey) warnings.push(`No auth configured for provider "${ref.provider}". The agent may fail until credentials are added. ${buildProviderAuthRecoveryHint({
		provider: ref.provider,
		config,
		includeEnvVar: true
	})}`);
	if (warnings.length > 0) await prompter.note(warnings.join("\n"), "Model check");
}
//#endregion
export { applyAuthChoice as n, warnIfModelConfigLooksOff as t };
