// openclaw/dist/bootstrap-budget-BZw4htx0.js
import { s as normalizeOptionalString } from "./string-coerce-CIT1-v_a.js";
import { c as normalizeAgentId, u as resolveAgentIdFromSessionKey } from "./session-key-BMb3Kc4r.js";
import { i as resolveAgentConfig, m as resolveDefaultAgentId, n as listAgentEntries, y as resolveSessionAgentIds } from "./agent-scope-BoZcgjP9.js";
import { f as filterBootstrapFilesForSession, h as loadWorkspaceBootstrapFiles, i as DEFAULT_HEARTBEAT_FILENAME } from "./workspace-4tKa-1bN.js";
import { n as createInternalHookEvent, p as triggerInternalHook } from "./internal-hooks-B5x-V_Su.js";
import { t as parseDurationMs } from "./parse-duration-Da8MNuvD.js";
import { o as resolveHeartbeatPrompt } from "./heartbeat-fvllvegq.js";
import { r as getOrLoadBootstrapFiles } from "./bootstrap-cache-B-BD03V5.js";
import { g as resolveBootstrapMaxChars, m as buildBootstrapContextFiles, v as resolveBootstrapTotalMaxChars } from "./pi-embedded-helpers-7aPSIElL.js";
import path from "node:path";
import fs from "node:fs/promises";
//#region src/agents/bootstrap-hooks.ts
async function applyBootstrapHookOverrides(params) {
	const sessionKey = params.sessionKey ?? params.sessionId ?? "unknown";
	const agentId = params.agentId ?? (params.sessionKey ? resolveAgentIdFromSessionKey(params.sessionKey) : void 0);
	const event = createInternalHookEvent("agent", "bootstrap", sessionKey, {
		workspaceDir: params.workspaceDir,
		bootstrapFiles: params.files,
		cfg: params.config,
		sessionKey: params.sessionKey,
		sessionId: params.sessionId,
		agentId
	});
	await triggerInternalHook(event);
	const updated = event.context.bootstrapFiles;
	return Array.isArray(updated) ? updated : params.files;
}
//#endregion
//#region src/agents/heartbeat-system-prompt.ts
function resolveHeartbeatConfigForSystemPrompt(config, agentId) {
	const defaults = config?.agents?.defaults?.heartbeat;
	if (!config || !agentId) return defaults;
	const overrides = resolveAgentConfig(config, agentId)?.heartbeat;
	if (!defaults && !overrides) return overrides;
	return {
		...defaults,
		...overrides
	};
}
function isHeartbeatEnabledByAgentPolicy(config, agentId) {
	const resolvedAgentId = normalizeAgentId(agentId);
	const agents = listAgentEntries(config);
	if (agents.some((entry) => Boolean(entry?.heartbeat))) return agents.some((entry) => Boolean(entry?.heartbeat) && normalizeAgentId(entry.id) === resolvedAgentId);
	return resolvedAgentId === resolveDefaultAgentId(config);
}
function isHeartbeatCadenceEnabled(heartbeat) {
	const trimmedEvery = normalizeOptionalString(heartbeat?.every ?? "30m") ?? "";
	if (!trimmedEvery) return false;
	try {
		return parseDurationMs(trimmedEvery, { defaultUnit: "m" }) > 0;
	} catch {
		return false;
	}
}
function shouldIncludeHeartbeatGuidanceForSystemPrompt(params) {
	const defaultAgentId = params.defaultAgentId ?? resolveDefaultAgentId(params.config ?? {});
	const agentId = params.agentId ?? defaultAgentId;
	if (!agentId || normalizeAgentId(agentId) !== normalizeAgentId(defaultAgentId)) return false;
	if (params.config && !isHeartbeatEnabledByAgentPolicy(params.config, agentId)) return false;
	const heartbeat = resolveHeartbeatConfigForSystemPrompt(params.config, agentId);
	if (heartbeat?.includeSystemPromptSection === false) return false;
	return isHeartbeatCadenceEnabled(heartbeat);
}
function resolveHeartbeatPromptForSystemPrompt(params) {
	const agentId = params.agentId ?? params.defaultAgentId ?? resolveDefaultAgentId(params.config ?? {});
	const heartbeat = resolveHeartbeatConfigForSystemPrompt(params.config, agentId);
	if (!shouldIncludeHeartbeatGuidanceForSystemPrompt(params)) return;
	return resolveHeartbeatPrompt(heartbeat?.prompt);
}
//#endregion
//#region src/agents/bootstrap-files.ts
const CONTINUATION_SCAN_MAX_TAIL_BYTES = 256 * 1024;
const CONTINUATION_SCAN_MAX_RECORDS = 500;
const FULL_BOOTSTRAP_COMPLETED_CUSTOM_TYPE = "openclaw:bootstrap-context:full";
function resolveContextInjectionMode(config) {
	return config?.agents?.defaults?.contextInjection ?? "always";
}
async function hasCompletedBootstrapTurn(sessionFile) {
	try {
		const stat = await fs.lstat(sessionFile);
		if (stat.isSymbolicLink()) return false;
		const fh = await fs.open(sessionFile, "r");
		try {
			const bytesToRead = Math.min(stat.size, CONTINUATION_SCAN_MAX_TAIL_BYTES);
			if (bytesToRead <= 0) return false;
			const start = stat.size - bytesToRead;
			const buffer = Buffer.allocUnsafe(bytesToRead);
			const { bytesRead } = await fh.read(buffer, 0, bytesToRead, start);
			let text = buffer.toString("utf-8", 0, bytesRead);
			if (start > 0) {
				const firstNewline = text.indexOf("\n");
				if (firstNewline === -1) return false;
				text = text.slice(firstNewline + 1);
			}
			const records = text.split(/\r?\n/u).filter((line) => line.trim().length > 0).slice(-CONTINUATION_SCAN_MAX_RECORDS);
			let compactedAfterLatestAssistant = false;
			for (let i = records.length - 1; i >= 0; i--) {
				const line = records[i];
				if (!line) continue;
				let entry;
				try {
					entry = JSON.parse(line);
				} catch {
					continue;
				}
				const record = entry;
				if (record?.type === "compaction") {
					compactedAfterLatestAssistant = true;
					continue;
				}
				if (record?.type === "custom" && record.customType === "openclaw:bootstrap-context:full") return !compactedAfterLatestAssistant;
			}
			return false;
		} finally {
			await fh.close();
		}
	} catch {
		return false;
	}
}
function makeBootstrapWarn(params) {
	if (!params.warn) return;
	return (message) => params.warn?.(`${message} (sessionKey=${params.sessionLabel})`);
}
function sanitizeBootstrapFiles(files, warn) {
	const sanitized = [];
	for (const file of files) {
		const pathValue = normalizeOptionalString(file.path) ?? "";
		if (!pathValue) {
			warn?.(`skipping bootstrap file "${file.name}" — missing or invalid "path" field (hook may have used "filePath" instead)`);
			continue;
		}
		sanitized.push({
			...file,
			path: pathValue
		});
	}
	return sanitized;
}
function applyContextModeFilter(params) {
	const contextMode = params.contextMode ?? "full";
	const runKind = params.runKind ?? "default";
	if (contextMode !== "lightweight") return params.files;
	if (runKind === "heartbeat") return params.files.filter((file) => file.name === "HEARTBEAT.md");
	return [];
}
function shouldExcludeHeartbeatBootstrapFile(params) {
	if (!params.config || params.runKind === "heartbeat") return false;
	const { defaultAgentId, sessionAgentId } = resolveSessionAgentIds({
		sessionKey: params.sessionKey ?? params.sessionId,
		config: params.config,
		agentId: params.agentId
	});
	if (sessionAgentId !== defaultAgentId) return false;
	return !shouldIncludeHeartbeatGuidanceForSystemPrompt({
		config: params.config,
		agentId: sessionAgentId,
		defaultAgentId
	});
}
function filterHeartbeatBootstrapFile(files, excludeHeartbeatBootstrapFile) {
	if (!excludeHeartbeatBootstrapFile) return files;
	return files.filter((file) => file.name !== DEFAULT_HEARTBEAT_FILENAME);
}
async function resolveBootstrapFilesForRun(params) {
	const excludeHeartbeatBootstrapFile = shouldExcludeHeartbeatBootstrapFile(params);
	const sessionKey = params.sessionKey ?? params.sessionId;
	return sanitizeBootstrapFiles(filterHeartbeatBootstrapFile(await applyBootstrapHookOverrides({
		files: applyContextModeFilter({
			files: filterBootstrapFilesForSession(params.sessionKey ? await getOrLoadBootstrapFiles({
				workspaceDir: params.workspaceDir,
				sessionKey: params.sessionKey
			}) : await loadWorkspaceBootstrapFiles(params.workspaceDir), sessionKey),
			contextMode: params.contextMode,
			runKind: params.runKind
		}),
		workspaceDir: params.workspaceDir,
		config: params.config,
		sessionKey: params.sessionKey,
		sessionId: params.sessionId,
		agentId: params.agentId
	}), excludeHeartbeatBootstrapFile), params.warn);
}
async function resolveBootstrapContextForRun(params) {
	const bootstrapFiles = await resolveBootstrapFilesForRun(params);
	return {
		bootstrapFiles,
		contextFiles: buildBootstrapContextFiles(bootstrapFiles, {
			maxChars: resolveBootstrapMaxChars(params.config),
			totalMaxChars: resolveBootstrapTotalMaxChars(params.config),
			warn: params.warn
		})
	};
}
//#endregion
//#region src/agents/bootstrap-budget.ts
const DEFAULT_BOOTSTRAP_NEAR_LIMIT_RATIO = .85;
function normalizePositiveLimit(value) {
	if (!Number.isFinite(value) || value <= 0) return 1;
	return Math.floor(value);
}
function formatWarningCause(cause) {
	return cause === "per-file-limit" ? "max/file" : "max/total";
}
function normalizeSeenSignatures(signatures) {
	if (!Array.isArray(signatures) || signatures.length === 0) return [];
	const seen = /* @__PURE__ */ new Set();
	const result = [];
	for (const signature of signatures) {
		const value = normalizeOptionalString(signature) ?? "";
		if (!value || seen.has(value)) continue;
		seen.add(value);
		result.push(value);
	}
	return result;
}
function appendSeenSignature(signatures, signature) {
	if (!signature.trim()) return signatures;
	if (signatures.includes(signature)) return signatures;
	const next = [...signatures, signature];
	if (next.length <= 32) return next;
	return next.slice(-32);
}
function resolveBootstrapWarningSignaturesSeen(report) {
	const truncation = report?.bootstrapTruncation;
	const seenFromReport = normalizeSeenSignatures(truncation?.warningSignaturesSeen);
	if (seenFromReport.length > 0) return seenFromReport;
	if (truncation?.warningMode === "off") return [];
	const single = typeof truncation?.promptWarningSignature === "string" ? normalizeOptionalString(truncation.promptWarningSignature) ?? "" : "";
	return single ? [single] : [];
}
function buildBootstrapInjectionStats(params) {
	const injectedByPath = /* @__PURE__ */ new Map();
	const injectedByBaseName = /* @__PURE__ */ new Map();
	for (const file of params.injectedFiles) {
		const pathValue = normalizeOptionalString(file.path) ?? "";
		if (!pathValue) continue;
		if (!injectedByPath.has(pathValue)) injectedByPath.set(pathValue, file.content);
		const normalizedPath = pathValue.replace(/\\/g, "/");
		const baseName = path.posix.basename(normalizedPath);
		if (!injectedByBaseName.has(baseName)) injectedByBaseName.set(baseName, file.content);
	}
	return params.bootstrapFiles.map((file) => {
		const pathValue = normalizeOptionalString(file.path) ?? "";
		const rawChars = file.missing ? 0 : (file.content ?? "").trimEnd().length;
		const injected = (pathValue ? injectedByPath.get(pathValue) : void 0) ?? injectedByPath.get(file.name) ?? injectedByBaseName.get(file.name);
		const injectedChars = injected ? injected.length : 0;
		const truncated = !file.missing && injectedChars < rawChars;
		return {
			name: file.name,
			path: pathValue || file.name,
			missing: file.missing,
			rawChars,
			injectedChars,
			truncated
		};
	});
}
function analyzeBootstrapBudget(params) {
	const bootstrapMaxChars = normalizePositiveLimit(params.bootstrapMaxChars);
	const bootstrapTotalMaxChars = normalizePositiveLimit(params.bootstrapTotalMaxChars);
	const nearLimitRatio = typeof params.nearLimitRatio === "number" && Number.isFinite(params.nearLimitRatio) && params.nearLimitRatio > 0 && params.nearLimitRatio < 1 ? params.nearLimitRatio : DEFAULT_BOOTSTRAP_NEAR_LIMIT_RATIO;
	const nonMissing = params.files.filter((file) => !file.missing);
	const rawChars = nonMissing.reduce((sum, file) => sum + file.rawChars, 0);
	const injectedChars = nonMissing.reduce((sum, file) => sum + file.injectedChars, 0);
	const totalNearLimit = injectedChars >= Math.ceil(bootstrapTotalMaxChars * nearLimitRatio);
	const totalOverLimit = injectedChars >= bootstrapTotalMaxChars;
	const files = params.files.map((file) => {
		if (file.missing) return {
			...file,
			nearLimit: false,
			causes: []
		};
		const perFileOverLimit = file.rawChars > bootstrapMaxChars;
		const nearLimit = file.rawChars >= Math.ceil(bootstrapMaxChars * nearLimitRatio);
		const causes = [];
		if (file.truncated) {
			if (perFileOverLimit) causes.push("per-file-limit");
			if (totalOverLimit) causes.push("total-limit");
		}
		return {
			...file,
			nearLimit,
			causes
		};
	});
	const truncatedFiles = files.filter((file) => file.truncated);
	return {
		files,
		truncatedFiles,
		nearLimitFiles: files.filter((file) => file.nearLimit),
		totalNearLimit,
		hasTruncation: truncatedFiles.length > 0,
		totals: {
			rawChars,
			injectedChars,
			truncatedChars: Math.max(0, rawChars - injectedChars),
			bootstrapMaxChars,
			bootstrapTotalMaxChars,
			nearLimitRatio
		}
	};
}
function buildBootstrapTruncationSignature(analysis) {
	if (!analysis.hasTruncation) return;
	const files = analysis.truncatedFiles.map((file) => ({
		path: file.path || file.name,
		rawChars: file.rawChars,
		injectedChars: file.injectedChars,
		causes: [...file.causes].toSorted()
	})).toSorted((a, b) => {
		const pathCmp = a.path.localeCompare(b.path);
		if (pathCmp !== 0) return pathCmp;
		if (a.rawChars !== b.rawChars) return a.rawChars - b.rawChars;
		if (a.injectedChars !== b.injectedChars) return a.injectedChars - b.injectedChars;
		return a.causes.join("+").localeCompare(b.causes.join("+"));
	});
	return JSON.stringify({
		bootstrapMaxChars: analysis.totals.bootstrapMaxChars,
		bootstrapTotalMaxChars: analysis.totals.bootstrapTotalMaxChars,
		files
	});
}
function formatBootstrapTruncationWarningLines(params) {
	if (!params.analysis.hasTruncation) return [];
	const maxFiles = typeof params.maxFiles === "number" && Number.isFinite(params.maxFiles) && params.maxFiles > 0 ? Math.floor(params.maxFiles) : 3;
	const lines = [];
	const duplicateNameCounts = params.analysis.truncatedFiles.reduce((acc, file) => {
		acc.set(file.name, (acc.get(file.name) ?? 0) + 1);
		return acc;
	}, /* @__PURE__ */ new Map());
	const topFiles = params.analysis.truncatedFiles.slice(0, maxFiles);
	for (const file of topFiles) {
		const pct = file.rawChars > 0 ? Math.round((file.rawChars - file.injectedChars) / file.rawChars * 100) : 0;
		const causeText = file.causes.length > 0 ? file.causes.map((cause) => formatWarningCause(cause)).join(", ") : "";
		const nameLabel = (duplicateNameCounts.get(file.name) ?? 0) > 1 && file.path.trim().length > 0 ? `${file.name} (${file.path})` : file.name;
		lines.push(`${nameLabel}: ${file.rawChars} raw -> ${file.injectedChars} injected (~${Math.max(0, pct)}% removed${causeText ? `; ${causeText}` : ""}).`);
	}
	if (params.analysis.truncatedFiles.length > topFiles.length) lines.push(`+${params.analysis.truncatedFiles.length - topFiles.length} more truncated file(s).`);
	lines.push("If unintentional, raise agents.defaults.bootstrapMaxChars and/or agents.defaults.bootstrapTotalMaxChars.");
	return lines;
}
function buildBootstrapPromptWarning(params) {
	const signature = buildBootstrapTruncationSignature(params.analysis);
	let seenSignatures = normalizeSeenSignatures(params.seenSignatures);
	if (params.previousSignature && !seenSignatures.includes(params.previousSignature)) seenSignatures = appendSeenSignature(seenSignatures, params.previousSignature);
	const hasSeenSignature = Boolean(signature && seenSignatures.includes(signature));
	const warningShown = params.mode !== "off" && Boolean(signature) && (params.mode === "always" || !hasSeenSignature);
	const warningSignaturesSeen = signature && params.mode !== "off" ? appendSeenSignature(seenSignatures, signature) : seenSignatures;
	return {
		signature,
		warningShown,
		lines: warningShown ? formatBootstrapTruncationWarningLines({
			analysis: params.analysis,
			maxFiles: params.maxFiles
		}) : [],
		warningSignaturesSeen
	};
}
function appendBootstrapPromptWarning(prompt, warningLines, options) {
	const normalizedLines = (warningLines ?? []).map((line) => line.trim()).filter(Boolean);
	if (normalizedLines.length === 0) return prompt;
	if (options?.preserveExactPrompt && prompt === options.preserveExactPrompt) return prompt;
	const warningBlock = [
		"[Bootstrap truncation warning]",
		"Some workspace bootstrap files were truncated before injection.",
		"Treat Project Context as partial and read the relevant files directly if details seem missing.",
		...normalizedLines.map((line) => `- ${line}`)
	].join("\n");
	return prompt ? `${prompt}\n\n${warningBlock}` : warningBlock;
}
const prependBootstrapPromptWarning = appendBootstrapPromptWarning;
function buildBootstrapTruncationReportMeta(params) {
	return {
		warningMode: params.warningMode,
		warningShown: params.warning.warningShown,
		promptWarningSignature: params.warning.signature,
		...params.warning.warningSignaturesSeen.length > 0 ? { warningSignaturesSeen: params.warning.warningSignaturesSeen } : {},
		truncatedFiles: params.analysis.truncatedFiles.length,
		nearLimitFiles: params.analysis.nearLimitFiles.length,
		totalNearLimit: params.analysis.totalNearLimit
	};
}
//#endregion
export { prependBootstrapPromptWarning as a, hasCompletedBootstrapTurn as c, resolveContextInjectionMode as d, resolveHeartbeatPromptForSystemPrompt as f, buildBootstrapTruncationReportMeta as i, makeBootstrapWarn as l, buildBootstrapInjectionStats as n, resolveBootstrapWarningSignaturesSeen as o, buildBootstrapPromptWarning as r, FULL_BOOTSTRAP_COMPLETED_CUSTOM_TYPE as s, analyzeBootstrapBudget as t, resolveBootstrapContextForRun as u };
