// openclaw/dist/clawhub-BFjxm1oA.js
import { r as formatErrorMessage } from "./errors-CGMew-1w.js";
import { s as normalizeOptionalString } from "./string-coerce-CIT1-v_a.js";
import { o as resolveCompatibilityHostVersion } from "./version-BnlvIG4z.js";
import { n as installPluginFromArchive } from "./install-OFjzq7oO.js";
import { i as DEFAULT_MAX_EXTRACTED_BYTES, r as DEFAULT_MAX_ENTRY_BYTES } from "./archive-Btdrfdmr.js";
import { a as fetchClawHubPackageVersion, c as normalizeClawHubSha256Integrity, d as resolveLatestVersionFromPackage, f as satisfiesGatewayMinimum, i as fetchClawHubPackageDetail, l as parseClawHubPluginSpec, n as downloadClawHubPackageArchive, p as satisfiesPluginApiRange, s as normalizeClawHubSha256Hex, t as ClawHubRequestError } from "./clawhub-DExQ87KK.js";
import fs from "node:fs/promises";
import { createHash } from "node:crypto";
import JSZip from "jszip";
//#region src/plugins/clawhub.ts
const CLAWHUB_INSTALL_ERROR_CODE = {
	INVALID_SPEC: "invalid_spec",
	PACKAGE_NOT_FOUND: "package_not_found",
	VERSION_NOT_FOUND: "version_not_found",
	NO_INSTALLABLE_VERSION: "no_installable_version",
	SKILL_PACKAGE: "skill_package",
	UNSUPPORTED_FAMILY: "unsupported_family",
	PRIVATE_PACKAGE: "private_package",
	INCOMPATIBLE_PLUGIN_API: "incompatible_plugin_api",
	INCOMPATIBLE_GATEWAY: "incompatible_gateway",
	MISSING_ARCHIVE_INTEGRITY: "missing_archive_integrity",
	ARCHIVE_INTEGRITY_MISMATCH: "archive_integrity_mismatch"
};
const CLAWHUB_GENERATED_ARCHIVE_METADATA_FILE = "_meta.json";
function formatClawHubSpecifier(params) {
	return `clawhub:${params.name}${params.version ? `@${params.version}` : ""}`;
}
function buildClawHubInstallFailure(error, code) {
	return {
		ok: false,
		error,
		code
	};
}
function isClawHubInstallFailure(value) {
	return Boolean(value && typeof value === "object" && "ok" in value && value.ok === false && "error" in value);
}
function mapClawHubRequestError(error, context) {
	if (error instanceof ClawHubRequestError && error.status === 404) {
		if (context.stage === "package") return buildClawHubInstallFailure("Package not found on ClawHub.", CLAWHUB_INSTALL_ERROR_CODE.PACKAGE_NOT_FOUND);
		return buildClawHubInstallFailure(`Version not found on ClawHub: ${context.name}@${context.version ?? "unknown"}.`, CLAWHUB_INSTALL_ERROR_CODE.VERSION_NOT_FOUND);
	}
	return buildClawHubInstallFailure(formatErrorMessage(error));
}
function resolveRequestedVersion(params) {
	if (params.requestedVersion) return params.requestedVersion;
	return resolveLatestVersionFromPackage(params.detail);
}
function readTrimmedString(value) {
	return normalizeOptionalString(value) ?? null;
}
function normalizeClawHubRelativePath(value) {
	if (typeof value !== "string" || value.length === 0) return null;
	if (value.trim() !== value || value.includes("\\")) return null;
	if (value.startsWith("/")) return null;
	if (value.split("/").some((segment) => segment.length === 0 || segment === "." || segment === "..")) return null;
	return value;
}
function describeInvalidClawHubRelativePath(value) {
	if (typeof value !== "string") return `non-string value of type ${typeof value}`;
	if (value.length === 0) return "empty string";
	if (value.trim() !== value) return `path "${value}" has leading or trailing whitespace`;
	if (value.includes("\\")) return `path "${value}" contains backslashes`;
	if (value.startsWith("/")) return `path "${value}" is absolute`;
	const segments = value.split("/");
	if (segments.some((segment) => segment.length === 0)) return `path "${value}" contains an empty segment`;
	if (segments.some((segment) => segment === "." || segment === "..")) return `path "${value}" contains dot segments`;
	return `path "${value}" failed validation for an unknown reason`;
}
function describeInvalidClawHubSha256(value) {
	if (typeof value !== "string") return `non-string value of type ${typeof value}`;
	if (value.length === 0) return "empty string";
	if (value.trim().length === 0) return "whitespace-only string";
	return `value "${value}" is not a 64-character hexadecimal SHA-256 digest`;
}
function resolveClawHubArchiveVerification(versionDetail, packageName, version) {
	const sha256hashValue = versionDetail.version?.sha256hash;
	const sha256hash = readTrimmedString(sha256hashValue);
	const integrity = sha256hash ? normalizeClawHubSha256Integrity(sha256hash) : null;
	if (integrity) return {
		ok: true,
		verification: {
			kind: "archive-integrity",
			integrity
		}
	};
	if (sha256hashValue !== void 0 && sha256hashValue !== null) return buildClawHubInstallFailure(`ClawHub version metadata for "${packageName}@${version}" has an invalid sha256hash (${typeof sha256hashValue === "string" && sha256hashValue.trim().length === 0 ? "empty string" : typeof sha256hashValue === "string" ? `unrecognized value "${sha256hashValue.trim()}"` : `non-string value of type ${typeof sha256hashValue}`}).`, CLAWHUB_INSTALL_ERROR_CODE.MISSING_ARCHIVE_INTEGRITY);
	const files = versionDetail.version?.files;
	if (!Array.isArray(files) || files.length === 0) return {
		ok: true,
		verification: null
	};
	const normalizedFiles = [];
	const seenPaths = /* @__PURE__ */ new Set();
	for (const [index, file] of files.entries()) {
		if (!file || typeof file !== "object") return buildClawHubInstallFailure(`ClawHub version metadata for "${packageName}@${version}" has an invalid files[${index}] entry (expected an object, got ${file === null ? "null" : typeof file}).`, CLAWHUB_INSTALL_ERROR_CODE.MISSING_ARCHIVE_INTEGRITY);
		const fileRecord = file;
		const filePath = normalizeClawHubRelativePath(fileRecord.path);
		const sha256Value = readTrimmedString(fileRecord.sha256);
		const sha256 = sha256Value ? normalizeClawHubSha256Hex(sha256Value) : null;
		if (!filePath) return buildClawHubInstallFailure(`ClawHub version metadata for "${packageName}@${version}" has an invalid files[${index}].path (${describeInvalidClawHubRelativePath(fileRecord.path)}).`, CLAWHUB_INSTALL_ERROR_CODE.MISSING_ARCHIVE_INTEGRITY);
		if (filePath === CLAWHUB_GENERATED_ARCHIVE_METADATA_FILE) return buildClawHubInstallFailure(`ClawHub version metadata for "${packageName}@${version}" must not include generated file "${filePath}" in files[].`, CLAWHUB_INSTALL_ERROR_CODE.MISSING_ARCHIVE_INTEGRITY);
		if (!sha256) return buildClawHubInstallFailure(`ClawHub version metadata for "${packageName}@${version}" has an invalid files[${index}].sha256 (${describeInvalidClawHubSha256(fileRecord.sha256)}).`, CLAWHUB_INSTALL_ERROR_CODE.MISSING_ARCHIVE_INTEGRITY);
		if (seenPaths.has(filePath)) return buildClawHubInstallFailure(`ClawHub version metadata for "${packageName}@${version}" has duplicate files[] path "${filePath}".`, CLAWHUB_INSTALL_ERROR_CODE.MISSING_ARCHIVE_INTEGRITY);
		seenPaths.add(filePath);
		normalizedFiles.push({
			path: filePath,
			sha256
		});
	}
	return {
		ok: true,
		verification: {
			kind: "file-list",
			files: normalizedFiles
		}
	};
}
async function readLimitedClawHubArchiveEntry(entry, limits, handlers) {
	const hintedSize = entry._data?.uncompressedSize;
	if (typeof hintedSize === "number" && Number.isFinite(hintedSize) && hintedSize > limits.maxEntryBytes) return buildClawHubInstallFailure(`ClawHub archive fallback verification rejected "${entry.name}" because it exceeds the per-file size limit.`, CLAWHUB_INSTALL_ERROR_CODE.ARCHIVE_INTEGRITY_MISMATCH);
	let entryBytes = 0;
	return await new Promise((resolve) => {
		let settled = false;
		const stream = entry.nodeStream("nodebuffer");
		stream.on("data", (chunk) => {
			if (settled) return;
			const buffer = typeof chunk === "string" ? Buffer.from(chunk) : Buffer.from(chunk);
			entryBytes += buffer.byteLength;
			if (entryBytes > limits.maxEntryBytes) {
				settled = true;
				stream.destroy?.();
				resolve(buildClawHubInstallFailure(`ClawHub archive fallback verification rejected "${entry.name}" because it exceeds the per-file size limit.`, CLAWHUB_INSTALL_ERROR_CODE.ARCHIVE_INTEGRITY_MISMATCH));
				return;
			}
			if (!limits.addArchiveBytes(buffer.byteLength)) {
				settled = true;
				stream.destroy?.();
				resolve(buildClawHubInstallFailure("ClawHub archive fallback verification exceeded the total extracted-size limit.", CLAWHUB_INSTALL_ERROR_CODE.ARCHIVE_INTEGRITY_MISMATCH));
				return;
			}
			handlers.onChunk(buffer);
		});
		stream.once("end", () => {
			if (settled) return;
			settled = true;
			resolve(handlers.onEnd());
		});
		stream.once("error", (error) => {
			if (settled) return;
			settled = true;
			resolve(buildClawHubInstallFailure(error instanceof Error ? error.message : String(error), CLAWHUB_INSTALL_ERROR_CODE.ARCHIVE_INTEGRITY_MISMATCH));
		});
	});
}
async function readClawHubArchiveEntryBuffer(entry, limits) {
	const chunks = [];
	return await readLimitedClawHubArchiveEntry(entry, limits, {
		onChunk(buffer) {
			chunks.push(buffer);
		},
		onEnd() {
			return Buffer.concat(chunks);
		}
	});
}
async function hashClawHubArchiveEntry(entry, limits) {
	const digest = createHash("sha256");
	return await readLimitedClawHubArchiveEntry(entry, limits, {
		onChunk(buffer) {
			digest.update(buffer);
		},
		onEnd() {
			return digest.digest("hex");
		}
	});
}
function validateClawHubArchiveMetaJson(params) {
	let parsed;
	try {
		parsed = JSON.parse(params.bytes.toString("utf8"));
	} catch {
		return buildClawHubInstallFailure(`ClawHub archive contents do not match files[] metadata for "${params.packageName}@${params.version}": _meta.json is not valid JSON.`, CLAWHUB_INSTALL_ERROR_CODE.ARCHIVE_INTEGRITY_MISMATCH);
	}
	if (!parsed || typeof parsed !== "object") return buildClawHubInstallFailure(`ClawHub archive contents do not match files[] metadata for "${params.packageName}@${params.version}": _meta.json is not a JSON object.`, CLAWHUB_INSTALL_ERROR_CODE.ARCHIVE_INTEGRITY_MISMATCH);
	const record = parsed;
	if (record.slug !== params.packageName) return buildClawHubInstallFailure(`ClawHub archive contents do not match files[] metadata for "${params.packageName}@${params.version}": _meta.json slug does not match the package name.`, CLAWHUB_INSTALL_ERROR_CODE.ARCHIVE_INTEGRITY_MISMATCH);
	if (record.version !== params.version) return buildClawHubInstallFailure(`ClawHub archive contents do not match files[] metadata for "${params.packageName}@${params.version}": _meta.json version does not match the package version.`, CLAWHUB_INSTALL_ERROR_CODE.ARCHIVE_INTEGRITY_MISMATCH);
	return null;
}
async function verifyClawHubArchiveFiles(params) {
	try {
		if ((await fs.stat(params.archivePath)).size > 268435456) return buildClawHubInstallFailure("ClawHub archive fallback verification rejected the downloaded archive because it exceeds the ZIP archive size limit.", CLAWHUB_INSTALL_ERROR_CODE.ARCHIVE_INTEGRITY_MISMATCH);
		const archiveBytes = await fs.readFile(params.archivePath);
		const zip = await JSZip.loadAsync(archiveBytes);
		const actualFiles = /* @__PURE__ */ new Map();
		const validatedGeneratedPaths = /* @__PURE__ */ new Set();
		let entryCount = 0;
		let extractedBytes = 0;
		const addArchiveBytes = (bytes) => {
			extractedBytes += bytes;
			return extractedBytes <= DEFAULT_MAX_EXTRACTED_BYTES;
		};
		for (const entry of Object.values(zip.files)) {
			entryCount += 1;
			if (entryCount > 5e4) return buildClawHubInstallFailure("ClawHub archive fallback verification exceeded the archive entry limit.", CLAWHUB_INSTALL_ERROR_CODE.ARCHIVE_INTEGRITY_MISMATCH);
			if (entry.dir) continue;
			const relativePath = normalizeClawHubRelativePath(entry.name);
			if (!relativePath) return buildClawHubInstallFailure(`ClawHub archive contents do not match files[] metadata for "${params.packageName}@${params.packageVersion}": invalid package file path "${entry.name}" (${describeInvalidClawHubRelativePath(entry.name)}).`, CLAWHUB_INSTALL_ERROR_CODE.ARCHIVE_INTEGRITY_MISMATCH);
			if (relativePath === CLAWHUB_GENERATED_ARCHIVE_METADATA_FILE) {
				const metaResult = await readClawHubArchiveEntryBuffer(entry, {
					maxEntryBytes: DEFAULT_MAX_ENTRY_BYTES,
					addArchiveBytes
				});
				if (isClawHubInstallFailure(metaResult)) return metaResult;
				const metaFailure = validateClawHubArchiveMetaJson({
					packageName: params.packageName,
					version: params.packageVersion,
					bytes: metaResult
				});
				if (metaFailure) return metaFailure;
				validatedGeneratedPaths.add(relativePath);
				continue;
			}
			const sha256 = await hashClawHubArchiveEntry(entry, {
				maxEntryBytes: DEFAULT_MAX_ENTRY_BYTES,
				addArchiveBytes
			});
			if (typeof sha256 !== "string") return sha256;
			actualFiles.set(relativePath, sha256);
		}
		for (const file of params.files) {
			const actualSha256 = actualFiles.get(file.path);
			if (!actualSha256) return buildClawHubInstallFailure(`ClawHub archive contents do not match files[] metadata for "${params.packageName}@${params.packageVersion}": missing "${file.path}".`, CLAWHUB_INSTALL_ERROR_CODE.ARCHIVE_INTEGRITY_MISMATCH);
			if (actualSha256 !== file.sha256) return buildClawHubInstallFailure(`ClawHub archive contents do not match files[] metadata for "${params.packageName}@${params.packageVersion}": expected ${file.path} to hash to ${file.sha256}, got ${actualSha256}.`, CLAWHUB_INSTALL_ERROR_CODE.ARCHIVE_INTEGRITY_MISMATCH);
			actualFiles.delete(file.path);
		}
		const unexpectedFile = [...actualFiles.keys()].toSorted()[0];
		if (unexpectedFile) return buildClawHubInstallFailure(`ClawHub archive contents do not match files[] metadata for "${params.packageName}@${params.packageVersion}": unexpected file "${unexpectedFile}".`, CLAWHUB_INSTALL_ERROR_CODE.ARCHIVE_INTEGRITY_MISMATCH);
		return {
			ok: true,
			validatedGeneratedPaths: [...validatedGeneratedPaths].toSorted()
		};
	} catch {
		return buildClawHubInstallFailure("ClawHub archive fallback verification failed while reading the downloaded archive.", CLAWHUB_INSTALL_ERROR_CODE.ARCHIVE_INTEGRITY_MISMATCH);
	}
}
async function resolveCompatiblePackageVersion(params) {
	const requestedVersion = resolveRequestedVersion(params);
	if (!requestedVersion) return buildClawHubInstallFailure(`ClawHub package "${params.detail.package?.name ?? "unknown"}" has no installable version.`, CLAWHUB_INSTALL_ERROR_CODE.NO_INSTALLABLE_VERSION);
	let versionDetail;
	try {
		versionDetail = await fetchClawHubPackageVersion({
			name: params.detail.package?.name ?? "",
			version: requestedVersion,
			baseUrl: params.baseUrl,
			token: params.token
		});
	} catch (error) {
		return mapClawHubRequestError(error, {
			stage: "version",
			name: params.detail.package?.name ?? "unknown",
			version: requestedVersion
		});
	}
	const resolvedVersion = versionDetail.version?.version ?? requestedVersion;
	if (params.detail.package?.family === "skill") return {
		ok: true,
		version: resolvedVersion,
		compatibility: versionDetail.version?.compatibility ?? params.detail.package?.compatibility ?? null,
		verification: null
	};
	const verificationState = resolveClawHubArchiveVerification(versionDetail, params.detail.package?.name ?? "unknown", resolvedVersion);
	if (!verificationState.ok) return verificationState;
	return {
		ok: true,
		version: resolvedVersion,
		compatibility: versionDetail.version?.compatibility ?? params.detail.package?.compatibility ?? null,
		verification: verificationState.verification
	};
}
function validateClawHubPluginPackage(params) {
	const pkg = params.detail.package;
	if (!pkg) return buildClawHubInstallFailure("Package not found on ClawHub.", CLAWHUB_INSTALL_ERROR_CODE.PACKAGE_NOT_FOUND);
	if (pkg.family === "skill") return buildClawHubInstallFailure(`"${pkg.name}" is a skill. Use "openclaw skills install ${pkg.name}" instead.`, CLAWHUB_INSTALL_ERROR_CODE.SKILL_PACKAGE);
	if (pkg.family !== "code-plugin" && pkg.family !== "bundle-plugin") return buildClawHubInstallFailure(`Unsupported ClawHub package family: ${String(pkg.family)}`, CLAWHUB_INSTALL_ERROR_CODE.UNSUPPORTED_FAMILY);
	if (pkg.channel === "private") return buildClawHubInstallFailure(`"${pkg.name}" is private on ClawHub and cannot be installed anonymously.`, CLAWHUB_INSTALL_ERROR_CODE.PRIVATE_PACKAGE);
	const compatibility = params.compatibility;
	const runtimeVersion = params.runtimeVersion;
	if (compatibility?.pluginApiRange && !satisfiesPluginApiRange(runtimeVersion, compatibility.pluginApiRange)) return buildClawHubInstallFailure(`Plugin "${pkg.name}" requires plugin API ${compatibility.pluginApiRange}, but this OpenClaw runtime exposes ${runtimeVersion}.`, CLAWHUB_INSTALL_ERROR_CODE.INCOMPATIBLE_PLUGIN_API);
	if (compatibility?.minGatewayVersion && !satisfiesGatewayMinimum(runtimeVersion, compatibility.minGatewayVersion)) return buildClawHubInstallFailure(`Plugin "${pkg.name}" requires OpenClaw >=${compatibility.minGatewayVersion}, but this host is ${runtimeVersion}.`, CLAWHUB_INSTALL_ERROR_CODE.INCOMPATIBLE_GATEWAY);
	return null;
}
function logClawHubPackageSummary(params) {
	const pkg = params.detail.package;
	if (!pkg) return;
	const verification = pkg.verification?.tier ? ` verification=${pkg.verification.tier}` : "";
	params.logger?.info?.(`ClawHub ${pkg.family} ${pkg.name}@${params.version} channel=${pkg.channel}${verification}`);
	const compatibilityParts = [params.compatibility?.pluginApiRange ? `pluginApi=${params.compatibility.pluginApiRange}` : null, params.compatibility?.minGatewayVersion ? `minGateway=${params.compatibility.minGatewayVersion}` : null].filter(Boolean);
	if (compatibilityParts.length > 0) params.logger?.info?.(`Compatibility: ${compatibilityParts.join(" ")}`);
	if (pkg.channel !== "official") params.logger?.warn?.(`ClawHub package "${pkg.name}" is ${pkg.channel}; review source and verification before enabling.`);
}
async function installPluginFromClawHub(params) {
	const parsed = parseClawHubPluginSpec(params.spec);
	if (!parsed?.name) return buildClawHubInstallFailure(`invalid ClawHub plugin spec: ${params.spec}`, CLAWHUB_INSTALL_ERROR_CODE.INVALID_SPEC);
	params.logger?.info?.(`Resolving ${formatClawHubSpecifier(parsed)}…`);
	let detail;
	try {
		detail = await fetchClawHubPackageDetail({
			name: parsed.name,
			baseUrl: params.baseUrl,
			token: params.token
		});
	} catch (error) {
		return mapClawHubRequestError(error, {
			stage: "package",
			name: parsed.name
		});
	}
	const versionState = await resolveCompatiblePackageVersion({
		detail,
		requestedVersion: parsed.version,
		baseUrl: params.baseUrl,
		token: params.token
	});
	if (!versionState.ok) return versionState;
	const runtimeVersion = resolveCompatibilityHostVersion();
	const validationFailure = validateClawHubPluginPackage({
		detail,
		compatibility: versionState.compatibility,
		runtimeVersion
	});
	if (validationFailure) return validationFailure;
	if (!versionState.verification) return buildClawHubInstallFailure(`ClawHub version metadata for "${parsed.name}@${versionState.version}" is missing sha256hash and usable files[] metadata for fallback archive verification.`, CLAWHUB_INSTALL_ERROR_CODE.MISSING_ARCHIVE_INTEGRITY);
	const canonicalPackageName = detail.package?.name ?? parsed.name;
	logClawHubPackageSummary({
		detail,
		version: versionState.version,
		compatibility: versionState.compatibility,
		logger: params.logger
	});
	let archive;
	try {
		archive = await downloadClawHubPackageArchive({
			name: parsed.name,
			version: versionState.version,
			baseUrl: params.baseUrl,
			token: params.token
		});
	} catch (error) {
		return buildClawHubInstallFailure(formatErrorMessage(error));
	}
	try {
		if (versionState.verification.kind === "archive-integrity") {
			if (archive.integrity !== versionState.verification.integrity) return buildClawHubInstallFailure(`ClawHub archive integrity mismatch for "${parsed.name}@${versionState.version}": expected ${versionState.verification.integrity}, got ${archive.integrity}.`, CLAWHUB_INSTALL_ERROR_CODE.ARCHIVE_INTEGRITY_MISMATCH);
		} else {
			const validatedPaths = versionState.verification.files.map((file) => file.path).toSorted().join(", ");
			const fallbackVerification = await verifyClawHubArchiveFiles({
				archivePath: archive.archivePath,
				packageName: canonicalPackageName,
				packageVersion: versionState.version,
				files: versionState.verification.files
			});
			if (!fallbackVerification.ok) return fallbackVerification;
			const validatedGeneratedPaths = fallbackVerification.validatedGeneratedPaths.length > 0 ? ` Validated generated metadata files present in archive: ${fallbackVerification.validatedGeneratedPaths.join(", ")} (JSON parse plus slug/version match only).` : "";
			params.logger?.warn?.(`ClawHub package "${canonicalPackageName}@${versionState.version}" is missing sha256hash; falling back to files[] verification. Validated files: ${validatedPaths}.${validatedGeneratedPaths}`);
		}
		params.logger?.info?.(`Downloading ${detail.package?.family === "bundle-plugin" ? "bundle" : "plugin"} ${parsed.name}@${versionState.version} from ClawHub…`);
		const installResult = await installPluginFromArchive({
			archivePath: archive.archivePath,
			dangerouslyForceUnsafeInstall: params.dangerouslyForceUnsafeInstall,
			logger: params.logger,
			mode: params.mode,
			dryRun: params.dryRun,
			expectedPluginId: params.expectedPluginId
		});
		if (!installResult.ok) return installResult;
		const pkg = detail.package;
		const clawhubFamily = pkg.family === "code-plugin" || pkg.family === "bundle-plugin" ? pkg.family : null;
		if (!clawhubFamily) return buildClawHubInstallFailure(`Unsupported ClawHub package family: ${pkg.family}`, CLAWHUB_INSTALL_ERROR_CODE.UNSUPPORTED_FAMILY);
		return {
			...installResult,
			packageName: parsed.name,
			clawhub: {
				source: "clawhub",
				clawhubUrl: normalizeOptionalString(params.baseUrl) || normalizeOptionalString(process.env.OPENCLAW_CLAWHUB_URL) || "https://clawhub.ai",
				clawhubPackage: parsed.name,
				clawhubFamily,
				clawhubChannel: pkg.channel,
				version: installResult.version ?? versionState.version,
				integrity: archive.integrity,
				resolvedAt: (/* @__PURE__ */ new Date()).toISOString()
			}
		};
	} finally {
		await archive.cleanup().catch(() => void 0);
	}
}
//#endregion
export { formatClawHubSpecifier as n, installPluginFromClawHub as r, CLAWHUB_INSTALL_ERROR_CODE as t };
