// openclaw/dist/cli.runtime-Cb2cQC63.js
import { r as formatErrorMessage } from "./errors-CGMew-1w.js";
import "./error-runtime-BNzEFRqy.js";
import { A as normalizeQaProviderMode, c as startQaGatewayChild, k as defaultQaModelForMode, l as startQaMockOpenAiServer, n as runQaSuite, p as startQaLabServer, r as resolveQaLiveTurnTimeoutMs } from "./suite-BW4kSK9C.js";
import { r as isQaFastModeModelRef } from "./model-selection-Dn2TIk7I.js";
import { i as normalizeQaThinkingLevel } from "./qa-gateway-config-Dt2kA3jj.js";
import { n as writeQaDockerHarnessFiles, t as buildQaDockerHarnessImage } from "./docker-harness-B2wtqHne.js";
import path from "node:path";
import { execFile } from "node:child_process";
import fs from "node:fs/promises";
import { randomUUID } from "node:crypto";
import { createServer } from "node:net";
import { setTimeout } from "node:timers/promises";
//#region extensions/qa-lab/src/manual-lane.runtime.ts
function resolveManualLaneTimeoutMs(params) {
	if (typeof params.timeoutMs === "number" && Number.isFinite(params.timeoutMs) && params.timeoutMs > 0) return params.timeoutMs;
	return resolveQaLiveTurnTimeoutMs({
		providerMode: params.providerMode,
		primaryModel: params.primaryModel,
		alternateModel: params.alternateModel
	}, 12e4, params.primaryModel);
}
async function runQaManualLane(params) {
	const sessionSuffix = params.primaryModel.replace(/[^a-z0-9._-]+/gi, "-");
	const lab = await startQaLabServer({
		repoRoot: params.repoRoot,
		embeddedGateway: "disabled"
	});
	const mock = params.providerMode === "mock-openai" ? await startQaMockOpenAiServer({
		host: "127.0.0.1",
		port: 0
	}) : null;
	const gateway = await startQaGatewayChild({
		repoRoot: params.repoRoot,
		providerBaseUrl: mock ? `${mock.baseUrl}/v1` : void 0,
		qaBusBaseUrl: lab.listenUrl,
		providerMode: params.providerMode,
		primaryModel: params.primaryModel,
		alternateModel: params.alternateModel,
		fastMode: params.fastMode,
		thinkingDefault: params.thinkingDefault,
		controlUiEnabled: false
	});
	const timeoutMs = resolveManualLaneTimeoutMs({
		providerMode: params.providerMode,
		primaryModel: params.primaryModel,
		alternateModel: params.alternateModel,
		timeoutMs: params.timeoutMs
	});
	try {
		const started = await gateway.call("agent", {
			idempotencyKey: randomUUID(),
			agentId: "qa",
			sessionKey: `agent:qa:manual:${sessionSuffix}`,
			message: params.message,
			deliver: true,
			channel: "qa-channel",
			to: "dm:qa-operator",
			replyChannel: "qa-channel",
			replyTo: "dm:qa-operator"
		}, { timeoutMs: 3e4 });
		if (!started.runId) throw new Error(`agent call did not return a runId: ${JSON.stringify(started)}`);
		const waited = await gateway.call("agent.wait", {
			runId: started.runId,
			timeoutMs
		}, { timeoutMs: timeoutMs + 5e3 });
		await setTimeout(500);
		const reply = lab.state.getSnapshot().messages.filter((candidate) => candidate.direction === "outbound" && candidate.conversation.id === "qa-operator").at(-1)?.text ?? null;
		return {
			model: params.primaryModel,
			waited,
			reply,
			watchUrl: lab.baseUrl
		};
	} catch (error) {
		throw new Error(formatErrorMessage(error), { cause: error });
	} finally {
		await gateway.stop();
		await mock?.stop();
		await lab.stop();
	}
}
//#endregion
//#region extensions/qa-lab/src/character-eval.ts
const DEFAULT_CHARACTER_SCENARIO_ID = "character-vibes-gollum";
const DEFAULT_CHARACTER_EVAL_MODELS = Object.freeze([
	"openai/gpt-5.4",
	"openai/gpt-5.2",
	"openai/gpt-5",
	"anthropic/claude-opus-4-6",
	"anthropic/claude-sonnet-4-6",
	"zai/glm-5.1",
	"moonshot/kimi-k2.5",
	"google/gemini-3.1-pro-preview"
]);
const DEFAULT_CHARACTER_THINKING = "high";
const DEFAULT_CHARACTER_EVAL_CONCURRENCY = 16;
const DEFAULT_CHARACTER_THINKING_BY_MODEL = Object.freeze({
	"openai/gpt-5.4": "xhigh",
	"openai/gpt-5.2": "xhigh",
	"openai/gpt-5": "xhigh"
});
const DEFAULT_JUDGE_MODELS = Object.freeze(["openai/gpt-5.4", "anthropic/claude-opus-4-6"]);
const DEFAULT_JUDGE_THINKING = "xhigh";
const DEFAULT_JUDGE_MODEL_OPTIONS = Object.freeze({
	"openai/gpt-5.4": {
		thinkingDefault: "xhigh",
		fastMode: true
	},
	"anthropic/claude-opus-4-6": { thinkingDefault: "high" }
});
function normalizeModelRefs(models) {
	return [...new Set(models.map((model) => model.trim()).filter((model) => model.length > 0))];
}
function resolveCandidateThinkingDefault(params) {
	return params.candidateModelOptions?.[params.model]?.thinkingDefault ?? params.candidateThinkingByModel?.[params.model] ?? params.candidateThinkingDefault ?? DEFAULT_CHARACTER_THINKING_BY_MODEL[params.model] ?? DEFAULT_CHARACTER_THINKING;
}
function resolveCandidateFastMode(params) {
	return params.candidateModelOptions?.[params.model]?.fastMode ?? params.candidateFastMode ?? isQaFastModeModelRef(params.model);
}
function resolveJudgeOptions(params) {
	const modelDefaults = DEFAULT_JUDGE_MODEL_OPTIONS[params.model];
	const modelOptions = params.judgeModelOptions?.[params.model];
	return {
		thinkingDefault: modelOptions?.thinkingDefault ?? params.judgeThinkingDefault ?? modelDefaults?.thinkingDefault ?? DEFAULT_JUDGE_THINKING,
		fastMode: modelOptions?.fastMode ?? modelDefaults?.fastMode ?? false
	};
}
function sanitizePathPart(value) {
	return value.replace(/[^a-z0-9._-]+/gi, "-").replace(/^-+|-+$/g, "") || "model";
}
function normalizeConcurrency(value, fallback = 1) {
	if (value === void 0) return fallback;
	if (!Number.isFinite(value)) return fallback;
	return Math.max(1, Math.floor(value));
}
async function mapWithConcurrency(items, concurrency, mapper) {
	const results = Array.from({ length: items.length });
	let nextIndex = 0;
	const workerCount = Math.min(normalizeConcurrency(concurrency), items.length);
	const workers = Array.from({ length: workerCount }, async () => {
		while (nextIndex < items.length) {
			const index = nextIndex;
			nextIndex += 1;
			results[index] = await mapper(items[index], index);
		}
	});
	await Promise.all(workers);
	return results;
}
function extractTranscript(result) {
	return result.scenarios.flatMap((scenario) => scenario.steps.map((step) => step.details).filter((detail) => Boolean(detail))).toSorted((left, right) => right.length - left.length)[0] ?? result.report;
}
function collectTranscriptStats(transcript) {
	return {
		transcriptChars: transcript.length,
		transcriptLines: transcript.length === 0 ? 0 : transcript.split(/\r?\n/).length,
		userTurns: transcript.match(/^USER\b/gm)?.length ?? 0,
		assistantTurns: transcript.match(/^ASSISTANT\b/gm)?.length ?? 0
	};
}
function detectTranscriptFailure(transcript) {
	return [
		[/\bmodel `[^`]+` is not supported\b/i, "model unsupported error leaked into transcript"],
		[/\binsufficient account balance\b/i, "account balance error leaked into transcript"],
		[/\b(?:backend|transport|internal) error\b/i, "backend error leaked into transcript"],
		[/\bsomething went wrong while processing your request\b/i, "generic request failure leaked into transcript"],
		[/\buse \/new to start a fresh session\b/i, "generic request failure leaked into transcript"],
		[/\bmodel did not produce a response before the LLM idle timeout\b/i, "LLM timeout leaked into transcript"],
		[/\btool failed\b/i, "tool failure leaked into transcript"],
		[/\b(?:read|write|edit|patch):[^\n]*\bfailed\b/i, "tool failure leaked into transcript"],
		[/\bnot configured\b/i, "configuration error leaked into transcript"]
	].find(([pattern]) => pattern.test(transcript))?.[1];
}
function formatDuration(ms) {
	if (!Number.isFinite(ms) || ms < 0) return "unknown";
	if (ms < 1e3) return `${Math.round(ms)}ms`;
	if (ms < 6e4) {
		const seconds = ms / 1e3;
		return `${seconds >= 10 ? Math.round(seconds) : Number(seconds.toFixed(1))}s`;
	}
	const totalSeconds = Math.round(ms / 1e3);
	const minutes = Math.floor(totalSeconds / 60);
	const seconds = totalSeconds % 60;
	return seconds === 0 ? `${minutes}m` : `${minutes}m ${seconds}s`;
}
function logCharacterEvalProgress(progress, message) {
	progress?.(`[qa-character] ${message}`);
}
function formatEvalIndex(index, total) {
	return `${index + 1}/${total}`;
}
function summarizeRunStats(run) {
	return [
		`status=${run.status}`,
		`duration=${formatDuration(run.durationMs)}`,
		`turns=${run.stats.userTurns}/${run.stats.assistantTurns}`,
		`chars=${run.stats.transcriptChars}`,
		...run.error ? [`error="${run.error}"`] : []
	].join(" ");
}
function formatBlindCandidateLabel(index) {
	return `candidate-${String(index + 1).padStart(2, "0")}`;
}
function buildJudgePrompt(params) {
	const labelToModel = /* @__PURE__ */ new Map();
	const runBlocks = params.runs.map((run, index) => {
		const label = params.blindModels ? formatBlindCandidateLabel(index) : run.model;
		labelToModel.set(label, run.model);
		return `## CANDIDATE ${label}

Status: ${run.status}
Duration ms (not used for ranking): ${run.durationMs}
Fast mode: ${run.fastMode ? "on" : "off"}
Thinking: ${run.thinkingDefault}
Transcript chars: ${run.stats.transcriptChars}
Assistant turns: ${run.stats.assistantTurns}
Error: ${run.error ?? "none"}

\`\`\`text
${run.transcript}
\`\`\``;
	}).join("\n\n");
	return {
		prompt: `You are grading OpenClaw natural character conversation transcripts for naturalness, vibes, and funniness.

Scenario id: ${params.scenarioId}

Rank the models by:
- natural conversational reaction
- playful character commitment
- funny, surprising details
- coherence across turns
- completing real user tasks without becoming generic
- not sounding aware of an eval or test
- avoiding tool/backend/error leakage

Treat candidate labels as opaque identifiers. Do not assume quality from the label.
Duration is recorded for separate benchmark analysis only. Do not rank models by speed.

Return strict JSON only with this shape:
{
  "rankings": [
    {
      "model": "same candidate label",
      "rank": 1,
      "score": 9.2,
      "summary": "one sentence",
      "strengths": ["short"],
      "weaknesses": ["short"]
    }
  ]
}

${runBlocks}`,
		labelToModel
	};
}
function normalizeJudgment(value, allowedModels) {
	const payload = value && typeof value === "object" ? value : {};
	return (Array.isArray(payload.rankings) ? payload.rankings : []).map((entry) => {
		if (!entry || typeof entry !== "object") return null;
		const record = entry;
		const model = typeof record.model === "string" ? record.model : "";
		if (!allowedModels.has(model)) return null;
		const rank = typeof record.rank === "number" ? record.rank : Number(record.rank);
		const score = typeof record.score === "number" ? record.score : Number(record.score);
		const summary = typeof record.summary === "string" ? record.summary : "";
		const strengths = Array.isArray(record.strengths) ? record.strengths.filter((item) => typeof item === "string") : [];
		const weaknesses = Array.isArray(record.weaknesses) ? record.weaknesses.filter((item) => typeof item === "string") : [];
		if (!Number.isFinite(rank) || !Number.isFinite(score)) return null;
		return {
			model,
			rank,
			score,
			summary,
			strengths,
			weaknesses
		};
	}).filter((entry) => Boolean(entry)).toSorted((left, right) => left.rank - right.rank || right.score - left.score);
}
function parseJudgeReply(reply, allowedModels) {
	if (!reply) throw new Error("judge did not return a reply");
	const trimmed = reply.trim();
	const jsonText = trimmed.match(/```(?:json)?\s*([\s\S]*?)```/)?.[1]?.trim() ?? trimmed.match(/\{[\s\S]*\}/)?.[0]?.trim() ?? trimmed;
	const rankings = normalizeJudgment(JSON.parse(jsonText), allowedModels);
	if (rankings.length === 0) throw new Error("judge reply did not contain valid rankings");
	return rankings;
}
async function defaultRunJudge(params) {
	return (await runQaManualLane({
		repoRoot: params.repoRoot,
		providerMode: "live-frontier",
		primaryModel: params.judgeModel,
		alternateModel: params.judgeModel,
		fastMode: params.judgeFastMode,
		thinkingDefault: params.judgeThinkingDefault,
		message: params.prompt,
		timeoutMs: params.timeoutMs
	})).reply;
}
function renderCharacterEvalReport(params) {
	const lines = [
		"# OpenClaw Character Eval Report",
		"",
		`- Started: ${params.startedAt.toISOString()}`,
		`- Finished: ${params.finishedAt.toISOString()}`,
		`- Duration: ${formatDuration(params.finishedAt.getTime() - params.startedAt.getTime())}`,
		`- Scenario: ${params.scenarioId}`,
		"- Execution: local QA gateway child processes, not Docker",
		`- Judges: ${params.judgments.map((judgment) => judgment.model).join(", ")}`,
		`- Judge thinking: ${params.judgments[0]?.thinkingDefault ?? DEFAULT_JUDGE_THINKING}`,
		`- Judge fast mode: ${params.judgments.every((judgment) => judgment.fastMode) ? "on" : "mixed"}`,
		`- Judge model labels: ${params.judgments.every((judgment) => judgment.blindModels) ? "blind" : "visible"}`,
		"",
		"## Judge Rankings",
		""
	];
	for (const judgment of params.judgments) {
		lines.push(`### ${judgment.model}`, "");
		lines.push(`- Duration: ${formatDuration(judgment.durationMs)}`, "");
		if (judgment.rankings.length > 0) for (const ranking of judgment.rankings) {
			lines.push(`${ranking.rank}. ${ranking.model} - ${ranking.score.toFixed(1)} - ${ranking.summary}`);
			if (ranking.strengths.length > 0) lines.push(`   Strengths: ${ranking.strengths.join("; ")}`);
			if (ranking.weaknesses.length > 0) lines.push(`   Weaknesses: ${ranking.weaknesses.join("; ")}`);
		}
		else {
			lines.push("- Judge ranking unavailable.");
			if (judgment.error) lines.push(`- Judge error: ${judgment.error}`);
		}
		lines.push("");
	}
	lines.push("## Run Stats", "");
	lines.push("| Model | Thinking | Fast mode | Status | Duration | User turns | Assistant turns | Transcript chars |");
	lines.push("| --- | --- | --- | --- | ---: | ---: | ---: | ---: |");
	for (const run of params.runs) lines.push(`| ${run.model} | ${run.thinkingDefault} | ${run.fastMode ? "on" : "off"} | ${run.status} | ${formatDuration(run.durationMs)} | ${run.stats.userTurns} | ${run.stats.assistantTurns} | ${run.stats.transcriptChars} |`);
	lines.push("", "## Transcripts", "");
	for (const run of params.runs) {
		lines.push(`### ${run.model}`, "");
		lines.push(`- Status: ${run.status}`);
		lines.push(`- Thinking: ${run.thinkingDefault}`);
		lines.push(`- Fast mode: ${run.fastMode ? "on" : "off"}`);
		lines.push(`- Duration: ${formatDuration(run.durationMs)}`);
		lines.push(`- Report: ${run.reportPath ?? "unavailable"}`);
		if (run.error) lines.push(`- Error: ${run.error}`);
		lines.push("", "```text", run.transcript.trim() || "(empty transcript)", "```", "");
	}
	return `${lines.join("\n")}\n`;
}
async function runQaCharacterEval(params) {
	const startedAt = /* @__PURE__ */ new Date();
	const repoRoot = path.resolve(params.repoRoot ?? process.cwd());
	const scenarioId = params.scenarioId?.trim() || DEFAULT_CHARACTER_SCENARIO_ID;
	const models = normalizeModelRefs(params.models.length > 0 ? params.models : DEFAULT_CHARACTER_EVAL_MODELS);
	if (models.length === 0) throw new Error("qa character-eval needs at least one --model <provider/model> ref");
	const outputDir = params.outputDir ?? path.join(repoRoot, ".artifacts", "qa-e2e", `character-eval-${Date.now().toString(36)}`);
	const runsDir = path.join(outputDir, "runs");
	await fs.mkdir(runsDir, { recursive: true });
	const runSuite = params.runSuite ?? runQaSuite;
	const candidateConcurrency = normalizeConcurrency(params.candidateConcurrency, DEFAULT_CHARACTER_EVAL_CONCURRENCY);
	logCharacterEvalProgress(params.progress, `start scenario=${scenarioId} candidates=${models.length} candidateConcurrency=${candidateConcurrency} output=${outputDir}`);
	const candidatesStartedAt = Date.now();
	const runs = await mapWithConcurrency(models, candidateConcurrency, async (model, index) => {
		const thinkingDefault = resolveCandidateThinkingDefault({
			model,
			candidateThinkingDefault: params.candidateThinkingDefault,
			candidateThinkingByModel: params.candidateThinkingByModel,
			candidateModelOptions: params.candidateModelOptions
		});
		const fastMode = resolveCandidateFastMode({
			model,
			candidateFastMode: params.candidateFastMode,
			candidateModelOptions: params.candidateModelOptions
		});
		const modelOutputDir = path.join(runsDir, sanitizePathPart(model));
		const runStartedAt = Date.now();
		logCharacterEvalProgress(params.progress, `candidate start ${formatEvalIndex(index, models.length)} model=${model} thinking=${thinkingDefault} fast=${fastMode ? "on" : "off"}`);
		try {
			const result = await runSuite({
				repoRoot,
				outputDir: modelOutputDir,
				providerMode: "live-frontier",
				primaryModel: model,
				alternateModel: model,
				fastMode,
				thinkingDefault,
				scenarioIds: [scenarioId]
			});
			const transcript = extractTranscript(result);
			const transcriptFailure = detectTranscriptFailure(transcript);
			const run = {
				model,
				status: result.scenarios.some((scenario) => scenario.status === "fail") || transcriptFailure ? "fail" : "pass",
				durationMs: Date.now() - runStartedAt,
				outputDir: modelOutputDir,
				thinkingDefault,
				fastMode,
				reportPath: result.reportPath,
				summaryPath: result.summaryPath,
				transcript,
				stats: collectTranscriptStats(transcript),
				...transcriptFailure ? { error: transcriptFailure } : {}
			};
			logCharacterEvalProgress(params.progress, `candidate done ${formatEvalIndex(index, models.length)} model=${model} ${summarizeRunStats(run)}`);
			return run;
		} catch (error) {
			const transcript = "";
			const run = {
				model,
				status: "fail",
				durationMs: Date.now() - runStartedAt,
				outputDir: modelOutputDir,
				thinkingDefault,
				fastMode,
				transcript,
				stats: collectTranscriptStats(transcript),
				error: formatErrorMessage(error)
			};
			logCharacterEvalProgress(params.progress, `candidate done ${formatEvalIndex(index, models.length)} model=${model} ${summarizeRunStats(run)}`);
			return run;
		}
	});
	const failedCandidateCount = runs.filter((run) => run.status === "fail").length;
	logCharacterEvalProgress(params.progress, `candidates done pass=${runs.length - failedCandidateCount} fail=${failedCandidateCount} duration=${formatDuration(Date.now() - candidatesStartedAt)}`);
	const judgeModels = normalizeModelRefs(params.judgeModels && params.judgeModels.length > 0 ? params.judgeModels : params.judgeModel ? [params.judgeModel] : DEFAULT_JUDGE_MODELS);
	const runJudge = params.runJudge ?? defaultRunJudge;
	const judgeConcurrency = normalizeConcurrency(params.judgeConcurrency, DEFAULT_CHARACTER_EVAL_CONCURRENCY);
	const judgeTimeoutMs = params.judgeTimeoutMs ?? 18e4;
	logCharacterEvalProgress(params.progress, `judges start judges=${judgeModels.length} judgeConcurrency=${judgeConcurrency} timeout=${formatDuration(judgeTimeoutMs)} labels=${params.judgeBlindModels === true ? "blind" : "visible"}`);
	const judgesStartedAt = Date.now();
	const judgments = await mapWithConcurrency(judgeModels, judgeConcurrency, async (judgeModel, index) => {
		const judgeOptions = resolveJudgeOptions({
			model: judgeModel,
			judgeThinkingDefault: params.judgeThinkingDefault,
			judgeModelOptions: params.judgeModelOptions
		});
		let rankings = [];
		let judgeError;
		const judgeStartedAt = Date.now();
		logCharacterEvalProgress(params.progress, `judge start ${formatEvalIndex(index, judgeModels.length)} model=${judgeModel} thinking=${judgeOptions.thinkingDefault} fast=${judgeOptions.fastMode ? "on" : "off"} timeout=${formatDuration(judgeTimeoutMs)}`);
		try {
			const judgePrompt = buildJudgePrompt({
				scenarioId,
				runs,
				blindModels: params.judgeBlindModels
			});
			rankings = parseJudgeReply(await runJudge({
				repoRoot,
				judgeModel,
				judgeThinkingDefault: judgeOptions.thinkingDefault,
				judgeFastMode: judgeOptions.fastMode,
				prompt: judgePrompt.prompt,
				timeoutMs: judgeTimeoutMs
			}), new Set(judgePrompt.labelToModel.keys())).map((ranking) => ({
				...ranking,
				model: judgePrompt.labelToModel.get(ranking.model) ?? ranking.model
			}));
		} catch (error) {
			judgeError = formatErrorMessage(error);
		}
		const judgment = {
			model: judgeModel,
			thinkingDefault: judgeOptions.thinkingDefault,
			fastMode: judgeOptions.fastMode,
			blindModels: params.judgeBlindModels === true,
			durationMs: Date.now() - judgeStartedAt,
			rankings,
			...judgeError ? { error: judgeError } : {}
		};
		logCharacterEvalProgress(params.progress, `judge done ${formatEvalIndex(index, judgeModels.length)} model=${judgeModel} rankings=${rankings.length} duration=${formatDuration(judgment.durationMs)}${judgeError ? ` error="${judgeError}"` : ""}`);
		return judgment;
	});
	const failedJudgeCount = judgments.filter((judgment) => judgment.rankings.length === 0).length;
	logCharacterEvalProgress(params.progress, `judges done ranked=${judgments.length - failedJudgeCount} failed=${failedJudgeCount} duration=${formatDuration(Date.now() - judgesStartedAt)}`);
	const finishedAt = /* @__PURE__ */ new Date();
	const report = renderCharacterEvalReport({
		scenarioId,
		startedAt,
		finishedAt,
		runs,
		judgments
	});
	const reportPath = path.join(outputDir, "character-eval-report.md");
	const summaryPath = path.join(outputDir, "character-eval-summary.json");
	await fs.writeFile(reportPath, report, "utf8");
	await fs.writeFile(summaryPath, `${JSON.stringify({
		scenarioId,
		runs,
		judgments
	}, null, 2)}\n`, "utf8");
	logCharacterEvalProgress(params.progress, `report written duration=${formatDuration(finishedAt.getTime() - startedAt.getTime())} report=${reportPath} summary=${summaryPath}`);
	return {
		outputDir,
		reportPath,
		summaryPath,
		runs,
		judgments
	};
}
//#endregion
//#region extensions/qa-lab/src/docker-up.runtime.ts
function resolveDefaultQaDockerDir(repoRoot) {
	return path.resolve(repoRoot, ".artifacts/qa-docker");
}
function describeError(error) {
	if (error instanceof Error) return error.message;
	if (typeof error === "string") return error;
	return JSON.stringify(error);
}
async function isPortFree(port) {
	return await new Promise((resolve) => {
		const server = createServer();
		server.once("error", () => resolve(false));
		server.listen(port, "127.0.0.1", () => {
			server.close(() => resolve(true));
		});
	});
}
async function findFreePort() {
	return await new Promise((resolve, reject) => {
		const server = createServer();
		server.once("error", reject);
		server.listen(0, () => {
			const address = server.address();
			if (!address || typeof address === "string") {
				server.close();
				reject(/* @__PURE__ */ new Error("failed to find free port"));
				return;
			}
			server.close((error) => {
				if (error) {
					reject(error);
					return;
				}
				resolve(address.port);
			});
		});
	});
}
async function resolveHostPort(preferredPort, pinned) {
	if (pinned || await isPortFree(preferredPort)) return preferredPort;
	return await findFreePort();
}
function trimCommandOutput(output) {
	const trimmed = output.trim();
	if (!trimmed) return "";
	const lines = trimmed.split("\n");
	return lines.length <= 120 ? trimmed : lines.slice(-120).join("\n");
}
async function execCommand(command, args, cwd) {
	return await new Promise((resolve, reject) => {
		execFile(command, args, {
			cwd,
			encoding: "utf8",
			maxBuffer: 10 * 1024 * 1024
		}, (error, stdout, stderr) => {
			if (error) {
				const renderedStdout = trimCommandOutput(stdout);
				const renderedStderr = trimCommandOutput(stderr);
				reject(new Error([
					`Command failed: ${[command, ...args].join(" ")}`,
					renderedStderr ? `stderr:\n${renderedStderr}` : "",
					renderedStdout ? `stdout:\n${renderedStdout}` : ""
				].filter(Boolean).join("\n\n")));
				return;
			}
			resolve({
				stdout,
				stderr
			});
		});
	});
}
async function waitForHealth(url, deps) {
	const timeoutMs = deps.timeoutMs ?? 36e4;
	const pollMs = deps.pollMs ?? 1e3;
	const startMs = Date.now();
	const deadline = startMs + timeoutMs;
	let lastError = null;
	while (Date.now() < deadline) {
		try {
			if ((await deps.fetchImpl(url)).ok) return;
			lastError = /* @__PURE__ */ new Error(`Health check returned non-OK for ${url}`);
		} catch (error) {
			lastError = error;
		}
		await deps.sleepImpl(pollMs);
	}
	const elapsedSec = Math.round((Date.now() - startMs) / 1e3);
	const lines = [
		`${deps.label ?? url} did not become healthy within ${elapsedSec}s (limit ${Math.round(timeoutMs / 1e3)}s).`,
		lastError ? `Last error: ${describeError(lastError)}` : "",
		`Hint: check container logs with \`docker compose -f ${deps.composeFile ?? "<compose-file>"} logs\` and verify the port is not already in use.`
	];
	throw new Error(lines.filter(Boolean).join("\n"));
}
async function isHealthy(url, fetchImpl) {
	try {
		return (await fetchImpl(url)).ok;
	} catch {
		return false;
	}
}
async function waitForDockerServiceHealth(service, composeFile, repoRoot, runCommand, sleepImpl, timeoutMs = 36e4, pollMs = 1e3) {
	const startMs = Date.now();
	const deadline = startMs + timeoutMs;
	let lastStatus = "unknown";
	while (Date.now() < deadline) {
		try {
			const { stdout } = await runCommand("docker", [
				"compose",
				"-f",
				composeFile,
				"ps",
				"--format",
				"json",
				service
			], repoRoot);
			const row = stdout.trim().split("\n").map((line) => line.trim()).filter(Boolean).map((line) => JSON.parse(line))[0];
			lastStatus = row?.Health ?? row?.State ?? "unknown";
			if (lastStatus === "healthy" || lastStatus === "running") return;
		} catch (error) {
			lastStatus = describeError(error);
		}
		await sleepImpl(pollMs);
	}
	const elapsedSec = Math.round((Date.now() - startMs) / 1e3);
	throw new Error([
		`${service} did not become healthy within ${elapsedSec}s (limit ${Math.round(timeoutMs / 1e3)}s).`,
		`Last status: ${lastStatus}`,
		`Hint: check container logs with \`docker compose -f ${composeFile} logs ${service}\`.`
	].join("\n"));
}
async function resolveComposeServiceUrl(service, port, composeFile, repoRoot, runCommand) {
	const { stdout: containerStdout } = await runCommand("docker", [
		"compose",
		"-f",
		composeFile,
		"ps",
		"-q",
		service
	], repoRoot);
	const containerId = containerStdout.trim();
	if (!containerId) return null;
	const { stdout: ipStdout } = await runCommand("docker", [
		"inspect",
		"--format",
		"{{range .NetworkSettings.Networks}}{{.IPAddress}}{{end}}",
		containerId
	], repoRoot);
	const ip = ipStdout.trim();
	if (!ip) return null;
	return `http://${ip}:${port}/`;
}
async function runQaDockerUp(params, deps) {
	const repoRoot = path.resolve(params.repoRoot ?? process.cwd());
	const resolveHostPortImpl = deps?.resolveHostPortImpl ?? resolveHostPort;
	const outputDir = path.resolve(params.outputDir ?? resolveDefaultQaDockerDir(repoRoot));
	const gatewayPort = await resolveHostPortImpl(params.gatewayPort ?? 18789, params.gatewayPort != null);
	const qaLabPort = await resolveHostPortImpl(params.qaLabPort ?? 43124, params.qaLabPort != null);
	const runCommand = deps?.runCommand ?? execCommand;
	const fetchImpl = deps?.fetchImpl ?? (async (input) => {
		return await fetch(input);
	});
	const sleepImpl = deps?.sleepImpl ?? setTimeout;
	if (!params.skipUiBuild) await runCommand("pnpm", ["qa:lab:build"], repoRoot);
	await writeQaDockerHarnessFiles({
		outputDir,
		repoRoot,
		gatewayPort,
		qaLabPort,
		providerBaseUrl: params.providerBaseUrl,
		imageName: params.image,
		usePrebuiltImage: params.usePrebuiltImage,
		bindUiDist: params.bindUiDist,
		includeQaLabUi: true
	});
	const composeFile = path.join(outputDir, "docker-compose.qa.yml");
	try {
		await runCommand("docker", [
			"compose",
			"-f",
			composeFile,
			"down",
			"--remove-orphans"
		], repoRoot);
	} catch {}
	const composeArgs = [
		"compose",
		"-f",
		composeFile,
		"up"
	];
	if (!params.usePrebuiltImage) composeArgs.push("--build");
	composeArgs.push("-d");
	await runCommand("docker", composeArgs, repoRoot);
	await sleepImpl(3e3);
	const qaLabUrl = `http://127.0.0.1:${qaLabPort}`;
	const hostGatewayUrl = `http://127.0.0.1:${gatewayPort}/`;
	await waitForHealth(`${qaLabUrl}/healthz`, {
		label: "QA Lab",
		fetchImpl,
		sleepImpl,
		composeFile
	});
	await waitForDockerServiceHealth("openclaw-qa-gateway", composeFile, repoRoot, runCommand, sleepImpl);
	let gatewayUrl = hostGatewayUrl;
	if (!await isHealthy(`${hostGatewayUrl}healthz`, fetchImpl)) {
		const containerGatewayUrl = await resolveComposeServiceUrl("openclaw-qa-gateway", 18789, composeFile, repoRoot, runCommand);
		if (containerGatewayUrl && await isHealthy(`${containerGatewayUrl}healthz`, fetchImpl)) gatewayUrl = containerGatewayUrl;
	}
	return {
		outputDir,
		composeFile,
		qaLabUrl,
		gatewayUrl,
		stopCommand: `docker compose -f ${composeFile} down`
	};
}
//#endregion
//#region extensions/qa-lab/src/cli.runtime.ts
function resolveQaManualLaneModels(opts) {
	const primaryModel = opts.primaryModel?.trim() || defaultQaModelForMode(opts.providerMode);
	const alternateModel = opts.alternateModel?.trim();
	return {
		primaryModel,
		alternateModel: alternateModel && alternateModel.length > 0 ? alternateModel : opts.primaryModel?.trim() ? primaryModel : defaultQaModelForMode(opts.providerMode, true)
	};
}
function parseQaThinkingLevel(label, value) {
	if (value === void 0) return;
	const normalized = normalizeQaThinkingLevel(value);
	if (!normalized) throw new Error(`${label} must be one of off, minimal, low, medium, high, xhigh, adaptive`);
	return normalized;
}
function parseQaModelThinkingOverrides(entries) {
	const overrides = {};
	for (const entry of entries ?? []) {
		const separatorIndex = entry.lastIndexOf("=");
		if (separatorIndex <= 0 || separatorIndex === entry.length - 1) throw new Error(`--model-thinking must use provider/model=level, got "${entry}"`);
		const model = entry.slice(0, separatorIndex).trim();
		const level = parseQaThinkingLevel("--model-thinking", entry.slice(separatorIndex + 1).trim());
		if (!model || !level) throw new Error(`--model-thinking must use provider/model=level, got "${entry}"`);
		overrides[model] = level;
	}
	return Object.keys(overrides).length > 0 ? overrides : void 0;
}
function parseQaBooleanModelOption(label, value) {
	switch (value.trim().toLowerCase()) {
		case "1":
		case "on":
		case "true":
		case "yes": return true;
		case "0":
		case "false":
		case "no":
		case "off": return false;
		default: throw new Error(`${label} fast must be one of true, false, on, off, yes, no, 1, 0`);
	}
}
function parseQaPositiveIntegerOption(label, value) {
	if (value === void 0) return;
	if (!Number.isFinite(value) || value < 1) throw new Error(`${label} must be a positive integer`);
	return Math.floor(value);
}
function parseQaModelSpecs(label, entries) {
	const models = [];
	const optionsByModel = {};
	for (const entry of entries ?? []) {
		const parts = entry.split(",").map((part) => part.trim());
		const model = parts[0];
		if (!model) throw new Error(`${label} must start with provider/model, got "${entry}"`);
		models.push(model);
		const options = {};
		for (const part of parts.slice(1)) {
			if (!part) throw new Error(`${label} option cannot be empty in "${entry}"`);
			if (part === "fast") {
				options.fastMode = true;
				continue;
			}
			if (part === "no-fast") {
				options.fastMode = false;
				continue;
			}
			const separatorIndex = part.indexOf("=");
			if (separatorIndex <= 0 || separatorIndex === part.length - 1) throw new Error(`${label} options must be thinking=<level>, fast, no-fast, or fast=<boolean>, got "${part}"`);
			const key = part.slice(0, separatorIndex).trim();
			const value = part.slice(separatorIndex + 1).trim();
			switch (key) {
				case "thinking": {
					const thinkingDefault = parseQaThinkingLevel(`${label} thinking`, value);
					if (!thinkingDefault) throw new Error(`${label} thinking must be one of off, minimal, low, medium, high, xhigh, adaptive`);
					options.thinkingDefault = thinkingDefault;
					break;
				}
				case "fast":
					options.fastMode = parseQaBooleanModelOption(label, value);
					break;
				default: throw new Error(`${label} does not support option "${key}" in "${entry}"`);
			}
		}
		if (Object.keys(options).length > 0) optionsByModel[model] = {
			...optionsByModel[model],
			...options
		};
	}
	return {
		models,
		optionsByModel: Object.keys(optionsByModel).length > 0 ? optionsByModel : void 0
	};
}
async function runInterruptibleServer(label, server) {
	process.stdout.write(`${label}: ${server.baseUrl}\n`);
	process.stdout.write("Press Ctrl+C to stop.\n");
	const shutdown = async () => {
		process.off("SIGINT", onSignal);
		process.off("SIGTERM", onSignal);
		await server.stop();
		process.exit(0);
	};
	const onSignal = () => {
		shutdown();
	};
	process.on("SIGINT", onSignal);
	process.on("SIGTERM", onSignal);
	await new Promise(() => void 0);
}
async function runQaLabSelfCheckCommand(opts) {
	const repoRoot = path.resolve(opts.repoRoot ?? process.cwd());
	const server = await startQaLabServer({
		repoRoot,
		outputPath: opts.output ? path.resolve(repoRoot, opts.output) : void 0
	});
	try {
		const result = await server.runSelfCheck();
		process.stdout.write(`QA self-check report: ${result.outputPath}\n`);
	} finally {
		await server.stop();
	}
}
async function runQaSuiteCommand(opts) {
	const repoRoot = path.resolve(opts.repoRoot ?? process.cwd());
	const providerMode = normalizeQaProviderMode(opts.providerMode);
	const result = await runQaSuite({
		repoRoot,
		outputDir: opts.outputDir ? path.resolve(repoRoot, opts.outputDir) : void 0,
		providerMode,
		primaryModel: opts.primaryModel,
		alternateModel: opts.alternateModel,
		fastMode: opts.fastMode,
		scenarioIds: opts.scenarioIds
	});
	process.stdout.write(`QA suite watch: ${result.watchUrl}\n`);
	process.stdout.write(`QA suite report: ${result.reportPath}\n`);
	process.stdout.write(`QA suite summary: ${result.summaryPath}\n`);
}
async function runQaCharacterEvalCommand(opts) {
	const repoRoot = path.resolve(opts.repoRoot ?? process.cwd());
	const candidates = parseQaModelSpecs("--model", opts.model);
	const judges = parseQaModelSpecs("--judge-model", opts.judgeModel);
	const result = await runQaCharacterEval({
		repoRoot,
		outputDir: opts.outputDir ? path.resolve(repoRoot, opts.outputDir) : void 0,
		models: candidates.models,
		scenarioId: opts.scenario,
		candidateFastMode: opts.fast,
		candidateThinkingDefault: parseQaThinkingLevel("--thinking", opts.thinking),
		candidateThinkingByModel: parseQaModelThinkingOverrides(opts.modelThinking),
		candidateModelOptions: candidates.optionsByModel,
		judgeModels: judges.models.length > 0 ? judges.models : void 0,
		judgeModelOptions: judges.optionsByModel,
		judgeTimeoutMs: opts.judgeTimeoutMs,
		judgeBlindModels: opts.blindJudgeModels === true ? true : void 0,
		candidateConcurrency: parseQaPositiveIntegerOption("--concurrency", opts.concurrency),
		judgeConcurrency: parseQaPositiveIntegerOption("--judge-concurrency", opts.judgeConcurrency),
		progress: (message) => process.stderr.write(`${message}\n`)
	});
	process.stdout.write(`QA character eval report: ${result.reportPath}\n`);
	process.stdout.write(`QA character eval summary: ${result.summaryPath}\n`);
}
async function runQaManualLaneCommand(opts) {
	const repoRoot = path.resolve(opts.repoRoot ?? process.cwd());
	const providerMode = opts.providerMode === void 0 ? "live-frontier" : normalizeQaProviderMode(opts.providerMode);
	const models = resolveQaManualLaneModels({
		providerMode,
		primaryModel: opts.primaryModel,
		alternateModel: opts.alternateModel
	});
	const result = await runQaManualLane({
		repoRoot,
		providerMode,
		primaryModel: models.primaryModel,
		alternateModel: models.alternateModel,
		fastMode: opts.fastMode,
		message: opts.message,
		timeoutMs: opts.timeoutMs
	});
	process.stdout.write(JSON.stringify(result, null, 2));
	process.stdout.write("\n");
}
async function runQaLabUiCommand(opts) {
	await runInterruptibleServer("QA Lab UI", await startQaLabServer({
		repoRoot: path.resolve(opts.repoRoot ?? process.cwd()),
		host: opts.host,
		port: Number.isFinite(opts.port) ? opts.port : void 0,
		advertiseHost: opts.advertiseHost,
		advertisePort: Number.isFinite(opts.advertisePort) ? opts.advertisePort : void 0,
		controlUiUrl: opts.controlUiUrl,
		controlUiToken: opts.controlUiToken,
		controlUiProxyTarget: opts.controlUiProxyTarget,
		uiDistDir: opts.uiDistDir,
		autoKickoffTarget: opts.autoKickoffTarget,
		embeddedGateway: opts.embeddedGateway,
		sendKickoffOnStart: opts.sendKickoffOnStart
	}));
}
async function runQaDockerScaffoldCommand(opts) {
	const repoRoot = path.resolve(opts.repoRoot ?? process.cwd());
	const result = await writeQaDockerHarnessFiles({
		outputDir: path.resolve(repoRoot, opts.outputDir),
		repoRoot,
		gatewayPort: Number.isFinite(opts.gatewayPort) ? opts.gatewayPort : void 0,
		qaLabPort: Number.isFinite(opts.qaLabPort) ? opts.qaLabPort : void 0,
		providerBaseUrl: opts.providerBaseUrl,
		imageName: opts.image,
		usePrebuiltImage: opts.usePrebuiltImage,
		bindUiDist: opts.bindUiDist
	});
	process.stdout.write(`QA docker scaffold: ${result.outputDir}\n`);
}
async function runQaDockerBuildImageCommand(opts) {
	const result = await buildQaDockerHarnessImage({
		repoRoot: path.resolve(opts.repoRoot ?? process.cwd()),
		imageName: opts.image
	});
	process.stdout.write(`QA docker image: ${result.imageName}\n`);
}
async function runQaDockerUpCommand(opts) {
	const repoRoot = path.resolve(opts.repoRoot ?? process.cwd());
	const result = await runQaDockerUp({
		repoRoot,
		outputDir: opts.outputDir ? path.resolve(repoRoot, opts.outputDir) : void 0,
		gatewayPort: Number.isFinite(opts.gatewayPort) ? opts.gatewayPort : void 0,
		qaLabPort: Number.isFinite(opts.qaLabPort) ? opts.qaLabPort : void 0,
		providerBaseUrl: opts.providerBaseUrl,
		image: opts.image,
		usePrebuiltImage: opts.usePrebuiltImage,
		bindUiDist: opts.bindUiDist,
		skipUiBuild: opts.skipUiBuild
	});
	process.stdout.write(`QA docker dir: ${result.outputDir}\n`);
	process.stdout.write(`QA Lab UI: ${result.qaLabUrl}\n`);
	process.stdout.write(`Gateway UI: ${result.gatewayUrl}\n`);
	process.stdout.write(`Stop: ${result.stopCommand}\n`);
}
async function runQaMockOpenAiCommand(opts) {
	await runInterruptibleServer("QA mock OpenAI", await startQaMockOpenAiServer({
		host: opts.host,
		port: Number.isFinite(opts.port) ? opts.port : void 0
	}));
}
//#endregion
export { runQaCharacterEvalCommand, runQaDockerBuildImageCommand, runQaDockerScaffoldCommand, runQaDockerUpCommand, runQaLabSelfCheckCommand, runQaLabUiCommand, runQaManualLaneCommand, runQaMockOpenAiCommand, runQaSuiteCommand };
