// openclaw/dist/comment-shared-OAmnMRRw.js
import { d as readStringValue, s as normalizeOptionalString, t as hasNonEmptyString$1 } from "./string-coerce-CIT1-v_a.js";
import { l as isRecord$1 } from "./utils-Supr9KvU.js";
import { i as asOptionalRecord } from "./record-coerce-BDH91-dk.js";
import "./text-runtime-BngW26kR.js";
//#region extensions/feishu/src/comment-shared.ts
function encodeQuery(params) {
	const query = new URLSearchParams();
	for (const [key, value] of Object.entries(params)) {
		const trimmed = value?.trim();
		if (trimmed) query.set(key, trimmed);
	}
	const queryString = query.toString();
	return queryString ? `?${queryString}` : "";
}
const readString = readStringValue;
const normalizeString = normalizeOptionalString;
const isRecord = isRecord$1;
const asRecord = asOptionalRecord;
const hasNonEmptyString = hasNonEmptyString$1;
function extractCommentElementText(element) {
	if (!isRecord(element)) return;
	const type = normalizeString(element.type);
	if (type === "text_run" && isRecord(element.text_run)) return normalizeString(element.text_run.content) || normalizeString(element.text_run.text);
	if (type === "mention") {
		const mention = isRecord(element.mention) ? element.mention : void 0;
		const mentionName = normalizeString(mention?.name) || normalizeString(mention?.display_name) || normalizeString(element.name);
		return mentionName ? `@${mentionName}` : "@mention";
	}
	if (type === "docs_link") {
		const docsLink = isRecord(element.docs_link) ? element.docs_link : void 0;
		return normalizeString(docsLink?.text) || normalizeString(docsLink?.url) || normalizeString(element.text) || normalizeString(element.url) || void 0;
	}
	return normalizeString(element.text) || normalizeString(element.content) || normalizeString(element.name) || void 0;
}
function extractReplyText(reply) {
	if (!reply || !isRecord(reply.content)) return;
	return (Array.isArray(reply.content.elements) ? reply.content.elements : []).map(extractCommentElementText).filter((part) => Boolean(part && part.trim())).join("").trim() || void 0;
}
//#endregion
export { isRecord as a, hasNonEmptyString as i, encodeQuery as n, normalizeString as o, extractReplyText as r, readString as s, asRecord as t };
