// openclaw/dist/config-presence-DX3v0oFJ.js
import { l as isRecord } from "./utils-Supr9KvU.js";
import { _ as resolveStateDir } from "./paths-nQmEmngS.js";
import { n as listBundledChannelIdsWithPersistedAuthState, t as hasBundledChannelPersistedAuthState } from "./persisted-auth-state-C_USvqgI.js";
import { i as hasNonEmptyString, u as listBundledChannelPluginIds } from "./channel-target-BYsT_Fvn.js";
import fs from "node:fs";
import os from "node:os";
//#region src/channels/config-presence.ts
const IGNORED_CHANNEL_CONFIG_KEYS = new Set(["defaults", "modelByChannel"]);
function hasMeaningfulChannelConfig(value) {
	if (!isRecord(value)) return false;
	return Object.keys(value).some((key) => key !== "enabled");
}
function listChannelEnvPrefixes(channelIds) {
	return channelIds.map((channelId) => [`${channelId.replace(/[^a-z0-9]+/gi, "_").toUpperCase()}_`, channelId]);
}
function hasPersistedChannelState(env) {
	return fs.existsSync(resolveStateDir(env, os.homedir));
}
const PERSISTED_AUTH_STATE_CHANNEL_IDS = listBundledChannelIdsWithPersistedAuthState();
function listPotentialConfiguredChannelIds(cfg, env = process.env, options = {}) {
	const configuredChannelIds = /* @__PURE__ */ new Set();
	const channelEnvPrefixes = listChannelEnvPrefixes(listBundledChannelPluginIds());
	const channels = isRecord(cfg.channels) ? cfg.channels : null;
	if (channels) for (const [key, value] of Object.entries(channels)) {
		if (IGNORED_CHANNEL_CONFIG_KEYS.has(key)) continue;
		if (hasMeaningfulChannelConfig(value)) configuredChannelIds.add(key);
	}
	for (const [key, value] of Object.entries(env)) {
		if (!hasNonEmptyString(value)) continue;
		for (const [prefix, channelId] of channelEnvPrefixes) if (key.startsWith(prefix)) configuredChannelIds.add(channelId);
	}
	if (options.includePersistedAuthState !== false && hasPersistedChannelState(env)) {
		for (const channelId of PERSISTED_AUTH_STATE_CHANNEL_IDS) if (hasBundledChannelPersistedAuthState({
			channelId,
			cfg,
			env
		})) configuredChannelIds.add(channelId);
	}
	return [...configuredChannelIds];
}
function hasEnvConfiguredChannel(cfg, env, options = {}) {
	const channelEnvPrefixes = listChannelEnvPrefixes(listBundledChannelPluginIds());
	for (const [key, value] of Object.entries(env)) {
		if (!hasNonEmptyString(value)) continue;
		if (channelEnvPrefixes.some(([prefix]) => key.startsWith(prefix))) return true;
	}
	if (options.includePersistedAuthState === false || !hasPersistedChannelState(env)) return false;
	return PERSISTED_AUTH_STATE_CHANNEL_IDS.some((channelId) => hasBundledChannelPersistedAuthState({
		channelId,
		cfg,
		env
	}));
}
function hasPotentialConfiguredChannels(cfg, env = process.env, options = {}) {
	const channels = isRecord(cfg?.channels) ? cfg.channels : null;
	if (channels) for (const [key, value] of Object.entries(channels)) {
		if (IGNORED_CHANNEL_CONFIG_KEYS.has(key)) continue;
		if (hasMeaningfulChannelConfig(value)) return true;
	}
	return hasEnvConfiguredChannel(cfg ?? {}, env, options);
}
//#endregion
export { listPotentialConfiguredChannelIds as n, hasPotentialConfiguredChannels as t };
