// openclaw/dist/conversation-binding-context-IXzVm0Hi.js
import { i as normalizeLowercaseStringOrEmpty, o as normalizeOptionalLowercaseString, s as normalizeOptionalString } from "./string-coerce-CIT1-v_a.js";
import { t as getActivePluginChannelRegistry } from "./runtime-D85K4f41.js";
import { C as normalizeChannelId, S as normalizeAnyChannelId } from "./message-channel-BzmQarxU.js";
import { n as parseExplicitTargetForChannel } from "./target-parsing-B0rjiTg_.js";
import { t as resolveConversationIdFromTargets } from "./conversation-id-CZWjKl2x.js";
//#region src/channels/conversation-binding-context.ts
const CANONICAL_TARGET_PREFIXES = [
	"user:",
	"channel:",
	"conversation:",
	"group:",
	"room:",
	"dm:",
	"spaces/"
];
function getLoadedChannelPlugin(rawChannel) {
	const normalized = normalizeAnyChannelId(rawChannel) ?? normalizeOptionalString(rawChannel);
	if (!normalized) return;
	return getActivePluginChannelRegistry()?.channels.find((entry) => entry.plugin.id === normalized)?.plugin;
}
function shouldDefaultParentConversationToSelf(plugin) {
	return plugin?.bindings?.selfParentConversationByDefault === true;
}
function resolveBindingAccountId(params) {
	return normalizeOptionalString(params.rawAccountId) || normalizeOptionalString(params.plugin?.config.defaultAccountId?.(params.cfg)) || "default";
}
function resolveChannelTargetId(params) {
	const target = normalizeOptionalString(params.target);
	if (!target) return;
	const lower = normalizeLowercaseStringOrEmpty(target);
	const channelPrefix = `${params.channel}:`;
	if (lower.startsWith(channelPrefix)) return resolveChannelTargetId({
		channel: params.channel,
		target: target.slice(channelPrefix.length)
	});
	if (CANONICAL_TARGET_PREFIXES.some((prefix) => lower.startsWith(prefix))) return target;
	const parsedTarget = normalizeOptionalString(parseExplicitTargetForChannel(params.channel, target)?.to);
	if (parsedTarget) return resolveConversationIdFromTargets({ targets: [parsedTarget] }) ?? parsedTarget;
	return resolveConversationIdFromTargets({ targets: [target] }) ?? target;
}
function buildThreadingContext(params) {
	const to = normalizeOptionalString(params.originatingTo) ?? normalizeOptionalString(params.fallbackTo);
	return {
		...to ? { To: to } : {},
		...params.from ? { From: params.from } : {},
		...params.chatType ? { ChatType: params.chatType } : {},
		...params.threadId ? { MessageThreadId: params.threadId } : {},
		...params.nativeChannelId ? { NativeChannelId: params.nativeChannelId } : {}
	};
}
function resolveConversationBindingContext(params) {
	const channel = normalizeAnyChannelId(params.channel) ?? normalizeChannelId(params.channel) ?? normalizeOptionalLowercaseString(params.channel);
	if (!channel) return null;
	const loadedPlugin = getLoadedChannelPlugin(channel);
	const accountId = resolveBindingAccountId({
		rawAccountId: params.accountId,
		plugin: loadedPlugin,
		cfg: params.cfg
	});
	const threadId = normalizeOptionalString(params.threadId != null ? String(params.threadId) : void 0);
	const resolvedByProvider = loadedPlugin?.bindings?.resolveCommandConversation?.({
		accountId,
		threadId,
		threadParentId: normalizeOptionalString(params.threadParentId),
		senderId: normalizeOptionalString(params.senderId),
		sessionKey: normalizeOptionalString(params.sessionKey),
		parentSessionKey: normalizeOptionalString(params.parentSessionKey),
		from: normalizeOptionalString(params.from),
		chatType: normalizeOptionalString(params.chatType),
		originatingTo: params.originatingTo ?? void 0,
		commandTo: params.commandTo ?? void 0,
		fallbackTo: params.fallbackTo ?? void 0
	});
	if (resolvedByProvider?.conversationId) {
		const resolvedParentConversationId = shouldDefaultParentConversationToSelf(loadedPlugin) && !threadId && !resolvedByProvider.parentConversationId ? resolvedByProvider.conversationId : resolvedByProvider.parentConversationId;
		return {
			channel,
			accountId,
			conversationId: resolvedByProvider.conversationId,
			...resolvedParentConversationId ? { parentConversationId: resolvedParentConversationId } : {},
			...threadId ? { threadId } : {}
		};
	}
	const focusedBinding = loadedPlugin?.threading?.resolveFocusedBinding?.({
		cfg: params.cfg,
		accountId,
		context: buildThreadingContext({
			fallbackTo: params.fallbackTo ?? void 0,
			originatingTo: params.originatingTo ?? void 0,
			threadId,
			from: normalizeOptionalString(params.from),
			chatType: normalizeOptionalString(params.chatType),
			nativeChannelId: normalizeOptionalString(params.nativeChannelId)
		})
	});
	if (focusedBinding?.conversationId) return {
		channel,
		accountId,
		conversationId: focusedBinding.conversationId,
		...focusedBinding.parentConversationId ? { parentConversationId: focusedBinding.parentConversationId } : {},
		...threadId ? { threadId } : {}
	};
	const baseConversationId = resolveChannelTargetId({
		channel,
		target: params.originatingTo
	}) ?? resolveChannelTargetId({
		channel,
		target: params.commandTo
	}) ?? resolveChannelTargetId({
		channel,
		target: params.fallbackTo
	});
	const parentConversationId = resolveChannelTargetId({
		channel,
		target: params.threadParentId
	}) ?? (threadId && baseConversationId && baseConversationId !== threadId ? baseConversationId : void 0);
	const conversationId = threadId || baseConversationId;
	if (!conversationId) return null;
	const normalizedParentConversationId = shouldDefaultParentConversationToSelf(loadedPlugin) && !threadId && !parentConversationId ? conversationId : parentConversationId;
	return {
		channel,
		accountId,
		conversationId,
		...normalizedParentConversationId ? { parentConversationId: normalizedParentConversationId } : {},
		...threadId ? { threadId } : {}
	};
}
//#endregion
export { resolveConversationBindingContext as t };
