// openclaw/dist/discovery-mFzRUsf-.js
import { r as formatErrorMessage } from "./errors-CGMew-1w.js";
import { i as normalizeLowercaseStringOrEmpty, o as normalizeOptionalLowercaseString } from "./string-coerce-CIT1-v_a.js";
import { t as createSubsystemLogger } from "./subsystem-C1arrdPy.js";
import { n as resolveAwsSdkEnvVarName } from "./model-auth-runtime-shared-DWhBSJYe.js";
import "./text-runtime-BngW26kR.js";
import "./core-CkOYV1-a.js";
import "./error-runtime-BNzEFRqy.js";
import "./provider-auth-runtime-tmnEM73m.js";
import { BedrockClient, ListFoundationModelsCommand, ListInferenceProfilesCommand } from "@aws-sdk/client-bedrock";
//#region extensions/amazon-bedrock/discovery.ts
const log = createSubsystemLogger("bedrock-discovery");
const DEFAULT_REFRESH_INTERVAL_SECONDS = 3600;
const DEFAULT_CONTEXT_WINDOW = 32e3;
const DEFAULT_MAX_TOKENS = 4096;
const DEFAULT_COST = {
	input: 0,
	output: 0,
	cacheRead: 0,
	cacheWrite: 0
};
const discoveryCache = /* @__PURE__ */ new Map();
let hasLoggedBedrockError = false;
function normalizeProviderFilter(filter) {
	if (!filter || filter.length === 0) return [];
	const normalized = new Set(filter.map((entry) => normalizeOptionalLowercaseString(entry)).filter((entry) => Boolean(entry)));
	return Array.from(normalized).toSorted();
}
function buildCacheKey(params) {
	return JSON.stringify(params);
}
function includesTextModalities(modalities) {
	return (modalities ?? []).some((entry) => normalizeOptionalLowercaseString(entry) === "text");
}
function isActive(summary) {
	const status = summary.modelLifecycle?.status;
	return typeof status === "string" ? status.toUpperCase() === "ACTIVE" : false;
}
function mapInputModalities(summary) {
	const inputs = summary.inputModalities ?? [];
	const mapped = /* @__PURE__ */ new Set();
	for (const modality of inputs) {
		const lower = normalizeOptionalLowercaseString(modality);
		if (lower === "text") mapped.add("text");
		if (lower === "image") mapped.add("image");
	}
	if (mapped.size === 0) mapped.add("text");
	return Array.from(mapped);
}
function inferReasoningSupport(summary) {
	const haystack = normalizeLowercaseStringOrEmpty(`${summary.modelId ?? ""} ${summary.modelName ?? ""}`);
	return haystack.includes("reasoning") || haystack.includes("thinking");
}
function resolveDefaultContextWindow(config) {
	const value = Math.floor(config?.defaultContextWindow ?? DEFAULT_CONTEXT_WINDOW);
	return value > 0 ? value : DEFAULT_CONTEXT_WINDOW;
}
function resolveDefaultMaxTokens(config) {
	const value = Math.floor(config?.defaultMaxTokens ?? DEFAULT_MAX_TOKENS);
	return value > 0 ? value : DEFAULT_MAX_TOKENS;
}
function matchesProviderFilter(summary, filter) {
	if (filter.length === 0) return true;
	const normalized = normalizeOptionalLowercaseString(summary.providerName ?? (typeof summary.modelId === "string" ? summary.modelId.split(".")[0] : void 0));
	if (!normalized) return false;
	return filter.includes(normalized);
}
function shouldIncludeSummary(summary, filter) {
	if (!summary.modelId?.trim()) return false;
	if (!matchesProviderFilter(summary, filter)) return false;
	if (summary.responseStreamingSupported !== true) return false;
	if (!includesTextModalities(summary.outputModalities)) return false;
	if (!isActive(summary)) return false;
	return true;
}
function toModelDefinition(summary, defaults) {
	const id = summary.modelId?.trim() ?? "";
	return {
		id,
		name: summary.modelName?.trim() || id,
		reasoning: inferReasoningSupport(summary),
		input: mapInputModalities(summary),
		cost: DEFAULT_COST,
		contextWindow: defaults.contextWindow,
		maxTokens: defaults.maxTokens
	};
}
/**
* Resolve the base foundation model ID from an inference profile.
*
* System-defined profiles use a region prefix:
*   "us.anthropic.claude-sonnet-4-6" → "anthropic.claude-sonnet-4-6"
*
* Application profiles carry the model ARN in their models[] array:
*   models[0].modelArn = "arn:aws:bedrock:us-east-1::foundation-model/anthropic.claude-sonnet-4-6"
*   → "anthropic.claude-sonnet-4-6"
*/
function resolveBaseModelId(profile) {
	const firstArn = profile.models?.[0]?.modelArn;
	if (firstArn) {
		const arnMatch = /foundation-model\/(.+)$/.exec(firstArn);
		if (arnMatch) return arnMatch[1];
	}
	if (profile.type === "SYSTEM_DEFINED") {
		const id = profile.inferenceProfileId ?? "";
		const prefixMatch = /^(?:us|eu|ap|jp|global)\.(.+)$/i.exec(id);
		if (prefixMatch) return prefixMatch[1];
	}
}
/**
* Fetch raw inference profile summaries from the Bedrock control plane.
* Handles pagination. Best-effort: silently returns empty array if IAM lacks
* bedrock:ListInferenceProfiles permission.
*/
async function fetchInferenceProfileSummaries(client) {
	try {
		const profiles = [];
		let nextToken;
		do {
			const response = await client.send(new ListInferenceProfilesCommand({ nextToken }));
			for (const summary of response.inferenceProfileSummaries ?? []) profiles.push(summary);
			nextToken = response.nextToken;
		} while (nextToken);
		return profiles;
	} catch (error) {
		log.debug?.("Skipping inference profile discovery", { error: formatErrorMessage(error) });
		return [];
	}
}
/**
* Convert raw inference profile summaries into model definitions.
*
* Each profile inherits capabilities (modalities, reasoning, context window,
* cost) from its underlying foundation model. This ensures that
* "us.anthropic.claude-sonnet-4-6" has the same capabilities as
* "anthropic.claude-sonnet-4-6" — including image input, reasoning support,
* and token limits.
*
* When the foundation model isn't found in the map (e.g. the model is only
* available via inference profiles in this region), safe defaults are used.
*/
function resolveInferenceProfiles(profiles, defaults, providerFilter, foundationModels) {
	const discovered = [];
	for (const profile of profiles) {
		if (!profile.inferenceProfileId?.trim()) continue;
		if (profile.status !== "ACTIVE") continue;
		if (providerFilter.length > 0) {
			if (!(profile.models ?? []).some((m) => {
				const provider = m.modelArn?.split("/")?.[1]?.split(".")?.[0];
				return provider ? providerFilter.includes(normalizeOptionalLowercaseString(provider) ?? "") : false;
			})) continue;
		}
		const baseModelId = resolveBaseModelId(profile);
		const baseModel = baseModelId ? foundationModels.get(normalizeLowercaseStringOrEmpty(baseModelId)) : void 0;
		discovered.push({
			id: profile.inferenceProfileId,
			name: profile.inferenceProfileName?.trim() || profile.inferenceProfileId,
			reasoning: baseModel?.reasoning ?? false,
			input: baseModel?.input ?? ["text"],
			cost: baseModel?.cost ?? DEFAULT_COST,
			contextWindow: baseModel?.contextWindow ?? defaults.contextWindow,
			maxTokens: baseModel?.maxTokens ?? defaults.maxTokens
		});
	}
	return discovered;
}
function resetBedrockDiscoveryCacheForTest() {
	discoveryCache.clear();
	hasLoggedBedrockError = false;
}
function resolveBedrockConfigApiKey(env = process.env) {
	return resolveAwsSdkEnvVarName(env);
}
async function discoverBedrockModels(params) {
	const refreshIntervalSeconds = Math.max(0, Math.floor(params.config?.refreshInterval ?? DEFAULT_REFRESH_INTERVAL_SECONDS));
	const providerFilter = normalizeProviderFilter(params.config?.providerFilter);
	const defaultContextWindow = resolveDefaultContextWindow(params.config);
	const defaultMaxTokens = resolveDefaultMaxTokens(params.config);
	const cacheKey = buildCacheKey({
		region: params.region,
		providerFilter,
		refreshIntervalSeconds,
		defaultContextWindow,
		defaultMaxTokens
	});
	const now = params.now?.() ?? Date.now();
	if (refreshIntervalSeconds > 0) {
		const cached = discoveryCache.get(cacheKey);
		if (cached?.value && cached.expiresAt > now) return cached.value;
		if (cached?.inFlight) return cached.inFlight;
	}
	const client = (params.clientFactory ?? ((region) => new BedrockClient({ region })))(params.region);
	const discoveryPromise = (async () => {
		const [foundationResponse, profileSummaries] = await Promise.all([client.send(new ListFoundationModelsCommand({})), fetchInferenceProfileSummaries(client)]);
		const discovered = [];
		const seenIds = /* @__PURE__ */ new Set();
		const foundationModels = /* @__PURE__ */ new Map();
		for (const summary of foundationResponse.modelSummaries ?? []) {
			if (!shouldIncludeSummary(summary, providerFilter)) continue;
			const def = toModelDefinition(summary, {
				contextWindow: defaultContextWindow,
				maxTokens: defaultMaxTokens
			});
			discovered.push(def);
			const normalizedId = normalizeLowercaseStringOrEmpty(def.id);
			seenIds.add(normalizedId);
			foundationModels.set(normalizedId, def);
		}
		const inferenceProfiles = resolveInferenceProfiles(profileSummaries, {
			contextWindow: defaultContextWindow,
			maxTokens: defaultMaxTokens
		}, providerFilter, foundationModels);
		for (const profile of inferenceProfiles) {
			const normalizedId = normalizeLowercaseStringOrEmpty(profile.id);
			if (!seenIds.has(normalizedId)) {
				discovered.push(profile);
				seenIds.add(normalizedId);
			}
		}
		return discovered.toSorted((a, b) => {
			const aGlobal = a.id.startsWith("global.") ? 0 : 1;
			const bGlobal = b.id.startsWith("global.") ? 0 : 1;
			if (aGlobal !== bGlobal) return aGlobal - bGlobal;
			return a.name.localeCompare(b.name);
		});
	})();
	if (refreshIntervalSeconds > 0) discoveryCache.set(cacheKey, {
		expiresAt: now + refreshIntervalSeconds * 1e3,
		inFlight: discoveryPromise
	});
	try {
		const value = await discoveryPromise;
		if (refreshIntervalSeconds > 0) discoveryCache.set(cacheKey, {
			expiresAt: now + refreshIntervalSeconds * 1e3,
			value
		});
		return value;
	} catch (error) {
		if (refreshIntervalSeconds > 0) discoveryCache.delete(cacheKey);
		if (!hasLoggedBedrockError) {
			hasLoggedBedrockError = true;
			log.warn("Failed to discover Bedrock models", { error: formatErrorMessage(error) });
		}
		return [];
	}
}
async function resolveImplicitBedrockProvider(params) {
	const env = params.env ?? process.env;
	const discoveryConfig = {
		...params.config?.models?.bedrockDiscovery,
		...params.pluginConfig?.discovery
	};
	const enabled = discoveryConfig?.enabled;
	const hasAwsCreds = resolveAwsSdkEnvVarName(env) !== void 0;
	if (enabled === false) return null;
	if (enabled !== true && !hasAwsCreds) return null;
	const region = discoveryConfig?.region ?? env.AWS_REGION ?? env.AWS_DEFAULT_REGION ?? "us-east-1";
	const models = await discoverBedrockModels({
		region,
		config: discoveryConfig,
		clientFactory: params.clientFactory
	});
	if (models.length === 0) return null;
	return {
		baseUrl: `https://bedrock-runtime.${region}.amazonaws.com`,
		api: "bedrock-converse-stream",
		auth: "aws-sdk",
		models
	};
}
function mergeImplicitBedrockProvider(params) {
	const { existing, implicit } = params;
	if (!existing) return implicit;
	return {
		...implicit,
		...existing,
		models: Array.isArray(existing.models) && existing.models.length > 0 ? existing.models : implicit.models
	};
}
//#endregion
export { resolveImplicitBedrockProvider as a, resolveBedrockConfigApiKey as i, mergeImplicitBedrockProvider as n, resetBedrockDiscoveryCacheForTest as r, discoverBedrockModels as t };
