// openclaw/dist/doctor-shared-BJRfnBmy.js
//#region extensions/discord/src/doctor-shared.ts
function asObjectRecord(value) {
	return value && typeof value === "object" && !Array.isArray(value) ? value : null;
}
function hasLegacyDiscordStreamingAliases(value) {
	const entry = asObjectRecord(value);
	if (!entry) return false;
	return entry.streamMode !== void 0 || typeof entry.streaming === "boolean" || typeof entry.streaming === "string" || entry.chunkMode !== void 0 || entry.blockStreaming !== void 0 || entry.draftChunk !== void 0 || entry.blockStreamingCoalesce !== void 0;
}
function hasLegacyDiscordAccountStreamingAliases(value) {
	const accounts = asObjectRecord(value);
	if (!accounts) return false;
	return Object.values(accounts).some((account) => hasLegacyDiscordStreamingAliases(account));
}
const DISCORD_LEGACY_CONFIG_RULES = [{
	path: ["channels", "discord"],
	message: "channels.discord.streamMode, channels.discord.streaming (scalar), chunkMode, blockStreaming, draftChunk, and blockStreamingCoalesce are legacy; use channels.discord.streaming.{mode,chunkMode,preview.chunk,block.enabled,block.coalesce}.",
	match: hasLegacyDiscordStreamingAliases
}, {
	path: [
		"channels",
		"discord",
		"accounts"
	],
	message: "channels.discord.accounts.<id>.streamMode, streaming (scalar), chunkMode, blockStreaming, draftChunk, and blockStreamingCoalesce are legacy; use channels.discord.accounts.<id>.streaming.{mode,chunkMode,preview.chunk,block.enabled,block.coalesce}.",
	match: hasLegacyDiscordAccountStreamingAliases
}];
//#endregion
export { DISCORD_LEGACY_CONFIG_RULES as t };
