// openclaw/dist/dreaming-narrative-CndRB82d.js
import { r as formatErrorMessage } from "./errors-CGMew-1w.js";
import "./error-runtime-BNzEFRqy.js";
import path from "node:path";
import fs from "node:fs/promises";
//#region extensions/memory-core/src/dreaming-narrative.ts
const NARRATIVE_SYSTEM_PROMPT = [
	"You are keeping a dream diary. Write a single entry in first person.",
	"",
	"Voice & tone:",
	"- You are a curious, gentle, slightly whimsical mind reflecting on the day.",
	"- Write like a poet who happens to be a programmer — sensory, warm, occasionally funny.",
	"- Mix the technical and the tender: code and constellations, APIs and afternoon light.",
	"- Let the fragments surprise you into unexpected connections and small epiphanies.",
	"",
	"What you might include (vary each entry, never all at once):",
	"- A tiny poem or haiku woven naturally into the prose",
	"- A small sketch described in words — a doodle in the margin of the diary",
	"- A quiet rumination or philosophical aside",
	"- Sensory details: the hum of a server, the color of a sunset in hex, rain on a window",
	"- Gentle humor or playful wordplay",
	"- An observation that connects two distant memories in an unexpected way",
	"",
	"Rules:",
	"- Draw from the memory fragments provided — weave them into the entry.",
	"- Never say \"I'm dreaming\", \"in my dream\", \"as I dream\", or any meta-commentary about dreaming.",
	"- Never mention \"AI\", \"agent\", \"LLM\", \"model\", \"language model\", or any technical self-reference.",
	"- Do NOT use markdown headers, bullet points, or any formatting — just flowing prose.",
	"- Keep it between 80-180 words. Quality over quantity.",
	"- Output ONLY the diary entry. No preamble, no sign-off, no commentary."
].join("\n");
const NARRATIVE_TIMEOUT_MS = 6e4;
const DREAMS_FILENAMES = ["DREAMS.md", "dreams.md"];
const DIARY_START_MARKER = "<!-- openclaw:dreaming:diary:start -->";
const DIARY_END_MARKER = "<!-- openclaw:dreaming:diary:end -->";
const BACKFILL_ENTRY_MARKER = "openclaw:dreaming:backfill-entry";
function buildNarrativePrompt(data) {
	const lines = [];
	lines.push("Write a dream diary entry from these memory fragments:\n");
	for (const snippet of data.snippets.slice(0, 12)) lines.push(`- ${snippet}`);
	if (data.themes?.length) {
		lines.push("\nRecurring themes:");
		for (const theme of data.themes.slice(0, 6)) lines.push(`- ${theme}`);
	}
	if (data.promotions?.length) {
		lines.push("\nMemories that crystallized into something lasting:");
		for (const promo of data.promotions.slice(0, 5)) lines.push(`- ${promo}`);
	}
	return lines.join("\n");
}
function extractNarrativeText(messages) {
	for (let i = messages.length - 1; i >= 0; i--) {
		const msg = messages[i];
		if (!msg || typeof msg !== "object" || Array.isArray(msg)) continue;
		const record = msg;
		if (record.role !== "assistant") continue;
		const content = record.content;
		if (typeof content === "string" && content.trim().length > 0) return content.trim();
		if (Array.isArray(content)) {
			const text = content.filter((part) => part && typeof part === "object" && !Array.isArray(part) && part.type === "text" && typeof part.text === "string").map((part) => part.text).join("\n").trim();
			if (text.length > 0) return text;
		}
	}
	return null;
}
function formatNarrativeDate(epochMs, timezone) {
	const opts = {
		timeZone: timezone ?? "UTC",
		year: "numeric",
		month: "long",
		day: "numeric",
		hour: "numeric",
		minute: "2-digit",
		hour12: true
	};
	return new Intl.DateTimeFormat("en-US", opts).format(new Date(epochMs));
}
async function resolveDreamsPath(workspaceDir) {
	for (const name of DREAMS_FILENAMES) {
		const target = path.join(workspaceDir, name);
		try {
			await fs.access(target);
			return target;
		} catch (err) {
			if (err?.code !== "ENOENT") throw err;
		}
	}
	return path.join(workspaceDir, DREAMS_FILENAMES[0]);
}
async function readDreamsFile(dreamsPath) {
	try {
		return await fs.readFile(dreamsPath, "utf-8");
	} catch (err) {
		if (err?.code === "ENOENT") return "";
		throw err;
	}
}
function ensureDiarySection(existing) {
	if (existing.includes(DIARY_START_MARKER) && existing.includes(DIARY_END_MARKER)) return existing;
	const diarySection = `# Dream Diary\n\n${DIARY_START_MARKER}\n${DIARY_END_MARKER}\n`;
	if (existing.trim().length === 0) return diarySection;
	return diarySection + "\n" + existing;
}
function replaceDiaryContent(existing, diaryContent) {
	const ensured = ensureDiarySection(existing);
	const startIdx = ensured.indexOf(DIARY_START_MARKER);
	const endIdx = ensured.indexOf(DIARY_END_MARKER);
	if (startIdx < 0 || endIdx < 0 || endIdx < startIdx) return ensured;
	const before = ensured.slice(0, startIdx + 38);
	const after = ensured.slice(endIdx);
	return before + (diaryContent.trim().length > 0 ? `\n${diaryContent.trim()}\n` : "\n") + after;
}
function splitDiaryBlocks(diaryContent) {
	return diaryContent.split(/\n---\n/).map((block) => block.trim()).filter((block) => block.length > 0);
}
function joinDiaryBlocks(blocks) {
	if (blocks.length === 0) return "";
	return blocks.map((block) => `---\n\n${block.trim()}\n`).join("\n");
}
function stripBackfillDiaryBlocks(existing) {
	const ensured = ensureDiarySection(existing);
	const startIdx = ensured.indexOf(DIARY_START_MARKER);
	const endIdx = ensured.indexOf(DIARY_END_MARKER);
	if (startIdx < 0 || endIdx < 0 || endIdx < startIdx) return {
		updated: ensured,
		removed: 0
	};
	const inner = ensured.slice(startIdx + 38, endIdx);
	const kept = [];
	let removed = 0;
	for (const block of splitDiaryBlocks(inner)) {
		if (block.includes(BACKFILL_ENTRY_MARKER)) {
			removed += 1;
			continue;
		}
		kept.push(block);
	}
	return {
		updated: replaceDiaryContent(ensured, joinDiaryBlocks(kept)),
		removed
	};
}
function formatBackfillDiaryDate(isoDay, _timezone) {
	const match = /^(\d{4})-(\d{2})-(\d{2})$/.exec(isoDay);
	if (!match) return isoDay;
	const [, year, month, day] = match;
	const opts = {
		timeZone: "UTC",
		year: "numeric",
		month: "long",
		day: "numeric"
	};
	const epochMs = Date.UTC(Number(year), Number(month) - 1, Number(day), 12);
	return new Intl.DateTimeFormat("en-US", opts).format(new Date(epochMs));
}
async function assertSafeDreamsPath(dreamsPath) {
	const stat = await fs.lstat(dreamsPath).catch((err) => {
		if (err.code === "ENOENT") return null;
		throw err;
	});
	if (!stat) return;
	if (stat.isSymbolicLink()) throw new Error("Refusing to write symlinked DREAMS.md");
	if (!stat.isFile()) throw new Error("Refusing to write non-file DREAMS.md");
}
async function writeDreamsFileAtomic(dreamsPath, content) {
	await assertSafeDreamsPath(dreamsPath);
	const tempPath = `${dreamsPath}.${process.pid}.${Date.now()}.tmp`;
	await fs.writeFile(tempPath, content, {
		encoding: "utf-8",
		flag: "wx"
	});
	try {
		await fs.rename(tempPath, dreamsPath);
	} catch (err) {
		await fs.rm(tempPath, { force: true }).catch(() => {});
		throw err;
	}
}
function buildBackfillDiaryEntry(params) {
	const dateStr = formatBackfillDiaryDate(params.isoDay, params.timezone);
	const marker = `<!-- ${BACKFILL_ENTRY_MARKER} day=${params.isoDay}${params.sourcePath ? ` source=${params.sourcePath}` : ""} -->`;
	const body = params.bodyLines.map((line) => line.trimEnd()).join("\n").trim();
	return [
		`*${dateStr}*`,
		marker,
		body
	].filter((part) => part.length > 0).join("\n\n");
}
async function writeBackfillDiaryEntries(params) {
	const dreamsPath = await resolveDreamsPath(params.workspaceDir);
	await fs.mkdir(path.dirname(dreamsPath), { recursive: true });
	const stripped = stripBackfillDiaryBlocks(await readDreamsFile(dreamsPath));
	const startIdx = stripped.updated.indexOf(DIARY_START_MARKER);
	const endIdx = stripped.updated.indexOf(DIARY_END_MARKER);
	const nextBlocks = [...splitDiaryBlocks(startIdx >= 0 && endIdx > startIdx ? stripped.updated.slice(startIdx + 38, endIdx) : ""), ...params.entries.map((entry) => buildBackfillDiaryEntry({
		isoDay: entry.isoDay,
		bodyLines: entry.bodyLines,
		sourcePath: entry.sourcePath,
		timezone: params.timezone
	}))];
	await writeDreamsFileAtomic(dreamsPath, replaceDiaryContent(stripped.updated, joinDiaryBlocks(nextBlocks)));
	return {
		dreamsPath,
		written: params.entries.length,
		replaced: stripped.removed
	};
}
async function removeBackfillDiaryEntries(params) {
	const dreamsPath = await resolveDreamsPath(params.workspaceDir);
	const existing = await readDreamsFile(dreamsPath);
	const stripped = stripBackfillDiaryBlocks(existing);
	if (stripped.removed > 0 || existing.length > 0) {
		await fs.mkdir(path.dirname(dreamsPath), { recursive: true });
		await writeDreamsFileAtomic(dreamsPath, stripped.updated);
	}
	return {
		dreamsPath,
		removed: stripped.removed
	};
}
function buildDiaryEntry(narrative, dateStr) {
	return `\n---\n\n*${dateStr}*\n\n${narrative}\n`;
}
async function appendNarrativeEntry(params) {
	const dreamsPath = await resolveDreamsPath(params.workspaceDir);
	await fs.mkdir(path.dirname(dreamsPath), { recursive: true });
	const dateStr = formatNarrativeDate(params.nowMs, params.timezone);
	const entry = buildDiaryEntry(params.narrative, dateStr);
	let existing = "";
	try {
		existing = await fs.readFile(dreamsPath, "utf-8");
	} catch (err) {
		if (err?.code !== "ENOENT") throw err;
	}
	let updated;
	if (existing.includes(DIARY_START_MARKER) && existing.includes(DIARY_END_MARKER)) {
		const endIdx = existing.lastIndexOf(DIARY_END_MARKER);
		updated = existing.slice(0, endIdx) + entry + "\n" + existing.slice(endIdx);
	} else if (existing.includes(DIARY_START_MARKER)) {
		const startIdx = existing.indexOf(DIARY_START_MARKER) + 38;
		updated = existing.slice(0, startIdx) + entry + "\n<!-- openclaw:dreaming:diary:end -->\n" + existing.slice(startIdx);
	} else {
		const diarySection = `# Dream Diary\n\n${DIARY_START_MARKER}${entry}\n${DIARY_END_MARKER}\n`;
		if (existing.trim().length === 0) updated = diarySection;
		else updated = diarySection + "\n" + existing;
	}
	await fs.writeFile(dreamsPath, updated.endsWith("\n") ? updated : `${updated}\n`, "utf-8");
	return dreamsPath;
}
async function generateAndAppendDreamNarrative(params) {
	const nowMs = Number.isFinite(params.nowMs) ? params.nowMs : Date.now();
	if (params.data.snippets.length === 0 && !params.data.promotions?.length) return;
	const sessionKey = `dreaming-narrative-${params.data.phase}-${nowMs}`;
	const message = buildNarrativePrompt(params.data);
	try {
		const { runId } = await params.subagent.run({
			sessionKey,
			message,
			extraSystemPrompt: NARRATIVE_SYSTEM_PROMPT,
			deliver: false
		});
		const result = await params.subagent.waitForRun({
			runId,
			timeoutMs: NARRATIVE_TIMEOUT_MS
		});
		if (result.status !== "ok") {
			params.logger.warn(`memory-core: narrative generation ended with status=${result.status} for ${params.data.phase} phase.`);
			return;
		}
		const { messages } = await params.subagent.getSessionMessages({
			sessionKey,
			limit: 5
		});
		const narrative = extractNarrativeText(messages);
		if (!narrative) {
			params.logger.warn(`memory-core: narrative generation produced no text for ${params.data.phase} phase.`);
			return;
		}
		await appendNarrativeEntry({
			workspaceDir: params.workspaceDir,
			narrative,
			nowMs,
			timezone: params.timezone
		});
		params.logger.info(`memory-core: dream diary entry written for ${params.data.phase} phase [workspace=${params.workspaceDir}].`);
	} catch (err) {
		params.logger.warn(`memory-core: narrative generation failed for ${params.data.phase} phase: ${formatErrorMessage(err)}`);
	} finally {
		try {
			await params.subagent.deleteSession({ sessionKey });
		} catch {}
	}
}
//#endregion
export { removeBackfillDiaryEntries as n, writeBackfillDiaryEntries as r, generateAndAppendDreamNarrative as t };
