// openclaw/dist/googlechat-CvNmYqYc.js
import "./message-channel-BzmQarxU.js";
import "./types.secrets-D-R9X_jm.js";
import "./config-schema-CiKOZuar.js";
import "./setup-helpers-HBBqrjBV.js";
import "./zod-schema.providers-core-WNCesmUy.js";
import "./common-WWiTrAeW.js";
import "./fetch-guard-Conjfhmz.js";
import "./fetch-AAz5jNDG.js";
import { n as resolveChannelGroupRequireMention } from "./group-policy-BvDNgSPT.js";
import "./channel-policy-CrbswhIK.js";
import "./dm-policy-shared-DSgdonEN.js";
import "./setup-wizard-helpers-BKnG96wL.js";
import "./channel-reply-pipeline-BqQuVf4k.js";
import "./channel-pairing-Dekkk_O_.js";
import "./status-helpers-B5Yxdpgu.js";
import "./webhook-ingress-CQD-nsi5.js";
import { t as createOptionalChannelSetupSurface } from "./channel-setup-BZ9bL1Nx.js";
import "./web-media-BNpd4ck7.js";
import "./outbound-media-VATKNny0.js";
//#region src/plugin-sdk/googlechat.ts
function resolveGoogleChatGroupRequireMention(params) {
	return resolveChannelGroupRequireMention({
		cfg: params.cfg,
		channel: "googlechat",
		groupId: params.groupId,
		accountId: params.accountId
	});
}
const googlechatSetup = createOptionalChannelSetupSurface({
	channel: "googlechat",
	label: "Google Chat",
	npmSpec: "@openclaw/googlechat",
	docsPath: "/channels/googlechat"
});
const googlechatSetupAdapter = googlechatSetup.setupAdapter;
const googlechatSetupWizard = googlechatSetup.setupWizard;
//#endregion
export { googlechatSetupWizard as n, resolveGoogleChatGroupRequireMention as r, googlechatSetupAdapter as t };
