// openclaw/dist/json-files-DgU46PWF.js
import path from "node:path";
import fs from "node:fs/promises";
import { randomUUID } from "node:crypto";
//#region src/infra/json-files.ts
function getErrorCode(err) {
	return err instanceof Error ? err.code : void 0;
}
async function replaceFileWithWindowsFallback(tempPath, filePath, mode) {
	try {
		await fs.rename(tempPath, filePath);
		return;
	} catch (err) {
		const code = getErrorCode(err);
		if (process.platform !== "win32" || code !== "EPERM" && code !== "EEXIST") throw err;
	}
	await fs.copyFile(tempPath, filePath);
	try {
		await fs.chmod(filePath, mode);
	} catch {}
	await fs.rm(tempPath, { force: true }).catch(() => void 0);
}
async function readJsonFile(filePath) {
	try {
		const raw = await fs.readFile(filePath, "utf8");
		return JSON.parse(raw);
	} catch {
		return null;
	}
}
async function writeJsonAtomic(filePath, value, options) {
	await writeTextAtomic(filePath, JSON.stringify(value, null, 2), {
		mode: options?.mode,
		ensureDirMode: options?.ensureDirMode,
		appendTrailingNewline: options?.trailingNewline
	});
}
async function writeTextAtomic(filePath, content, options) {
	const mode = options?.mode ?? 384;
	const payload = options?.appendTrailingNewline && !content.endsWith("\n") ? `${content}\n` : content;
	const mkdirOptions = { recursive: true };
	if (typeof options?.ensureDirMode === "number") mkdirOptions.mode = options.ensureDirMode;
	await fs.mkdir(path.dirname(filePath), mkdirOptions);
	const parentDir = path.dirname(filePath);
	const tmp = `${filePath}.${randomUUID()}.tmp`;
	try {
		const tmpHandle = await fs.open(tmp, "w", mode);
		try {
			await tmpHandle.writeFile(payload, { encoding: "utf8" });
			await tmpHandle.sync();
		} finally {
			await tmpHandle.close().catch(() => void 0);
		}
		try {
			await fs.chmod(tmp, mode);
		} catch {}
		await replaceFileWithWindowsFallback(tmp, filePath, mode);
		try {
			const dirHandle = await fs.open(parentDir, "r");
			try {
				await dirHandle.sync();
			} finally {
				await dirHandle.close().catch(() => void 0);
			}
		} catch {}
		try {
			await fs.chmod(filePath, mode);
		} catch {}
	} finally {
		await fs.rm(tmp, { force: true }).catch(() => void 0);
	}
}
function createAsyncLock() {
	let lock = Promise.resolve();
	return async function withLock(fn) {
		const prev = lock;
		let release;
		lock = new Promise((resolve) => {
			release = resolve;
		});
		await prev;
		try {
			return await fn();
		} finally {
			release?.();
		}
	};
}
//#endregion
export { writeTextAtomic as i, readJsonFile as n, writeJsonAtomic as r, createAsyncLock as t };
