// openclaw/dist/memory-core-engine-runtime-qe1msgzM.js
import { n as createLazyFacadeObjectValue } from "./facade-loader-BswmYjZd.js";
import { t as loadActivatedBundledPluginPublicSurfaceModuleSync } from "./facade-runtime-BWwtAwNg.js";
//#region src/plugin-sdk/memory-core-engine-runtime.ts
function loadFacadeModule() {
	return loadActivatedBundledPluginPublicSurfaceModuleSync({
		dirName: "memory-core",
		artifactBasename: "runtime-api.js"
	});
}
const auditShortTermPromotionArtifacts = ((...args) => loadFacadeModule()["auditShortTermPromotionArtifacts"](...args));
const getBuiltinMemoryEmbeddingProviderDoctorMetadata = ((...args) => loadFacadeModule()["getBuiltinMemoryEmbeddingProviderDoctorMetadata"](...args));
const getMemorySearchManager = ((...args) => loadFacadeModule()["getMemorySearchManager"](...args));
const listBuiltinAutoSelectMemoryEmbeddingProviderDoctorMetadata = ((...args) => loadFacadeModule()["listBuiltinAutoSelectMemoryEmbeddingProviderDoctorMetadata"](...args));
const MemoryIndexManager = createLazyFacadeObjectValue(() => loadFacadeModule()["MemoryIndexManager"]);
const repairShortTermPromotionArtifacts = ((...args) => loadFacadeModule()["repairShortTermPromotionArtifacts"](...args));
//#endregion
export { listBuiltinAutoSelectMemoryEmbeddingProviderDoctorMetadata as a, getMemorySearchManager as i, auditShortTermPromotionArtifacts as n, repairShortTermPromotionArtifacts as o, getBuiltinMemoryEmbeddingProviderDoctorMetadata as r, MemoryIndexManager as t };
