// openclaw/dist/monitor-polling.runtime-DZuLuNHo.js
import { r as formatErrorMessage } from "./errors-CGMew-1w.js";
import { i as normalizeLowercaseStringOrEmpty } from "./string-coerce-CIT1-v_a.js";
import { n as sleepWithAbort, t as computeBackoff } from "./backoff-iIsw4jvJ.js";
import { r as formatDurationPrecise } from "./format-duration-6LP5Dzpt.js";
import "./text-runtime-BngW26kR.js";
import "./runtime-env-DxA7Qhwh.js";
import "./ssrf-runtime-BTBCC-nu.js";
import { s as isRecoverableTelegramNetworkError } from "./format-CYgK41Z7.js";
import { n as readTelegramUpdateOffset, r as writeTelegramUpdateOffset } from "./update-offset-store-DaXtC-k8.js";
import { b as withTelegramApiErrorLogging } from "./send-xf4W8Rq0.js";
import { t as createTelegramBot } from "./bot-ClNJQIfx.js";
import { run } from "@grammyjs/runner";
//#region extensions/telegram/src/polling-transport-state.ts
var TelegramPollingTransportState = class {
	#telegramTransport;
	#transportDirty = false;
	constructor(opts) {
		this.opts = opts;
		this.#telegramTransport = opts.initialTransport;
	}
	markDirty() {
		this.#transportDirty = true;
	}
	acquireForNextCycle() {
		const nextTransport = this.#transportDirty || !this.#telegramTransport ? this.opts.createTelegramTransport?.() ?? this.#telegramTransport : this.#telegramTransport;
		if (this.#transportDirty && nextTransport) this.opts.log("[telegram][diag] rebuilding transport for next polling cycle");
		this.#telegramTransport = nextTransport;
		this.#transportDirty = false;
		return nextTransport;
	}
};
//#endregion
//#region extensions/telegram/src/polling-session.ts
const TELEGRAM_POLL_RESTART_POLICY = {
	initialMs: 2e3,
	maxMs: 3e4,
	factor: 1.8,
	jitter: .25
};
const POLL_STALL_THRESHOLD_MS = 9e4;
const POLL_WATCHDOG_INTERVAL_MS = 3e4;
const POLL_STOP_GRACE_MS = 15e3;
const waitForGracefulStop = async (stop) => {
	let timer;
	try {
		await Promise.race([stop(), new Promise((resolve) => {
			timer = setTimeout(resolve, POLL_STOP_GRACE_MS);
			timer.unref?.();
		})]);
	} finally {
		if (timer) clearTimeout(timer);
	}
};
var TelegramPollingSession = class {
	#restartAttempts = 0;
	#webhookCleared = false;
	#forceRestarted = false;
	#activeRunner;
	#activeFetchAbort;
	#transportState;
	constructor(opts) {
		this.opts = opts;
		this.#transportState = new TelegramPollingTransportState({
			log: opts.log,
			initialTransport: opts.telegramTransport,
			createTelegramTransport: opts.createTelegramTransport
		});
	}
	get activeRunner() {
		return this.#activeRunner;
	}
	markForceRestarted() {
		this.#forceRestarted = true;
	}
	markTransportDirty() {
		this.#transportState.markDirty();
	}
	abortActiveFetch() {
		this.#activeFetchAbort?.abort();
	}
	async runUntilAbort() {
		while (!this.opts.abortSignal?.aborted) {
			const bot = await this.#createPollingBot();
			if (!bot) continue;
			const cleanupState = await this.#ensureWebhookCleanup(bot);
			if (cleanupState === "retry") continue;
			if (cleanupState === "exit") return;
			if (await this.#runPollingCycle(bot) === "exit") return;
		}
	}
	async #waitBeforeRestart(buildLine) {
		this.#restartAttempts += 1;
		const delayMs = computeBackoff(TELEGRAM_POLL_RESTART_POLICY, this.#restartAttempts);
		const delay = formatDurationPrecise(delayMs);
		this.opts.log(buildLine(delay));
		try {
			await sleepWithAbort(delayMs, this.opts.abortSignal);
		} catch (sleepErr) {
			if (this.opts.abortSignal?.aborted) return false;
			throw sleepErr;
		}
		return true;
	}
	async #waitBeforeRetryOnRecoverableSetupError(err, logPrefix) {
		if (this.opts.abortSignal?.aborted) return false;
		if (!isRecoverableTelegramNetworkError(err, { context: "unknown" })) throw err;
		return this.#waitBeforeRestart((delay) => `${logPrefix}: ${formatErrorMessage(err)}; retrying in ${delay}.`);
	}
	async #createPollingBot() {
		const fetchAbortController = new AbortController();
		this.#activeFetchAbort = fetchAbortController;
		const telegramTransport = this.#transportState.acquireForNextCycle();
		try {
			return createTelegramBot({
				token: this.opts.token,
				runtime: this.opts.runtime,
				proxyFetch: this.opts.proxyFetch,
				config: this.opts.config,
				accountId: this.opts.accountId,
				fetchAbortSignal: fetchAbortController.signal,
				updateOffset: {
					lastUpdateId: this.opts.getLastUpdateId(),
					onUpdateId: this.opts.persistUpdateId
				},
				telegramTransport
			});
		} catch (err) {
			await this.#waitBeforeRetryOnRecoverableSetupError(err, "Telegram setup network error");
			if (this.#activeFetchAbort === fetchAbortController) this.#activeFetchAbort = void 0;
			return;
		}
	}
	async #ensureWebhookCleanup(bot) {
		if (this.#webhookCleared) return "ready";
		try {
			await withTelegramApiErrorLogging({
				operation: "deleteWebhook",
				runtime: this.opts.runtime,
				fn: () => bot.api.deleteWebhook({ drop_pending_updates: false })
			});
			this.#webhookCleared = true;
			return "ready";
		} catch (err) {
			return await this.#waitBeforeRetryOnRecoverableSetupError(err, "Telegram webhook cleanup failed") ? "retry" : "exit";
		}
	}
	async #confirmPersistedOffset(bot) {
		const lastUpdateId = this.opts.getLastUpdateId();
		if (lastUpdateId === null || lastUpdateId >= Number.MAX_SAFE_INTEGER) return;
		try {
			await bot.api.getUpdates({
				offset: lastUpdateId + 1,
				limit: 1,
				timeout: 0
			});
		} catch {}
	}
	async #runPollingCycle(bot) {
		await this.#confirmPersistedOffset(bot);
		let lastGetUpdatesAt = Date.now();
		let lastApiActivityAt = Date.now();
		let nextInFlightApiCallId = 0;
		let latestInFlightApiStartedAt = null;
		const inFlightApiStartedAt = /* @__PURE__ */ new Map();
		let lastGetUpdatesStartedAt = null;
		let lastGetUpdatesFinishedAt = null;
		let lastGetUpdatesDurationMs = null;
		let lastGetUpdatesOutcome = "not-started";
		let lastGetUpdatesError = null;
		let lastGetUpdatesOffset = null;
		let inFlightGetUpdates = 0;
		let stallDiagLoggedAt = 0;
		bot.api.config.use(async (prev, method, payload, signal) => {
			if (method !== "getUpdates") {
				const startedAt = Date.now();
				const callId = nextInFlightApiCallId;
				nextInFlightApiCallId += 1;
				inFlightApiStartedAt.set(callId, startedAt);
				latestInFlightApiStartedAt = latestInFlightApiStartedAt == null ? startedAt : Math.max(latestInFlightApiStartedAt, startedAt);
				try {
					const result = await prev(method, payload, signal);
					lastApiActivityAt = Date.now();
					return result;
				} finally {
					inFlightApiStartedAt.delete(callId);
					if (latestInFlightApiStartedAt === startedAt) {
						let newestStartedAt = null;
						for (const activeStartedAt of inFlightApiStartedAt.values()) newestStartedAt = newestStartedAt == null ? activeStartedAt : Math.max(newestStartedAt, activeStartedAt);
						latestInFlightApiStartedAt = newestStartedAt;
					}
				}
			}
			const startedAt = Date.now();
			lastGetUpdatesAt = startedAt;
			lastGetUpdatesStartedAt = startedAt;
			lastGetUpdatesOffset = payload && typeof payload === "object" && "offset" in payload ? payload.offset ?? null : null;
			inFlightGetUpdates += 1;
			lastGetUpdatesOutcome = "started";
			lastGetUpdatesError = null;
			try {
				const result = await prev(method, payload, signal);
				const finishedAt = Date.now();
				lastGetUpdatesFinishedAt = finishedAt;
				lastGetUpdatesDurationMs = finishedAt - startedAt;
				lastGetUpdatesOutcome = Array.isArray(result) ? `ok:${result.length}` : "ok";
				return result;
			} catch (err) {
				const finishedAt = Date.now();
				lastGetUpdatesFinishedAt = finishedAt;
				lastGetUpdatesDurationMs = finishedAt - startedAt;
				lastGetUpdatesOutcome = "error";
				lastGetUpdatesError = formatErrorMessage(err);
				throw err;
			} finally {
				inFlightGetUpdates = Math.max(0, inFlightGetUpdates - 1);
			}
		});
		const runner = run(bot, this.opts.runnerOptions);
		this.#activeRunner = runner;
		const fetchAbortController = this.#activeFetchAbort;
		const abortFetch = () => {
			fetchAbortController?.abort();
		};
		if (this.opts.abortSignal && fetchAbortController) this.opts.abortSignal.addEventListener("abort", abortFetch, { once: true });
		let stopPromise;
		let stalledRestart = false;
		let forceCycleTimer;
		let forceCycleResolve;
		const forceCyclePromise = new Promise((resolve) => {
			forceCycleResolve = resolve;
		});
		const stopRunner = () => {
			fetchAbortController?.abort();
			stopPromise ??= Promise.resolve(runner.stop()).then(() => void 0).catch(() => {});
			return stopPromise;
		};
		const stopBot = () => {
			return Promise.resolve(bot.stop()).then(() => void 0).catch(() => {});
		};
		const stopOnAbort = () => {
			if (this.opts.abortSignal?.aborted) stopRunner();
		};
		const watchdog = setInterval(() => {
			if (this.opts.abortSignal?.aborted) return;
			const now = Date.now();
			const activeElapsed = inFlightGetUpdates > 0 && lastGetUpdatesStartedAt != null ? now - lastGetUpdatesStartedAt : 0;
			const idleElapsed = inFlightGetUpdates > 0 ? 0 : now - (lastGetUpdatesFinishedAt ?? lastGetUpdatesAt);
			const elapsed = inFlightGetUpdates > 0 ? activeElapsed : idleElapsed;
			const apiElapsed = now - (latestInFlightApiStartedAt == null ? lastApiActivityAt : Math.max(lastApiActivityAt, latestInFlightApiStartedAt));
			if (elapsed > POLL_STALL_THRESHOLD_MS && apiElapsed > POLL_STALL_THRESHOLD_MS && runner.isRunning()) {
				if (stallDiagLoggedAt && now - stallDiagLoggedAt < POLL_STALL_THRESHOLD_MS / 2) return;
				stallDiagLoggedAt = now;
				this.#transportState.markDirty();
				stalledRestart = true;
				const elapsedLabel = inFlightGetUpdates > 0 ? `active getUpdates stuck for ${formatDurationPrecise(elapsed)}` : `no completed getUpdates for ${formatDurationPrecise(elapsed)}`;
				this.opts.log(`[telegram] Polling stall detected (${elapsedLabel}); forcing restart. [diag inFlight=${inFlightGetUpdates} outcome=${lastGetUpdatesOutcome} startedAt=${lastGetUpdatesStartedAt ?? "n/a"} finishedAt=${lastGetUpdatesFinishedAt ?? "n/a"} durationMs=${lastGetUpdatesDurationMs ?? "n/a"} offset=${lastGetUpdatesOffset ?? "n/a"}${lastGetUpdatesError ? ` error=${lastGetUpdatesError}` : ""}]`);
				stopRunner();
				stopBot();
				if (!forceCycleTimer) forceCycleTimer = setTimeout(() => {
					if (this.opts.abortSignal?.aborted) return;
					this.opts.log(`[telegram] Polling runner stop timed out after ${formatDurationPrecise(POLL_STOP_GRACE_MS)}; forcing restart cycle.`);
					forceCycleResolve?.();
				}, POLL_STOP_GRACE_MS);
			}
		}, POLL_WATCHDOG_INTERVAL_MS);
		this.opts.abortSignal?.addEventListener("abort", stopOnAbort, { once: true });
		try {
			await Promise.race([runner.task(), forceCyclePromise]);
			if (this.opts.abortSignal?.aborted) return "exit";
			const reason = stalledRestart ? "polling stall detected" : this.#forceRestarted ? "unhandled network error" : "runner stopped (maxRetryTime exceeded or graceful stop)";
			this.#forceRestarted = false;
			this.opts.log(`[telegram][diag] polling cycle finished reason=${reason} inFlight=${inFlightGetUpdates} outcome=${lastGetUpdatesOutcome} startedAt=${lastGetUpdatesStartedAt ?? "n/a"} finishedAt=${lastGetUpdatesFinishedAt ?? "n/a"} durationMs=${lastGetUpdatesDurationMs ?? "n/a"} offset=${lastGetUpdatesOffset ?? "n/a"}${lastGetUpdatesError ? ` error=${String(lastGetUpdatesError)}` : ""}`);
			return await this.#waitBeforeRestart((delay) => `Telegram polling runner stopped (${reason}); restarting in ${delay}.`) ? "continue" : "exit";
		} catch (err) {
			this.#forceRestarted = false;
			if (this.opts.abortSignal?.aborted) throw err;
			const isConflict = isGetUpdatesConflict(err);
			if (isConflict) this.#webhookCleared = false;
			const isRecoverable = isRecoverableTelegramNetworkError(err, { context: "polling" });
			if (isRecoverable) this.#transportState.markDirty();
			if (!isConflict && !isRecoverable) throw err;
			const reason = isConflict ? "getUpdates conflict" : "network error";
			const errMsg = formatErrorMessage(err);
			this.opts.log(`[telegram][diag] polling cycle error reason=${reason} inFlight=${inFlightGetUpdates} outcome=${lastGetUpdatesOutcome} startedAt=${lastGetUpdatesStartedAt ?? "n/a"} finishedAt=${lastGetUpdatesFinishedAt ?? "n/a"} durationMs=${lastGetUpdatesDurationMs ?? "n/a"} offset=${lastGetUpdatesOffset ?? "n/a"} err=${errMsg}${lastGetUpdatesError ? ` lastGetUpdatesError=${String(lastGetUpdatesError)}` : ""}`);
			return await this.#waitBeforeRestart((delay) => `Telegram ${reason}: ${errMsg}; retrying in ${delay}.`) ? "continue" : "exit";
		} finally {
			clearInterval(watchdog);
			if (forceCycleTimer) clearTimeout(forceCycleTimer);
			this.opts.abortSignal?.removeEventListener("abort", abortFetch);
			this.opts.abortSignal?.removeEventListener("abort", stopOnAbort);
			await waitForGracefulStop(stopRunner);
			await waitForGracefulStop(stopBot);
			this.#activeRunner = void 0;
			if (this.#activeFetchAbort === fetchAbortController) this.#activeFetchAbort = void 0;
		}
	}
};
const isGetUpdatesConflict = (err) => {
	if (!err || typeof err !== "object") return false;
	const typed = err;
	if ((typed.error_code ?? typed.errorCode) !== 409) return false;
	return normalizeLowercaseStringOrEmpty([
		typed.method,
		typed.description,
		typed.message
	].filter((value) => typeof value === "string").join(" ")).includes("getupdates");
};
//#endregion
export { TelegramPollingSession, readTelegramUpdateOffset, writeTelegramUpdateOffset };
