// openclaw/dist/onboard-BkY6XWC_.js
import { p as createModelCatalogPresetAppliers } from "./provider-onboard-Kw4E2oms.js";
import { n as ARCEE_MODEL_CATALOG, r as buildArceeModelDefinition, t as ARCEE_BASE_URL } from "./models-BLQuhZLl.js";
//#region extensions/arcee/provider-catalog.ts
const OPENROUTER_BASE_URL = "https://openrouter.ai/api/v1";
function normalizeBaseUrl(baseUrl) {
	return String(baseUrl ?? "").trim().replace(/\/+$/, "");
}
function isArceeOpenRouterBaseUrl(baseUrl) {
	return normalizeBaseUrl(baseUrl) === OPENROUTER_BASE_URL;
}
function toArceeOpenRouterModelId(modelId) {
	const normalized = modelId.trim();
	if (!normalized || normalized.startsWith("arcee/")) return normalized;
	return `arcee/${normalized}`;
}
function buildArceeCatalogModels() {
	return ARCEE_MODEL_CATALOG.map(buildArceeModelDefinition);
}
function buildArceeOpenRouterCatalogModels() {
	return buildArceeCatalogModels().map((model) => ({
		...model,
		id: toArceeOpenRouterModelId(model.id)
	}));
}
function buildArceeProvider() {
	return {
		baseUrl: ARCEE_BASE_URL,
		api: "openai-completions",
		models: buildArceeCatalogModels()
	};
}
function buildArceeOpenRouterProvider() {
	return {
		baseUrl: OPENROUTER_BASE_URL,
		api: "openai-completions",
		models: buildArceeOpenRouterCatalogModels()
	};
}
//#endregion
//#region extensions/arcee/onboard.ts
const ARCEE_DEFAULT_MODEL_REF = "arcee/trinity-large-thinking";
const ARCEE_OPENROUTER_DEFAULT_MODEL_REF = "arcee/trinity-large-thinking";
const arceePresetAppliers = createModelCatalogPresetAppliers({
	primaryModelRef: ARCEE_DEFAULT_MODEL_REF,
	resolveParams: (_cfg) => ({
		providerId: "arcee",
		api: "openai-completions",
		baseUrl: ARCEE_BASE_URL,
		catalogModels: buildArceeCatalogModels(),
		aliases: [{
			modelRef: ARCEE_DEFAULT_MODEL_REF,
			alias: "Arcee AI"
		}]
	})
});
const arceeOpenRouterPresetAppliers = createModelCatalogPresetAppliers({
	primaryModelRef: ARCEE_OPENROUTER_DEFAULT_MODEL_REF,
	resolveParams: (_cfg) => ({
		providerId: "arcee",
		api: "openai-completions",
		baseUrl: OPENROUTER_BASE_URL,
		catalogModels: buildArceeOpenRouterCatalogModels(),
		aliases: [{
			modelRef: ARCEE_OPENROUTER_DEFAULT_MODEL_REF,
			alias: "Arcee AI (OpenRouter)"
		}]
	})
});
function applyArceeProviderConfig(cfg) {
	return arceePresetAppliers.applyProviderConfig(cfg);
}
function applyArceeConfig(cfg) {
	return arceePresetAppliers.applyConfig(cfg);
}
function applyArceeOpenRouterConfig(cfg) {
	return arceeOpenRouterPresetAppliers.applyConfig(cfg);
}
//#endregion
export { applyArceeProviderConfig as a, buildArceeOpenRouterCatalogModels as c, isArceeOpenRouterBaseUrl as d, toArceeOpenRouterModelId as f, applyArceeOpenRouterConfig as i, buildArceeOpenRouterProvider as l, ARCEE_OPENROUTER_DEFAULT_MODEL_REF as n, OPENROUTER_BASE_URL as o, applyArceeConfig as r, buildArceeCatalogModels as s, ARCEE_DEFAULT_MODEL_REF as t, buildArceeProvider as u };
