// openclaw/dist/register.sync.runtime-CdGhhgB0.js
import { o as resolveImplicitMantleProvider, r as mergeImplicitMantleProvider, s as resolveMantleBearerToken } from "./discovery-D64L03sw.js";
//#region extensions/amazon-bedrock-mantle/register.sync.runtime.ts
function registerBedrockMantlePlugin(api) {
	const providerId = "amazon-bedrock-mantle";
	api.registerProvider({
		id: providerId,
		label: "Amazon Bedrock Mantle (OpenAI-compatible)",
		docsPath: "/providers/bedrock-mantle",
		auth: [],
		catalog: {
			order: "simple",
			run: async (ctx) => {
				const implicit = await resolveImplicitMantleProvider({ env: ctx.env });
				if (!implicit) return null;
				return { provider: mergeImplicitMantleProvider({
					existing: ctx.config.models?.providers?.[providerId],
					implicit
				}) };
			}
		},
		resolveConfigApiKey: ({ env }) => resolveMantleBearerToken(env) ? "AWS_BEARER_TOKEN_BEDROCK" : void 0,
		matchesContextOverflowError: ({ errorMessage }) => /context_length_exceeded|max.*tokens.*exceeded/i.test(errorMessage),
		classifyFailoverReason: ({ errorMessage }) => {
			if (/rate_limit|too many requests|429/i.test(errorMessage)) return "rate_limit";
			if (/overloaded|503/i.test(errorMessage)) return "overloaded";
		}
	});
}
//#endregion
export { registerBedrockMantlePlugin as t };
