// openclaw/dist/runtime-registry-loader-Bgr336E_.js
import { r as getActivePluginRegistry } from "./runtime-D85K4f41.js";
import { $t as resolvePluginRuntimeLoadContext, Xt as buildPluginRuntimeLoadOptions } from "./config-nRyLQtXB.js";
import { r as loadOpenClawPlugins } from "./loader-BSIqIOsD.js";
import { n as resolveConfiguredChannelPluginIds, t as resolveChannelPluginIds } from "./channel-plugin-ids-DFL3s5jv.js";
//#region src/plugins/runtime/runtime-registry-loader.ts
let pluginRegistryLoaded = "none";
function scopeRank(scope) {
	switch (scope) {
		case "none": return 0;
		case "configured-channels": return 1;
		case "channels": return 2;
		case "all": return 3;
	}
}
function activeRegistrySatisfiesScope(scope, active, expectedChannelPluginIds, requestedPluginIds) {
	if (!active) return false;
	if (requestedPluginIds.length > 0) {
		const activePluginIds = new Set(active.plugins.filter((plugin) => plugin.status === "loaded").map((plugin) => plugin.id));
		return requestedPluginIds.every((pluginId) => activePluginIds.has(pluginId));
	}
	const activeChannelPluginIds = new Set(active.channels.map((entry) => entry.plugin.id));
	switch (scope) {
		case "configured-channels":
		case "channels": return active.channels.length > 0 && expectedChannelPluginIds.every((pluginId) => activeChannelPluginIds.has(pluginId));
		case "all": return false;
	}
}
function ensurePluginRegistryLoaded(options) {
	const scope = options?.scope ?? "all";
	const requestedPluginIds = options?.onlyPluginIds?.map((pluginId) => pluginId.trim()).filter(Boolean) ?? [];
	const scopedLoad = requestedPluginIds.length > 0;
	const context = resolvePluginRuntimeLoadContext(options);
	const expectedChannelPluginIds = scopedLoad ? requestedPluginIds : scope === "configured-channels" ? resolveConfiguredChannelPluginIds({
		config: context.config,
		workspaceDir: context.workspaceDir,
		env: context.env
	}) : scope === "channels" ? resolveChannelPluginIds({
		config: context.config,
		workspaceDir: context.workspaceDir,
		env: context.env
	}) : [];
	const active = getActivePluginRegistry();
	if (!scopedLoad && scopeRank(pluginRegistryLoaded) >= scopeRank(scope) && activeRegistrySatisfiesScope(scope, active, expectedChannelPluginIds, expectedChannelPluginIds)) return;
	if ((pluginRegistryLoaded === "none" || scopedLoad) && activeRegistrySatisfiesScope(scope, active, expectedChannelPluginIds, expectedChannelPluginIds)) {
		if (!scopedLoad) pluginRegistryLoaded = scope;
		return;
	}
	loadOpenClawPlugins(buildPluginRuntimeLoadOptions(context, {
		throwOnLoadError: true,
		...expectedChannelPluginIds.length > 0 ? { onlyPluginIds: expectedChannelPluginIds } : {}
	}));
	if (!scopedLoad) pluginRegistryLoaded = scope;
}
const __testing = { resetPluginRegistryLoadedForTests() {
	pluginRegistryLoaded = "none";
} };
//#endregion
export { ensurePluginRegistryLoaded as n, __testing as t };
