// openclaw/dist/sessions-ChPyTFUE.js
import { c as loadConfig } from "./config-nRyLQtXB.js";
import "./store-Cgl8QMzI.js";
import { i as resolveMainSessionKey } from "./main-session-_52dIe-Q.js";
import { l as resolveStorePath } from "./paths-zYrymkDo.js";
import "./reset-SzHn7hUh.js";
import "./session-key-BLQuAiCF.js";
import { t as loadSessionStore } from "./store-load-Bte4ho7k.js";
import "./session-file-DSggyxFr.js";
import { t as parseSessionThreadInfo } from "./thread-info-Bp_C3DYn.js";
import "./transcript-BdAYvPnf.js";
import "./targets-shZzytX9.js";
//#region src/config/sessions/main-session.runtime.ts
function resolveMainSessionKeyFromConfig() {
	return resolveMainSessionKey(loadConfig());
}
//#endregion
//#region src/config/sessions/delivery-info.ts
function extractDeliveryInfo(sessionKey) {
	const { baseSessionKey, threadId } = parseSessionThreadInfo(sessionKey);
	if (!sessionKey || !baseSessionKey) return {
		deliveryContext: void 0,
		threadId
	};
	let deliveryContext;
	try {
		const store = loadSessionStore(resolveStorePath(loadConfig().session?.store));
		let entry = store[sessionKey];
		if (!entry?.deliveryContext && baseSessionKey !== sessionKey) entry = store[baseSessionKey];
		if (entry?.deliveryContext) {
			const resolvedThreadId = entry.deliveryContext.threadId ?? entry.lastThreadId ?? entry.origin?.threadId;
			deliveryContext = {
				channel: entry.deliveryContext.channel,
				to: entry.deliveryContext.to,
				accountId: entry.deliveryContext.accountId,
				threadId: resolvedThreadId != null ? String(resolvedThreadId) : void 0
			};
		}
	} catch {}
	return {
		deliveryContext,
		threadId
	};
}
//#endregion
export { resolveMainSessionKeyFromConfig as n, extractDeliveryInfo as t };
