// openclaw/dist/slash-dispatch.runtime-IfhbXa13.js
import { t as resolveConversationLabel$1 } from "./conversation-label-DKlwjdHD.js";
import { s as resolveChunkMode$1 } from "./chunk-DS9PgIig.js";
import { i as resolveAgentRoute$1 } from "./resolve-route-Cv_oyKns.js";
import "./routing-CC84LXmn.js";
import { t as finalizeInboundContext$1 } from "./inbound-context-Bi-BPgBI.js";
import { n as dispatchReplyWithDispatcher$1 } from "./provider-dispatcher-jJpsCDDr.js";
import { t as resolveMarkdownTableMode$1 } from "./markdown-tables-BLsPOjqk.js";
import "./config-runtime-B7RP_kUw.js";
import "./reply-runtime-Bj9CGlMM.js";
import { t as recordInboundSessionMetaSafe$1 } from "./conversation-runtime-C2R31CkT.js";
import { r as deliverSlackSlashReplies$1 } from "./replies-Dv7Xeg_O.js";
//#region extensions/slack/src/monitor/slash-dispatch.runtime.ts
function resolveChunkMode(...args) {
	return resolveChunkMode$1(...args);
}
function finalizeInboundContext(...args) {
	return finalizeInboundContext$1(...args);
}
function dispatchReplyWithDispatcher(...args) {
	return dispatchReplyWithDispatcher$1(...args);
}
function resolveConversationLabel(...args) {
	return resolveConversationLabel$1(...args);
}
function recordInboundSessionMetaSafe(...args) {
	return recordInboundSessionMetaSafe$1(...args);
}
function resolveMarkdownTableMode(...args) {
	return resolveMarkdownTableMode$1(...args);
}
function resolveAgentRoute(...args) {
	return resolveAgentRoute$1(...args);
}
function deliverSlackSlashReplies(...args) {
	return deliverSlackSlashReplies$1(...args);
}
//#endregion
export { deliverSlackSlashReplies, dispatchReplyWithDispatcher, finalizeInboundContext, recordInboundSessionMetaSafe, resolveAgentRoute, resolveChunkMode, resolveConversationLabel, resolveMarkdownTableMode };
