// openclaw/dist/store-_YWeF8Qv.js
import { s as normalizeOptionalString } from "./string-coerce-CIT1-v_a.js";
import { f as resolveConfigDir } from "./utils-Supr9KvU.js";
import { l as readLocalFileSafely$1, t as SafeOpenError } from "./fs-safe-B2XplAF8.js";
import { n as extensionForMime, t as detectMime } from "./mime-CQnQ252G.js";
import { t as retainSafeHeadersForCrossOriginRedirect } from "./redirect-headers-Ck4cgDgU.js";
import { u as resolvePinnedHostname } from "./ssrf-TOfbRtf4.js";
import { createWriteStream } from "node:fs";
import path from "node:path";
import fs$1 from "node:fs/promises";
import crypto from "node:crypto";
import { pipeline } from "node:stream/promises";
import { request } from "node:http";
import { request as request$1 } from "node:https";
//#region src/media/store.runtime.ts
const readLocalFileSafely = readLocalFileSafely$1;
function isSafeOpenError(error) {
	return error instanceof SafeOpenError;
}
//#endregion
//#region src/media/store.ts
const resolveMediaDir = () => path.join(resolveConfigDir(), "media");
const MEDIA_MAX_BYTES = 5 * 1024 * 1024;
const MAX_BYTES = MEDIA_MAX_BYTES;
const DEFAULT_TTL_MS = 120 * 1e3;
const MEDIA_FILE_MODE = 420;
const defaultHttpRequestImpl = request;
const defaultHttpsRequestImpl = request$1;
const defaultResolvePinnedHostnameImpl = resolvePinnedHostname;
let httpRequestImpl = defaultHttpRequestImpl;
let httpsRequestImpl = defaultHttpsRequestImpl;
let resolvePinnedHostnameImpl = defaultResolvePinnedHostnameImpl;
function setMediaStoreNetworkDepsForTest(deps) {
	httpRequestImpl = deps?.httpRequest ?? defaultHttpRequestImpl;
	httpsRequestImpl = deps?.httpsRequest ?? defaultHttpsRequestImpl;
	resolvePinnedHostnameImpl = deps?.resolvePinnedHostname ?? defaultResolvePinnedHostnameImpl;
}
/**
* Sanitize a filename for cross-platform safety.
* Removes chars unsafe on Windows/SharePoint/all platforms.
* Keeps: alphanumeric, dots, hyphens, underscores, Unicode letters/numbers.
*/
function sanitizeFilename(name) {
	const trimmed = name.trim();
	if (!trimmed) return "";
	return trimmed.replace(/[^\p{L}\p{N}._-]+/gu, "_").replace(/_+/g, "_").replace(/^_|_$/g, "").slice(0, 60);
}
/**
* Extract original filename from path if it matches the embedded format.
* Pattern: {original}---{uuid}.{ext} → returns "{original}.{ext}"
* Falls back to basename if no pattern match, or "file.bin" if empty.
*/
function extractOriginalFilename(filePath) {
	const basename = path.basename(filePath);
	if (!basename) return "file.bin";
	const ext = path.extname(basename);
	const match = path.basename(basename, ext).match(/^(.+)---[a-f0-9]{8}-[a-f0-9]{4}-[a-f0-9]{4}-[a-f0-9]{4}-[a-f0-9]{12}$/i);
	if (match?.[1]) return `${match[1]}${ext}`;
	return basename;
}
function getMediaDir() {
	return resolveMediaDir();
}
async function ensureMediaDir() {
	const mediaDir = resolveMediaDir();
	await fs$1.mkdir(mediaDir, {
		recursive: true,
		mode: 448
	});
	return mediaDir;
}
function isMissingPathError(err) {
	return err instanceof Error && "code" in err && err.code === "ENOENT";
}
async function retryAfterRecreatingDir(dir, run) {
	try {
		return await run();
	} catch (err) {
		if (!isMissingPathError(err)) throw err;
		await fs$1.mkdir(dir, {
			recursive: true,
			mode: 448
		});
		return await run();
	}
}
async function cleanOldMedia(ttlMs = DEFAULT_TTL_MS, options = {}) {
	const mediaDir = await ensureMediaDir();
	const now = Date.now();
	const recursive = options.recursive ?? false;
	const pruneEmptyDirs = recursive && (options.pruneEmptyDirs ?? false);
	const removeExpiredFilesInDir = async (dir) => {
		const dirEntries = await fs$1.readdir(dir).catch(() => null);
		if (!dirEntries) return false;
		for (const entry of dirEntries) {
			const fullPath = path.join(dir, entry);
			const stat = await fs$1.lstat(fullPath).catch(() => null);
			if (!stat || stat.isSymbolicLink()) continue;
			if (stat.isDirectory()) {
				if (recursive) {
					if (await removeExpiredFilesInDir(fullPath)) await fs$1.rmdir(fullPath).catch(() => {});
				}
				continue;
			}
			if (!stat.isFile()) continue;
			if (now - stat.mtimeMs > ttlMs) await fs$1.rm(fullPath, { force: true }).catch(() => {});
		}
		if (!pruneEmptyDirs) return false;
		const remainingEntries = await fs$1.readdir(dir).catch(() => null);
		return remainingEntries !== null && remainingEntries.length === 0;
	};
	const entries = await fs$1.readdir(mediaDir).catch(() => []);
	for (const file of entries) {
		const full = path.join(mediaDir, file);
		const stat = await fs$1.lstat(full).catch(() => null);
		if (!stat || stat.isSymbolicLink()) continue;
		if (stat.isDirectory()) {
			if (await removeExpiredFilesInDir(full)) await fs$1.rmdir(full).catch(() => {});
			continue;
		}
		if (stat.isFile() && now - stat.mtimeMs > ttlMs) await fs$1.rm(full, { force: true }).catch(() => {});
	}
}
function looksLikeUrl(src) {
	return /^https?:\/\//i.test(src);
}
/**
* Download media to disk while capturing the first few KB for mime sniffing.
*/
async function downloadToFile(url, dest, headers, maxRedirects = 5) {
	return await new Promise((resolve, reject) => {
		let parsedUrl;
		try {
			parsedUrl = new URL(url);
		} catch {
			reject(/* @__PURE__ */ new Error("Invalid URL"));
			return;
		}
		if (!["http:", "https:"].includes(parsedUrl.protocol)) {
			reject(/* @__PURE__ */ new Error(`Invalid URL protocol: ${parsedUrl.protocol}. Only HTTP/HTTPS allowed.`));
			return;
		}
		const requestImpl = parsedUrl.protocol === "https:" ? httpsRequestImpl : httpRequestImpl;
		resolvePinnedHostnameImpl(parsedUrl.hostname).then((pinned) => {
			const req = requestImpl(parsedUrl, {
				headers,
				lookup: pinned.lookup
			}, (res) => {
				if (res.statusCode && res.statusCode >= 300 && res.statusCode < 400) {
					const location = res.headers.location;
					if (!location || maxRedirects <= 0) {
						reject(/* @__PURE__ */ new Error(`Redirect loop or missing Location header`));
						return;
					}
					const redirectUrl = new URL(location, url).href;
					resolve(downloadToFile(redirectUrl, dest, new URL(redirectUrl).origin === parsedUrl.origin ? headers : retainSafeHeadersForCrossOriginRedirect(headers), maxRedirects - 1));
					return;
				}
				if (!res.statusCode || res.statusCode >= 400) {
					reject(/* @__PURE__ */ new Error(`HTTP ${res.statusCode ?? "?"} downloading media`));
					return;
				}
				let total = 0;
				const sniffChunks = [];
				let sniffLen = 0;
				const out = createWriteStream(dest, { mode: MEDIA_FILE_MODE });
				res.on("data", (chunk) => {
					total += chunk.length;
					if (sniffLen < 16384) {
						sniffChunks.push(chunk);
						sniffLen += chunk.length;
					}
					if (total > MAX_BYTES) req.destroy(/* @__PURE__ */ new Error("Media exceeds 5MB limit"));
				});
				pipeline(res, out).then(() => {
					const sniffBuffer = Buffer.concat(sniffChunks, Math.min(sniffLen, 16384));
					const rawHeader = res.headers["content-type"];
					resolve({
						headerMime: Array.isArray(rawHeader) ? rawHeader[0] : rawHeader,
						sniffBuffer,
						size: total
					});
				}).catch(async (err) => {
					await fs$1.rm(dest, { force: true }).catch(() => {});
					reject(err);
				});
			});
			req.on("error", reject);
			req.end();
		}).catch(reject);
	});
}
function buildSavedMediaId(params) {
	if (!params.originalFilename) return params.ext ? `${params.baseId}${params.ext}` : params.baseId;
	const base = path.parse(params.originalFilename).name;
	const sanitized = sanitizeFilename(base);
	return sanitized ? `${sanitized}---${params.baseId}${params.ext}` : `${params.baseId}${params.ext}`;
}
function buildSavedMediaResult(params) {
	return {
		id: params.id,
		path: path.join(params.dir, params.id),
		size: params.size,
		contentType: params.contentType
	};
}
async function writeSavedMediaBuffer(params) {
	const dest = path.join(params.dir, params.id);
	await retryAfterRecreatingDir(params.dir, () => fs$1.writeFile(dest, params.buffer, { mode: MEDIA_FILE_MODE }));
	return dest;
}
var SaveMediaSourceError = class extends Error {
	constructor(code, message, options) {
		super(message, options);
		this.code = code;
		this.name = "SaveMediaSourceError";
	}
};
function toSaveMediaSourceError(err) {
	switch (err.code) {
		case "symlink": return new SaveMediaSourceError("invalid-path", "Media path must not be a symlink", { cause: err });
		case "not-file": return new SaveMediaSourceError("not-file", "Media path is not a file", { cause: err });
		case "path-mismatch": return new SaveMediaSourceError("path-mismatch", "Media path changed during read", { cause: err });
		case "too-large": return new SaveMediaSourceError("too-large", "Media exceeds 5MB limit", { cause: err });
		case "not-found": return new SaveMediaSourceError("not-found", "Media path does not exist", { cause: err });
		case "outside-workspace": return new SaveMediaSourceError("invalid-path", "Media path is outside workspace root", { cause: err });
		default: return new SaveMediaSourceError("invalid-path", "Media path is not safe to read", { cause: err });
	}
}
async function saveMediaSource(source, headers, subdir = "") {
	const baseDir = resolveMediaDir();
	const dir = subdir ? path.join(baseDir, subdir) : baseDir;
	await fs$1.mkdir(dir, {
		recursive: true,
		mode: 448
	});
	await cleanOldMedia(DEFAULT_TTL_MS, { recursive: false });
	const baseId = crypto.randomUUID();
	if (looksLikeUrl(source)) {
		const tempDest = path.join(dir, `${baseId}.tmp`);
		const { headerMime, sniffBuffer, size } = await retryAfterRecreatingDir(dir, () => downloadToFile(source, tempDest, headers));
		const mime = await detectMime({
			buffer: sniffBuffer,
			headerMime,
			filePath: source
		});
		const id = buildSavedMediaId({
			baseId,
			ext: extensionForMime(mime) ?? path.extname(new URL(source).pathname)
		});
		const finalDest = path.join(dir, id);
		await fs$1.rename(tempDest, finalDest);
		return buildSavedMediaResult({
			dir,
			id,
			size,
			contentType: mime
		});
	}
	try {
		const { buffer, stat } = await readLocalFileSafely({
			filePath: source,
			maxBytes: MAX_BYTES
		});
		const mime = await detectMime({
			buffer,
			filePath: source
		});
		const id = buildSavedMediaId({
			baseId,
			ext: extensionForMime(mime) ?? path.extname(source)
		});
		await writeSavedMediaBuffer({
			dir,
			id,
			buffer
		});
		return buildSavedMediaResult({
			dir,
			id,
			size: stat.size,
			contentType: mime
		});
	} catch (err) {
		if (isSafeOpenError(err)) throw toSaveMediaSourceError(err);
		throw err;
	}
}
async function saveMediaBuffer(buffer, contentType, subdir = "inbound", maxBytes = MAX_BYTES, originalFilename) {
	if (buffer.byteLength > maxBytes) throw new Error(`Media exceeds ${(maxBytes / (1024 * 1024)).toFixed(0)}MB limit`);
	const dir = path.join(resolveMediaDir(), subdir);
	await fs$1.mkdir(dir, {
		recursive: true,
		mode: 448
	});
	const uuid = crypto.randomUUID();
	const headerExt = extensionForMime(normalizeOptionalString(contentType?.split(";")[0]));
	const mime = await detectMime({
		buffer,
		headerMime: contentType
	});
	const id = buildSavedMediaId({
		baseId: uuid,
		ext: headerExt ?? extensionForMime(mime) ?? "",
		originalFilename
	});
	await writeSavedMediaBuffer({
		dir,
		id,
		buffer
	});
	return buildSavedMediaResult({
		dir,
		id,
		size: buffer.byteLength,
		contentType: mime
	});
}
/**
* Resolves a media ID saved by saveMediaBuffer to its absolute physical path.
*
* This is the read-side counterpart to saveMediaBuffer and is used by the
* agent runner to hydrate opaque `media://inbound/<id>` URIs written by the
* Gateway's claim-check offload path.
*
* Security:
* - Rejects IDs containing path separators, "..", or null bytes to prevent
*   directory traversal and path injection outside the resolved subdir.
* - Verifies the resolved path is a regular file (not a symlink or directory)
*   before returning it, matching the write-side MEDIA_FILE_MODE policy.
*
* @param id      The media ID as returned by SavedMedia.id (may include
*                extension and original-filename prefix,
*                e.g. "photo---<uuid>.png" or "图片---<uuid>.png").
* @param subdir  The subdirectory the file was saved into (default "inbound").
* @returns       Absolute path to the file on disk.
* @throws        If the ID is unsafe, the file does not exist, or is not a
*                regular file.
*/
async function resolveMediaBufferPath(id, subdir = "inbound") {
	if (!id || id.includes("/") || id.includes("\\") || id.includes("\0") || id === "..") throw new Error(`resolveMediaBufferPath: unsafe media ID: ${JSON.stringify(id)}`);
	const dir = path.join(resolveMediaDir(), subdir);
	const resolved = path.join(dir, id);
	if (!resolved.startsWith(dir + path.sep) && resolved !== dir) throw new Error(`resolveMediaBufferPath: path escapes media directory: ${JSON.stringify(id)}`);
	const stat = await fs$1.lstat(resolved);
	if (stat.isSymbolicLink()) throw new Error(`resolveMediaBufferPath: refusing to follow symlink for media ID: ${JSON.stringify(id)}`);
	if (!stat.isFile()) throw new Error(`resolveMediaBufferPath: media ID does not resolve to a file: ${JSON.stringify(id)}`);
	return resolved;
}
/**
* Deletes a file previously saved by saveMediaBuffer.
*
* This is used by parseMessageWithAttachments to clean up files that were
* successfully offloaded earlier in the same request when a later attachment
* fails validation and the entire parse is aborted, preventing orphaned files
* from accumulating on disk ahead of the periodic TTL sweep.
*
* Uses resolveMediaBufferPath to apply the same path-safety guards as the
* read path (separator checks, symlink rejection, etc.) before unlinking.
*
* Errors are intentionally not suppressed — callers that want best-effort
* cleanup should catch and discard exceptions themselves (e.g. via
* Promise.allSettled).
*
* @param id     The media ID as returned by SavedMedia.id.
* @param subdir The subdirectory the file was saved into (default "inbound").
*/
async function deleteMediaBuffer(id, subdir = "inbound") {
	const physicalPath = await resolveMediaBufferPath(id, subdir);
	await fs$1.unlink(physicalPath);
}
//#endregion
export { ensureMediaDir as a, resolveMediaBufferPath as c, setMediaStoreNetworkDepsForTest as d, deleteMediaBuffer as i, saveMediaBuffer as l, SaveMediaSourceError as n, extractOriginalFilename as o, cleanOldMedia as r, getMediaDir as s, MEDIA_MAX_BYTES as t, saveMediaSource as u };
