// openclaw/dist/thread-bindings.manager-Dki468EV.js
import { s as normalizeOptionalString } from "./string-coerce-CIT1-v_a.js";
import { u as resolveAgentIdFromSessionKey } from "./session-key-BMb3Kc4r.js";
import { n as normalizeAccountId } from "./account-id-BfovDWrj.js";
import { r as logVerbose } from "./globals-B2UyK3pD.js";
import { r as getRuntimeConfigSnapshot } from "./runtime-snapshot-BVr5fQzj.js";
import { a as registerSessionBindingAdapter, o as unregisterSessionBindingAdapter } from "./session-binding-service-Bwa7BQEQ.js";
import { _ as resolveThreadBindingThreadName, h as resolveThreadBindingFarewellText } from "./thread-bindings-policy-Vf1M899j.js";
import "./text-runtime-BngW26kR.js";
import "./routing-CC84LXmn.js";
import "./runtime-env-DxA7Qhwh.js";
import "./conversation-runtime-C2R31CkT.js";
import { t as resolveThreadBindingConversationIdFromBindingId } from "./thread-binding-id-DYdRH0Im.js";
import { O as createDiscordRestClient } from "./send.shared-Bx-dGYH9.js";
import { A as DEFAULT_THREAD_BINDING_IDLE_TIMEOUT_MS, C as resolveThreadBindingMaxAgeMs, D as shouldDefaultPersist, E as setBindingRecord, M as THREAD_BINDINGS_SWEEP_INTERVAL_MS, S as resolveThreadBindingMaxAgeExpiresAt, T as saveBindingsToDisk, _ as resetThreadBindingsForTests, a as THREAD_BINDING_TOUCH_PERSIST_MIN_INTERVAL_MS, b as resolveThreadBindingIdleTimeoutMs, c as getThreadBindingToken, d as normalizeThreadBindingDurationMs, f as normalizeThreadId, g as removeBindingRecord, h as rememberThreadBindingToken, n as MANAGERS_BY_ACCOUNT_ID, o as ensureBindingsLoaded, p as rememberRecentUnboundWebhookEcho, r as PERSIST_BY_ACCOUNT_ID, s as forgetThreadBindingToken, t as BINDINGS_BY_THREAD_ID, u as normalizeTargetKind, v as resolveBindingIdsForSession, w as resolveThreadBindingsPath, x as resolveThreadBindingInactivityExpiresAt, y as resolveBindingRecordKey } from "./thread-bindings.state-DYxR1cgm.js";
import { a as isThreadArchived, c as summarizeDiscordError, i as isDiscordThreadGoneError, n as createWebhookForChannel, o as maybeSendBindingMessage, r as findReusableWebhook, s as resolveChannelIdForBinding, t as createThreadForBinding } from "./thread-bindings.discord-api-qBdIytJp.js";
import { Routes } from "discord-api-types/v10";
//#region extensions/discord/src/monitor/thread-bindings.manager.ts
function registerManager(manager) {
	MANAGERS_BY_ACCOUNT_ID.set(manager.accountId, manager);
}
function unregisterManager(accountId, manager) {
	if (MANAGERS_BY_ACCOUNT_ID.get(accountId) === manager) MANAGERS_BY_ACCOUNT_ID.delete(accountId);
}
const SWEEPERS_BY_ACCOUNT_ID = /* @__PURE__ */ new Map();
function resolveEffectiveBindingExpiresAt(params) {
	const inactivityExpiresAt = resolveThreadBindingInactivityExpiresAt({
		record: params.record,
		defaultIdleTimeoutMs: params.defaultIdleTimeoutMs
	});
	const maxAgeExpiresAt = resolveThreadBindingMaxAgeExpiresAt({
		record: params.record,
		defaultMaxAgeMs: params.defaultMaxAgeMs
	});
	if (inactivityExpiresAt != null && maxAgeExpiresAt != null) return Math.min(inactivityExpiresAt, maxAgeExpiresAt);
	return inactivityExpiresAt ?? maxAgeExpiresAt;
}
function createNoopManager(accountIdRaw) {
	return {
		accountId: normalizeAccountId(accountIdRaw),
		getIdleTimeoutMs: () => DEFAULT_THREAD_BINDING_IDLE_TIMEOUT_MS,
		getMaxAgeMs: () => 0,
		getByThreadId: () => void 0,
		getBySessionKey: () => void 0,
		listBySessionKey: () => [],
		listBindings: () => [],
		touchThread: () => null,
		bindTarget: async () => null,
		unbindThread: () => null,
		unbindBySessionKey: () => [],
		stop: () => {}
	};
}
function toSessionBindingTargetKind(raw) {
	return raw === "subagent" ? "subagent" : "session";
}
function toThreadBindingTargetKind(raw) {
	return raw === "subagent" ? "subagent" : "acp";
}
function isDirectConversationBindingId(value) {
	const trimmed = normalizeOptionalString(value);
	return Boolean(trimmed && /^(user:|channel:)/i.test(trimmed));
}
function toSessionBindingRecord(record, defaults) {
	return {
		bindingId: resolveBindingRecordKey({
			accountId: record.accountId,
			threadId: record.threadId
		}) ?? `${record.accountId}:${record.threadId}`,
		targetSessionKey: record.targetSessionKey,
		targetKind: toSessionBindingTargetKind(record.targetKind),
		conversation: {
			channel: "discord",
			accountId: record.accountId,
			conversationId: record.threadId,
			parentConversationId: record.channelId
		},
		status: "active",
		boundAt: record.boundAt,
		expiresAt: resolveEffectiveBindingExpiresAt({
			record,
			defaultIdleTimeoutMs: defaults.idleTimeoutMs,
			defaultMaxAgeMs: defaults.maxAgeMs
		}),
		metadata: {
			agentId: record.agentId,
			label: record.label,
			webhookId: record.webhookId,
			webhookToken: record.webhookToken,
			boundBy: record.boundBy,
			lastActivityAt: record.lastActivityAt,
			idleTimeoutMs: resolveThreadBindingIdleTimeoutMs({
				record,
				defaultIdleTimeoutMs: defaults.idleTimeoutMs
			}),
			maxAgeMs: resolveThreadBindingMaxAgeMs({
				record,
				defaultMaxAgeMs: defaults.maxAgeMs
			}),
			...record.metadata
		}
	};
}
function createThreadBindingManager(params = {}) {
	ensureBindingsLoaded();
	const accountId = normalizeAccountId(params.accountId);
	const existing = MANAGERS_BY_ACCOUNT_ID.get(accountId);
	if (existing) {
		rememberThreadBindingToken({
			accountId,
			token: params.token
		});
		return existing;
	}
	rememberThreadBindingToken({
		accountId,
		token: params.token
	});
	const persist = params.persist ?? shouldDefaultPersist();
	PERSIST_BY_ACCOUNT_ID.set(accountId, persist);
	const idleTimeoutMs = normalizeThreadBindingDurationMs(params.idleTimeoutMs, DEFAULT_THREAD_BINDING_IDLE_TIMEOUT_MS);
	const maxAgeMs = normalizeThreadBindingDurationMs(params.maxAgeMs, 0);
	const resolveCurrentCfg = () => getRuntimeConfigSnapshot() ?? params.cfg;
	const resolveCurrentToken = () => getThreadBindingToken(accountId) ?? params.token;
	let sweepTimer = null;
	const runSweepOnce = async () => {
		const bindings = manager.listBindings();
		if (bindings.length === 0) return;
		let rest = null;
		for (const snapshotBinding of bindings) {
			const binding = manager.getByThreadId(snapshotBinding.threadId);
			if (!binding) continue;
			const now = Date.now();
			const inactivityExpiresAt = resolveThreadBindingInactivityExpiresAt({
				record: binding,
				defaultIdleTimeoutMs: idleTimeoutMs
			});
			const maxAgeExpiresAt = resolveThreadBindingMaxAgeExpiresAt({
				record: binding,
				defaultMaxAgeMs: maxAgeMs
			});
			const expirationCandidates = [];
			if (inactivityExpiresAt != null && now >= inactivityExpiresAt) expirationCandidates.push({
				reason: "idle-expired",
				at: inactivityExpiresAt
			});
			if (maxAgeExpiresAt != null && now >= maxAgeExpiresAt) expirationCandidates.push({
				reason: "max-age-expired",
				at: maxAgeExpiresAt
			});
			if (expirationCandidates.length > 0) {
				expirationCandidates.sort((a, b) => a.at - b.at);
				const reason = expirationCandidates[0]?.reason ?? "idle-expired";
				manager.unbindThread({
					threadId: binding.threadId,
					reason,
					sendFarewell: true,
					farewellText: resolveThreadBindingFarewellText({
						reason,
						idleTimeoutMs: resolveThreadBindingIdleTimeoutMs({
							record: binding,
							defaultIdleTimeoutMs: idleTimeoutMs
						}),
						maxAgeMs: resolveThreadBindingMaxAgeMs({
							record: binding,
							defaultMaxAgeMs: maxAgeMs
						})
					})
				});
				continue;
			}
			if (isDirectConversationBindingId(binding.threadId)) continue;
			if (!rest) try {
				const cfg = resolveCurrentCfg();
				rest = createDiscordRestClient({
					accountId,
					token: resolveCurrentToken()
				}, cfg).rest;
			} catch {
				return;
			}
			try {
				const channel = await rest.get(Routes.channel(binding.threadId));
				if (!channel || typeof channel !== "object") {
					logVerbose(`discord thread binding sweep probe returned invalid payload for ${binding.threadId}`);
					continue;
				}
				if (isThreadArchived(channel)) manager.unbindThread({
					threadId: binding.threadId,
					reason: "thread-archived",
					sendFarewell: true
				});
			} catch (err) {
				if (isDiscordThreadGoneError(err)) {
					logVerbose(`discord thread binding sweep removing stale binding ${binding.threadId}: ${summarizeDiscordError(err)}`);
					manager.unbindThread({
						threadId: binding.threadId,
						reason: "thread-delete",
						sendFarewell: false
					});
					continue;
				}
				logVerbose(`discord thread binding sweep probe failed for ${binding.threadId}: ${summarizeDiscordError(err)}`);
			}
		}
	};
	SWEEPERS_BY_ACCOUNT_ID.set(accountId, runSweepOnce);
	const manager = {
		accountId,
		getIdleTimeoutMs: () => idleTimeoutMs,
		getMaxAgeMs: () => maxAgeMs,
		getByThreadId: (threadId) => {
			const key = resolveBindingRecordKey({
				accountId,
				threadId
			});
			if (!key) return;
			const entry = BINDINGS_BY_THREAD_ID.get(key);
			if (!entry || entry.accountId !== accountId) return;
			return entry;
		},
		getBySessionKey: (targetSessionKey) => {
			return manager.listBySessionKey(targetSessionKey)[0];
		},
		listBySessionKey: (targetSessionKey) => {
			return resolveBindingIdsForSession({
				targetSessionKey,
				accountId
			}).map((bindingKey) => BINDINGS_BY_THREAD_ID.get(bindingKey)).filter((entry) => Boolean(entry));
		},
		listBindings: () => [...BINDINGS_BY_THREAD_ID.values()].filter((entry) => entry.accountId === accountId),
		touchThread: (touchParams) => {
			const key = resolveBindingRecordKey({
				accountId,
				threadId: touchParams.threadId
			});
			if (!key) return null;
			const existing = BINDINGS_BY_THREAD_ID.get(key);
			if (!existing || existing.accountId !== accountId) return null;
			const now = Date.now();
			const at = typeof touchParams.at === "number" && Number.isFinite(touchParams.at) ? Math.max(0, Math.floor(touchParams.at)) : now;
			const nextRecord = {
				...existing,
				lastActivityAt: Math.max(existing.lastActivityAt || 0, at)
			};
			setBindingRecord(nextRecord);
			if (touchParams.persist ?? persist) saveBindingsToDisk({ minIntervalMs: THREAD_BINDING_TOUCH_PERSIST_MIN_INTERVAL_MS });
			return nextRecord;
		},
		bindTarget: async (bindParams) => {
			const cfg = resolveCurrentCfg();
			let threadId = normalizeThreadId(bindParams.threadId);
			let channelId = normalizeOptionalString(bindParams.channelId) ?? "";
			const directConversationBinding = isDirectConversationBindingId(threadId) || isDirectConversationBindingId(channelId);
			if (!threadId && bindParams.createThread) {
				if (!channelId) return null;
				const threadName = resolveThreadBindingThreadName({
					agentId: bindParams.agentId,
					label: bindParams.label
				});
				threadId = await createThreadForBinding({
					cfg,
					accountId,
					token: resolveCurrentToken(),
					channelId,
					threadName: normalizeOptionalString(bindParams.threadName) ?? threadName
				}) ?? void 0;
			}
			if (!threadId) return null;
			if (!channelId && directConversationBinding) channelId = threadId;
			if (!channelId) channelId = await resolveChannelIdForBinding({
				cfg,
				accountId,
				token: resolveCurrentToken(),
				threadId,
				channelId: bindParams.channelId
			}) ?? "";
			if (!channelId) return null;
			const targetSessionKey = normalizeOptionalString(bindParams.targetSessionKey) ?? "";
			if (!targetSessionKey) return null;
			const targetKind = normalizeTargetKind(bindParams.targetKind, targetSessionKey);
			let webhookId = normalizeOptionalString(bindParams.webhookId) ?? "";
			let webhookToken = normalizeOptionalString(bindParams.webhookToken) ?? "";
			if (!directConversationBinding && (!webhookId || !webhookToken)) {
				const cachedWebhook = findReusableWebhook({
					accountId,
					channelId
				});
				webhookId = cachedWebhook.webhookId ?? "";
				webhookToken = cachedWebhook.webhookToken ?? "";
			}
			if (!directConversationBinding && (!webhookId || !webhookToken)) {
				const createdWebhook = await createWebhookForChannel({
					cfg,
					accountId,
					token: resolveCurrentToken(),
					channelId
				});
				webhookId = createdWebhook.webhookId ?? "";
				webhookToken = createdWebhook.webhookToken ?? "";
			}
			const now = Date.now();
			const record = {
				accountId,
				channelId,
				threadId,
				targetKind,
				targetSessionKey,
				agentId: normalizeOptionalString(bindParams.agentId) ?? resolveAgentIdFromSessionKey(targetSessionKey),
				label: normalizeOptionalString(bindParams.label),
				webhookId: webhookId || void 0,
				webhookToken: webhookToken || void 0,
				boundBy: normalizeOptionalString(bindParams.boundBy) || "system",
				boundAt: now,
				lastActivityAt: now,
				idleTimeoutMs,
				maxAgeMs,
				metadata: bindParams.metadata && typeof bindParams.metadata === "object" ? { ...bindParams.metadata } : void 0
			};
			setBindingRecord(record);
			if (persist) saveBindingsToDisk();
			const introText = bindParams.introText?.trim();
			if (introText) maybeSendBindingMessage({
				cfg,
				record,
				text: introText
			});
			return record;
		},
		unbindThread: (unbindParams) => {
			const bindingKey = resolveBindingRecordKey({
				accountId,
				threadId: unbindParams.threadId
			});
			if (!bindingKey) return null;
			const existing = BINDINGS_BY_THREAD_ID.get(bindingKey);
			if (!existing || existing.accountId !== accountId) return null;
			const removed = removeBindingRecord(bindingKey);
			if (!removed) return null;
			rememberRecentUnboundWebhookEcho(removed);
			if (persist) saveBindingsToDisk();
			if (unbindParams.sendFarewell !== false) maybeSendBindingMessage({
				cfg: resolveCurrentCfg(),
				record: removed,
				text: resolveThreadBindingFarewellText({
					reason: unbindParams.reason,
					farewellText: unbindParams.farewellText,
					idleTimeoutMs: resolveThreadBindingIdleTimeoutMs({
						record: removed,
						defaultIdleTimeoutMs: idleTimeoutMs
					}),
					maxAgeMs: resolveThreadBindingMaxAgeMs({
						record: removed,
						defaultMaxAgeMs: maxAgeMs
					})
				}),
				preferWebhook: false
			});
			return removed;
		},
		unbindBySessionKey: (unbindParams) => {
			const ids = resolveBindingIdsForSession({
				targetSessionKey: unbindParams.targetSessionKey,
				accountId,
				targetKind: unbindParams.targetKind
			});
			if (ids.length === 0) return [];
			const removed = [];
			for (const bindingKey of ids) {
				const binding = BINDINGS_BY_THREAD_ID.get(bindingKey);
				if (!binding) continue;
				const entry = manager.unbindThread({
					threadId: binding.threadId,
					reason: unbindParams.reason,
					sendFarewell: unbindParams.sendFarewell,
					farewellText: unbindParams.farewellText
				});
				if (entry) removed.push(entry);
			}
			return removed;
		},
		stop: () => {
			if (sweepTimer) {
				clearInterval(sweepTimer);
				sweepTimer = null;
			}
			SWEEPERS_BY_ACCOUNT_ID.delete(accountId);
			unregisterManager(accountId, manager);
			unregisterSessionBindingAdapter({
				channel: "discord",
				accountId,
				adapter: sessionBindingAdapter
			});
			forgetThreadBindingToken(accountId);
		}
	};
	if (params.enableSweeper !== false) {
		sweepTimer = setInterval(() => {
			runSweepOnce();
		}, THREAD_BINDINGS_SWEEP_INTERVAL_MS);
		if (!(process.env.VITEST || false)) sweepTimer.unref?.();
	}
	const sessionBindingAdapter = {
		channel: "discord",
		accountId,
		capabilities: { placements: ["current", "child"] },
		bind: async (input) => {
			if (input.conversation.channel !== "discord") return null;
			const targetSessionKey = input.targetSessionKey.trim();
			if (!targetSessionKey) return null;
			const conversationId = normalizeOptionalString(input.conversation.conversationId) ?? "";
			const placement = input.placement === "child" ? "child" : "current";
			const metadata = input.metadata ?? {};
			const label = normalizeOptionalString(metadata.label);
			const threadName = typeof metadata.threadName === "string" ? normalizeOptionalString(metadata.threadName) : void 0;
			const introText = typeof metadata.introText === "string" ? normalizeOptionalString(metadata.introText) : void 0;
			const boundBy = typeof metadata.boundBy === "string" ? normalizeOptionalString(metadata.boundBy) : void 0;
			const agentId = typeof metadata.agentId === "string" ? normalizeOptionalString(metadata.agentId) : void 0;
			let threadId;
			let channelId = normalizeOptionalString(input.conversation.parentConversationId);
			let createThread = false;
			if (placement === "child") {
				createThread = true;
				if (!channelId && conversationId) channelId = await resolveChannelIdForBinding({
					cfg: resolveCurrentCfg(),
					accountId,
					token: resolveCurrentToken(),
					threadId: conversationId
				}) ?? void 0;
			} else threadId = conversationId || void 0;
			const bound = await manager.bindTarget({
				threadId,
				channelId,
				createThread,
				threadName,
				targetKind: toThreadBindingTargetKind(input.targetKind),
				targetSessionKey,
				agentId,
				label,
				boundBy,
				introText,
				metadata
			});
			return bound ? toSessionBindingRecord(bound, {
				idleTimeoutMs,
				maxAgeMs
			}) : null;
		},
		listBySession: (targetSessionKey) => manager.listBySessionKey(targetSessionKey).map((entry) => toSessionBindingRecord(entry, {
			idleTimeoutMs,
			maxAgeMs
		})),
		resolveByConversation: (ref) => {
			if (ref.channel !== "discord") return null;
			const binding = manager.getByThreadId(ref.conversationId);
			return binding ? toSessionBindingRecord(binding, {
				idleTimeoutMs,
				maxAgeMs
			}) : null;
		},
		touch: (bindingId, at) => {
			const threadId = resolveThreadBindingConversationIdFromBindingId({
				accountId,
				bindingId
			});
			if (!threadId) return;
			manager.touchThread({
				threadId,
				at,
				persist: true
			});
		},
		unbind: async (input) => {
			if (input.targetSessionKey?.trim()) return manager.unbindBySessionKey({
				targetSessionKey: input.targetSessionKey,
				reason: input.reason
			}).map((entry) => toSessionBindingRecord(entry, {
				idleTimeoutMs,
				maxAgeMs
			}));
			const threadId = resolveThreadBindingConversationIdFromBindingId({
				accountId,
				bindingId: input.bindingId
			});
			if (!threadId) return [];
			const removed = manager.unbindThread({
				threadId,
				reason: input.reason
			});
			return removed ? [toSessionBindingRecord(removed, {
				idleTimeoutMs,
				maxAgeMs
			})] : [];
		}
	};
	registerSessionBindingAdapter(sessionBindingAdapter);
	registerManager(manager);
	return manager;
}
function createNoopThreadBindingManager(accountId) {
	return createNoopManager(accountId);
}
function getThreadBindingManager(accountId) {
	const normalized = normalizeAccountId(accountId);
	return MANAGERS_BY_ACCOUNT_ID.get(normalized) ?? null;
}
const __testing = {
	resolveThreadBindingsPath,
	resolveThreadBindingThreadName,
	resetThreadBindingsForTests,
	runThreadBindingSweepForAccount: async (accountId) => {
		const sweep = SWEEPERS_BY_ACCOUNT_ID.get(normalizeAccountId(accountId));
		if (sweep) await sweep();
	}
};
//#endregion
export { getThreadBindingManager as i, createNoopThreadBindingManager as n, createThreadBindingManager as r, __testing as t };
