// openclaw/dist/video-generation-provider-Bzu42GKw.js
import { r as postJsonRequest, t as assertOkOrThrowHttpError } from "./shared-BGaxsxBu.js";
import "./provider-http-Dk5isK1Q.js";
import { t as isProviderApiKeyConfigured } from "./provider-auth-BLK7aiS1.js";
import { a as DEFAULT_VYDRA_VIDEO_MODEL, c as extractVydraResultUrls, h as resolveVydraResponseStatus, m as resolveVydraResponseJobId, p as resolveVydraRequestContext, s as downloadVydraAsset, u as resolveCompletedVydraPayload } from "./shared-3hZRCDrZ.js";
//#region extensions/vydra/video-generation-provider.ts
const VYDRA_KLING_MODEL = "kling";
function resolveVydraVideoRequestBody(req) {
	const model = req.model?.trim() || "veo3";
	if (model === VYDRA_KLING_MODEL) {
		const imageUrl = (req.inputImages?.[0])?.url?.trim();
		if (!imageUrl) throw new Error("Vydra kling currently requires a remote image URL reference.");
		return {
			model,
			body: {
				prompt: req.prompt,
				image_url: imageUrl,
				video_url: imageUrl
			}
		};
	}
	if ((req.inputImages?.length ?? 0) > 0) throw new Error(`Vydra ${model} does not support image reference inputs in the bundled plugin.`);
	return {
		model,
		body: { prompt: req.prompt }
	};
}
function buildVydraVideoGenerationProvider() {
	return {
		id: "vydra",
		label: "Vydra",
		defaultModel: DEFAULT_VYDRA_VIDEO_MODEL,
		models: [DEFAULT_VYDRA_VIDEO_MODEL, VYDRA_KLING_MODEL],
		isConfigured: ({ agentDir }) => isProviderApiKeyConfigured({
			provider: "vydra",
			agentDir
		}),
		capabilities: {
			generate: { maxVideos: 1 },
			imageToVideo: {
				enabled: true,
				maxVideos: 1,
				maxInputImages: 1
			},
			videoToVideo: { enabled: false }
		},
		async generateVideo(req) {
			if ((req.inputVideos?.length ?? 0) > 0) throw new Error("Vydra video generation does not support video reference inputs.");
			const { fetchFn, baseUrl, allowPrivateNetwork, headers, dispatcherPolicy } = await resolveVydraRequestContext({
				cfg: req.cfg,
				agentDir: req.agentDir,
				authStore: req.authStore,
				capability: "video"
			});
			const { model, body } = resolveVydraVideoRequestBody(req);
			const { response, release } = await postJsonRequest({
				url: `${baseUrl}/models/${model}`,
				headers,
				body,
				timeoutMs: req.timeoutMs,
				fetchFn,
				allowPrivateNetwork,
				dispatcherPolicy
			});
			try {
				await assertOkOrThrowHttpError(response, "Vydra video generation failed");
				const submitted = await response.json();
				const completedPayload = await resolveCompletedVydraPayload({
					submitted,
					baseUrl,
					headers,
					timeoutMs: req.timeoutMs,
					fetchFn,
					kind: "video",
					missingJobIdMessage: "Vydra video generation response missing job id"
				});
				const videoUrl = extractVydraResultUrls(completedPayload, "video")[0];
				if (!videoUrl) throw new Error("Vydra video generation completed without a video URL");
				const video = await downloadVydraAsset({
					url: videoUrl,
					kind: "video",
					timeoutMs: req.timeoutMs,
					fetchFn
				});
				return {
					videos: [{
						buffer: video.buffer,
						mimeType: video.mimeType,
						fileName: video.fileName
					}],
					model,
					metadata: {
						jobId: resolveVydraResponseJobId(completedPayload) ?? resolveVydraResponseJobId(submitted),
						videoUrl,
						status: resolveVydraResponseStatus(completedPayload) ?? "completed"
					}
				};
			} finally {
				await release();
			}
		}
	};
}
//#endregion
export { buildVydraVideoGenerationProvider as t };
