// openshell/dist/config.js
import dotenv from "dotenv";
import { isAbsolute, join, resolve } from "node:path";
dotenv.config();
function normalizeProvider(value) {
    return value?.toLowerCase() === "claude" ? "claude" : "codex";
}
function normalizeString(value) {
    const trimmed = value?.trim();
    return trimmed ? trimmed : undefined;
}
function parseMaxHistory(value) {
    const parsed = Number.parseInt(value ?? "20", 10);
    if (!Number.isFinite(parsed) || parsed < 2 || parsed > 200) {
        throw new Error("MAX_HISTORY must be an integer between 2 and 200.");
    }
    return parsed;
}
function parseEditInterval(value) {
    const parsed = Number.parseInt(value ?? "900", 10);
    if (!Number.isFinite(parsed) || parsed < 200 || parsed > 10000) {
        throw new Error("STREAM_EDIT_INTERVAL_MS must be 200-10000.");
    }
    return parsed;
}
function parseSessionStorePath(value) {
    const input = normalizeString(value) ?? ".openshell/session-store.json";
    const expanded = input === "~"
        ? (process.env.HOME ?? input)
        : input.startsWith("~/")
            ? (process.env.HOME ? join(process.env.HOME, input.slice(2)) : input)
            : input;
    return isAbsolute(expanded) ? expanded : resolve(process.cwd(), expanded);
}
export function loadConfig() {
    const telegramBotToken = normalizeString(process.env.TELEGRAM_BOT_TOKEN);
    if (!telegramBotToken) {
        throw new Error("TELEGRAM_BOT_TOKEN is required.");
    }
    const defaultProvider = normalizeProvider(process.env.DEFAULT_PROVIDER);
    return {
        telegramBotToken,
        defaultProvider,
        maxHistory: parseMaxHistory(process.env.MAX_HISTORY),
        systemPrompt: normalizeString(process.env.SYSTEM_PROMPT),
        cliWorkdir: normalizeString(process.env.CLI_WORKDIR) ?? process.cwd(),
        sessionStorePath: parseSessionStorePath(process.env.SESSION_STORE_PATH),
        streamEditIntervalMs: parseEditInterval(process.env.STREAM_EDIT_INTERVAL_MS),
        codex: {
            command: normalizeString(process.env.CODEX_COMMAND) ?? "codex",
            model: normalizeString(process.env.CODEX_MODEL),
        },
        claude: {
            command: normalizeString(process.env.CLAUDE_COMMAND) ?? "claude",
            model: normalizeString(process.env.CLAUDE_MODEL),
        },
    };
}
