// @pierre/diffs/dist/react/index.js
import { GutterUtilitySlotStyles, MergeConflictSlotStyles, noopRender } from "./constants.js";
import { renderFileChildren } from "./utils/renderFileChildren.js";
import { templateRender } from "./utils/templateRender.js";
import { Virtualizer, VirtualizerContext, useVirtualizer } from "./Virtualizer.js";
import { WorkerPoolContext, WorkerPoolContextProvider, useWorkerPool } from "./WorkerPoolContext.js";
import { useStableCallback } from "./utils/useStableCallback.js";
import { useFileInstance } from "./utils/useFileInstance.js";
import { File } from "./File.js";
import { renderDiffChildren } from "./utils/renderDiffChildren.js";
import { useFileDiffInstance } from "./utils/useFileDiffInstance.js";
import { FileDiff } from "./FileDiff.js";
import { MultiFileDiff } from "./MultiFileDiff.js";
import { PatchDiff } from "./PatchDiff.js";
import { UnresolvedFile } from "./UnresolvedFile.js";

export { File, FileDiff, GutterUtilitySlotStyles, MergeConflictSlotStyles, MultiFileDiff, PatchDiff, UnresolvedFile, Virtualizer, VirtualizerContext, WorkerPoolContext, WorkerPoolContextProvider, noopRender, renderDiffChildren, renderFileChildren, templateRender, useFileDiffInstance, useFileInstance, useStableCallback, useVirtualizer, useWorkerPool };
