// openclaw/dist/extensions/whatsapp/test-api.js
import { a as shouldLogVerbose } from "../../globals-B2UyK3pD.js";
import { b as sendTextMediaPayload, p as resolveSendableOutboundReplyParts } from "../../reply-payload-CcbNO0Ax.js";
import { a as chunkText } from "../../chunk-DS9PgIig.js";
import { t as resolveOutboundSendDep } from "../../send-deps-B8hkliIX.js";
import "../../runtime-env-DxA7Qhwh.js";
import "../../reply-runtime-Bj9CGlMM.js";
import { t as sanitizeForPlainText } from "../../sanitize-text-D1JP-tV-.js";
import "../../outbound-runtime-DxyEM-0h.js";
import { t as globalExpect } from "../../test.p_J6dB8a-sNRqGTRJ.js";
import { a as createEmptyChannelResult, i as createAttachedChannelResultAdapter } from "../../channel-send-result-COHJCNM4.js";
import { a as WHATSAPP_LEGACY_OUTBOUND_SEND_DEP_KEYS, c as resolveWhatsAppOutboundTarget, n as sendPollWhatsApp } from "../../send-CRc-uCHW.js";
import { t as resolveWhatsAppRuntimeGroupPolicy } from "../../runtime-group-policy-DiLtBJ4r.js";
//#region extensions/whatsapp/src/outbound-adapter.ts
function trimLeadingWhitespace(text) {
	return text?.trimStart() ?? "";
}
const whatsappOutbound = {
	deliveryMode: "gateway",
	chunker: chunkText,
	chunkerMode: "text",
	textChunkLimit: 4e3,
	sanitizeText: ({ text }) => sanitizeForPlainText(text),
	pollMaxOptions: 12,
	resolveTarget: ({ to, allowFrom, mode }) => resolveWhatsAppOutboundTarget({
		to,
		allowFrom,
		mode
	}),
	sendPayload: async (ctx) => {
		const text = trimLeadingWhitespace(ctx.payload.text);
		const hasMedia = resolveSendableOutboundReplyParts(ctx.payload).hasMedia;
		if (!text && !hasMedia) return createEmptyChannelResult("whatsapp");
		return await sendTextMediaPayload({
			channel: "whatsapp",
			ctx: {
				...ctx,
				payload: {
					...ctx.payload,
					text
				}
			},
			adapter: whatsappOutbound
		});
	},
	...createAttachedChannelResultAdapter({
		channel: "whatsapp",
		sendText: async ({ cfg, to, text, accountId, deps, gifPlayback }) => {
			const normalizedText = trimLeadingWhitespace(text);
			if (!normalizedText) return createEmptyChannelResult("whatsapp");
			return await (resolveOutboundSendDep(deps, "whatsapp", { legacyKeys: WHATSAPP_LEGACY_OUTBOUND_SEND_DEP_KEYS }) ?? (await import("../../send-3UIWngH9.js")).sendMessageWhatsApp)(to, normalizedText, {
				verbose: false,
				cfg,
				accountId: accountId ?? void 0,
				gifPlayback
			});
		},
		sendMedia: async ({ cfg, to, text, mediaUrl, mediaLocalRoots, mediaReadFile, accountId, deps, gifPlayback }) => {
			const normalizedText = trimLeadingWhitespace(text);
			return await (resolveOutboundSendDep(deps, "whatsapp", { legacyKeys: WHATSAPP_LEGACY_OUTBOUND_SEND_DEP_KEYS }) ?? (await import("../../send-3UIWngH9.js")).sendMessageWhatsApp)(to, normalizedText, {
				verbose: false,
				cfg,
				mediaUrl,
				mediaLocalRoots,
				mediaReadFile,
				accountId: accountId ?? void 0,
				gifPlayback
			});
		},
		sendPoll: async ({ cfg, to, poll, accountId }) => await sendPollWhatsApp(to, poll, {
			verbose: shouldLogVerbose(),
			accountId: accountId ?? void 0,
			cfg
		})
	})
};
//#endregion
//#region extensions/whatsapp/src/outbound-test-support.ts
function createWhatsAppPollFixture() {
	return {
		cfg: { marker: "resolved-cfg" },
		poll: {
			question: "Lunch?",
			options: ["Pizza", "Sushi"],
			maxSelections: 1
		},
		to: "+1555",
		accountId: "work"
	};
}
function expectWhatsAppPollSent(sendPollWhatsApp, params) {
	globalExpect(sendPollWhatsApp).toHaveBeenCalledWith(params.to ?? "+1555", params.poll, {
		verbose: false,
		accountId: params.accountId ?? "work",
		cfg: params.cfg
	});
}
//#endregion
export { createWhatsAppPollFixture, expectWhatsAppPollSent, resolveWhatsAppRuntimeGroupPolicy, whatsappOutbound };
